// rclnodejs/generated/gps_msgs/gps_msgs__msg__GPSFix.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GPSStatusWrapper = require('../../generated/gps_msgs/gps_msgs__msg__GPSStatus.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const GPSFixRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  status: GPSStatusWrapper.refObjectType,
  latitude: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  altitude: primitiveTypes.float64,
  track: primitiveTypes.float64,
  speed: primitiveTypes.float64,
  climb: primitiveTypes.float64,
  pitch: primitiveTypes.float64,
  roll: primitiveTypes.float64,
  dip: primitiveTypes.float64,
  time: primitiveTypes.float64,
  gdop: primitiveTypes.float64,
  pdop: primitiveTypes.float64,
  hdop: primitiveTypes.float64,
  vdop: primitiveTypes.float64,
  tdop: primitiveTypes.float64,
  err: primitiveTypes.float64,
  err_horz: primitiveTypes.float64,
  err_vert: primitiveTypes.float64,
  err_track: primitiveTypes.float64,
  err_speed: primitiveTypes.float64,
  err_climb: primitiveTypes.float64,
  err_time: primitiveTypes.float64,
  err_pitch: primitiveTypes.float64,
  err_roll: primitiveTypes.float64,
  err_dip: primitiveTypes.float64,
  position_covariance: ArrayType(primitiveTypes.float64, 9),
  position_covariance_type: primitiveTypes.uint8,
});
const GPSFixRefArray = ArrayType(GPSFixRefStruct);
const GPSFixRefStructArray = StructType({
  data: GPSFixRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GPSFixWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPSFixRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.status =  new GPSStatusWrapper(other._wrapperFields.status);
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
      this._trackIntialized = true;
      this._speedIntialized = true;
      this._climbIntialized = true;
      this._pitchIntialized = true;
      this._rollIntialized = true;
      this._dipIntialized = true;
      this._timeIntialized = true;
      this._gdopIntialized = true;
      this._pdopIntialized = true;
      this._hdopIntialized = true;
      this._vdopIntialized = true;
      this._tdopIntialized = true;
      this._errIntialized = true;
      this._err_horzIntialized = true;
      this._err_vertIntialized = true;
      this._err_trackIntialized = true;
      this._err_speedIntialized = true;
      this._err_climbIntialized = true;
      this._err_timeIntialized = true;
      this._err_pitchIntialized = true;
      this._err_rollIntialized = true;
      this._err_dipIntialized = true;
      this._wrapperFields.position_covariance = Float64Wrapper.createArray();
      this._wrapperFields.position_covariance.copy(other._wrapperFields.position_covariance);
      this._position_covariance_typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GPSFixRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.status = new GPSStatusWrapper();
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
    this._trackIntialized = false;
    this._speedIntialized = false;
    this._climbIntialized = false;
    this._pitchIntialized = false;
    this._rollIntialized = false;
    this._dipIntialized = false;
    this._timeIntialized = false;
    this._gdopIntialized = false;
    this._pdopIntialized = false;
    this._hdopIntialized = false;
    this._vdopIntialized = false;
    this._tdopIntialized = false;
    this._errIntialized = false;
    this._err_horzIntialized = false;
    this._err_vertIntialized = false;
    this._err_trackIntialized = false;
    this._err_speedIntialized = false;
    this._err_climbIntialized = false;
    this._err_timeIntialized = false;
    this._err_pitchIntialized = false;
    this._err_rollIntialized = false;
    this._err_dipIntialized = false;
    this._wrapperFields.position_covariance = Float64Wrapper.createArray();
    this._position_covariance_typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GPSFixWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPSFixArrayWrapper;
  }
  static get ArrayType() {
    return GPSFixArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPSFixRefStructArray
  }
  static get refObjectType() {
    return GPSFixRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in GPSFix');
      }
      if (checkConsistency && !this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in GPSFix');
      }
      if (checkConsistency && !this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in GPSFix');
      }
      if (checkConsistency && !this._trackIntialized) {
        throw new TypeError('Invalid argument: track in GPSFix');
      }
      if (checkConsistency && !this._speedIntialized) {
        throw new TypeError('Invalid argument: speed in GPSFix');
      }
      if (checkConsistency && !this._climbIntialized) {
        throw new TypeError('Invalid argument: climb in GPSFix');
      }
      if (checkConsistency && !this._pitchIntialized) {
        throw new TypeError('Invalid argument: pitch in GPSFix');
      }
      if (checkConsistency && !this._rollIntialized) {
        throw new TypeError('Invalid argument: roll in GPSFix');
      }
      if (checkConsistency && !this._dipIntialized) {
        throw new TypeError('Invalid argument: dip in GPSFix');
      }
      if (checkConsistency && !this._timeIntialized) {
        throw new TypeError('Invalid argument: time in GPSFix');
      }
      if (checkConsistency && !this._gdopIntialized) {
        throw new TypeError('Invalid argument: gdop in GPSFix');
      }
      if (checkConsistency && !this._pdopIntialized) {
        throw new TypeError('Invalid argument: pdop in GPSFix');
      }
      if (checkConsistency && !this._hdopIntialized) {
        throw new TypeError('Invalid argument: hdop in GPSFix');
      }
      if (checkConsistency && !this._vdopIntialized) {
        throw new TypeError('Invalid argument: vdop in GPSFix');
      }
      if (checkConsistency && !this._tdopIntialized) {
        throw new TypeError('Invalid argument: tdop in GPSFix');
      }
      if (checkConsistency && !this._errIntialized) {
        throw new TypeError('Invalid argument: err in GPSFix');
      }
      if (checkConsistency && !this._err_horzIntialized) {
        throw new TypeError('Invalid argument: err_horz in GPSFix');
      }
      if (checkConsistency && !this._err_vertIntialized) {
        throw new TypeError('Invalid argument: err_vert in GPSFix');
      }
      if (checkConsistency && !this._err_trackIntialized) {
        throw new TypeError('Invalid argument: err_track in GPSFix');
      }
      if (checkConsistency && !this._err_speedIntialized) {
        throw new TypeError('Invalid argument: err_speed in GPSFix');
      }
      if (checkConsistency && !this._err_climbIntialized) {
        throw new TypeError('Invalid argument: err_climb in GPSFix');
      }
      if (checkConsistency && !this._err_timeIntialized) {
        throw new TypeError('Invalid argument: err_time in GPSFix');
      }
      if (checkConsistency && !this._err_pitchIntialized) {
        throw new TypeError('Invalid argument: err_pitch in GPSFix');
      }
      if (checkConsistency && !this._err_rollIntialized) {
        throw new TypeError('Invalid argument: err_roll in GPSFix');
      }
      if (checkConsistency && !this._err_dipIntialized) {
        throw new TypeError('Invalid argument: err_dip in GPSFix');
      }
      if (checkConsistency && !this._position_covariance_typeIntialized) {
        throw new TypeError('Invalid argument: position_covariance_type in GPSFix');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
    this._refObject.position_covariance = Array.from(this._wrapperFields.position_covariance.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.status.copyRefObject(refObject.status);
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._trackIntialized = true;
    this._refObject.track = refObject.track;
    this._speedIntialized = true;
    this._refObject.speed = refObject.speed;
    this._climbIntialized = true;
    this._refObject.climb = refObject.climb;
    this._pitchIntialized = true;
    this._refObject.pitch = refObject.pitch;
    this._rollIntialized = true;
    this._refObject.roll = refObject.roll;
    this._dipIntialized = true;
    this._refObject.dip = refObject.dip;
    this._timeIntialized = true;
    this._refObject.time = refObject.time;
    this._gdopIntialized = true;
    this._refObject.gdop = refObject.gdop;
    this._pdopIntialized = true;
    this._refObject.pdop = refObject.pdop;
    this._hdopIntialized = true;
    this._refObject.hdop = refObject.hdop;
    this._vdopIntialized = true;
    this._refObject.vdop = refObject.vdop;
    this._tdopIntialized = true;
    this._refObject.tdop = refObject.tdop;
    this._errIntialized = true;
    this._refObject.err = refObject.err;
    this._err_horzIntialized = true;
    this._refObject.err_horz = refObject.err_horz;
    this._err_vertIntialized = true;
    this._refObject.err_vert = refObject.err_vert;
    this._err_trackIntialized = true;
    this._refObject.err_track = refObject.err_track;
    this._err_speedIntialized = true;
    this._refObject.err_speed = refObject.err_speed;
    this._err_climbIntialized = true;
    this._refObject.err_climb = refObject.err_climb;
    this._err_timeIntialized = true;
    this._refObject.err_time = refObject.err_time;
    this._err_pitchIntialized = true;
    this._refObject.err_pitch = refObject.err_pitch;
    this._err_rollIntialized = true;
    this._refObject.err_roll = refObject.err_roll;
    this._err_dipIntialized = true;
    this._refObject.err_dip = refObject.err_dip;
    this._wrapperFields.position_covariance.fill(refObject.position_covariance.toArray());
    this._position_covariance_typeIntialized = true;
    this._refObject.position_covariance_type = refObject.position_covariance_type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    GPSStatusWrapper.freeStruct(refObject.status);
  }
  static destoryRawROS(msg) {
    GPSFixWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gps_msgs', subFolder: 'msg', interfaceName: 'GPSFix'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof GPSStatusWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new GPSStatusWrapper(value));
    }
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get track() {
    if (!this._trackIntialized) {
      return undefined;
    }
    return this._refObject.track;
  }
  set track(value) {
    this._trackIntialized = true;
    this._refObject.track = value;
  }
  get speed() {
    if (!this._speedIntialized) {
      return undefined;
    }
    return this._refObject.speed;
  }
  set speed(value) {
    this._speedIntialized = true;
    this._refObject.speed = value;
  }
  get climb() {
    if (!this._climbIntialized) {
      return undefined;
    }
    return this._refObject.climb;
  }
  set climb(value) {
    this._climbIntialized = true;
    this._refObject.climb = value;
  }
  get pitch() {
    if (!this._pitchIntialized) {
      return undefined;
    }
    return this._refObject.pitch;
  }
  set pitch(value) {
    this._pitchIntialized = true;
    this._refObject.pitch = value;
  }
  get roll() {
    if (!this._rollIntialized) {
      return undefined;
    }
    return this._refObject.roll;
  }
  set roll(value) {
    this._rollIntialized = true;
    this._refObject.roll = value;
  }
  get dip() {
    if (!this._dipIntialized) {
      return undefined;
    }
    return this._refObject.dip;
  }
  set dip(value) {
    this._dipIntialized = true;
    this._refObject.dip = value;
  }
  get time() {
    if (!this._timeIntialized) {
      return undefined;
    }
    return this._refObject.time;
  }
  set time(value) {
    this._timeIntialized = true;
    this._refObject.time = value;
  }
  get gdop() {
    if (!this._gdopIntialized) {
      return undefined;
    }
    return this._refObject.gdop;
  }
  set gdop(value) {
    this._gdopIntialized = true;
    this._refObject.gdop = value;
  }
  get pdop() {
    if (!this._pdopIntialized) {
      return undefined;
    }
    return this._refObject.pdop;
  }
  set pdop(value) {
    this._pdopIntialized = true;
    this._refObject.pdop = value;
  }
  get hdop() {
    if (!this._hdopIntialized) {
      return undefined;
    }
    return this._refObject.hdop;
  }
  set hdop(value) {
    this._hdopIntialized = true;
    this._refObject.hdop = value;
  }
  get vdop() {
    if (!this._vdopIntialized) {
      return undefined;
    }
    return this._refObject.vdop;
  }
  set vdop(value) {
    this._vdopIntialized = true;
    this._refObject.vdop = value;
  }
  get tdop() {
    if (!this._tdopIntialized) {
      return undefined;
    }
    return this._refObject.tdop;
  }
  set tdop(value) {
    this._tdopIntialized = true;
    this._refObject.tdop = value;
  }
  get err() {
    if (!this._errIntialized) {
      return undefined;
    }
    return this._refObject.err;
  }
  set err(value) {
    this._errIntialized = true;
    this._refObject.err = value;
  }
  get err_horz() {
    if (!this._err_horzIntialized) {
      return undefined;
    }
    return this._refObject.err_horz;
  }
  set err_horz(value) {
    this._err_horzIntialized = true;
    this._refObject.err_horz = value;
  }
  get err_vert() {
    if (!this._err_vertIntialized) {
      return undefined;
    }
    return this._refObject.err_vert;
  }
  set err_vert(value) {
    this._err_vertIntialized = true;
    this._refObject.err_vert = value;
  }
  get err_track() {
    if (!this._err_trackIntialized) {
      return undefined;
    }
    return this._refObject.err_track;
  }
  set err_track(value) {
    this._err_trackIntialized = true;
    this._refObject.err_track = value;
  }
  get err_speed() {
    if (!this._err_speedIntialized) {
      return undefined;
    }
    return this._refObject.err_speed;
  }
  set err_speed(value) {
    this._err_speedIntialized = true;
    this._refObject.err_speed = value;
  }
  get err_climb() {
    if (!this._err_climbIntialized) {
      return undefined;
    }
    return this._refObject.err_climb;
  }
  set err_climb(value) {
    this._err_climbIntialized = true;
    this._refObject.err_climb = value;
  }
  get err_time() {
    if (!this._err_timeIntialized) {
      return undefined;
    }
    return this._refObject.err_time;
  }
  set err_time(value) {
    this._err_timeIntialized = true;
    this._refObject.err_time = value;
  }
  get err_pitch() {
    if (!this._err_pitchIntialized) {
      return undefined;
    }
    return this._refObject.err_pitch;
  }
  set err_pitch(value) {
    this._err_pitchIntialized = true;
    this._refObject.err_pitch = value;
  }
  get err_roll() {
    if (!this._err_rollIntialized) {
      return undefined;
    }
    return this._refObject.err_roll;
  }
  set err_roll(value) {
    this._err_rollIntialized = true;
    this._refObject.err_roll = value;
  }
  get err_dip() {
    if (!this._err_dipIntialized) {
      return undefined;
    }
    return this._refObject.err_dip;
  }
  set err_dip(value) {
    this._err_dipIntialized = true;
    this._refObject.err_dip = value;
  }
  get position_covariance() {
    return this._wrapperFields['position_covariance'].data;
  }
  set position_covariance(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['position_covariance'].fill(value);
  }
  get position_covariance_type() {
    if (!this._position_covariance_typeIntialized) {
      return undefined;
    }
    return this._refObject.position_covariance_type;
  }
  set position_covariance_type(value) {
    this._position_covariance_typeIntialized = true;
    this._refObject.position_covariance_type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPSFixRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._trackIntialized = true;
    this._speedIntialized = true;
    this._climbIntialized = true;
    this._pitchIntialized = true;
    this._rollIntialized = true;
    this._dipIntialized = true;
    this._timeIntialized = true;
    this._gdopIntialized = true;
    this._pdopIntialized = true;
    this._hdopIntialized = true;
    this._vdopIntialized = true;
    this._tdopIntialized = true;
    this._errIntialized = true;
    this._err_horzIntialized = true;
    this._err_vertIntialized = true;
    this._err_trackIntialized = true;
    this._err_speedIntialized = true;
    this._err_climbIntialized = true;
    this._err_timeIntialized = true;
    this._err_pitchIntialized = true;
    this._err_rollIntialized = true;
    this._err_dipIntialized = true;
    this._wrapperFields.position_covariance.fill(refObject.position_covariance.toArray());
    this._position_covariance_typeIntialized = true;
  }
  copy(other) {
    this._refObject = new GPSFixRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._trackIntialized = true;
    this._speedIntialized = true;
    this._climbIntialized = true;
    this._pitchIntialized = true;
    this._rollIntialized = true;
    this._dipIntialized = true;
    this._timeIntialized = true;
    this._gdopIntialized = true;
    this._pdopIntialized = true;
    this._hdopIntialized = true;
    this._vdopIntialized = true;
    this._tdopIntialized = true;
    this._errIntialized = true;
    this._err_horzIntialized = true;
    this._err_vertIntialized = true;
    this._err_trackIntialized = true;
    this._err_speedIntialized = true;
    this._err_climbIntialized = true;
    this._err_timeIntialized = true;
    this._err_pitchIntialized = true;
    this._err_rollIntialized = true;
    this._err_dipIntialized = true;
    this._wrapperFields.position_covariance.copy(other._wrapperFields.position_covariance);
    this._position_covariance_typeIntialized = true;
  }
  static get classType() {
    return GPSFixWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"COVARIANCE_TYPE_UNKNOWN","value":0},{"type":"uint8","name":"COVARIANCE_TYPE_APPROXIMATED","value":1},{"type":"uint8","name":"COVARIANCE_TYPE_DIAGONAL_KNOWN","value":2},{"type":"uint8","name":"COVARIANCE_TYPE_KNOWN","value":3}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"gps_msgs","type":"GPSStatus","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"track","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"climb","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tdop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_horz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_vert","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_track","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_climb","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_roll","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"err_dip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_covariance","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_covariance_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gps_msgs","type":"GPSFix","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPSFix"};
  }
  hasMember(name) {
    let memberNames = ["header","status","latitude","longitude","altitude","track","speed","climb","pitch","roll","dip","time","gdop","pdop","hdop","vdop","tdop","err","err_horz","err_vert","err_track","err_speed","err_climb","err_time","err_pitch","err_roll","err_dip","position_covariance","position_covariance_type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GPSFixArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPSFixWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPSFixWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSFixArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPSFixArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GPSFixRefArray(size);
    this._refObject = new GPSFixRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPSFixWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPSFixArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPSFixArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPSFixWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPSFixWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPSFixArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(GPSFixWrapper, "COVARIANCE_TYPE_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSFixWrapper, "COVARIANCE_TYPE_APPROXIMATED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSFixWrapper, "COVARIANCE_TYPE_DIAGONAL_KNOWN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GPSFixWrapper, "COVARIANCE_TYPE_KNOWN", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = GPSFixWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_APPROXIMATED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_DIAGONAL_KNOWN",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "COVARIANCE_TYPE_KNOWN",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "gps_msgs",
        "type": "GPSStatus",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "track",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "climb",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tdop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_horz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_vert",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_track",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_climb",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_roll",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "err_dip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_covariance",
      "type": {
        "isArray": true,
        "arraySize": 9,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_covariance_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gps_msgs",
    "type": "GPSFix",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GPSFix"
}
*/
