// rclnodejs/generated/geometry_msgs/geometry_msgs__msg__VelocityStamped.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const VelocityStampedRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  body_frame_id: primitiveTypes.string,
  reference_frame_id: primitiveTypes.string,
  velocity: TwistWrapper.refObjectType,
});
const VelocityStampedRefArray = ArrayType(VelocityStampedRefStruct);
const VelocityStampedRefStructArray = StructType({
  data: VelocityStampedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VelocityStampedWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VelocityStampedRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._body_frame_idIntialized = true;
      this._wrapperFields.body_frame_id =  new StringWrapper(other._wrapperFields.body_frame_id);
      this._reference_frame_idIntialized = true;
      this._wrapperFields.reference_frame_id =  new StringWrapper(other._wrapperFields.reference_frame_id);
      this._wrapperFields.velocity =  new TwistWrapper(other._wrapperFields.velocity);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VelocityStampedRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._body_frame_idIntialized = false;
    this._wrapperFields.body_frame_id = new StringWrapper();
    this._reference_frame_idIntialized = false;
    this._wrapperFields.reference_frame_id = new StringWrapper();
    this._wrapperFields.velocity = new TwistWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new VelocityStampedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VelocityStampedArrayWrapper;
  }
  static get ArrayType() {
    return VelocityStampedArrayWrapper;
  }
  static get refObjectArrayType() {
    return VelocityStampedRefStructArray
  }
  static get refObjectType() {
    return VelocityStampedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._body_frame_idIntialized) {
        throw new TypeError('Invalid argument: body_frame_id in VelocityStamped');
      }
      if (checkConsistency && !this._reference_frame_idIntialized) {
        throw new TypeError('Invalid argument: reference_frame_id in VelocityStamped');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.body_frame_id.freeze(own, checkConsistency);
    }
    this._refObject.body_frame_id = this._wrapperFields.body_frame_id.refObject;
    if (own) {
      this._wrapperFields.reference_frame_id.freeze(own, checkConsistency);
    }
    this._refObject.reference_frame_id = this._wrapperFields.reference_frame_id.refObject;
    this._wrapperFields.velocity.freeze(own, checkConsistency);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._body_frame_idIntialized = true;
    this._wrapperFields.body_frame_id.data = refObject.body_frame_id.data;
    this._reference_frame_idIntialized = true;
    this._wrapperFields.reference_frame_id.data = refObject.reference_frame_id.data;
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.body_frame_id);
    StringWrapper.freeStruct(refObject.reference_frame_id);
    TwistWrapper.freeStruct(refObject.velocity);
  }
  static destoryRawROS(msg) {
    VelocityStampedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geometry_msgs', subFolder: 'msg', interfaceName: 'VelocityStamped'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get body_frame_id() {
    if (!this._body_frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.body_frame_id.data;
  }
  set body_frame_id(value) {
    this._body_frame_idIntialized = true;
    this._wrapperFields.body_frame_id.data = value;
  }
  get reference_frame_id() {
    if (!this._reference_frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.reference_frame_id.data;
  }
  set reference_frame_id(value) {
    this._reference_frame_idIntialized = true;
    this._wrapperFields.reference_frame_id.data = value;
  }
  get velocity() {
    return this._wrapperFields.velocity;
  }
  set velocity(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.velocity.copy(value);
    } else {
      this._wrapperFields.velocity.copy(new TwistWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new VelocityStampedRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._body_frame_idIntialized = true;
    this._wrapperFields.body_frame_id.copyRefObject(this._refObject.body_frame_id);
    this._reference_frame_idIntialized = true;
    this._wrapperFields.reference_frame_id.copyRefObject(this._refObject.reference_frame_id);
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
  }
  copy(other) {
    this._refObject = new VelocityStampedRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._body_frame_idIntialized = true;
    this._wrapperFields.body_frame_id.copy(other._wrapperFields.body_frame_id);
    this._reference_frame_idIntialized = true;
    this._wrapperFields.reference_frame_id.copy(other._wrapperFields.reference_frame_id);
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
  }
  static get classType() {
    return VelocityStampedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference_frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geometry_msgs","type":"VelocityStamped","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VelocityStamped"};
  }
  hasMember(name) {
    let memberNames = ["header","body_frame_id","reference_frame_id","velocity"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VelocityStampedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VelocityStampedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VelocityStampedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VelocityStampedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VelocityStampedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VelocityStampedRefArray(size);
    this._refObject = new VelocityStampedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VelocityStampedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VelocityStampedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VelocityStampedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VelocityStampedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VelocityStampedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VelocityStampedArrayWrapper;
  }
}
module.exports = VelocityStampedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reference_frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geometry_msgs",
    "type": "VelocityStamped",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VelocityStamped"
}
*/
