// rclnodejs/generated/gazebo_msgs/gazebo_msgs__msg__WheelSlip.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const WheelSlipRefStruct = StructType({
  name: StringWrapper.refObjectArrayType,
  lateral_slip: Float64Wrapper.refObjectArrayType,
  longitudinal_slip: Float64Wrapper.refObjectArrayType,
});
const WheelSlipRefArray = ArrayType(WheelSlipRefStruct);
const WheelSlipRefStructArray = StructType({
  data: WheelSlipRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WheelSlipWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._nameArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WheelSlipRefStruct(other._refObject.toObject());
      this._wrapperFields.name = StringWrapper.createArray();
      this._wrapperFields.name.copy(other._wrapperFields.name);
      this.name = other.name;
      this._wrapperFields.lateral_slip = Float64Wrapper.createArray();
      this._wrapperFields.lateral_slip.copy(other._wrapperFields.lateral_slip);
      this._wrapperFields.longitudinal_slip = Float64Wrapper.createArray();
      this._wrapperFields.longitudinal_slip.copy(other._wrapperFields.longitudinal_slip);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WheelSlipRefStruct();
    this._wrapperFields.name = StringWrapper.createArray();
    this._wrapperFields.lateral_slip = Float64Wrapper.createArray();
    this._wrapperFields.longitudinal_slip = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new WheelSlipWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WheelSlipArrayWrapper;
  }
  static get ArrayType() {
    return WheelSlipArrayWrapper;
  }
  static get refObjectArrayType() {
    return WheelSlipRefStructArray
  }
  static get refObjectType() {
    return WheelSlipRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (!own) {
      this._wrapperFields.name.fill(this._nameArray);
      this._wrapperFields.name.freeze(own, checkConsistency);
      this._refObject.name = this._wrapperFields.name.refObject;
    } else {
      this._wrapperFields.name.fill([]);
      this._wrapperFields.name.freeze(own, checkConsistency);
      this._refObject.name = this._wrapperFields.name.refObject;
    }
    this._wrapperFields.lateral_slip.freeze(own, checkConsistency);
    this._refObject.lateral_slip = this._wrapperFields.lateral_slip.refObject;
    if (own) {
      this._wrapperFields.lateral_slip.fill(Float64Array.from([]));
      this._wrapperFields.lateral_slip.freeze(own, checkConsistency);
      this._refObject.lateral_slip = this._wrapperFields.lateral_slip.refObject;
    }
    this._wrapperFields.longitudinal_slip.freeze(own, checkConsistency);
    this._refObject.longitudinal_slip = this._wrapperFields.longitudinal_slip.refObject;
    if (own) {
      this._wrapperFields.longitudinal_slip.fill(Float64Array.from([]));
      this._wrapperFields.longitudinal_slip.freeze(own, checkConsistency);
      this._refObject.longitudinal_slip = this._wrapperFields.longitudinal_slip.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    this._wrapperFields.lateral_slip.copyRefObject(refObject.lateral_slip);
    this._wrapperFields.longitudinal_slip.copyRefObject(refObject.longitudinal_slip);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.name.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.name);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.name, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.lateral_slip.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.lateral_slip);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.lateral_slip, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.longitudinal_slip.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.longitudinal_slip);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.longitudinal_slip, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    WheelSlipWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'msg', interfaceName: 'WheelSlip'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._nameArray;
  }
  set name(value) {
    this._nameArray = value;
  }
  get lateral_slip() {
    return this._wrapperFields['lateral_slip'].data;
  }
  set lateral_slip(value) {
    this._wrapperFields['lateral_slip'].fill(value);
  }
  get longitudinal_slip() {
    return this._wrapperFields['longitudinal_slip'].data;
  }
  set longitudinal_slip(value) {
    this._wrapperFields['longitudinal_slip'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new WheelSlipRefStruct(refObject.toObject());
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    this._wrapperFields.lateral_slip.copyRefObject(this._refObject.lateral_slip);
    this._wrapperFields.longitudinal_slip.copyRefObject(this._refObject.longitudinal_slip);
  }
  copy(other) {
    this._refObject = new WheelSlipRefStruct(other._refObject.toObject());
    this._nameArray = other._nameArray.slice();
    this._wrapperFields.lateral_slip.copy(other._wrapperFields.lateral_slip);
    this._wrapperFields.longitudinal_slip.copy(other._wrapperFields.longitudinal_slip);
  }
  static get classType() {
    return WheelSlipWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lateral_slip","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitudinal_slip","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"WheelSlip","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WheelSlip"};
  }
  hasMember(name) {
    let memberNames = ["name","lateral_slip","longitudinal_slip"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WheelSlipArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WheelSlipWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WheelSlipWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WheelSlipArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WheelSlipArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WheelSlipRefArray(size);
    this._refObject = new WheelSlipRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WheelSlipWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WheelSlipArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WheelSlipArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WheelSlipWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WheelSlipWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WheelSlipArrayWrapper;
  }
}
module.exports = WheelSlipWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lateral_slip",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitudinal_slip",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "WheelSlip",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WheelSlip"
}
*/
