// rclnodejs/generated/theora_image_transport/theora_image_transport__msg__Packet.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const PacketRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  data: UInt8Wrapper.refObjectArrayType,
  b_o_s: primitiveTypes.int32,
  e_o_s: primitiveTypes.int32,
  granulepos: primitiveTypes.int64,
  packetno: primitiveTypes.int64,
});
const PacketRefArray = ArrayType(PacketRefStruct);
const PacketRefStructArray = StructType({
  data: PacketRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PacketWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PacketRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
      this._b_o_sIntialized = true;
      this._e_o_sIntialized = true;
      this._granuleposIntialized = true;
      this._packetnoIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PacketRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.data = UInt8Wrapper.createArray();
    this._b_o_sIntialized = false;
    this._e_o_sIntialized = false;
    this._granuleposIntialized = false;
    this._packetnoIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PacketWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PacketArrayWrapper;
  }
  static get ArrayType() {
    return PacketArrayWrapper;
  }
  static get refObjectArrayType() {
    return PacketRefStructArray
  }
  static get refObjectType() {
    return PacketRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._b_o_sIntialized) {
        throw new TypeError('Invalid argument: b_o_s in Packet');
      }
      if (checkConsistency && !this._e_o_sIntialized) {
        throw new TypeError('Invalid argument: e_o_s in Packet');
      }
      if (checkConsistency && !this._granuleposIntialized) {
        throw new TypeError('Invalid argument: granulepos in Packet');
      }
      if (checkConsistency && !this._packetnoIntialized) {
        throw new TypeError('Invalid argument: packetno in Packet');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Uint8Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.data.copyRefObject(refObject.data);
    this._b_o_sIntialized = true;
    this._refObject.b_o_s = refObject.b_o_s;
    this._e_o_sIntialized = true;
    this._refObject.e_o_s = refObject.e_o_s;
    this._granuleposIntialized = true;
    this._refObject.granulepos = refObject.granulepos;
    this._packetnoIntialized = true;
    this._refObject.packetno = refObject.packetno;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.data.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.data);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    PacketWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'theora_image_transport', subFolder: 'msg', interfaceName: 'Packet'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  get b_o_s() {
    if (!this._b_o_sIntialized) {
      return undefined;
    }
    return this._refObject.b_o_s;
  }
  set b_o_s(value) {
    this._b_o_sIntialized = true;
    this._refObject.b_o_s = value;
  }
  get e_o_s() {
    if (!this._e_o_sIntialized) {
      return undefined;
    }
    return this._refObject.e_o_s;
  }
  set e_o_s(value) {
    this._e_o_sIntialized = true;
    this._refObject.e_o_s = value;
  }
  get granulepos() {
    if (!this._granuleposIntialized) {
      return undefined;
    }
    return this._refObject.granulepos;
  }
  set granulepos(value) {
    this._granuleposIntialized = true;
    this._refObject.granulepos = value;
  }
  get packetno() {
    if (!this._packetnoIntialized) {
      return undefined;
    }
    return this._refObject.packetno;
  }
  set packetno(value) {
    this._packetnoIntialized = true;
    this._refObject.packetno = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PacketRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.data.copyRefObject(this._refObject.data);
    this._b_o_sIntialized = true;
    this._e_o_sIntialized = true;
    this._granuleposIntialized = true;
    this._packetnoIntialized = true;
  }
  copy(other) {
    this._refObject = new PacketRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.data.copy(other._wrapperFields.data);
    this._b_o_sIntialized = true;
    this._e_o_sIntialized = true;
    this._granuleposIntialized = true;
    this._packetnoIntialized = true;
  }
  static get classType() {
    return PacketWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"b_o_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"e_o_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"granulepos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"packetno","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"theora_image_transport","type":"Packet","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Packet"};
  }
  hasMember(name) {
    let memberNames = ["header","data","b_o_s","e_o_s","granulepos","packetno"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PacketArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PacketWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PacketWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PacketArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PacketArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PacketRefArray(size);
    this._refObject = new PacketRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PacketWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PacketArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PacketArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PacketWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PacketWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PacketArrayWrapper;
  }
}
module.exports = PacketWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "b_o_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "e_o_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "granulepos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "packetno",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "theora_image_transport",
    "type": "Packet",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Packet"
}
*/
