// rclnodejs/generated/turtlebot4_msgs/turtlebot4_msgs__msg__UserButton.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const UserButtonRefStruct = StructType({
  button: ArrayType(primitiveTypes.bool, 4),
});
const UserButtonRefArray = ArrayType(UserButtonRefStruct);
const UserButtonRefStructArray = StructType({
  data: UserButtonRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UserButtonWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._buttonArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UserButtonRefStruct(other._refObject.toObject());
      this._wrapperFields.button = BoolWrapper.createArray();
      this._wrapperFields.button.copy(other._wrapperFields.button);
      this.button = other.button;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UserButtonRefStruct();
    this._wrapperFields.button = BoolWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new UserButtonWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UserButtonArrayWrapper;
  }
  static get ArrayType() {
    return UserButtonArrayWrapper;
  }
  static get refObjectArrayType() {
    return UserButtonRefStructArray
  }
  static get refObjectType() {
    return UserButtonRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._refObject.button = this._buttonArray;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._buttonArray = refObject.button.toArray();
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    UserButtonWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlebot4_msgs', subFolder: 'msg', interfaceName: 'UserButton'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get button() {
    return this._buttonArray;
  }
  set button(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._buttonArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UserButtonRefStruct(refObject.toObject());
    refObject.button.data.length = refObject.button.size;
    for (let index = 0; index < refObject.button.size; index++) {
      this._buttonArray[index] = refObject.button.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new UserButtonRefStruct(other._refObject.toObject());
    this._buttonArray = other._buttonArray.slice();
  }
  static get classType() {
    return UserButtonWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"button","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlebot4_msgs","type":"UserButton","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UserButton"};
  }
  hasMember(name) {
    let memberNames = ["button"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UserButtonArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UserButtonWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UserButtonWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UserButtonArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UserButtonArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UserButtonRefArray(size);
    this._refObject = new UserButtonRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UserButtonWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UserButtonArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UserButtonArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UserButtonWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UserButtonWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UserButtonArrayWrapper;
  }
}
module.exports = UserButtonWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "button",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlebot4_msgs",
    "type": "UserButton",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UserButton"
}
*/
