// rclnodejs/generated/turtlebot4_msgs/turtlebot4_msgs__msg__UserLed.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UserLedRefStruct = StructType({
  led: primitiveTypes.uint8,
  color: primitiveTypes.uint8,
  blink_period: primitiveTypes.uint32,
  duty_cycle: primitiveTypes.float64,
});
const UserLedRefArray = ArrayType(UserLedRefStruct);
const UserLedRefStructArray = StructType({
  data: UserLedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UserLedWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UserLedRefStruct(other._refObject.toObject());
      this._ledIntialized = true;
      this._colorIntialized = true;
      this._blink_periodIntialized = true;
      this._duty_cycleIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UserLedRefStruct();
    this._ledIntialized = false;
    this._colorIntialized = false;
    this._blink_periodIntialized = false;
    this._duty_cycleIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new UserLedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UserLedArrayWrapper;
  }
  static get ArrayType() {
    return UserLedArrayWrapper;
  }
  static get refObjectArrayType() {
    return UserLedRefStructArray
  }
  static get refObjectType() {
    return UserLedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._ledIntialized) {
        throw new TypeError('Invalid argument: led in UserLed');
      }
      if (checkConsistency && !this._colorIntialized) {
        throw new TypeError('Invalid argument: color in UserLed');
      }
      if (checkConsistency && !this._blink_periodIntialized) {
        throw new TypeError('Invalid argument: blink_period in UserLed');
      }
      if (checkConsistency && !this._duty_cycleIntialized) {
        throw new TypeError('Invalid argument: duty_cycle in UserLed');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._ledIntialized = true;
    this._refObject.led = refObject.led;
    this._colorIntialized = true;
    this._refObject.color = refObject.color;
    this._blink_periodIntialized = true;
    this._refObject.blink_period = refObject.blink_period;
    this._duty_cycleIntialized = true;
    this._refObject.duty_cycle = refObject.duty_cycle;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    UserLedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlebot4_msgs', subFolder: 'msg', interfaceName: 'UserLed'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get led() {
    if (!this._ledIntialized) {
      return undefined;
    }
    return this._refObject.led;
  }
  set led(value) {
    this._ledIntialized = true;
    this._refObject.led = value;
  }
  get color() {
    if (!this._colorIntialized) {
      return undefined;
    }
    return this._refObject.color;
  }
  set color(value) {
    this._colorIntialized = true;
    this._refObject.color = value;
  }
  get blink_period() {
    if (!this._blink_periodIntialized) {
      return undefined;
    }
    return this._refObject.blink_period;
  }
  set blink_period(value) {
    this._blink_periodIntialized = true;
    this._refObject.blink_period = value;
  }
  get duty_cycle() {
    if (!this._duty_cycleIntialized) {
      return undefined;
    }
    return this._refObject.duty_cycle;
  }
  set duty_cycle(value) {
    this._duty_cycleIntialized = true;
    this._refObject.duty_cycle = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UserLedRefStruct(refObject.toObject());
    this._ledIntialized = true;
    this._colorIntialized = true;
    this._blink_periodIntialized = true;
    this._duty_cycleIntialized = true;
  }
  copy(other) {
    this._refObject = new UserLedRefStruct(other._refObject.toObject());
    this._ledIntialized = true;
    this._colorIntialized = true;
    this._blink_periodIntialized = true;
    this._duty_cycleIntialized = true;
  }
  static get classType() {
    return UserLedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"USER_LED_1","value":0},{"type":"uint8","name":"USER_LED_2","value":1},{"type":"uint8","name":"COLOR_OFF","value":0},{"type":"uint8","name":"COLOR_GREEN","value":1},{"type":"uint8","name":"COLOR_RED","value":2},{"type":"uint8","name":"COLOR_YELLOW","value":3}],"fields":[{"name":"led","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"color","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"blink_period","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"duty_cycle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlebot4_msgs","type":"UserLed","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UserLed"};
  }
  hasMember(name) {
    let memberNames = ["led","color","blink_period","duty_cycle"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UserLedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UserLedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UserLedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UserLedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UserLedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UserLedRefArray(size);
    this._refObject = new UserLedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UserLedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UserLedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UserLedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UserLedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UserLedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UserLedArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(UserLedWrapper, "USER_LED_1", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(UserLedWrapper, "USER_LED_2", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(UserLedWrapper, "COLOR_OFF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(UserLedWrapper, "COLOR_GREEN", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(UserLedWrapper, "COLOR_RED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(UserLedWrapper, "COLOR_YELLOW", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = UserLedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "USER_LED_1",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "USER_LED_2",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "COLOR_OFF",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "COLOR_GREEN",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "COLOR_RED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "COLOR_YELLOW",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "led",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "color",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "blink_period",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "duty_cycle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlebot4_msgs",
    "type": "UserLed",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UserLed"
}
*/
