// rclnodejs/generated/ros_ign_interfaces/ros_ign_interfaces__msg__WorldControl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WorldResetWrapper = require('../../generated/ros_gz_interfaces/ros_gz_interfaces__msg__WorldReset.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const WorldControlRefStruct = StructType({
  pause: primitiveTypes.bool,
  step: primitiveTypes.bool,
  multi_step: primitiveTypes.uint32,
  reset: WorldResetWrapper.refObjectType,
  seed: primitiveTypes.uint32,
  run_to_sim_time: TimeWrapper.refObjectType,
});
const WorldControlRefArray = ArrayType(WorldControlRefStruct);
const WorldControlRefStructArray = StructType({
  data: WorldControlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WorldControlWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WorldControlRefStruct(other._refObject.toObject());
      this._pauseIntialized = true;
      this._stepIntialized = true;
      this._multi_stepIntialized = true;
      this._wrapperFields.reset =  new WorldResetWrapper(other._wrapperFields.reset);
      this._seedIntialized = true;
      this._wrapperFields.run_to_sim_time =  new TimeWrapper(other._wrapperFields.run_to_sim_time);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WorldControlRefStruct();
    this._pauseIntialized = false;
    this._stepIntialized = false;
    this._refObject.multi_step = 0;
    this._multi_stepIntialized = true;
    this._wrapperFields.reset = new WorldResetWrapper();
    this._seedIntialized = false;
    this._wrapperFields.run_to_sim_time = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new WorldControlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WorldControlArrayWrapper;
  }
  static get ArrayType() {
    return WorldControlArrayWrapper;
  }
  static get refObjectArrayType() {
    return WorldControlRefStructArray
  }
  static get refObjectType() {
    return WorldControlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._pauseIntialized) {
        throw new TypeError('Invalid argument: pause in WorldControl');
      }
      if (checkConsistency && !this._stepIntialized) {
        throw new TypeError('Invalid argument: step in WorldControl');
      }
      if (checkConsistency && !this._multi_stepIntialized) {
        throw new TypeError('Invalid argument: multi_step in WorldControl');
      }
      if (checkConsistency && !this._seedIntialized) {
        throw new TypeError('Invalid argument: seed in WorldControl');
      }
    this._wrapperFields.reset.freeze(own, checkConsistency);
    this._refObject.reset = this._wrapperFields.reset.refObject;
    this._wrapperFields.run_to_sim_time.freeze(own, checkConsistency);
    this._refObject.run_to_sim_time = this._wrapperFields.run_to_sim_time.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._pauseIntialized = true;
    this._refObject.pause = refObject.pause;
    this._stepIntialized = true;
    this._refObject.step = refObject.step;
    this._multi_stepIntialized = true;
    this._refObject.multi_step = refObject.multi_step;
    this._wrapperFields.reset.copyRefObject(refObject.reset);
    this._seedIntialized = true;
    this._refObject.seed = refObject.seed;
    this._wrapperFields.run_to_sim_time.copyRefObject(refObject.run_to_sim_time);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    WorldResetWrapper.freeStruct(refObject.reset);
    TimeWrapper.freeStruct(refObject.run_to_sim_time);
  }
  static destoryRawROS(msg) {
    WorldControlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_ign_interfaces', subFolder: 'msg', interfaceName: 'WorldControl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pause() {
    if (!this._pauseIntialized) {
      return undefined;
    }
    return this._refObject.pause;
  }
  set pause(value) {
    this._pauseIntialized = true;
    this._refObject.pause = value;
  }
  get step() {
    if (!this._stepIntialized) {
      return undefined;
    }
    return this._refObject.step;
  }
  set step(value) {
    this._stepIntialized = true;
    this._refObject.step = value;
  }
  get multi_step() {
    if (!this._multi_stepIntialized) {
      return undefined;
    }
    return this._refObject.multi_step;
  }
  set multi_step(value) {
    this._multi_stepIntialized = true;
    this._refObject.multi_step = value;
  }
  get reset() {
    return this._wrapperFields.reset;
  }
  set reset(value) {
    if (value instanceof WorldResetWrapper) {
      this._wrapperFields.reset.copy(value);
    } else {
      this._wrapperFields.reset.copy(new WorldResetWrapper(value));
    }
  }
  get seed() {
    if (!this._seedIntialized) {
      return undefined;
    }
    return this._refObject.seed;
  }
  set seed(value) {
    this._seedIntialized = true;
    this._refObject.seed = value;
  }
  get run_to_sim_time() {
    return this._wrapperFields.run_to_sim_time;
  }
  set run_to_sim_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.run_to_sim_time.copy(value);
    } else {
      this._wrapperFields.run_to_sim_time.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new WorldControlRefStruct(refObject.toObject());
    this._pauseIntialized = true;
    this._stepIntialized = true;
    this._multi_stepIntialized = true;
    this._wrapperFields.reset.copyRefObject(this._refObject.reset);
    this._seedIntialized = true;
    this._wrapperFields.run_to_sim_time.copyRefObject(this._refObject.run_to_sim_time);
  }
  copy(other) {
    this._refObject = new WorldControlRefStruct(other._refObject.toObject());
    this._pauseIntialized = true;
    this._stepIntialized = true;
    this._multi_stepIntialized = true;
    this._wrapperFields.reset.copy(other._wrapperFields.reset);
    this._seedIntialized = true;
    this._wrapperFields.run_to_sim_time.copy(other._wrapperFields.run_to_sim_time);
  }
  static get classType() {
    return WorldControlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pause","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"multi_step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":0},{"name":"reset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"ros_gz_interfaces","type":"WorldReset","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"seed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"run_to_sim_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ros_ign_interfaces","type":"WorldControl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WorldControl"};
  }
  hasMember(name) {
    let memberNames = ["pause","step","multi_step","reset","seed","run_to_sim_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WorldControlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WorldControlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WorldControlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WorldControlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WorldControlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WorldControlRefArray(size);
    this._refObject = new WorldControlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WorldControlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WorldControlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WorldControlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WorldControlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WorldControlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WorldControlArrayWrapper;
  }
}
module.exports = WorldControlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pause",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "step",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "multi_step",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 0
    },
    {
      "name": "reset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "ros_gz_interfaces",
        "type": "WorldReset",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "seed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "run_to_sim_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_ign_interfaces",
    "type": "WorldControl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WorldControl"
}
*/
