// rclnodejs/generated/ros_ign_interfaces/ros_ign_interfaces__srv__SpawnEntity_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EntityFactoryWrapper = require('../../generated/ros_gz_interfaces/ros_gz_interfaces__msg__EntityFactory.js');
const SpawnEntity_RequestRefStruct = StructType({
  entity_factory: EntityFactoryWrapper.refObjectType,
});
const SpawnEntity_RequestRefArray = ArrayType(SpawnEntity_RequestRefStruct);
const SpawnEntity_RequestRefStructArray = StructType({
  data: SpawnEntity_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SpawnEntity_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SpawnEntity_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.entity_factory =  new EntityFactoryWrapper(other._wrapperFields.entity_factory);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SpawnEntity_RequestRefStruct();
    this._wrapperFields.entity_factory = new EntityFactoryWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SpawnEntity_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SpawnEntity_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SpawnEntity_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SpawnEntity_RequestRefStructArray
  }
  static get refObjectType() {
    return SpawnEntity_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.entity_factory.freeze(own, checkConsistency);
    this._refObject.entity_factory = this._wrapperFields.entity_factory.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.entity_factory.copyRefObject(refObject.entity_factory);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    EntityFactoryWrapper.freeStruct(refObject.entity_factory);
  }
  static destoryRawROS(msg) {
    SpawnEntity_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_ign_interfaces', subFolder: 'srv', interfaceName: 'SpawnEntity_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get entity_factory() {
    return this._wrapperFields.entity_factory;
  }
  set entity_factory(value) {
    if (value instanceof EntityFactoryWrapper) {
      this._wrapperFields.entity_factory.copy(value);
    } else {
      this._wrapperFields.entity_factory.copy(new EntityFactoryWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SpawnEntity_RequestRefStruct(refObject.toObject());
    this._wrapperFields.entity_factory.copyRefObject(this._refObject.entity_factory);
  }
  copy(other) {
    this._refObject = new SpawnEntity_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.entity_factory.copy(other._wrapperFields.entity_factory);
  }
  static get classType() {
    return SpawnEntity_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"entity_factory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"ros_gz_interfaces","type":"EntityFactory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ros_ign_interfaces","type":"SpawnEntity_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SpawnEntity_Request"};
  }
  hasMember(name) {
    let memberNames = ["entity_factory"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SpawnEntity_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SpawnEntity_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SpawnEntity_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SpawnEntity_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SpawnEntity_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SpawnEntity_RequestRefArray(size);
    this._refObject = new SpawnEntity_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SpawnEntity_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SpawnEntity_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SpawnEntity_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SpawnEntity_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SpawnEntity_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SpawnEntity_RequestArrayWrapper;
  }
}
module.exports = SpawnEntity_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "entity_factory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "ros_gz_interfaces",
        "type": "EntityFactory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_ign_interfaces",
    "type": "SpawnEntity_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SpawnEntity_Request"
}
*/
