// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__msg__RGBD.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const CameraInfoWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__CameraInfo.js');
const ImageWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__Image.js');
const RGBDRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  rgb_camera_info: CameraInfoWrapper.refObjectType,
  depth_camera_info: CameraInfoWrapper.refObjectType,
  rgb: ImageWrapper.refObjectType,
  depth: ImageWrapper.refObjectType,
});
const RGBDRefArray = ArrayType(RGBDRefStruct);
const RGBDRefStructArray = StructType({
  data: RGBDRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RGBDWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RGBDRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.rgb_camera_info =  new CameraInfoWrapper(other._wrapperFields.rgb_camera_info);
      this._wrapperFields.depth_camera_info =  new CameraInfoWrapper(other._wrapperFields.depth_camera_info);
      this._wrapperFields.rgb =  new ImageWrapper(other._wrapperFields.rgb);
      this._wrapperFields.depth =  new ImageWrapper(other._wrapperFields.depth);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RGBDRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.rgb_camera_info = new CameraInfoWrapper();
    this._wrapperFields.depth_camera_info = new CameraInfoWrapper();
    this._wrapperFields.rgb = new ImageWrapper();
    this._wrapperFields.depth = new ImageWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RGBDWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RGBDArrayWrapper;
  }
  static get ArrayType() {
    return RGBDArrayWrapper;
  }
  static get refObjectArrayType() {
    return RGBDRefStructArray
  }
  static get refObjectType() {
    return RGBDRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.rgb_camera_info.freeze(own, checkConsistency);
    this._refObject.rgb_camera_info = this._wrapperFields.rgb_camera_info.refObject;
    this._wrapperFields.depth_camera_info.freeze(own, checkConsistency);
    this._refObject.depth_camera_info = this._wrapperFields.depth_camera_info.refObject;
    this._wrapperFields.rgb.freeze(own, checkConsistency);
    this._refObject.rgb = this._wrapperFields.rgb.refObject;
    this._wrapperFields.depth.freeze(own, checkConsistency);
    this._refObject.depth = this._wrapperFields.depth.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.rgb_camera_info.copyRefObject(refObject.rgb_camera_info);
    this._wrapperFields.depth_camera_info.copyRefObject(refObject.depth_camera_info);
    this._wrapperFields.rgb.copyRefObject(refObject.rgb);
    this._wrapperFields.depth.copyRefObject(refObject.depth);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    CameraInfoWrapper.freeStruct(refObject.rgb_camera_info);
    CameraInfoWrapper.freeStruct(refObject.depth_camera_info);
    ImageWrapper.freeStruct(refObject.rgb);
    ImageWrapper.freeStruct(refObject.depth);
  }
  static destoryRawROS(msg) {
    RGBDWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'msg', interfaceName: 'RGBD'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get rgb_camera_info() {
    return this._wrapperFields.rgb_camera_info;
  }
  set rgb_camera_info(value) {
    if (value instanceof CameraInfoWrapper) {
      this._wrapperFields.rgb_camera_info.copy(value);
    } else {
      this._wrapperFields.rgb_camera_info.copy(new CameraInfoWrapper(value));
    }
  }
  get depth_camera_info() {
    return this._wrapperFields.depth_camera_info;
  }
  set depth_camera_info(value) {
    if (value instanceof CameraInfoWrapper) {
      this._wrapperFields.depth_camera_info.copy(value);
    } else {
      this._wrapperFields.depth_camera_info.copy(new CameraInfoWrapper(value));
    }
  }
  get rgb() {
    return this._wrapperFields.rgb;
  }
  set rgb(value) {
    if (value instanceof ImageWrapper) {
      this._wrapperFields.rgb.copy(value);
    } else {
      this._wrapperFields.rgb.copy(new ImageWrapper(value));
    }
  }
  get depth() {
    return this._wrapperFields.depth;
  }
  set depth(value) {
    if (value instanceof ImageWrapper) {
      this._wrapperFields.depth.copy(value);
    } else {
      this._wrapperFields.depth.copy(new ImageWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new RGBDRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.rgb_camera_info.copyRefObject(this._refObject.rgb_camera_info);
    this._wrapperFields.depth_camera_info.copyRefObject(this._refObject.depth_camera_info);
    this._wrapperFields.rgb.copyRefObject(this._refObject.rgb);
    this._wrapperFields.depth.copyRefObject(this._refObject.depth);
  }
  copy(other) {
    this._refObject = new RGBDRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.rgb_camera_info.copy(other._wrapperFields.rgb_camera_info);
    this._wrapperFields.depth_camera_info.copy(other._wrapperFields.depth_camera_info);
    this._wrapperFields.rgb.copy(other._wrapperFields.rgb);
    this._wrapperFields.depth.copy(other._wrapperFields.depth);
  }
  static get classType() {
    return RGBDWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"rgb_camera_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"CameraInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"depth_camera_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"CameraInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"rgb","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"Image","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"depth","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"Image","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"RGBD","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RGBD"};
  }
  hasMember(name) {
    let memberNames = ["header","rgb_camera_info","depth_camera_info","rgb","depth"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RGBDArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RGBDWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RGBDWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RGBDArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RGBDArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RGBDRefArray(size);
    this._refObject = new RGBDRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RGBDWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RGBDArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RGBDArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RGBDWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RGBDWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RGBDArrayWrapper;
  }
}
module.exports = RGBDWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "rgb_camera_info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "CameraInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "depth_camera_info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "CameraInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "rgb",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "Image",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "depth",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "Image",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "realsense2_camera_msgs",
    "type": "RGBD",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RGBD"
}
*/
