// rclnodejs/generated/control_msgs/control_msgs__action__ParallelGripperCommand_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JointStateWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__JointState.js');
const ParallelGripperCommand_FeedbackRefStruct = StructType({
  state: JointStateWrapper.refObjectType,
});
const ParallelGripperCommand_FeedbackRefArray = ArrayType(ParallelGripperCommand_FeedbackRefStruct);
const ParallelGripperCommand_FeedbackRefStructArray = StructType({
  data: ParallelGripperCommand_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ParallelGripperCommand_FeedbackWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParallelGripperCommand_FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.state =  new JointStateWrapper(other._wrapperFields.state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ParallelGripperCommand_FeedbackRefStruct();
    this._wrapperFields.state = new JointStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ParallelGripperCommand_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParallelGripperCommand_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return ParallelGripperCommand_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParallelGripperCommand_FeedbackRefStructArray
  }
  static get refObjectType() {
    return ParallelGripperCommand_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.state.copyRefObject(refObject.state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    JointStateWrapper.freeStruct(refObject.state);
  }
  static destoryRawROS(msg) {
    ParallelGripperCommand_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'ParallelGripperCommand_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof JointStateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new JointStateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ParallelGripperCommand_FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.state.copyRefObject(this._refObject.state);
  }
  copy(other) {
    this._refObject = new ParallelGripperCommand_FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
  }
  static get classType() {
    return ParallelGripperCommand_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"JointState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"ParallelGripperCommand_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParallelGripperCommand_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ParallelGripperCommand_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParallelGripperCommand_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParallelGripperCommand_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParallelGripperCommand_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParallelGripperCommand_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ParallelGripperCommand_FeedbackRefArray(size);
    this._refObject = new ParallelGripperCommand_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParallelGripperCommand_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParallelGripperCommand_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParallelGripperCommand_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParallelGripperCommand_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParallelGripperCommand_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParallelGripperCommand_FeedbackArrayWrapper;
  }
}
module.exports = ParallelGripperCommand_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "JointState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "ParallelGripperCommand_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ParallelGripperCommand_Feedback"
}
*/
