// rclnodejs/generated/control_msgs/control_msgs__action__FollowJointTrajectory_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectory.js');
const MultiDOFJointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__MultiDOFJointTrajectory.js');
const JointToleranceWrapper = require('../../generated/control_msgs/control_msgs__msg__JointTolerance.js');
const JointComponentToleranceWrapper = require('../../generated/control_msgs/control_msgs__msg__JointComponentTolerance.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const FollowJointTrajectory_GoalRefStruct = StructType({
  trajectory: JointTrajectoryWrapper.refObjectType,
  multi_dof_trajectory: MultiDOFJointTrajectoryWrapper.refObjectType,
  path_tolerance: JointToleranceWrapper.refObjectArrayType,
  component_path_tolerance: JointComponentToleranceWrapper.refObjectArrayType,
  goal_tolerance: JointToleranceWrapper.refObjectArrayType,
  component_goal_tolerance: JointComponentToleranceWrapper.refObjectArrayType,
  goal_time_tolerance: DurationWrapper.refObjectType,
});
const FollowJointTrajectory_GoalRefArray = ArrayType(FollowJointTrajectory_GoalRefStruct);
const FollowJointTrajectory_GoalRefStructArray = StructType({
  data: FollowJointTrajectory_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowJointTrajectory_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowJointTrajectory_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.trajectory =  new JointTrajectoryWrapper(other._wrapperFields.trajectory);
      this._wrapperFields.multi_dof_trajectory =  new MultiDOFJointTrajectoryWrapper(other._wrapperFields.multi_dof_trajectory);
      this._wrapperFields.path_tolerance = JointToleranceWrapper.createArray();
      this._wrapperFields.path_tolerance.copy(other._wrapperFields.path_tolerance);
      this._wrapperFields.component_path_tolerance = JointComponentToleranceWrapper.createArray();
      this._wrapperFields.component_path_tolerance.copy(other._wrapperFields.component_path_tolerance);
      this._wrapperFields.goal_tolerance = JointToleranceWrapper.createArray();
      this._wrapperFields.goal_tolerance.copy(other._wrapperFields.goal_tolerance);
      this._wrapperFields.component_goal_tolerance = JointComponentToleranceWrapper.createArray();
      this._wrapperFields.component_goal_tolerance.copy(other._wrapperFields.component_goal_tolerance);
      this._wrapperFields.goal_time_tolerance =  new DurationWrapper(other._wrapperFields.goal_time_tolerance);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowJointTrajectory_GoalRefStruct();
    this._wrapperFields.trajectory = new JointTrajectoryWrapper();
    this._wrapperFields.multi_dof_trajectory = new MultiDOFJointTrajectoryWrapper();
    this._wrapperFields.path_tolerance = JointToleranceWrapper.createArray();
    this._wrapperFields.component_path_tolerance = JointComponentToleranceWrapper.createArray();
    this._wrapperFields.goal_tolerance = JointToleranceWrapper.createArray();
    this._wrapperFields.component_goal_tolerance = JointComponentToleranceWrapper.createArray();
    this._wrapperFields.goal_time_tolerance = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FollowJointTrajectory_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowJointTrajectory_GoalArrayWrapper;
  }
  static get ArrayType() {
    return FollowJointTrajectory_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowJointTrajectory_GoalRefStructArray
  }
  static get refObjectType() {
    return FollowJointTrajectory_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
    this._wrapperFields.multi_dof_trajectory.freeze(own, checkConsistency);
    this._refObject.multi_dof_trajectory = this._wrapperFields.multi_dof_trajectory.refObject;
    this._wrapperFields.path_tolerance.freeze(own, checkConsistency);
    this._refObject.path_tolerance = this._wrapperFields.path_tolerance.refObject;
    this._wrapperFields.component_path_tolerance.freeze(own, checkConsistency);
    this._refObject.component_path_tolerance = this._wrapperFields.component_path_tolerance.refObject;
    this._wrapperFields.goal_tolerance.freeze(own, checkConsistency);
    this._refObject.goal_tolerance = this._wrapperFields.goal_tolerance.refObject;
    this._wrapperFields.component_goal_tolerance.freeze(own, checkConsistency);
    this._refObject.component_goal_tolerance = this._wrapperFields.component_goal_tolerance.refObject;
    this._wrapperFields.goal_time_tolerance.freeze(own, checkConsistency);
    this._refObject.goal_time_tolerance = this._wrapperFields.goal_time_tolerance.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
    this._wrapperFields.multi_dof_trajectory.copyRefObject(refObject.multi_dof_trajectory);
    this._wrapperFields.path_tolerance.copyRefObject(refObject.path_tolerance);
    this._wrapperFields.component_path_tolerance.copyRefObject(refObject.component_path_tolerance);
    this._wrapperFields.goal_tolerance.copyRefObject(refObject.goal_tolerance);
    this._wrapperFields.component_goal_tolerance.copyRefObject(refObject.component_goal_tolerance);
    this._wrapperFields.goal_time_tolerance.copyRefObject(refObject.goal_time_tolerance);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    JointTrajectoryWrapper.freeStruct(refObject.trajectory);
    MultiDOFJointTrajectoryWrapper.freeStruct(refObject.multi_dof_trajectory);
    if (refObject.path_tolerance.size != 0) {
      JointToleranceWrapper.ArrayType.freeArray(refObject.path_tolerance);
      if (JointToleranceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.path_tolerance, JointToleranceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.component_path_tolerance.size != 0) {
      JointComponentToleranceWrapper.ArrayType.freeArray(refObject.component_path_tolerance);
      if (JointComponentToleranceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.component_path_tolerance, JointComponentToleranceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.goal_tolerance.size != 0) {
      JointToleranceWrapper.ArrayType.freeArray(refObject.goal_tolerance);
      if (JointToleranceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.goal_tolerance, JointToleranceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.component_goal_tolerance.size != 0) {
      JointComponentToleranceWrapper.ArrayType.freeArray(refObject.component_goal_tolerance);
      if (JointComponentToleranceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.component_goal_tolerance, JointComponentToleranceWrapper.refObjectArrayType, 'data');
      }
    }
    DurationWrapper.freeStruct(refObject.goal_time_tolerance);
  }
  static destoryRawROS(msg) {
    FollowJointTrajectory_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'FollowJointTrajectory_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    if (value instanceof JointTrajectoryWrapper) {
      this._wrapperFields.trajectory.copy(value);
    } else {
      this._wrapperFields.trajectory.copy(new JointTrajectoryWrapper(value));
    }
  }
  get multi_dof_trajectory() {
    return this._wrapperFields.multi_dof_trajectory;
  }
  set multi_dof_trajectory(value) {
    if (value instanceof MultiDOFJointTrajectoryWrapper) {
      this._wrapperFields.multi_dof_trajectory.copy(value);
    } else {
      this._wrapperFields.multi_dof_trajectory.copy(new MultiDOFJointTrajectoryWrapper(value));
    }
  }
  get path_tolerance() {
    return this._wrapperFields.path_tolerance;
  }
  set path_tolerance(value) {
    this._wrapperFields.path_tolerance.fill(value);
  }
  get component_path_tolerance() {
    return this._wrapperFields.component_path_tolerance;
  }
  set component_path_tolerance(value) {
    this._wrapperFields.component_path_tolerance.fill(value);
  }
  get goal_tolerance() {
    return this._wrapperFields.goal_tolerance;
  }
  set goal_tolerance(value) {
    this._wrapperFields.goal_tolerance.fill(value);
  }
  get component_goal_tolerance() {
    return this._wrapperFields.component_goal_tolerance;
  }
  set component_goal_tolerance(value) {
    this._wrapperFields.component_goal_tolerance.fill(value);
  }
  get goal_time_tolerance() {
    return this._wrapperFields.goal_time_tolerance;
  }
  set goal_time_tolerance(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.goal_time_tolerance.copy(value);
    } else {
      this._wrapperFields.goal_time_tolerance.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new FollowJointTrajectory_GoalRefStruct(refObject.toObject());
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
    this._wrapperFields.multi_dof_trajectory.copyRefObject(this._refObject.multi_dof_trajectory);
    this._wrapperFields.path_tolerance.copyRefObject(this._refObject.path_tolerance);
    this._wrapperFields.component_path_tolerance.copyRefObject(this._refObject.component_path_tolerance);
    this._wrapperFields.goal_tolerance.copyRefObject(this._refObject.goal_tolerance);
    this._wrapperFields.component_goal_tolerance.copyRefObject(this._refObject.component_goal_tolerance);
    this._wrapperFields.goal_time_tolerance.copyRefObject(this._refObject.goal_time_tolerance);
  }
  copy(other) {
    this._refObject = new FollowJointTrajectory_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
    this._wrapperFields.multi_dof_trajectory.copy(other._wrapperFields.multi_dof_trajectory);
    this._wrapperFields.path_tolerance.copy(other._wrapperFields.path_tolerance);
    this._wrapperFields.component_path_tolerance.copy(other._wrapperFields.component_path_tolerance);
    this._wrapperFields.goal_tolerance.copy(other._wrapperFields.goal_tolerance);
    this._wrapperFields.component_goal_tolerance.copy(other._wrapperFields.component_goal_tolerance);
    this._wrapperFields.goal_time_tolerance.copy(other._wrapperFields.goal_time_tolerance);
  }
  static get classType() {
    return FollowJointTrajectory_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"multi_dof_trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"MultiDOFJointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"path_tolerance","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"JointTolerance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"component_path_tolerance","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"JointComponentTolerance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal_tolerance","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"JointTolerance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"component_goal_tolerance","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"JointComponentTolerance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal_time_tolerance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"FollowJointTrajectory_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowJointTrajectory_Goal"};
  }
  hasMember(name) {
    let memberNames = ["trajectory","multi_dof_trajectory","path_tolerance","component_path_tolerance","goal_tolerance","component_goal_tolerance","goal_time_tolerance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowJointTrajectory_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowJointTrajectory_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowJointTrajectory_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowJointTrajectory_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowJointTrajectory_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowJointTrajectory_GoalRefArray(size);
    this._refObject = new FollowJointTrajectory_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowJointTrajectory_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowJointTrajectory_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowJointTrajectory_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowJointTrajectory_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowJointTrajectory_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowJointTrajectory_GoalArrayWrapper;
  }
}
module.exports = FollowJointTrajectory_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "multi_dof_trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "MultiDOFJointTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "path_tolerance",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "control_msgs",
        "type": "JointTolerance",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "component_path_tolerance",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "control_msgs",
        "type": "JointComponentTolerance",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal_tolerance",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "control_msgs",
        "type": "JointTolerance",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "component_goal_tolerance",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "control_msgs",
        "type": "JointComponentTolerance",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal_time_tolerance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "FollowJointTrajectory_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowJointTrajectory_Goal"
}
*/
