// rclnodejs/generated/control_msgs/control_msgs__msg__SingleDOFStateStamped.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const SingleDOFStateWrapper = require('../../generated/control_msgs/control_msgs__msg__SingleDOFState.js');
const SingleDOFStateStampedRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  dof_state: SingleDOFStateWrapper.refObjectType,
});
const SingleDOFStateStampedRefArray = ArrayType(SingleDOFStateStampedRefStruct);
const SingleDOFStateStampedRefStructArray = StructType({
  data: SingleDOFStateStampedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SingleDOFStateStampedWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SingleDOFStateStampedRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.dof_state =  new SingleDOFStateWrapper(other._wrapperFields.dof_state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SingleDOFStateStampedRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.dof_state = new SingleDOFStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SingleDOFStateStampedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SingleDOFStateStampedArrayWrapper;
  }
  static get ArrayType() {
    return SingleDOFStateStampedArrayWrapper;
  }
  static get refObjectArrayType() {
    return SingleDOFStateStampedRefStructArray
  }
  static get refObjectType() {
    return SingleDOFStateStampedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.dof_state.freeze(own, checkConsistency);
    this._refObject.dof_state = this._wrapperFields.dof_state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.dof_state.copyRefObject(refObject.dof_state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    SingleDOFStateWrapper.freeStruct(refObject.dof_state);
  }
  static destoryRawROS(msg) {
    SingleDOFStateStampedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'SingleDOFStateStamped'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get dof_state() {
    return this._wrapperFields.dof_state;
  }
  set dof_state(value) {
    if (value instanceof SingleDOFStateWrapper) {
      this._wrapperFields.dof_state.copy(value);
    } else {
      this._wrapperFields.dof_state.copy(new SingleDOFStateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SingleDOFStateStampedRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.dof_state.copyRefObject(this._refObject.dof_state);
  }
  copy(other) {
    this._refObject = new SingleDOFStateStampedRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.dof_state.copy(other._wrapperFields.dof_state);
  }
  static get classType() {
    return SingleDOFStateStampedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"dof_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"control_msgs","type":"SingleDOFState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"SingleDOFStateStamped","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SingleDOFStateStamped"};
  }
  hasMember(name) {
    let memberNames = ["header","dof_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SingleDOFStateStampedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SingleDOFStateStampedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SingleDOFStateStampedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleDOFStateStampedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleDOFStateStampedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SingleDOFStateStampedRefArray(size);
    this._refObject = new SingleDOFStateStampedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SingleDOFStateStampedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SingleDOFStateStampedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SingleDOFStateStampedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SingleDOFStateStampedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SingleDOFStateStampedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SingleDOFStateStampedArrayWrapper;
  }
}
module.exports = SingleDOFStateStampedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "dof_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "control_msgs",
        "type": "SingleDOFState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "SingleDOFStateStamped",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SingleDOFStateStamped"
}
*/
