// rclnodejs/generated/control_msgs/control_msgs__msg__SingleDOFState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SingleDOFStateRefStruct = StructType({
  name: primitiveTypes.string,
  reference: primitiveTypes.float64,
  feedback: primitiveTypes.float64,
  feedback_dot: primitiveTypes.float64,
  error: primitiveTypes.float64,
  error_dot: primitiveTypes.float64,
  time_step: primitiveTypes.float64,
  output: primitiveTypes.float64,
});
const SingleDOFStateRefArray = ArrayType(SingleDOFStateRefStruct);
const SingleDOFStateRefStructArray = StructType({
  data: SingleDOFStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SingleDOFStateWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SingleDOFStateRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._referenceIntialized = true;
      this._feedbackIntialized = true;
      this._feedback_dotIntialized = true;
      this._errorIntialized = true;
      this._error_dotIntialized = true;
      this._time_stepIntialized = true;
      this._outputIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SingleDOFStateRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._referenceIntialized = false;
    this._feedbackIntialized = false;
    this._feedback_dotIntialized = false;
    this._errorIntialized = false;
    this._error_dotIntialized = false;
    this._time_stepIntialized = false;
    this._outputIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SingleDOFStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SingleDOFStateArrayWrapper;
  }
  static get ArrayType() {
    return SingleDOFStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return SingleDOFStateRefStructArray
  }
  static get refObjectType() {
    return SingleDOFStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in SingleDOFState');
      }
      if (checkConsistency && !this._referenceIntialized) {
        throw new TypeError('Invalid argument: reference in SingleDOFState');
      }
      if (checkConsistency && !this._feedbackIntialized) {
        throw new TypeError('Invalid argument: feedback in SingleDOFState');
      }
      if (checkConsistency && !this._feedback_dotIntialized) {
        throw new TypeError('Invalid argument: feedback_dot in SingleDOFState');
      }
      if (checkConsistency && !this._errorIntialized) {
        throw new TypeError('Invalid argument: error in SingleDOFState');
      }
      if (checkConsistency && !this._error_dotIntialized) {
        throw new TypeError('Invalid argument: error_dot in SingleDOFState');
      }
      if (checkConsistency && !this._time_stepIntialized) {
        throw new TypeError('Invalid argument: time_step in SingleDOFState');
      }
      if (checkConsistency && !this._outputIntialized) {
        throw new TypeError('Invalid argument: output in SingleDOFState');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._referenceIntialized = true;
    this._refObject.reference = refObject.reference;
    this._feedbackIntialized = true;
    this._refObject.feedback = refObject.feedback;
    this._feedback_dotIntialized = true;
    this._refObject.feedback_dot = refObject.feedback_dot;
    this._errorIntialized = true;
    this._refObject.error = refObject.error;
    this._error_dotIntialized = true;
    this._refObject.error_dot = refObject.error_dot;
    this._time_stepIntialized = true;
    this._refObject.time_step = refObject.time_step;
    this._outputIntialized = true;
    this._refObject.output = refObject.output;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destoryRawROS(msg) {
    SingleDOFStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'SingleDOFState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get reference() {
    if (!this._referenceIntialized) {
      return undefined;
    }
    return this._refObject.reference;
  }
  set reference(value) {
    this._referenceIntialized = true;
    this._refObject.reference = value;
  }
  get feedback() {
    if (!this._feedbackIntialized) {
      return undefined;
    }
    return this._refObject.feedback;
  }
  set feedback(value) {
    this._feedbackIntialized = true;
    this._refObject.feedback = value;
  }
  get feedback_dot() {
    if (!this._feedback_dotIntialized) {
      return undefined;
    }
    return this._refObject.feedback_dot;
  }
  set feedback_dot(value) {
    this._feedback_dotIntialized = true;
    this._refObject.feedback_dot = value;
  }
  get error() {
    if (!this._errorIntialized) {
      return undefined;
    }
    return this._refObject.error;
  }
  set error(value) {
    this._errorIntialized = true;
    this._refObject.error = value;
  }
  get error_dot() {
    if (!this._error_dotIntialized) {
      return undefined;
    }
    return this._refObject.error_dot;
  }
  set error_dot(value) {
    this._error_dotIntialized = true;
    this._refObject.error_dot = value;
  }
  get time_step() {
    if (!this._time_stepIntialized) {
      return undefined;
    }
    return this._refObject.time_step;
  }
  set time_step(value) {
    this._time_stepIntialized = true;
    this._refObject.time_step = value;
  }
  get output() {
    if (!this._outputIntialized) {
      return undefined;
    }
    return this._refObject.output;
  }
  set output(value) {
    this._outputIntialized = true;
    this._refObject.output = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SingleDOFStateRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._referenceIntialized = true;
    this._feedbackIntialized = true;
    this._feedback_dotIntialized = true;
    this._errorIntialized = true;
    this._error_dotIntialized = true;
    this._time_stepIntialized = true;
    this._outputIntialized = true;
  }
  copy(other) {
    this._refObject = new SingleDOFStateRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._referenceIntialized = true;
    this._feedbackIntialized = true;
    this._feedback_dotIntialized = true;
    this._errorIntialized = true;
    this._error_dotIntialized = true;
    this._time_stepIntialized = true;
    this._outputIntialized = true;
  }
  static get classType() {
    return SingleDOFStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"feedback","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"feedback_dot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_dot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"output","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"SingleDOFState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SingleDOFState"};
  }
  hasMember(name) {
    let memberNames = ["name","reference","feedback","feedback_dot","error","error_dot","time_step","output"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SingleDOFStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SingleDOFStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SingleDOFStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleDOFStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleDOFStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SingleDOFStateRefArray(size);
    this._refObject = new SingleDOFStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SingleDOFStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SingleDOFStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SingleDOFStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SingleDOFStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SingleDOFStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SingleDOFStateArrayWrapper;
  }
}
module.exports = SingleDOFStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reference",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "feedback",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "feedback_dot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_dot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_step",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "output",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "SingleDOFState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SingleDOFState"
}
*/
