// rclnodejs/generated/vision_msgs/vision_msgs__msg__Detection2DArray.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Detection2DWrapper = require('../../generated/vision_msgs/vision_msgs__msg__Detection2D.js');
const Detection2DArrayRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  detections: Detection2DWrapper.refObjectArrayType,
});
const Detection2DArrayRefArray = ArrayType(Detection2DArrayRefStruct);
const Detection2DArrayRefStructArray = StructType({
  data: Detection2DArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Detection2DArrayWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Detection2DArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.detections = Detection2DWrapper.createArray();
      this._wrapperFields.detections.copy(other._wrapperFields.detections);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Detection2DArrayRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.detections = Detection2DWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new Detection2DArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Detection2DArrayArrayWrapper;
  }
  static get ArrayType() {
    return Detection2DArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return Detection2DArrayRefStructArray
  }
  static get refObjectType() {
    return Detection2DArrayRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.detections.freeze(own, checkConsistency);
    this._refObject.detections = this._wrapperFields.detections.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.detections.copyRefObject(refObject.detections);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.detections.size != 0) {
      Detection2DWrapper.ArrayType.freeArray(refObject.detections);
      if (Detection2DWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.detections, Detection2DWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    Detection2DArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'Detection2DArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get detections() {
    return this._wrapperFields.detections;
  }
  set detections(value) {
    this._wrapperFields.detections.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new Detection2DArrayRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.detections.copyRefObject(this._refObject.detections);
  }
  copy(other) {
    this._refObject = new Detection2DArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.detections.copy(other._wrapperFields.detections);
  }
  static get classType() {
    return Detection2DArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"detections","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"vision_msgs","type":"Detection2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"Detection2DArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Detection2DArray"};
  }
  hasMember(name) {
    let memberNames = ["header","detections"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Detection2DArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Detection2DArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Detection2DArrayWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Detection2DArrayArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Detection2DArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Detection2DArrayRefArray(size);
    this._refObject = new Detection2DArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Detection2DArrayWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Detection2DArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Detection2DArrayArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Detection2DArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Detection2DArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Detection2DArrayArrayWrapper;
  }
}
module.exports = Detection2DArrayWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "detections",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "vision_msgs",
        "type": "Detection2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "Detection2DArray",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Detection2DArray"
}
*/
