// rclnodejs/generated/vision_msgs/vision_msgs__msg__ObjectHypothesisWithPose.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ObjectHypothesisWrapper = require('../../generated/vision_msgs/vision_msgs__msg__ObjectHypothesis.js');
const PoseWithCovarianceWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseWithCovariance.js');
const ObjectHypothesisWithPoseRefStruct = StructType({
  hypothesis: ObjectHypothesisWrapper.refObjectType,
  pose: PoseWithCovarianceWrapper.refObjectType,
});
const ObjectHypothesisWithPoseRefArray = ArrayType(ObjectHypothesisWithPoseRefStruct);
const ObjectHypothesisWithPoseRefStructArray = StructType({
  data: ObjectHypothesisWithPoseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObjectHypothesisWithPoseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectHypothesisWithPoseRefStruct(other._refObject.toObject());
      this._wrapperFields.hypothesis =  new ObjectHypothesisWrapper(other._wrapperFields.hypothesis);
      this._wrapperFields.pose =  new PoseWithCovarianceWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObjectHypothesisWithPoseRefStruct();
    this._wrapperFields.hypothesis = new ObjectHypothesisWrapper();
    this._wrapperFields.pose = new PoseWithCovarianceWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectHypothesisWithPoseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectHypothesisWithPoseArrayWrapper;
  }
  static get ArrayType() {
    return ObjectHypothesisWithPoseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectHypothesisWithPoseRefStructArray
  }
  static get refObjectType() {
    return ObjectHypothesisWithPoseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.hypothesis.freeze(own, checkConsistency);
    this._refObject.hypothesis = this._wrapperFields.hypothesis.refObject;
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.hypothesis.copyRefObject(refObject.hypothesis);
    this._wrapperFields.pose.copyRefObject(refObject.pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    ObjectHypothesisWrapper.freeStruct(refObject.hypothesis);
    PoseWithCovarianceWrapper.freeStruct(refObject.pose);
  }
  static destoryRawROS(msg) {
    ObjectHypothesisWithPoseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'ObjectHypothesisWithPose'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get hypothesis() {
    return this._wrapperFields.hypothesis;
  }
  set hypothesis(value) {
    if (value instanceof ObjectHypothesisWrapper) {
      this._wrapperFields.hypothesis.copy(value);
    } else {
      this._wrapperFields.hypothesis.copy(new ObjectHypothesisWrapper(value));
    }
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWithCovarianceWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWithCovarianceWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectHypothesisWithPoseRefStruct(refObject.toObject());
    this._wrapperFields.hypothesis.copyRefObject(this._refObject.hypothesis);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new ObjectHypothesisWithPoseRefStruct(other._refObject.toObject());
    this._wrapperFields.hypothesis.copy(other._wrapperFields.hypothesis);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return ObjectHypothesisWithPoseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"hypothesis","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"vision_msgs","type":"ObjectHypothesis","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseWithCovariance","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"ObjectHypothesisWithPose","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectHypothesisWithPose"};
  }
  hasMember(name) {
    let memberNames = ["hypothesis","pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObjectHypothesisWithPoseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectHypothesisWithPoseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectHypothesisWithPoseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectHypothesisWithPoseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectHypothesisWithPoseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObjectHypothesisWithPoseRefArray(size);
    this._refObject = new ObjectHypothesisWithPoseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectHypothesisWithPoseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectHypothesisWithPoseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectHypothesisWithPoseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectHypothesisWithPoseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectHypothesisWithPoseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectHypothesisWithPoseArrayWrapper;
  }
}
module.exports = ObjectHypothesisWithPoseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "hypothesis",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "vision_msgs",
        "type": "ObjectHypothesis",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseWithCovariance",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "ObjectHypothesisWithPose",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObjectHypothesisWithPose"
}
*/
