// rclnodejs/generated/vision_msgs/vision_msgs__msg__BoundingBox2D.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Pose2DWrapper = require('../../generated/vision_msgs/vision_msgs__msg__Pose2D.js');
const BoundingBox2DRefStruct = StructType({
  center: Pose2DWrapper.refObjectType,
  size_x: primitiveTypes.float64,
  size_y: primitiveTypes.float64,
});
const BoundingBox2DRefArray = ArrayType(BoundingBox2DRefStruct);
const BoundingBox2DRefStructArray = StructType({
  data: BoundingBox2DRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BoundingBox2DWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BoundingBox2DRefStruct(other._refObject.toObject());
      this._wrapperFields.center =  new Pose2DWrapper(other._wrapperFields.center);
      this._size_xIntialized = true;
      this._size_yIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BoundingBox2DRefStruct();
    this._wrapperFields.center = new Pose2DWrapper();
    this._size_xIntialized = false;
    this._size_yIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new BoundingBox2DWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BoundingBox2DArrayWrapper;
  }
  static get ArrayType() {
    return BoundingBox2DArrayWrapper;
  }
  static get refObjectArrayType() {
    return BoundingBox2DRefStructArray
  }
  static get refObjectType() {
    return BoundingBox2DRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._size_xIntialized) {
        throw new TypeError('Invalid argument: size_x in BoundingBox2D');
      }
      if (checkConsistency && !this._size_yIntialized) {
        throw new TypeError('Invalid argument: size_y in BoundingBox2D');
      }
    this._wrapperFields.center.freeze(own, checkConsistency);
    this._refObject.center = this._wrapperFields.center.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.center.copyRefObject(refObject.center);
    this._size_xIntialized = true;
    this._refObject.size_x = refObject.size_x;
    this._size_yIntialized = true;
    this._refObject.size_y = refObject.size_y;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Pose2DWrapper.freeStruct(refObject.center);
  }
  static destoryRawROS(msg) {
    BoundingBox2DWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'BoundingBox2D'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get center() {
    return this._wrapperFields.center;
  }
  set center(value) {
    if (value instanceof Pose2DWrapper) {
      this._wrapperFields.center.copy(value);
    } else {
      this._wrapperFields.center.copy(new Pose2DWrapper(value));
    }
  }
  get size_x() {
    if (!this._size_xIntialized) {
      return undefined;
    }
    return this._refObject.size_x;
  }
  set size_x(value) {
    this._size_xIntialized = true;
    this._refObject.size_x = value;
  }
  get size_y() {
    if (!this._size_yIntialized) {
      return undefined;
    }
    return this._refObject.size_y;
  }
  set size_y(value) {
    this._size_yIntialized = true;
    this._refObject.size_y = value;
  }
  copyRefObject(refObject) {
    this._refObject = new BoundingBox2DRefStruct(refObject.toObject());
    this._wrapperFields.center.copyRefObject(this._refObject.center);
    this._size_xIntialized = true;
    this._size_yIntialized = true;
  }
  copy(other) {
    this._refObject = new BoundingBox2DRefStruct(other._refObject.toObject());
    this._wrapperFields.center.copy(other._wrapperFields.center);
    this._size_xIntialized = true;
    this._size_yIntialized = true;
  }
  static get classType() {
    return BoundingBox2DWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"center","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"vision_msgs","type":"Pose2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"size_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"BoundingBox2D","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BoundingBox2D"};
  }
  hasMember(name) {
    let memberNames = ["center","size_x","size_y"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BoundingBox2DArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BoundingBox2DWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BoundingBox2DWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBox2DArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBox2DArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BoundingBox2DRefArray(size);
    this._refObject = new BoundingBox2DRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BoundingBox2DWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BoundingBox2DArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BoundingBox2DArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BoundingBox2DWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BoundingBox2DWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BoundingBox2DArrayWrapper;
  }
}
module.exports = BoundingBox2DWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "center",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "vision_msgs",
        "type": "Pose2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "size_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "BoundingBox2D",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BoundingBox2D"
}
*/
