// rclnodejs/generated/vision_msgs/vision_msgs__msg__BoundingBox3D.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const BoundingBox3DRefStruct = StructType({
  center: PoseWrapper.refObjectType,
  size: Vector3Wrapper.refObjectType,
});
const BoundingBox3DRefArray = ArrayType(BoundingBox3DRefStruct);
const BoundingBox3DRefStructArray = StructType({
  data: BoundingBox3DRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BoundingBox3DWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BoundingBox3DRefStruct(other._refObject.toObject());
      this._wrapperFields.center =  new PoseWrapper(other._wrapperFields.center);
      this._wrapperFields.size =  new Vector3Wrapper(other._wrapperFields.size);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BoundingBox3DRefStruct();
    this._wrapperFields.center = new PoseWrapper();
    this._wrapperFields.size = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new BoundingBox3DWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BoundingBox3DArrayWrapper;
  }
  static get ArrayType() {
    return BoundingBox3DArrayWrapper;
  }
  static get refObjectArrayType() {
    return BoundingBox3DRefStructArray
  }
  static get refObjectType() {
    return BoundingBox3DRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.center.freeze(own, checkConsistency);
    this._refObject.center = this._wrapperFields.center.refObject;
    this._wrapperFields.size.freeze(own, checkConsistency);
    this._refObject.size = this._wrapperFields.size.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.center.copyRefObject(refObject.center);
    this._wrapperFields.size.copyRefObject(refObject.size);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.center);
    Vector3Wrapper.freeStruct(refObject.size);
  }
  static destoryRawROS(msg) {
    BoundingBox3DWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'BoundingBox3D'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get center() {
    return this._wrapperFields.center;
  }
  set center(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.center.copy(value);
    } else {
      this._wrapperFields.center.copy(new PoseWrapper(value));
    }
  }
  get size() {
    return this._wrapperFields.size;
  }
  set size(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.size.copy(value);
    } else {
      this._wrapperFields.size.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new BoundingBox3DRefStruct(refObject.toObject());
    this._wrapperFields.center.copyRefObject(this._refObject.center);
    this._wrapperFields.size.copyRefObject(this._refObject.size);
  }
  copy(other) {
    this._refObject = new BoundingBox3DRefStruct(other._refObject.toObject());
    this._wrapperFields.center.copy(other._wrapperFields.center);
    this._wrapperFields.size.copy(other._wrapperFields.size);
  }
  static get classType() {
    return BoundingBox3DWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"center","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"BoundingBox3D","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BoundingBox3D"};
  }
  hasMember(name) {
    let memberNames = ["center","size"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BoundingBox3DArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BoundingBox3DWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BoundingBox3DWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBox3DArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBox3DArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BoundingBox3DRefArray(size);
    this._refObject = new BoundingBox3DRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BoundingBox3DWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BoundingBox3DArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BoundingBox3DArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BoundingBox3DWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BoundingBox3DWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BoundingBox3DArrayWrapper;
  }
}
module.exports = BoundingBox3DWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "center",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "size",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "BoundingBox3D",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BoundingBox3D"
}
*/
