// rclnodejs/generated/vision_msgs/vision_msgs__msg__Detection3D.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const ObjectHypothesisWithPoseWrapper = require('../../generated/vision_msgs/vision_msgs__msg__ObjectHypothesisWithPose.js');
const BoundingBox3DWrapper = require('../../generated/vision_msgs/vision_msgs__msg__BoundingBox3D.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Detection3DRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  results: ObjectHypothesisWithPoseWrapper.refObjectArrayType,
  bbox: BoundingBox3DWrapper.refObjectType,
  id: primitiveTypes.string,
});
const Detection3DRefArray = ArrayType(Detection3DRefStruct);
const Detection3DRefStructArray = StructType({
  data: Detection3DRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Detection3DWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Detection3DRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.results = ObjectHypothesisWithPoseWrapper.createArray();
      this._wrapperFields.results.copy(other._wrapperFields.results);
      this._wrapperFields.bbox =  new BoundingBox3DWrapper(other._wrapperFields.bbox);
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Detection3DRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.results = ObjectHypothesisWithPoseWrapper.createArray();
    this._wrapperFields.bbox = new BoundingBox3DWrapper();
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Detection3DWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Detection3DArrayWrapper;
  }
  static get ArrayType() {
    return Detection3DArrayWrapper;
  }
  static get refObjectArrayType() {
    return Detection3DRefStructArray
  }
  static get refObjectType() {
    return Detection3DRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in Detection3D');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.results.freeze(own, checkConsistency);
    this._refObject.results = this._wrapperFields.results.refObject;
    this._wrapperFields.bbox.freeze(own, checkConsistency);
    this._refObject.bbox = this._wrapperFields.bbox.refObject;
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.results.copyRefObject(refObject.results);
    this._wrapperFields.bbox.copyRefObject(refObject.bbox);
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.results.size != 0) {
      ObjectHypothesisWithPoseWrapper.ArrayType.freeArray(refObject.results);
      if (ObjectHypothesisWithPoseWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.results, ObjectHypothesisWithPoseWrapper.refObjectArrayType, 'data');
      }
    }
    BoundingBox3DWrapper.freeStruct(refObject.bbox);
    StringWrapper.freeStruct(refObject.id);
  }
  static destoryRawROS(msg) {
    Detection3DWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'Detection3D'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get results() {
    return this._wrapperFields.results;
  }
  set results(value) {
    this._wrapperFields.results.fill(value);
  }
  get bbox() {
    return this._wrapperFields.bbox;
  }
  set bbox(value) {
    if (value instanceof BoundingBox3DWrapper) {
      this._wrapperFields.bbox.copy(value);
    } else {
      this._wrapperFields.bbox.copy(new BoundingBox3DWrapper(value));
    }
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Detection3DRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.results.copyRefObject(this._refObject.results);
    this._wrapperFields.bbox.copyRefObject(this._refObject.bbox);
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
  }
  copy(other) {
    this._refObject = new Detection3DRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.results.copy(other._wrapperFields.results);
    this._wrapperFields.bbox.copy(other._wrapperFields.bbox);
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
  }
  static get classType() {
    return Detection3DWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"results","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"vision_msgs","type":"ObjectHypothesisWithPose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"bbox","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"vision_msgs","type":"BoundingBox3D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"Detection3D","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Detection3D"};
  }
  hasMember(name) {
    let memberNames = ["header","results","bbox","id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Detection3DArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Detection3DWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Detection3DWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Detection3DArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Detection3DArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Detection3DRefArray(size);
    this._refObject = new Detection3DRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Detection3DWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Detection3DArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Detection3DArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Detection3DWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Detection3DWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Detection3DArrayWrapper;
  }
}
module.exports = Detection3DWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "results",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "vision_msgs",
        "type": "ObjectHypothesisWithPose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "bbox",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "vision_msgs",
        "type": "BoundingBox3D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "Detection3D",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Detection3D"
}
*/
