// rclnodejs/generated/vision_msgs/vision_msgs__msg__VisionClass.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VisionClassRefStruct = StructType({
  class_id: primitiveTypes.uint16,
  class_name: primitiveTypes.string,
});
const VisionClassRefArray = ArrayType(VisionClassRefStruct);
const VisionClassRefStructArray = StructType({
  data: VisionClassRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VisionClassWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VisionClassRefStruct(other._refObject.toObject());
      this._class_idIntialized = true;
      this._class_nameIntialized = true;
      this._wrapperFields.class_name =  new StringWrapper(other._wrapperFields.class_name);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VisionClassRefStruct();
    this._class_idIntialized = false;
    this._class_nameIntialized = false;
    this._wrapperFields.class_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new VisionClassWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VisionClassArrayWrapper;
  }
  static get ArrayType() {
    return VisionClassArrayWrapper;
  }
  static get refObjectArrayType() {
    return VisionClassRefStructArray
  }
  static get refObjectType() {
    return VisionClassRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._class_idIntialized) {
        throw new TypeError('Invalid argument: class_id in VisionClass');
      }
      if (checkConsistency && !this._class_nameIntialized) {
        throw new TypeError('Invalid argument: class_name in VisionClass');
      }
    if (own) {
      this._wrapperFields.class_name.freeze(own, checkConsistency);
    }
    this._refObject.class_name = this._wrapperFields.class_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._class_idIntialized = true;
    this._refObject.class_id = refObject.class_id;
    this._class_nameIntialized = true;
    this._wrapperFields.class_name.data = refObject.class_name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.class_name);
  }
  static destoryRawROS(msg) {
    VisionClassWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'VisionClass'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get class_id() {
    if (!this._class_idIntialized) {
      return undefined;
    }
    return this._refObject.class_id;
  }
  set class_id(value) {
    this._class_idIntialized = true;
    this._refObject.class_id = value;
  }
  get class_name() {
    if (!this._class_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.class_name.data;
  }
  set class_name(value) {
    this._class_nameIntialized = true;
    this._wrapperFields.class_name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VisionClassRefStruct(refObject.toObject());
    this._class_idIntialized = true;
    this._class_nameIntialized = true;
    this._wrapperFields.class_name.copyRefObject(this._refObject.class_name);
  }
  copy(other) {
    this._refObject = new VisionClassRefStruct(other._refObject.toObject());
    this._class_idIntialized = true;
    this._class_nameIntialized = true;
    this._wrapperFields.class_name.copy(other._wrapperFields.class_name);
  }
  static get classType() {
    return VisionClassWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"class_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"class_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"VisionClass","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VisionClass"};
  }
  hasMember(name) {
    let memberNames = ["class_id","class_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VisionClassArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VisionClassWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VisionClassWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VisionClassArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VisionClassArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VisionClassRefArray(size);
    this._refObject = new VisionClassRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VisionClassWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VisionClassArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VisionClassArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VisionClassWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VisionClassWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VisionClassArrayWrapper;
  }
}
module.exports = VisionClassWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "class_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "class_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "VisionClass",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VisionClass"
}
*/
