// rclnodejs/generated/octomap_msgs/octomap_msgs__msg__OctomapWithPose.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const OctomapWrapper = require('../../generated/octomap_msgs/octomap_msgs__msg__Octomap.js');
const OctomapWithPoseRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  origin: PoseWrapper.refObjectType,
  octomap: OctomapWrapper.refObjectType,
});
const OctomapWithPoseRefArray = ArrayType(OctomapWithPoseRefStruct);
const OctomapWithPoseRefStructArray = StructType({
  data: OctomapWithPoseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OctomapWithPoseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OctomapWithPoseRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.origin =  new PoseWrapper(other._wrapperFields.origin);
      this._wrapperFields.octomap =  new OctomapWrapper(other._wrapperFields.octomap);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OctomapWithPoseRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.origin = new PoseWrapper();
    this._wrapperFields.octomap = new OctomapWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new OctomapWithPoseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OctomapWithPoseArrayWrapper;
  }
  static get ArrayType() {
    return OctomapWithPoseArrayWrapper;
  }
  static get refObjectArrayType() {
    return OctomapWithPoseRefStructArray
  }
  static get refObjectType() {
    return OctomapWithPoseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.origin.freeze(own, checkConsistency);
    this._refObject.origin = this._wrapperFields.origin.refObject;
    this._wrapperFields.octomap.freeze(own, checkConsistency);
    this._refObject.octomap = this._wrapperFields.octomap.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.origin.copyRefObject(refObject.origin);
    this._wrapperFields.octomap.copyRefObject(refObject.octomap);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PoseWrapper.freeStruct(refObject.origin);
    OctomapWrapper.freeStruct(refObject.octomap);
  }
  static destoryRawROS(msg) {
    OctomapWithPoseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'octomap_msgs', subFolder: 'msg', interfaceName: 'OctomapWithPose'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get origin() {
    return this._wrapperFields.origin;
  }
  set origin(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.origin.copy(value);
    } else {
      this._wrapperFields.origin.copy(new PoseWrapper(value));
    }
  }
  get octomap() {
    return this._wrapperFields.octomap;
  }
  set octomap(value) {
    if (value instanceof OctomapWrapper) {
      this._wrapperFields.octomap.copy(value);
    } else {
      this._wrapperFields.octomap.copy(new OctomapWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new OctomapWithPoseRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.origin.copyRefObject(this._refObject.origin);
    this._wrapperFields.octomap.copyRefObject(this._refObject.octomap);
  }
  copy(other) {
    this._refObject = new OctomapWithPoseRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.origin.copy(other._wrapperFields.origin);
    this._wrapperFields.octomap.copy(other._wrapperFields.octomap);
  }
  static get classType() {
    return OctomapWithPoseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"origin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"octomap","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"octomap_msgs","type":"Octomap","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"octomap_msgs","type":"OctomapWithPose","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OctomapWithPose"};
  }
  hasMember(name) {
    let memberNames = ["header","origin","octomap"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OctomapWithPoseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OctomapWithPoseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OctomapWithPoseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OctomapWithPoseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OctomapWithPoseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OctomapWithPoseRefArray(size);
    this._refObject = new OctomapWithPoseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OctomapWithPoseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OctomapWithPoseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OctomapWithPoseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OctomapWithPoseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OctomapWithPoseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OctomapWithPoseArrayWrapper;
  }
}
module.exports = OctomapWithPoseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "origin",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "octomap",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "octomap_msgs",
        "type": "Octomap",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "octomap_msgs",
    "type": "OctomapWithPose",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OctomapWithPose"
}
*/
