// rclnodejs/generated/octomap_msgs/octomap_msgs__srv__BoundingBoxQuery_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const BoundingBoxQuery_RequestRefStruct = StructType({
  min: PointWrapper.refObjectType,
  max: PointWrapper.refObjectType,
});
const BoundingBoxQuery_RequestRefArray = ArrayType(BoundingBoxQuery_RequestRefStruct);
const BoundingBoxQuery_RequestRefStructArray = StructType({
  data: BoundingBoxQuery_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BoundingBoxQuery_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BoundingBoxQuery_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.min =  new PointWrapper(other._wrapperFields.min);
      this._wrapperFields.max =  new PointWrapper(other._wrapperFields.max);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BoundingBoxQuery_RequestRefStruct();
    this._wrapperFields.min = new PointWrapper();
    this._wrapperFields.max = new PointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new BoundingBoxQuery_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BoundingBoxQuery_RequestArrayWrapper;
  }
  static get ArrayType() {
    return BoundingBoxQuery_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return BoundingBoxQuery_RequestRefStructArray
  }
  static get refObjectType() {
    return BoundingBoxQuery_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.min.freeze(own, checkConsistency);
    this._refObject.min = this._wrapperFields.min.refObject;
    this._wrapperFields.max.freeze(own, checkConsistency);
    this._refObject.max = this._wrapperFields.max.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.min.copyRefObject(refObject.min);
    this._wrapperFields.max.copyRefObject(refObject.max);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointWrapper.freeStruct(refObject.min);
    PointWrapper.freeStruct(refObject.max);
  }
  static destoryRawROS(msg) {
    BoundingBoxQuery_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'octomap_msgs', subFolder: 'srv', interfaceName: 'BoundingBoxQuery_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get min() {
    return this._wrapperFields.min;
  }
  set min(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.min.copy(value);
    } else {
      this._wrapperFields.min.copy(new PointWrapper(value));
    }
  }
  get max() {
    return this._wrapperFields.max;
  }
  set max(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.max.copy(value);
    } else {
      this._wrapperFields.max.copy(new PointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new BoundingBoxQuery_RequestRefStruct(refObject.toObject());
    this._wrapperFields.min.copyRefObject(this._refObject.min);
    this._wrapperFields.max.copyRefObject(this._refObject.max);
  }
  copy(other) {
    this._refObject = new BoundingBoxQuery_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.min.copy(other._wrapperFields.min);
    this._wrapperFields.max.copy(other._wrapperFields.max);
  }
  static get classType() {
    return BoundingBoxQuery_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"octomap_msgs","type":"BoundingBoxQuery_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BoundingBoxQuery_Request"};
  }
  hasMember(name) {
    let memberNames = ["min","max"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BoundingBoxQuery_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BoundingBoxQuery_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BoundingBoxQuery_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBoxQuery_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBoxQuery_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BoundingBoxQuery_RequestRefArray(size);
    this._refObject = new BoundingBoxQuery_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BoundingBoxQuery_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BoundingBoxQuery_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BoundingBoxQuery_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BoundingBoxQuery_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BoundingBoxQuery_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BoundingBoxQuery_RequestArrayWrapper;
  }
}
module.exports = BoundingBoxQuery_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "octomap_msgs",
    "type": "BoundingBoxQuery_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BoundingBoxQuery_Request"
}
*/
