// rclnodejs/generated/octomap_msgs/octomap_msgs__msg__Octomap.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Int8Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int8.js');
const OctomapRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  binary: primitiveTypes.bool,
  id: primitiveTypes.string,
  resolution: primitiveTypes.float64,
  data: Int8Wrapper.refObjectArrayType,
});
const OctomapRefArray = ArrayType(OctomapRefStruct);
const OctomapRefStructArray = StructType({
  data: OctomapRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OctomapWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OctomapRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._binaryIntialized = true;
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
      this._resolutionIntialized = true;
      this._wrapperFields.data = Int8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OctomapRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._binaryIntialized = false;
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
    this._resolutionIntialized = false;
    this._wrapperFields.data = Int8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new OctomapWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OctomapArrayWrapper;
  }
  static get ArrayType() {
    return OctomapArrayWrapper;
  }
  static get refObjectArrayType() {
    return OctomapRefStructArray
  }
  static get refObjectType() {
    return OctomapRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._binaryIntialized) {
        throw new TypeError('Invalid argument: binary in Octomap');
      }
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in Octomap');
      }
      if (checkConsistency && !this._resolutionIntialized) {
        throw new TypeError('Invalid argument: resolution in Octomap');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Int8Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._binaryIntialized = true;
    this._refObject.binary = refObject.binary;
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
    this._resolutionIntialized = true;
    this._refObject.resolution = refObject.resolution;
    this._wrapperFields.data.copyRefObject(refObject.data);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.id);
    if (refObject.data.size != 0) {
      Int8Wrapper.ArrayType.freeArray(refObject.data);
      if (Int8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, Int8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    OctomapWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'octomap_msgs', subFolder: 'msg', interfaceName: 'Octomap'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get binary() {
    if (!this._binaryIntialized) {
      return undefined;
    }
    return this._refObject.binary;
  }
  set binary(value) {
    this._binaryIntialized = true;
    this._refObject.binary = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  get resolution() {
    if (!this._resolutionIntialized) {
      return undefined;
    }
    return this._refObject.resolution;
  }
  set resolution(value) {
    this._resolutionIntialized = true;
    this._refObject.resolution = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new OctomapRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._binaryIntialized = true;
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._resolutionIntialized = true;
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new OctomapRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._binaryIntialized = true;
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._resolutionIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return OctomapWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"binary","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"octomap_msgs","type":"Octomap","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Octomap"};
  }
  hasMember(name) {
    let memberNames = ["header","binary","id","resolution","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OctomapArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OctomapWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OctomapWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OctomapArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OctomapArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OctomapRefArray(size);
    this._refObject = new OctomapRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OctomapWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OctomapArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OctomapArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OctomapWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OctomapWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OctomapArrayWrapper;
  }
}
module.exports = OctomapWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "binary",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "octomap_msgs",
    "type": "Octomap",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Octomap"
}
*/
