// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__GetTrajectoryStates_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatusResponseWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__StatusResponse.js');
const TrajectoryStatesWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__TrajectoryStates.js');
const GetTrajectoryStates_ResponseRefStruct = StructType({
  status: StatusResponseWrapper.refObjectType,
  trajectory_states: TrajectoryStatesWrapper.refObjectType,
});
const GetTrajectoryStates_ResponseRefArray = ArrayType(GetTrajectoryStates_ResponseRefStruct);
const GetTrajectoryStates_ResponseRefStructArray = StructType({
  data: GetTrajectoryStates_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetTrajectoryStates_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetTrajectoryStates_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.status =  new StatusResponseWrapper(other._wrapperFields.status);
      this._wrapperFields.trajectory_states =  new TrajectoryStatesWrapper(other._wrapperFields.trajectory_states);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetTrajectoryStates_ResponseRefStruct();
    this._wrapperFields.status = new StatusResponseWrapper();
    this._wrapperFields.trajectory_states = new TrajectoryStatesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetTrajectoryStates_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetTrajectoryStates_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetTrajectoryStates_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetTrajectoryStates_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetTrajectoryStates_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
    this._wrapperFields.trajectory_states.freeze(own, checkConsistency);
    this._refObject.trajectory_states = this._wrapperFields.trajectory_states.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.status.copyRefObject(refObject.status);
    this._wrapperFields.trajectory_states.copyRefObject(refObject.trajectory_states);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StatusResponseWrapper.freeStruct(refObject.status);
    TrajectoryStatesWrapper.freeStruct(refObject.trajectory_states);
  }
  static destoryRawROS(msg) {
    GetTrajectoryStates_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'GetTrajectoryStates_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof StatusResponseWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new StatusResponseWrapper(value));
    }
  }
  get trajectory_states() {
    return this._wrapperFields.trajectory_states;
  }
  set trajectory_states(value) {
    if (value instanceof TrajectoryStatesWrapper) {
      this._wrapperFields.trajectory_states.copy(value);
    } else {
      this._wrapperFields.trajectory_states.copy(new TrajectoryStatesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetTrajectoryStates_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._wrapperFields.trajectory_states.copyRefObject(this._refObject.trajectory_states);
  }
  copy(other) {
    this._refObject = new GetTrajectoryStates_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._wrapperFields.trajectory_states.copy(other._wrapperFields.trajectory_states);
  }
  static get classType() {
    return GetTrajectoryStates_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"StatusResponse","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"trajectory_states","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"TrajectoryStates","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"GetTrajectoryStates_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetTrajectoryStates_Response"};
  }
  hasMember(name) {
    let memberNames = ["status","trajectory_states"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetTrajectoryStates_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetTrajectoryStates_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetTrajectoryStates_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetTrajectoryStates_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetTrajectoryStates_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetTrajectoryStates_ResponseRefArray(size);
    this._refObject = new GetTrajectoryStates_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetTrajectoryStates_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetTrajectoryStates_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetTrajectoryStates_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetTrajectoryStates_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetTrajectoryStates_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetTrajectoryStates_ResponseArrayWrapper;
  }
}
module.exports = GetTrajectoryStates_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "StatusResponse",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "trajectory_states",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "TrajectoryStates",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "GetTrajectoryStates_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetTrajectoryStates_Response"
}
*/
