// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__ReadMetrics_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatusResponseWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__StatusResponse.js');
const MetricFamilyWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__MetricFamily.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const ReadMetrics_ResponseRefStruct = StructType({
  status: StatusResponseWrapper.refObjectType,
  metric_families: MetricFamilyWrapper.refObjectArrayType,
  timestamp: TimeWrapper.refObjectType,
});
const ReadMetrics_ResponseRefArray = ArrayType(ReadMetrics_ResponseRefStruct);
const ReadMetrics_ResponseRefStructArray = StructType({
  data: ReadMetrics_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ReadMetrics_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ReadMetrics_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.status =  new StatusResponseWrapper(other._wrapperFields.status);
      this._wrapperFields.metric_families = MetricFamilyWrapper.createArray();
      this._wrapperFields.metric_families.copy(other._wrapperFields.metric_families);
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ReadMetrics_ResponseRefStruct();
    this._wrapperFields.status = new StatusResponseWrapper();
    this._wrapperFields.metric_families = MetricFamilyWrapper.createArray();
    this._wrapperFields.timestamp = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ReadMetrics_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ReadMetrics_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ReadMetrics_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ReadMetrics_ResponseRefStructArray
  }
  static get refObjectType() {
    return ReadMetrics_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
    this._wrapperFields.metric_families.freeze(own, checkConsistency);
    this._refObject.metric_families = this._wrapperFields.metric_families.refObject;
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.status.copyRefObject(refObject.status);
    this._wrapperFields.metric_families.copyRefObject(refObject.metric_families);
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StatusResponseWrapper.freeStruct(refObject.status);
    if (refObject.metric_families.size != 0) {
      MetricFamilyWrapper.ArrayType.freeArray(refObject.metric_families);
      if (MetricFamilyWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.metric_families, MetricFamilyWrapper.refObjectArrayType, 'data');
      }
    }
    TimeWrapper.freeStruct(refObject.timestamp);
  }
  static destoryRawROS(msg) {
    ReadMetrics_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'ReadMetrics_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof StatusResponseWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new StatusResponseWrapper(value));
    }
  }
  get metric_families() {
    return this._wrapperFields.metric_families;
  }
  set metric_families(value) {
    this._wrapperFields.metric_families.fill(value);
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ReadMetrics_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._wrapperFields.metric_families.copyRefObject(this._refObject.metric_families);
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
  }
  copy(other) {
    this._refObject = new ReadMetrics_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._wrapperFields.metric_families.copy(other._wrapperFields.metric_families);
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
  }
  static get classType() {
    return ReadMetrics_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"StatusResponse","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"metric_families","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"cartographer_ros_msgs","type":"MetricFamily","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"ReadMetrics_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ReadMetrics_Response"};
  }
  hasMember(name) {
    let memberNames = ["status","metric_families","timestamp"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ReadMetrics_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ReadMetrics_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ReadMetrics_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ReadMetrics_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ReadMetrics_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ReadMetrics_ResponseRefArray(size);
    this._refObject = new ReadMetrics_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ReadMetrics_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ReadMetrics_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ReadMetrics_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ReadMetrics_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ReadMetrics_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ReadMetrics_ResponseArrayWrapper;
  }
}
module.exports = ReadMetrics_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "StatusResponse",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "metric_families",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "cartographer_ros_msgs",
        "type": "MetricFamily",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "ReadMetrics_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ReadMetrics_Response"
}
*/
