// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__Metric.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MetricLabelWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__MetricLabel.js');
const HistogramBucketWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__HistogramBucket.js');
const MetricRefStruct = StructType({
  type: primitiveTypes.uint8,
  labels: MetricLabelWrapper.refObjectArrayType,
  value: primitiveTypes.float64,
  counts_by_bucket: HistogramBucketWrapper.refObjectArrayType,
});
const MetricRefArray = ArrayType(MetricRefStruct);
const MetricRefStructArray = StructType({
  data: MetricRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MetricWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MetricRefStruct(other._refObject.toObject());
      this._typeIntialized = true;
      this._wrapperFields.labels = MetricLabelWrapper.createArray();
      this._wrapperFields.labels.copy(other._wrapperFields.labels);
      this._valueIntialized = true;
      this._wrapperFields.counts_by_bucket = HistogramBucketWrapper.createArray();
      this._wrapperFields.counts_by_bucket.copy(other._wrapperFields.counts_by_bucket);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MetricRefStruct();
    this._typeIntialized = false;
    this._wrapperFields.labels = MetricLabelWrapper.createArray();
    this._valueIntialized = false;
    this._wrapperFields.counts_by_bucket = HistogramBucketWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MetricWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MetricArrayWrapper;
  }
  static get ArrayType() {
    return MetricArrayWrapper;
  }
  static get refObjectArrayType() {
    return MetricRefStructArray
  }
  static get refObjectType() {
    return MetricRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._typeIntialized) {
        throw new TypeError('Invalid argument: type in Metric');
      }
      if (checkConsistency && !this._valueIntialized) {
        throw new TypeError('Invalid argument: value in Metric');
      }
    this._wrapperFields.labels.freeze(own, checkConsistency);
    this._refObject.labels = this._wrapperFields.labels.refObject;
    this._wrapperFields.counts_by_bucket.freeze(own, checkConsistency);
    this._refObject.counts_by_bucket = this._wrapperFields.counts_by_bucket.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._wrapperFields.labels.copyRefObject(refObject.labels);
    this._valueIntialized = true;
    this._refObject.value = refObject.value;
    this._wrapperFields.counts_by_bucket.copyRefObject(refObject.counts_by_bucket);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.labels.size != 0) {
      MetricLabelWrapper.ArrayType.freeArray(refObject.labels);
      if (MetricLabelWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.labels, MetricLabelWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.counts_by_bucket.size != 0) {
      HistogramBucketWrapper.ArrayType.freeArray(refObject.counts_by_bucket);
      if (HistogramBucketWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.counts_by_bucket, HistogramBucketWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MetricWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'Metric'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get labels() {
    return this._wrapperFields.labels;
  }
  set labels(value) {
    this._wrapperFields.labels.fill(value);
  }
  get value() {
    if (!this._valueIntialized) {
      return undefined;
    }
    return this._refObject.value;
  }
  set value(value) {
    this._valueIntialized = true;
    this._refObject.value = value;
  }
  get counts_by_bucket() {
    return this._wrapperFields.counts_by_bucket;
  }
  set counts_by_bucket(value) {
    this._wrapperFields.counts_by_bucket.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MetricRefStruct(refObject.toObject());
    this._typeIntialized = true;
    this._wrapperFields.labels.copyRefObject(this._refObject.labels);
    this._valueIntialized = true;
    this._wrapperFields.counts_by_bucket.copyRefObject(this._refObject.counts_by_bucket);
  }
  copy(other) {
    this._refObject = new MetricRefStruct(other._refObject.toObject());
    this._typeIntialized = true;
    this._wrapperFields.labels.copy(other._wrapperFields.labels);
    this._valueIntialized = true;
    this._wrapperFields.counts_by_bucket.copy(other._wrapperFields.counts_by_bucket);
  }
  static get classType() {
    return MetricWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"TYPE_COUNTER","value":0},{"type":"uint8","name":"TYPE_GAUGE","value":1},{"type":"uint8","name":"TYPE_HISTOGRAM","value":2}],"fields":[{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"labels","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"cartographer_ros_msgs","type":"MetricLabel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"counts_by_bucket","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"cartographer_ros_msgs","type":"HistogramBucket","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"Metric","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Metric"};
  }
  hasMember(name) {
    let memberNames = ["type","labels","value","counts_by_bucket"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MetricArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MetricWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MetricWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MetricArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MetricArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MetricRefArray(size);
    this._refObject = new MetricRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MetricWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MetricArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MetricArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MetricWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MetricWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MetricArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(MetricWrapper, "TYPE_COUNTER", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MetricWrapper, "TYPE_GAUGE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MetricWrapper, "TYPE_HISTOGRAM", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = MetricWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "TYPE_COUNTER",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "TYPE_GAUGE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "TYPE_HISTOGRAM",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "labels",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "cartographer_ros_msgs",
        "type": "MetricLabel",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "counts_by_bucket",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "cartographer_ros_msgs",
        "type": "HistogramBucket",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "Metric",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Metric"
}
*/
