// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__MetricFamily.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MetricWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__Metric.js');
const MetricFamilyRefStruct = StructType({
  name: primitiveTypes.string,
  description: primitiveTypes.string,
  metrics: MetricWrapper.refObjectArrayType,
});
const MetricFamilyRefArray = ArrayType(MetricFamilyRefStruct);
const MetricFamilyRefStructArray = StructType({
  data: MetricFamilyRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MetricFamilyWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MetricFamilyRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._descriptionIntialized = true;
      this._wrapperFields.description =  new StringWrapper(other._wrapperFields.description);
      this._wrapperFields.metrics = MetricWrapper.createArray();
      this._wrapperFields.metrics.copy(other._wrapperFields.metrics);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MetricFamilyRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._descriptionIntialized = false;
    this._wrapperFields.description = new StringWrapper();
    this._wrapperFields.metrics = MetricWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MetricFamilyWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MetricFamilyArrayWrapper;
  }
  static get ArrayType() {
    return MetricFamilyArrayWrapper;
  }
  static get refObjectArrayType() {
    return MetricFamilyRefStructArray
  }
  static get refObjectType() {
    return MetricFamilyRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in MetricFamily');
      }
      if (checkConsistency && !this._descriptionIntialized) {
        throw new TypeError('Invalid argument: description in MetricFamily');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.description.freeze(own, checkConsistency);
    }
    this._refObject.description = this._wrapperFields.description.refObject;
    this._wrapperFields.metrics.freeze(own, checkConsistency);
    this._refObject.metrics = this._wrapperFields.metrics.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._descriptionIntialized = true;
    this._wrapperFields.description.data = refObject.description.data;
    this._wrapperFields.metrics.copyRefObject(refObject.metrics);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    StringWrapper.freeStruct(refObject.description);
    if (refObject.metrics.size != 0) {
      MetricWrapper.ArrayType.freeArray(refObject.metrics);
      if (MetricWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.metrics, MetricWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MetricFamilyWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'MetricFamily'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get description() {
    if (!this._descriptionIntialized) {
      return undefined;
    }
    return this._wrapperFields.description.data;
  }
  set description(value) {
    this._descriptionIntialized = true;
    this._wrapperFields.description.data = value;
  }
  get metrics() {
    return this._wrapperFields.metrics;
  }
  set metrics(value) {
    this._wrapperFields.metrics.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MetricFamilyRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._descriptionIntialized = true;
    this._wrapperFields.description.copyRefObject(this._refObject.description);
    this._wrapperFields.metrics.copyRefObject(this._refObject.metrics);
  }
  copy(other) {
    this._refObject = new MetricFamilyRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._descriptionIntialized = true;
    this._wrapperFields.description.copy(other._wrapperFields.description);
    this._wrapperFields.metrics.copy(other._wrapperFields.metrics);
  }
  static get classType() {
    return MetricFamilyWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"description","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"metrics","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"cartographer_ros_msgs","type":"Metric","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"MetricFamily","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MetricFamily"};
  }
  hasMember(name) {
    let memberNames = ["name","description","metrics"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MetricFamilyArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MetricFamilyWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MetricFamilyWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MetricFamilyArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MetricFamilyArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MetricFamilyRefArray(size);
    this._refObject = new MetricFamilyRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MetricFamilyWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MetricFamilyArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MetricFamilyArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MetricFamilyWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MetricFamilyWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MetricFamilyArrayWrapper;
  }
}
module.exports = MetricFamilyWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "description",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "metrics",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "cartographer_ros_msgs",
        "type": "Metric",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "MetricFamily",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MetricFamily"
}
*/
