// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__StartTrajectory_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const StartTrajectory_RequestRefStruct = StructType({
  configuration_directory: primitiveTypes.string,
  configuration_basename: primitiveTypes.string,
  use_initial_pose: primitiveTypes.bool,
  initial_pose: PoseWrapper.refObjectType,
  relative_to_trajectory_id: primitiveTypes.int32,
});
const StartTrajectory_RequestRefArray = ArrayType(StartTrajectory_RequestRefStruct);
const StartTrajectory_RequestRefStructArray = StructType({
  data: StartTrajectory_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StartTrajectory_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StartTrajectory_RequestRefStruct(other._refObject.toObject());
      this._configuration_directoryIntialized = true;
      this._wrapperFields.configuration_directory =  new StringWrapper(other._wrapperFields.configuration_directory);
      this._configuration_basenameIntialized = true;
      this._wrapperFields.configuration_basename =  new StringWrapper(other._wrapperFields.configuration_basename);
      this._use_initial_poseIntialized = true;
      this._wrapperFields.initial_pose =  new PoseWrapper(other._wrapperFields.initial_pose);
      this._relative_to_trajectory_idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StartTrajectory_RequestRefStruct();
    this._configuration_directoryIntialized = false;
    this._wrapperFields.configuration_directory = new StringWrapper();
    this._configuration_basenameIntialized = false;
    this._wrapperFields.configuration_basename = new StringWrapper();
    this._use_initial_poseIntialized = false;
    this._wrapperFields.initial_pose = new PoseWrapper();
    this._relative_to_trajectory_idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StartTrajectory_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StartTrajectory_RequestArrayWrapper;
  }
  static get ArrayType() {
    return StartTrajectory_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return StartTrajectory_RequestRefStructArray
  }
  static get refObjectType() {
    return StartTrajectory_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._configuration_directoryIntialized) {
        throw new TypeError('Invalid argument: configuration_directory in StartTrajectory_Request');
      }
      if (checkConsistency && !this._configuration_basenameIntialized) {
        throw new TypeError('Invalid argument: configuration_basename in StartTrajectory_Request');
      }
      if (checkConsistency && !this._use_initial_poseIntialized) {
        throw new TypeError('Invalid argument: use_initial_pose in StartTrajectory_Request');
      }
      if (checkConsistency && !this._relative_to_trajectory_idIntialized) {
        throw new TypeError('Invalid argument: relative_to_trajectory_id in StartTrajectory_Request');
      }
    if (own) {
      this._wrapperFields.configuration_directory.freeze(own, checkConsistency);
    }
    this._refObject.configuration_directory = this._wrapperFields.configuration_directory.refObject;
    if (own) {
      this._wrapperFields.configuration_basename.freeze(own, checkConsistency);
    }
    this._refObject.configuration_basename = this._wrapperFields.configuration_basename.refObject;
    this._wrapperFields.initial_pose.freeze(own, checkConsistency);
    this._refObject.initial_pose = this._wrapperFields.initial_pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._configuration_directoryIntialized = true;
    this._wrapperFields.configuration_directory.data = refObject.configuration_directory.data;
    this._configuration_basenameIntialized = true;
    this._wrapperFields.configuration_basename.data = refObject.configuration_basename.data;
    this._use_initial_poseIntialized = true;
    this._refObject.use_initial_pose = refObject.use_initial_pose;
    this._wrapperFields.initial_pose.copyRefObject(refObject.initial_pose);
    this._relative_to_trajectory_idIntialized = true;
    this._refObject.relative_to_trajectory_id = refObject.relative_to_trajectory_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.configuration_directory);
    StringWrapper.freeStruct(refObject.configuration_basename);
    PoseWrapper.freeStruct(refObject.initial_pose);
  }
  static destoryRawROS(msg) {
    StartTrajectory_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'StartTrajectory_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get configuration_directory() {
    if (!this._configuration_directoryIntialized) {
      return undefined;
    }
    return this._wrapperFields.configuration_directory.data;
  }
  set configuration_directory(value) {
    this._configuration_directoryIntialized = true;
    this._wrapperFields.configuration_directory.data = value;
  }
  get configuration_basename() {
    if (!this._configuration_basenameIntialized) {
      return undefined;
    }
    return this._wrapperFields.configuration_basename.data;
  }
  set configuration_basename(value) {
    this._configuration_basenameIntialized = true;
    this._wrapperFields.configuration_basename.data = value;
  }
  get use_initial_pose() {
    if (!this._use_initial_poseIntialized) {
      return undefined;
    }
    return this._refObject.use_initial_pose;
  }
  set use_initial_pose(value) {
    this._use_initial_poseIntialized = true;
    this._refObject.use_initial_pose = value;
  }
  get initial_pose() {
    return this._wrapperFields.initial_pose;
  }
  set initial_pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.initial_pose.copy(value);
    } else {
      this._wrapperFields.initial_pose.copy(new PoseWrapper(value));
    }
  }
  get relative_to_trajectory_id() {
    if (!this._relative_to_trajectory_idIntialized) {
      return undefined;
    }
    return this._refObject.relative_to_trajectory_id;
  }
  set relative_to_trajectory_id(value) {
    this._relative_to_trajectory_idIntialized = true;
    this._refObject.relative_to_trajectory_id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StartTrajectory_RequestRefStruct(refObject.toObject());
    this._configuration_directoryIntialized = true;
    this._wrapperFields.configuration_directory.copyRefObject(this._refObject.configuration_directory);
    this._configuration_basenameIntialized = true;
    this._wrapperFields.configuration_basename.copyRefObject(this._refObject.configuration_basename);
    this._use_initial_poseIntialized = true;
    this._wrapperFields.initial_pose.copyRefObject(this._refObject.initial_pose);
    this._relative_to_trajectory_idIntialized = true;
  }
  copy(other) {
    this._refObject = new StartTrajectory_RequestRefStruct(other._refObject.toObject());
    this._configuration_directoryIntialized = true;
    this._wrapperFields.configuration_directory.copy(other._wrapperFields.configuration_directory);
    this._configuration_basenameIntialized = true;
    this._wrapperFields.configuration_basename.copy(other._wrapperFields.configuration_basename);
    this._use_initial_poseIntialized = true;
    this._wrapperFields.initial_pose.copy(other._wrapperFields.initial_pose);
    this._relative_to_trajectory_idIntialized = true;
  }
  static get classType() {
    return StartTrajectory_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"configuration_directory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"configuration_basename","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"use_initial_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"initial_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"relative_to_trajectory_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"StartTrajectory_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StartTrajectory_Request"};
  }
  hasMember(name) {
    let memberNames = ["configuration_directory","configuration_basename","use_initial_pose","initial_pose","relative_to_trajectory_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StartTrajectory_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StartTrajectory_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StartTrajectory_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StartTrajectory_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StartTrajectory_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StartTrajectory_RequestRefArray(size);
    this._refObject = new StartTrajectory_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StartTrajectory_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StartTrajectory_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StartTrajectory_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StartTrajectory_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StartTrajectory_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StartTrajectory_RequestArrayWrapper;
  }
}
module.exports = StartTrajectory_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "configuration_directory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "configuration_basename",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "use_initial_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "initial_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "relative_to_trajectory_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "StartTrajectory_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StartTrajectory_Request"
}
*/
