// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__TrajectoryStates.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const TrajectoryStatesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  trajectory_id: Int32Wrapper.refObjectArrayType,
  trajectory_state: UInt8Wrapper.refObjectArrayType,
});
const TrajectoryStatesRefArray = ArrayType(TrajectoryStatesRefStruct);
const TrajectoryStatesRefStructArray = StructType({
  data: TrajectoryStatesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TrajectoryStatesWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TrajectoryStatesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.trajectory_id = Int32Wrapper.createArray();
      this._wrapperFields.trajectory_id.copy(other._wrapperFields.trajectory_id);
      this._wrapperFields.trajectory_state = UInt8Wrapper.createArray();
      this._wrapperFields.trajectory_state.copy(other._wrapperFields.trajectory_state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TrajectoryStatesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.trajectory_id = Int32Wrapper.createArray();
    this._wrapperFields.trajectory_state = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TrajectoryStatesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TrajectoryStatesArrayWrapper;
  }
  static get ArrayType() {
    return TrajectoryStatesArrayWrapper;
  }
  static get refObjectArrayType() {
    return TrajectoryStatesRefStructArray
  }
  static get refObjectType() {
    return TrajectoryStatesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.trajectory_id.freeze(own, checkConsistency);
    this._refObject.trajectory_id = this._wrapperFields.trajectory_id.refObject;
    if (own) {
      this._wrapperFields.trajectory_id.fill(Int32Array.from([]));
      this._wrapperFields.trajectory_id.freeze(own, checkConsistency);
      this._refObject.trajectory_id = this._wrapperFields.trajectory_id.refObject;
    }
    this._wrapperFields.trajectory_state.freeze(own, checkConsistency);
    this._refObject.trajectory_state = this._wrapperFields.trajectory_state.refObject;
    if (own) {
      this._wrapperFields.trajectory_state.fill(Uint8Array.from([]));
      this._wrapperFields.trajectory_state.freeze(own, checkConsistency);
      this._refObject.trajectory_state = this._wrapperFields.trajectory_state.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.trajectory_id.copyRefObject(refObject.trajectory_id);
    this._wrapperFields.trajectory_state.copyRefObject(refObject.trajectory_state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.trajectory_id.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.trajectory_id);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.trajectory_id, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.trajectory_state.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.trajectory_state);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.trajectory_state, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    TrajectoryStatesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'TrajectoryStates'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get trajectory_id() {
    return this._wrapperFields['trajectory_id'].data;
  }
  set trajectory_id(value) {
    this._wrapperFields['trajectory_id'].fill(value);
  }
  get trajectory_state() {
    return this._wrapperFields['trajectory_state'].data;
  }
  set trajectory_state(value) {
    this._wrapperFields['trajectory_state'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TrajectoryStatesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.trajectory_id.copyRefObject(this._refObject.trajectory_id);
    this._wrapperFields.trajectory_state.copyRefObject(this._refObject.trajectory_state);
  }
  copy(other) {
    this._refObject = new TrajectoryStatesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.trajectory_id.copy(other._wrapperFields.trajectory_id);
    this._wrapperFields.trajectory_state.copy(other._wrapperFields.trajectory_state);
  }
  static get classType() {
    return TrajectoryStatesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ACTIVE","value":0},{"type":"uint8","name":"FINISHED","value":1},{"type":"uint8","name":"FROZEN","value":2},{"type":"uint8","name":"DELETED","value":3}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"trajectory_id","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trajectory_state","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"TrajectoryStates","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TrajectoryStates"};
  }
  hasMember(name) {
    let memberNames = ["header","trajectory_id","trajectory_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TrajectoryStatesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TrajectoryStatesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TrajectoryStatesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryStatesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryStatesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TrajectoryStatesRefArray(size);
    this._refObject = new TrajectoryStatesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TrajectoryStatesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TrajectoryStatesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TrajectoryStatesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TrajectoryStatesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TrajectoryStatesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TrajectoryStatesArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(TrajectoryStatesWrapper, "ACTIVE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TrajectoryStatesWrapper, "FINISHED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TrajectoryStatesWrapper, "FROZEN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TrajectoryStatesWrapper, "DELETED", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = TrajectoryStatesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ACTIVE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "FINISHED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FROZEN",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "DELETED",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "trajectory_id",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trajectory_state",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "TrajectoryStates",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TrajectoryStates"
}
*/
