// rclnodejs/generated/vision_msgs/vision_msgs__msg__LabelInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const VisionClassWrapper = require('../../generated/vision_msgs/vision_msgs__msg__VisionClass.js');
const LabelInfoRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  class_map: VisionClassWrapper.refObjectArrayType,
  threshold: primitiveTypes.float32,
});
const LabelInfoRefArray = ArrayType(LabelInfoRefStruct);
const LabelInfoRefStructArray = StructType({
  data: LabelInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LabelInfoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LabelInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.class_map = VisionClassWrapper.createArray();
      this._wrapperFields.class_map.copy(other._wrapperFields.class_map);
      this._thresholdIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LabelInfoRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.class_map = VisionClassWrapper.createArray();
    this._thresholdIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LabelInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LabelInfoArrayWrapper;
  }
  static get ArrayType() {
    return LabelInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return LabelInfoRefStructArray
  }
  static get refObjectType() {
    return LabelInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._thresholdIntialized) {
        throw new TypeError('Invalid argument: threshold in LabelInfo');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.class_map.freeze(own, checkConsistency);
    this._refObject.class_map = this._wrapperFields.class_map.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.class_map.copyRefObject(refObject.class_map);
    this._thresholdIntialized = true;
    this._refObject.threshold = refObject.threshold;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.class_map.size != 0) {
      VisionClassWrapper.ArrayType.freeArray(refObject.class_map);
      if (VisionClassWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.class_map, VisionClassWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LabelInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'LabelInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get class_map() {
    return this._wrapperFields.class_map;
  }
  set class_map(value) {
    this._wrapperFields.class_map.fill(value);
  }
  get threshold() {
    if (!this._thresholdIntialized) {
      return undefined;
    }
    return this._refObject.threshold;
  }
  set threshold(value) {
    this._thresholdIntialized = true;
    this._refObject.threshold = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LabelInfoRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.class_map.copyRefObject(this._refObject.class_map);
    this._thresholdIntialized = true;
  }
  copy(other) {
    this._refObject = new LabelInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.class_map.copy(other._wrapperFields.class_map);
    this._thresholdIntialized = true;
  }
  static get classType() {
    return LabelInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"class_map","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"vision_msgs","type":"VisionClass","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"threshold","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"LabelInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LabelInfo"};
  }
  hasMember(name) {
    let memberNames = ["header","class_map","threshold"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LabelInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LabelInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LabelInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LabelInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LabelInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LabelInfoRefArray(size);
    this._refObject = new LabelInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LabelInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LabelInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LabelInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LabelInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LabelInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LabelInfoArrayWrapper;
  }
}
module.exports = LabelInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "class_map",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "vision_msgs",
        "type": "VisionClass",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "threshold",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "LabelInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LabelInfo"
}
*/
