// rclnodejs/generated/vision_msgs/vision_msgs__msg__ObjectHypothesis.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ObjectHypothesisRefStruct = StructType({
  class_id: primitiveTypes.string,
  score: primitiveTypes.float64,
});
const ObjectHypothesisRefArray = ArrayType(ObjectHypothesisRefStruct);
const ObjectHypothesisRefStructArray = StructType({
  data: ObjectHypothesisRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObjectHypothesisWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectHypothesisRefStruct(other._refObject.toObject());
      this._class_idIntialized = true;
      this._wrapperFields.class_id =  new StringWrapper(other._wrapperFields.class_id);
      this._scoreIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObjectHypothesisRefStruct();
    this._class_idIntialized = false;
    this._wrapperFields.class_id = new StringWrapper();
    this._scoreIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ObjectHypothesisWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectHypothesisArrayWrapper;
  }
  static get ArrayType() {
    return ObjectHypothesisArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectHypothesisRefStructArray
  }
  static get refObjectType() {
    return ObjectHypothesisRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._class_idIntialized) {
        throw new TypeError('Invalid argument: class_id in ObjectHypothesis');
      }
      if (checkConsistency && !this._scoreIntialized) {
        throw new TypeError('Invalid argument: score in ObjectHypothesis');
      }
    if (own) {
      this._wrapperFields.class_id.freeze(own, checkConsistency);
    }
    this._refObject.class_id = this._wrapperFields.class_id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._class_idIntialized = true;
    this._wrapperFields.class_id.data = refObject.class_id.data;
    this._scoreIntialized = true;
    this._refObject.score = refObject.score;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.class_id);
  }
  static destoryRawROS(msg) {
    ObjectHypothesisWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'ObjectHypothesis'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get class_id() {
    if (!this._class_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.class_id.data;
  }
  set class_id(value) {
    this._class_idIntialized = true;
    this._wrapperFields.class_id.data = value;
  }
  get score() {
    if (!this._scoreIntialized) {
      return undefined;
    }
    return this._refObject.score;
  }
  set score(value) {
    this._scoreIntialized = true;
    this._refObject.score = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectHypothesisRefStruct(refObject.toObject());
    this._class_idIntialized = true;
    this._wrapperFields.class_id.copyRefObject(this._refObject.class_id);
    this._scoreIntialized = true;
  }
  copy(other) {
    this._refObject = new ObjectHypothesisRefStruct(other._refObject.toObject());
    this._class_idIntialized = true;
    this._wrapperFields.class_id.copy(other._wrapperFields.class_id);
    this._scoreIntialized = true;
  }
  static get classType() {
    return ObjectHypothesisWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"class_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"score","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"ObjectHypothesis","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectHypothesis"};
  }
  hasMember(name) {
    let memberNames = ["class_id","score"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObjectHypothesisArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectHypothesisWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectHypothesisWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectHypothesisArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectHypothesisArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObjectHypothesisRefArray(size);
    this._refObject = new ObjectHypothesisRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectHypothesisWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectHypothesisArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectHypothesisArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectHypothesisWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectHypothesisWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectHypothesisArrayWrapper;
  }
}
module.exports = ObjectHypothesisWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "class_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "score",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "ObjectHypothesis",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObjectHypothesis"
}
*/
