// rclnodejs/generated/vision_msgs/vision_msgs__msg__VisionInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VisionInfoRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  method: primitiveTypes.string,
  database_location: primitiveTypes.string,
  database_version: primitiveTypes.int32,
});
const VisionInfoRefArray = ArrayType(VisionInfoRefStruct);
const VisionInfoRefStructArray = StructType({
  data: VisionInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VisionInfoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VisionInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._methodIntialized = true;
      this._wrapperFields.method =  new StringWrapper(other._wrapperFields.method);
      this._database_locationIntialized = true;
      this._wrapperFields.database_location =  new StringWrapper(other._wrapperFields.database_location);
      this._database_versionIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VisionInfoRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._methodIntialized = false;
    this._wrapperFields.method = new StringWrapper();
    this._database_locationIntialized = false;
    this._wrapperFields.database_location = new StringWrapper();
    this._database_versionIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VisionInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VisionInfoArrayWrapper;
  }
  static get ArrayType() {
    return VisionInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return VisionInfoRefStructArray
  }
  static get refObjectType() {
    return VisionInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._methodIntialized) {
        throw new TypeError('Invalid argument: method in VisionInfo');
      }
      if (checkConsistency && !this._database_locationIntialized) {
        throw new TypeError('Invalid argument: database_location in VisionInfo');
      }
      if (checkConsistency && !this._database_versionIntialized) {
        throw new TypeError('Invalid argument: database_version in VisionInfo');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.method.freeze(own, checkConsistency);
    }
    this._refObject.method = this._wrapperFields.method.refObject;
    if (own) {
      this._wrapperFields.database_location.freeze(own, checkConsistency);
    }
    this._refObject.database_location = this._wrapperFields.database_location.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._methodIntialized = true;
    this._wrapperFields.method.data = refObject.method.data;
    this._database_locationIntialized = true;
    this._wrapperFields.database_location.data = refObject.database_location.data;
    this._database_versionIntialized = true;
    this._refObject.database_version = refObject.database_version;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.method);
    StringWrapper.freeStruct(refObject.database_location);
  }
  static destoryRawROS(msg) {
    VisionInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'vision_msgs', subFolder: 'msg', interfaceName: 'VisionInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get method() {
    if (!this._methodIntialized) {
      return undefined;
    }
    return this._wrapperFields.method.data;
  }
  set method(value) {
    this._methodIntialized = true;
    this._wrapperFields.method.data = value;
  }
  get database_location() {
    if (!this._database_locationIntialized) {
      return undefined;
    }
    return this._wrapperFields.database_location.data;
  }
  set database_location(value) {
    this._database_locationIntialized = true;
    this._wrapperFields.database_location.data = value;
  }
  get database_version() {
    if (!this._database_versionIntialized) {
      return undefined;
    }
    return this._refObject.database_version;
  }
  set database_version(value) {
    this._database_versionIntialized = true;
    this._refObject.database_version = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VisionInfoRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._methodIntialized = true;
    this._wrapperFields.method.copyRefObject(this._refObject.method);
    this._database_locationIntialized = true;
    this._wrapperFields.database_location.copyRefObject(this._refObject.database_location);
    this._database_versionIntialized = true;
  }
  copy(other) {
    this._refObject = new VisionInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._methodIntialized = true;
    this._wrapperFields.method.copy(other._wrapperFields.method);
    this._database_locationIntialized = true;
    this._wrapperFields.database_location.copy(other._wrapperFields.database_location);
    this._database_versionIntialized = true;
  }
  static get classType() {
    return VisionInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"method","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"database_location","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"database_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"vision_msgs","type":"VisionInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VisionInfo"};
  }
  hasMember(name) {
    let memberNames = ["header","method","database_location","database_version"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VisionInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VisionInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VisionInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VisionInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VisionInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VisionInfoRefArray(size);
    this._refObject = new VisionInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VisionInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VisionInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VisionInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VisionInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VisionInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VisionInfoArrayWrapper;
  }
}
module.exports = VisionInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "method",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "database_location",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "database_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "vision_msgs",
    "type": "VisionInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VisionInfo"
}
*/
