// rclnodejs/generated/mocap4r2_control_msgs/mocap4r2_control_msgs__msg__MocapInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MocapInfoRefStruct = StructType({
  mocap4r2_source: primitiveTypes.string,
  ros_version_source: primitiveTypes.int8,
  topics: StringWrapper.refObjectArrayType,
});
const MocapInfoRefArray = ArrayType(MocapInfoRefStruct);
const MocapInfoRefStructArray = StructType({
  data: MocapInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MocapInfoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._topicsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MocapInfoRefStruct(other._refObject.toObject());
      this._mocap4r2_sourceIntialized = true;
      this._wrapperFields.mocap4r2_source =  new StringWrapper(other._wrapperFields.mocap4r2_source);
      this._ros_version_sourceIntialized = true;
      this._wrapperFields.topics = StringWrapper.createArray();
      this._wrapperFields.topics.copy(other._wrapperFields.topics);
      this.topics = other.topics;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MocapInfoRefStruct();
    this._mocap4r2_sourceIntialized = false;
    this._wrapperFields.mocap4r2_source = new StringWrapper();
    this._ros_version_sourceIntialized = false;
    this._wrapperFields.topics = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MocapInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MocapInfoArrayWrapper;
  }
  static get ArrayType() {
    return MocapInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return MocapInfoRefStructArray
  }
  static get refObjectType() {
    return MocapInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._mocap4r2_sourceIntialized) {
        throw new TypeError('Invalid argument: mocap4r2_source in MocapInfo');
      }
      if (checkConsistency && !this._ros_version_sourceIntialized) {
        throw new TypeError('Invalid argument: ros_version_source in MocapInfo');
      }
    if (own) {
      this._wrapperFields.mocap4r2_source.freeze(own, checkConsistency);
    }
    this._refObject.mocap4r2_source = this._wrapperFields.mocap4r2_source.refObject;
    if (!own) {
      this._wrapperFields.topics.fill(this._topicsArray);
      this._wrapperFields.topics.freeze(own, checkConsistency);
      this._refObject.topics = this._wrapperFields.topics.refObject;
    } else {
      this._wrapperFields.topics.fill([]);
      this._wrapperFields.topics.freeze(own, checkConsistency);
      this._refObject.topics = this._wrapperFields.topics.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._mocap4r2_sourceIntialized = true;
    this._wrapperFields.mocap4r2_source.data = refObject.mocap4r2_source.data;
    this._ros_version_sourceIntialized = true;
    this._refObject.ros_version_source = refObject.ros_version_source;
    refObject.topics.data.length = refObject.topics.size;
    for (let index = 0; index < refObject.topics.size; index++) {
      this._topicsArray[index] = refObject.topics.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.mocap4r2_source);
    if (refObject.topics.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.topics);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.topics, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MocapInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'mocap4r2_control_msgs', subFolder: 'msg', interfaceName: 'MocapInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mocap4r2_source() {
    if (!this._mocap4r2_sourceIntialized) {
      return undefined;
    }
    return this._wrapperFields.mocap4r2_source.data;
  }
  set mocap4r2_source(value) {
    this._mocap4r2_sourceIntialized = true;
    this._wrapperFields.mocap4r2_source.data = value;
  }
  get ros_version_source() {
    if (!this._ros_version_sourceIntialized) {
      return undefined;
    }
    return this._refObject.ros_version_source;
  }
  set ros_version_source(value) {
    this._ros_version_sourceIntialized = true;
    this._refObject.ros_version_source = value;
  }
  get topics() {
    return this._topicsArray;
  }
  set topics(value) {
    this._topicsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MocapInfoRefStruct(refObject.toObject());
    this._mocap4r2_sourceIntialized = true;
    this._wrapperFields.mocap4r2_source.copyRefObject(this._refObject.mocap4r2_source);
    this._ros_version_sourceIntialized = true;
    refObject.topics.data.length = refObject.topics.size;
    for (let index = 0; index < refObject.topics.size; index++) {
      this._topicsArray[index] = refObject.topics.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new MocapInfoRefStruct(other._refObject.toObject());
    this._mocap4r2_sourceIntialized = true;
    this._wrapperFields.mocap4r2_source.copy(other._wrapperFields.mocap4r2_source);
    this._ros_version_sourceIntialized = true;
    this._topicsArray = other._topicsArray.slice();
  }
  static get classType() {
    return MocapInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"ROS2","value":0},{"type":"int8","name":"ROS1","value":1}],"fields":[{"name":"mocap4r2_source","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ros_version_source","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"topics","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"mocap4r2_control_msgs","type":"MocapInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MocapInfo"};
  }
  hasMember(name) {
    let memberNames = ["mocap4r2_source","ros_version_source","topics"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MocapInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MocapInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MocapInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MocapInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MocapInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MocapInfoRefArray(size);
    this._refObject = new MocapInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MocapInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MocapInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MocapInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MocapInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MocapInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MocapInfoArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(MocapInfoWrapper, "ROS2", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MocapInfoWrapper, "ROS1", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = MocapInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int8",
      "name": "ROS2",
      "value": 0
    },
    {
      "type": "int8",
      "name": "ROS1",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "mocap4r2_source",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ros_version_source",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "topics",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "mocap4r2_control_msgs",
    "type": "MocapInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MocapInfo"
}
*/
