// rclnodejs/generated/graph_msgs/graph_msgs__msg__Edges.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const EdgesRefStruct = StructType({
  node_ids: UInt32Wrapper.refObjectArrayType,
  weights: Float64Wrapper.refObjectArrayType,
});
const EdgesRefArray = ArrayType(EdgesRefStruct);
const EdgesRefStructArray = StructType({
  data: EdgesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EdgesWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EdgesRefStruct(other._refObject.toObject());
      this._wrapperFields.node_ids = UInt32Wrapper.createArray();
      this._wrapperFields.node_ids.copy(other._wrapperFields.node_ids);
      this._wrapperFields.weights = Float64Wrapper.createArray();
      this._wrapperFields.weights.copy(other._wrapperFields.weights);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EdgesRefStruct();
    this._wrapperFields.node_ids = UInt32Wrapper.createArray();
    this._wrapperFields.weights = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new EdgesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EdgesArrayWrapper;
  }
  static get ArrayType() {
    return EdgesArrayWrapper;
  }
  static get refObjectArrayType() {
    return EdgesRefStructArray
  }
  static get refObjectType() {
    return EdgesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.node_ids.freeze(own, checkConsistency);
    this._refObject.node_ids = this._wrapperFields.node_ids.refObject;
    if (own) {
      this._wrapperFields.node_ids.fill(Uint32Array.from([]));
      this._wrapperFields.node_ids.freeze(own, checkConsistency);
      this._refObject.node_ids = this._wrapperFields.node_ids.refObject;
    }
    this._wrapperFields.weights.freeze(own, checkConsistency);
    this._refObject.weights = this._wrapperFields.weights.refObject;
    if (own) {
      this._wrapperFields.weights.fill(Float64Array.from([]));
      this._wrapperFields.weights.freeze(own, checkConsistency);
      this._refObject.weights = this._wrapperFields.weights.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.node_ids.copyRefObject(refObject.node_ids);
    this._wrapperFields.weights.copyRefObject(refObject.weights);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.node_ids.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.node_ids);
      if (UInt32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.node_ids, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.weights.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.weights);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.weights, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    EdgesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'graph_msgs', subFolder: 'msg', interfaceName: 'Edges'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node_ids() {
    return this._wrapperFields['node_ids'].data;
  }
  set node_ids(value) {
    this._wrapperFields['node_ids'].fill(value);
  }
  get weights() {
    return this._wrapperFields['weights'].data;
  }
  set weights(value) {
    this._wrapperFields['weights'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new EdgesRefStruct(refObject.toObject());
    this._wrapperFields.node_ids.copyRefObject(this._refObject.node_ids);
    this._wrapperFields.weights.copyRefObject(this._refObject.weights);
  }
  copy(other) {
    this._refObject = new EdgesRefStruct(other._refObject.toObject());
    this._wrapperFields.node_ids.copy(other._wrapperFields.node_ids);
    this._wrapperFields.weights.copy(other._wrapperFields.weights);
  }
  static get classType() {
    return EdgesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node_ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"weights","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"graph_msgs","type":"Edges","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Edges"};
  }
  hasMember(name) {
    let memberNames = ["node_ids","weights"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EdgesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EdgesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EdgesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EdgesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EdgesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EdgesRefArray(size);
    this._refObject = new EdgesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EdgesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EdgesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EdgesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EdgesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EdgesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EdgesArrayWrapper;
  }
}
module.exports = EdgesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "node_ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "weights",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "graph_msgs",
    "type": "Edges",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Edges"
}
*/
