// rclnodejs/generated/graph_msgs/graph_msgs__msg__GeometryGraph.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const EdgesWrapper = require('../../generated/graph_msgs/graph_msgs__msg__Edges.js');
const GeometryGraphRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  nodes: PointWrapper.refObjectArrayType,
  edges: EdgesWrapper.refObjectArrayType,
});
const GeometryGraphRefArray = ArrayType(GeometryGraphRefStruct);
const GeometryGraphRefStructArray = StructType({
  data: GeometryGraphRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GeometryGraphWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GeometryGraphRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.nodes = PointWrapper.createArray();
      this._wrapperFields.nodes.copy(other._wrapperFields.nodes);
      this._wrapperFields.edges = EdgesWrapper.createArray();
      this._wrapperFields.edges.copy(other._wrapperFields.edges);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GeometryGraphRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.nodes = PointWrapper.createArray();
    this._wrapperFields.edges = EdgesWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GeometryGraphWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GeometryGraphArrayWrapper;
  }
  static get ArrayType() {
    return GeometryGraphArrayWrapper;
  }
  static get refObjectArrayType() {
    return GeometryGraphRefStructArray
  }
  static get refObjectType() {
    return GeometryGraphRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.nodes.freeze(own, checkConsistency);
    this._refObject.nodes = this._wrapperFields.nodes.refObject;
    this._wrapperFields.edges.freeze(own, checkConsistency);
    this._refObject.edges = this._wrapperFields.edges.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.nodes.copyRefObject(refObject.nodes);
    this._wrapperFields.edges.copyRefObject(refObject.edges);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.nodes.size != 0) {
      PointWrapper.ArrayType.freeArray(refObject.nodes);
      if (PointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.nodes, PointWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.edges.size != 0) {
      EdgesWrapper.ArrayType.freeArray(refObject.edges);
      if (EdgesWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.edges, EdgesWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GeometryGraphWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'graph_msgs', subFolder: 'msg', interfaceName: 'GeometryGraph'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get nodes() {
    return this._wrapperFields.nodes;
  }
  set nodes(value) {
    this._wrapperFields.nodes.fill(value);
  }
  get edges() {
    return this._wrapperFields.edges;
  }
  set edges(value) {
    this._wrapperFields.edges.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GeometryGraphRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.nodes.copyRefObject(this._refObject.nodes);
    this._wrapperFields.edges.copyRefObject(this._refObject.edges);
  }
  copy(other) {
    this._refObject = new GeometryGraphRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.nodes.copy(other._wrapperFields.nodes);
    this._wrapperFields.edges.copy(other._wrapperFields.edges);
  }
  static get classType() {
    return GeometryGraphWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"nodes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"edges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"graph_msgs","type":"Edges","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"graph_msgs","type":"GeometryGraph","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GeometryGraph"};
  }
  hasMember(name) {
    let memberNames = ["header","nodes","edges"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GeometryGraphArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GeometryGraphWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GeometryGraphWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeometryGraphArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeometryGraphArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GeometryGraphRefArray(size);
    this._refObject = new GeometryGraphRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GeometryGraphWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GeometryGraphArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GeometryGraphArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GeometryGraphWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GeometryGraphWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GeometryGraphArrayWrapper;
  }
}
module.exports = GeometryGraphWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "nodes",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "edges",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "graph_msgs",
        "type": "Edges",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "graph_msgs",
    "type": "GeometryGraph",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GeometryGraph"
}
*/
