// rclnodejs/generated/actuator_msgs/actuator_msgs__msg__ActuatorsAngularPosition.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const ActuatorsAngularPositionRefStruct = StructType({
  position: Float64Wrapper.refObjectArrayType,
});
const ActuatorsAngularPositionRefArray = ArrayType(ActuatorsAngularPositionRefStruct);
const ActuatorsAngularPositionRefStructArray = StructType({
  data: ActuatorsAngularPositionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorsAngularPositionWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorsAngularPositionRefStruct(other._refObject.toObject());
      this._wrapperFields.position = Float64Wrapper.createArray();
      this._wrapperFields.position.copy(other._wrapperFields.position);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorsAngularPositionRefStruct();
    this._wrapperFields.position = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorsAngularPositionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorsAngularPositionArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorsAngularPositionArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorsAngularPositionRefStructArray
  }
  static get refObjectType() {
    return ActuatorsAngularPositionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.position.freeze(own, checkConsistency);
    this._refObject.position = this._wrapperFields.position.refObject;
    if (own) {
      this._wrapperFields.position.fill(Float64Array.from([]));
      this._wrapperFields.position.freeze(own, checkConsistency);
      this._refObject.position = this._wrapperFields.position.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.position.copyRefObject(refObject.position);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.position.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.position);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.position, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ActuatorsAngularPositionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'actuator_msgs', subFolder: 'msg', interfaceName: 'ActuatorsAngularPosition'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get position() {
    return this._wrapperFields['position'].data;
  }
  set position(value) {
    this._wrapperFields['position'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorsAngularPositionRefStruct(refObject.toObject());
    this._wrapperFields.position.copyRefObject(this._refObject.position);
  }
  copy(other) {
    this._refObject = new ActuatorsAngularPositionRefStruct(other._refObject.toObject());
    this._wrapperFields.position.copy(other._wrapperFields.position);
  }
  static get classType() {
    return ActuatorsAngularPositionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"position","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"actuator_msgs","type":"ActuatorsAngularPosition","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorsAngularPosition"};
  }
  hasMember(name) {
    let memberNames = ["position"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorsAngularPositionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorsAngularPositionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorsAngularPositionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorsAngularPositionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorsAngularPositionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorsAngularPositionRefArray(size);
    this._refObject = new ActuatorsAngularPositionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorsAngularPositionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorsAngularPositionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorsAngularPositionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorsAngularPositionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorsAngularPositionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorsAngularPositionArrayWrapper;
  }
}
module.exports = ActuatorsAngularPositionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "position",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "actuator_msgs",
    "type": "ActuatorsAngularPosition",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorsAngularPosition"
}
*/
