// rclnodejs/generated/actuator_msgs/actuator_msgs__msg__ActuatorsVelocity.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const ActuatorsAngularVelocityWrapper = require('../../generated/actuator_msgs/actuator_msgs__msg__ActuatorsAngularVelocity.js');
const ActuatorsLinearVelocityWrapper = require('../../generated/actuator_msgs/actuator_msgs__msg__ActuatorsLinearVelocity.js');
const ActuatorsVelocityRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  angular: ActuatorsAngularVelocityWrapper.refObjectType,
  linear: ActuatorsLinearVelocityWrapper.refObjectType,
});
const ActuatorsVelocityRefArray = ArrayType(ActuatorsVelocityRefStruct);
const ActuatorsVelocityRefStructArray = StructType({
  data: ActuatorsVelocityRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorsVelocityWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorsVelocityRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.angular =  new ActuatorsAngularVelocityWrapper(other._wrapperFields.angular);
      this._wrapperFields.linear =  new ActuatorsLinearVelocityWrapper(other._wrapperFields.linear);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorsVelocityRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.angular = new ActuatorsAngularVelocityWrapper();
    this._wrapperFields.linear = new ActuatorsLinearVelocityWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorsVelocityWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorsVelocityArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorsVelocityArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorsVelocityRefStructArray
  }
  static get refObjectType() {
    return ActuatorsVelocityRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.angular.freeze(own, checkConsistency);
    this._refObject.angular = this._wrapperFields.angular.refObject;
    this._wrapperFields.linear.freeze(own, checkConsistency);
    this._refObject.linear = this._wrapperFields.linear.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.angular.copyRefObject(refObject.angular);
    this._wrapperFields.linear.copyRefObject(refObject.linear);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    ActuatorsAngularVelocityWrapper.freeStruct(refObject.angular);
    ActuatorsLinearVelocityWrapper.freeStruct(refObject.linear);
  }
  static destoryRawROS(msg) {
    ActuatorsVelocityWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'actuator_msgs', subFolder: 'msg', interfaceName: 'ActuatorsVelocity'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get angular() {
    return this._wrapperFields.angular;
  }
  set angular(value) {
    if (value instanceof ActuatorsAngularVelocityWrapper) {
      this._wrapperFields.angular.copy(value);
    } else {
      this._wrapperFields.angular.copy(new ActuatorsAngularVelocityWrapper(value));
    }
  }
  get linear() {
    return this._wrapperFields.linear;
  }
  set linear(value) {
    if (value instanceof ActuatorsLinearVelocityWrapper) {
      this._wrapperFields.linear.copy(value);
    } else {
      this._wrapperFields.linear.copy(new ActuatorsLinearVelocityWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorsVelocityRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.angular.copyRefObject(this._refObject.angular);
    this._wrapperFields.linear.copyRefObject(this._refObject.linear);
  }
  copy(other) {
    this._refObject = new ActuatorsVelocityRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.angular.copy(other._wrapperFields.angular);
    this._wrapperFields.linear.copy(other._wrapperFields.linear);
  }
  static get classType() {
    return ActuatorsVelocityWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"angular","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"actuator_msgs","type":"ActuatorsAngularVelocity","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"linear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"actuator_msgs","type":"ActuatorsLinearVelocity","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"actuator_msgs","type":"ActuatorsVelocity","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorsVelocity"};
  }
  hasMember(name) {
    let memberNames = ["header","angular","linear"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorsVelocityArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorsVelocityWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorsVelocityWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorsVelocityArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorsVelocityArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorsVelocityRefArray(size);
    this._refObject = new ActuatorsVelocityRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorsVelocityWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorsVelocityArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorsVelocityArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorsVelocityWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorsVelocityWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorsVelocityArrayWrapper;
  }
}
module.exports = ActuatorsVelocityWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "angular",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "actuator_msgs",
        "type": "ActuatorsAngularVelocity",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "linear",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "actuator_msgs",
        "type": "ActuatorsLinearVelocity",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "actuator_msgs",
    "type": "ActuatorsVelocity",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorsVelocity"
}
*/
