// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__action__ObjectRecognition.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class object_recognition_msgs__action__ObjectRecognition {
  static get Goal() {
    return require('./object_recognition_msgs__action__ObjectRecognition_Goal.js');
  }
  static get Result() {
    return require('./object_recognition_msgs__action__ObjectRecognition_Result.js');
  }
  static get Feedback() {
    return require('./object_recognition_msgs__action__ObjectRecognition_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./object_recognition_msgs__action__ObjectRecognition_SendGoal.js');
      },
      get GetResultService() {
        return require('./object_recognition_msgs__action__ObjectRecognition_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./object_recognition_msgs__action__ObjectRecognition_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'action', interfaceName: 'ObjectRecognition'};
  }
}
module.exports = object_recognition_msgs__action__ObjectRecognition;
