// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__action__ObjectRecognition_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const ObjectRecognition_GoalRefStruct = StructType({
  use_roi: primitiveTypes.bool,
  filter_limits: Float32Wrapper.refObjectArrayType,
});
const ObjectRecognition_GoalRefArray = ArrayType(ObjectRecognition_GoalRefStruct);
const ObjectRecognition_GoalRefStructArray = StructType({
  data: ObjectRecognition_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObjectRecognition_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectRecognition_GoalRefStruct(other._refObject.toObject());
      this._use_roiIntialized = true;
      this._wrapperFields.filter_limits = Float32Wrapper.createArray();
      this._wrapperFields.filter_limits.copy(other._wrapperFields.filter_limits);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObjectRecognition_GoalRefStruct();
    this._use_roiIntialized = false;
    this._wrapperFields.filter_limits = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectRecognition_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectRecognition_GoalArrayWrapper;
  }
  static get ArrayType() {
    return ObjectRecognition_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectRecognition_GoalRefStructArray
  }
  static get refObjectType() {
    return ObjectRecognition_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._use_roiIntialized) {
        throw new TypeError('Invalid argument: use_roi in ObjectRecognition_Goal');
      }
    this._wrapperFields.filter_limits.freeze(own, checkConsistency);
    this._refObject.filter_limits = this._wrapperFields.filter_limits.refObject;
    if (own) {
      this._wrapperFields.filter_limits.fill(Float32Array.from([]));
      this._wrapperFields.filter_limits.freeze(own, checkConsistency);
      this._refObject.filter_limits = this._wrapperFields.filter_limits.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._use_roiIntialized = true;
    this._refObject.use_roi = refObject.use_roi;
    this._wrapperFields.filter_limits.copyRefObject(refObject.filter_limits);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.filter_limits.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.filter_limits);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.filter_limits, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ObjectRecognition_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'action', interfaceName: 'ObjectRecognition_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get use_roi() {
    if (!this._use_roiIntialized) {
      return undefined;
    }
    return this._refObject.use_roi;
  }
  set use_roi(value) {
    this._use_roiIntialized = true;
    this._refObject.use_roi = value;
  }
  get filter_limits() {
    return this._wrapperFields['filter_limits'].data;
  }
  set filter_limits(value) {
    this._wrapperFields['filter_limits'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectRecognition_GoalRefStruct(refObject.toObject());
    this._use_roiIntialized = true;
    this._wrapperFields.filter_limits.copyRefObject(this._refObject.filter_limits);
  }
  copy(other) {
    this._refObject = new ObjectRecognition_GoalRefStruct(other._refObject.toObject());
    this._use_roiIntialized = true;
    this._wrapperFields.filter_limits.copy(other._wrapperFields.filter_limits);
  }
  static get classType() {
    return ObjectRecognition_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"use_roi","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"filter_limits","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"object_recognition_msgs","type":"ObjectRecognition_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectRecognition_Goal"};
  }
  hasMember(name) {
    let memberNames = ["use_roi","filter_limits"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObjectRecognition_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectRecognition_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectRecognition_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectRecognition_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectRecognition_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObjectRecognition_GoalRefArray(size);
    this._refObject = new ObjectRecognition_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectRecognition_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectRecognition_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectRecognition_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectRecognition_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectRecognition_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectRecognition_GoalArrayWrapper;
  }
}
module.exports = ObjectRecognition_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "use_roi",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "filter_limits",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "object_recognition_msgs",
    "type": "ObjectRecognition_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObjectRecognition_Goal"
}
*/
