// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__srv__GetObjectInformation_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ObjectInformationWrapper = require('../../generated/object_recognition_msgs/object_recognition_msgs__msg__ObjectInformation.js');
const GetObjectInformation_ResponseRefStruct = StructType({
  information: ObjectInformationWrapper.refObjectType,
});
const GetObjectInformation_ResponseRefArray = ArrayType(GetObjectInformation_ResponseRefStruct);
const GetObjectInformation_ResponseRefStructArray = StructType({
  data: GetObjectInformation_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetObjectInformation_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetObjectInformation_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.information =  new ObjectInformationWrapper(other._wrapperFields.information);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetObjectInformation_ResponseRefStruct();
    this._wrapperFields.information = new ObjectInformationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetObjectInformation_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetObjectInformation_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetObjectInformation_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetObjectInformation_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetObjectInformation_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.information.freeze(own, checkConsistency);
    this._refObject.information = this._wrapperFields.information.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.information.copyRefObject(refObject.information);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    ObjectInformationWrapper.freeStruct(refObject.information);
  }
  static destoryRawROS(msg) {
    GetObjectInformation_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'srv', interfaceName: 'GetObjectInformation_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get information() {
    return this._wrapperFields.information;
  }
  set information(value) {
    if (value instanceof ObjectInformationWrapper) {
      this._wrapperFields.information.copy(value);
    } else {
      this._wrapperFields.information.copy(new ObjectInformationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetObjectInformation_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.information.copyRefObject(this._refObject.information);
  }
  copy(other) {
    this._refObject = new GetObjectInformation_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.information.copy(other._wrapperFields.information);
  }
  static get classType() {
    return GetObjectInformation_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"information","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"object_recognition_msgs","type":"ObjectInformation","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"object_recognition_msgs","type":"GetObjectInformation_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetObjectInformation_Response"};
  }
  hasMember(name) {
    let memberNames = ["information"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetObjectInformation_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetObjectInformation_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetObjectInformation_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectInformation_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectInformation_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetObjectInformation_ResponseRefArray(size);
    this._refObject = new GetObjectInformation_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetObjectInformation_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetObjectInformation_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetObjectInformation_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetObjectInformation_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetObjectInformation_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetObjectInformation_ResponseArrayWrapper;
  }
}
module.exports = GetObjectInformation_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "information",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "object_recognition_msgs",
        "type": "ObjectInformation",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "object_recognition_msgs",
    "type": "GetObjectInformation_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetObjectInformation_Response"
}
*/
