// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__srv__SetHardwareComponentState_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StateWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__State.js');
const SetHardwareComponentState_ResponseRefStruct = StructType({
  ok: primitiveTypes.bool,
  state: StateWrapper.refObjectType,
});
const SetHardwareComponentState_ResponseRefArray = ArrayType(SetHardwareComponentState_ResponseRefStruct);
const SetHardwareComponentState_ResponseRefStructArray = StructType({
  data: SetHardwareComponentState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetHardwareComponentState_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetHardwareComponentState_ResponseRefStruct(other._refObject.toObject());
      this._okIntialized = true;
      this._wrapperFields.state =  new StateWrapper(other._wrapperFields.state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetHardwareComponentState_ResponseRefStruct();
    this._okIntialized = false;
    this._wrapperFields.state = new StateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetHardwareComponentState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetHardwareComponentState_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return SetHardwareComponentState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetHardwareComponentState_ResponseRefStructArray
  }
  static get refObjectType() {
    return SetHardwareComponentState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._okIntialized) {
        throw new TypeError('Invalid argument: ok in SetHardwareComponentState_Response');
      }
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._okIntialized = true;
    this._refObject.ok = refObject.ok;
    this._wrapperFields.state.copyRefObject(refObject.state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StateWrapper.freeStruct(refObject.state);
  }
  static destoryRawROS(msg) {
    SetHardwareComponentState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'srv', interfaceName: 'SetHardwareComponentState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ok() {
    if (!this._okIntialized) {
      return undefined;
    }
    return this._refObject.ok;
  }
  set ok(value) {
    this._okIntialized = true;
    this._refObject.ok = value;
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof StateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new StateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetHardwareComponentState_ResponseRefStruct(refObject.toObject());
    this._okIntialized = true;
    this._wrapperFields.state.copyRefObject(this._refObject.state);
  }
  copy(other) {
    this._refObject = new SetHardwareComponentState_ResponseRefStruct(other._refObject.toObject());
    this._okIntialized = true;
    this._wrapperFields.state.copy(other._wrapperFields.state);
  }
  static get classType() {
    return SetHardwareComponentState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ok","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"SetHardwareComponentState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetHardwareComponentState_Response"};
  }
  hasMember(name) {
    let memberNames = ["ok","state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetHardwareComponentState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetHardwareComponentState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetHardwareComponentState_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetHardwareComponentState_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetHardwareComponentState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetHardwareComponentState_ResponseRefArray(size);
    this._refObject = new SetHardwareComponentState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetHardwareComponentState_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetHardwareComponentState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetHardwareComponentState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetHardwareComponentState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetHardwareComponentState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetHardwareComponentState_ResponseArrayWrapper;
  }
}
module.exports = SetHardwareComponentState_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ok",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "lifecycle_msgs",
        "type": "State",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "controller_manager_msgs",
    "type": "SetHardwareComponentState_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetHardwareComponentState_Response"
}
*/
