// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__msg__HardwareComponentState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StateWrapper = require('../../generated/lifecycle_msgs/lifecycle_msgs__msg__State.js');
const HardwareInterfaceWrapper = require('../../generated/controller_manager_msgs/controller_manager_msgs__msg__HardwareInterface.js');
const HardwareComponentStateRefStruct = StructType({
  name: primitiveTypes.string,
  type: primitiveTypes.string,
  class_type: primitiveTypes.string,
  state: StateWrapper.refObjectType,
  command_interfaces: HardwareInterfaceWrapper.refObjectArrayType,
  state_interfaces: HardwareInterfaceWrapper.refObjectArrayType,
});
const HardwareComponentStateRefArray = ArrayType(HardwareComponentStateRefStruct);
const HardwareComponentStateRefStructArray = StructType({
  data: HardwareComponentStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HardwareComponentStateWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HardwareComponentStateRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._typeIntialized = true;
      this._wrapperFields.type =  new StringWrapper(other._wrapperFields.type);
      this._class_typeIntialized = true;
      this._wrapperFields.class_type =  new StringWrapper(other._wrapperFields.class_type);
      this._wrapperFields.state =  new StateWrapper(other._wrapperFields.state);
      this._wrapperFields.command_interfaces = HardwareInterfaceWrapper.createArray();
      this._wrapperFields.command_interfaces.copy(other._wrapperFields.command_interfaces);
      this._wrapperFields.state_interfaces = HardwareInterfaceWrapper.createArray();
      this._wrapperFields.state_interfaces.copy(other._wrapperFields.state_interfaces);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HardwareComponentStateRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._typeIntialized = false;
    this._wrapperFields.type = new StringWrapper();
    this._class_typeIntialized = false;
    this._wrapperFields.class_type = new StringWrapper();
    this._wrapperFields.state = new StateWrapper();
    this._wrapperFields.command_interfaces = HardwareInterfaceWrapper.createArray();
    this._wrapperFields.state_interfaces = HardwareInterfaceWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new HardwareComponentStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HardwareComponentStateArrayWrapper;
  }
  static get ArrayType() {
    return HardwareComponentStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return HardwareComponentStateRefStructArray
  }
  static get refObjectType() {
    return HardwareComponentStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in HardwareComponentState');
      }
      if (checkConsistency && !this._typeIntialized) {
        throw new TypeError('Invalid argument: type in HardwareComponentState');
      }
      if (checkConsistency && !this._class_typeIntialized) {
        throw new TypeError('Invalid argument: class_type in HardwareComponentState');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.type.freeze(own, checkConsistency);
    }
    this._refObject.type = this._wrapperFields.type.refObject;
    if (own) {
      this._wrapperFields.class_type.freeze(own, checkConsistency);
    }
    this._refObject.class_type = this._wrapperFields.class_type.refObject;
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
    this._wrapperFields.command_interfaces.freeze(own, checkConsistency);
    this._refObject.command_interfaces = this._wrapperFields.command_interfaces.refObject;
    this._wrapperFields.state_interfaces.freeze(own, checkConsistency);
    this._refObject.state_interfaces = this._wrapperFields.state_interfaces.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._typeIntialized = true;
    this._wrapperFields.type.data = refObject.type.data;
    this._class_typeIntialized = true;
    this._wrapperFields.class_type.data = refObject.class_type.data;
    this._wrapperFields.state.copyRefObject(refObject.state);
    this._wrapperFields.command_interfaces.copyRefObject(refObject.command_interfaces);
    this._wrapperFields.state_interfaces.copyRefObject(refObject.state_interfaces);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    StringWrapper.freeStruct(refObject.type);
    StringWrapper.freeStruct(refObject.class_type);
    StateWrapper.freeStruct(refObject.state);
    if (refObject.command_interfaces.size != 0) {
      HardwareInterfaceWrapper.ArrayType.freeArray(refObject.command_interfaces);
      if (HardwareInterfaceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.command_interfaces, HardwareInterfaceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.state_interfaces.size != 0) {
      HardwareInterfaceWrapper.ArrayType.freeArray(refObject.state_interfaces);
      if (HardwareInterfaceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.state_interfaces, HardwareInterfaceWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    HardwareComponentStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'msg', interfaceName: 'HardwareComponentState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._wrapperFields.type.data;
  }
  set type(value) {
    this._typeIntialized = true;
    this._wrapperFields.type.data = value;
  }
  get class_type() {
    if (!this._class_typeIntialized) {
      return undefined;
    }
    return this._wrapperFields.class_type.data;
  }
  set class_type(value) {
    this._class_typeIntialized = true;
    this._wrapperFields.class_type.data = value;
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof StateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new StateWrapper(value));
    }
  }
  get command_interfaces() {
    return this._wrapperFields.command_interfaces;
  }
  set command_interfaces(value) {
    this._wrapperFields.command_interfaces.fill(value);
  }
  get state_interfaces() {
    return this._wrapperFields.state_interfaces;
  }
  set state_interfaces(value) {
    this._wrapperFields.state_interfaces.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new HardwareComponentStateRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._typeIntialized = true;
    this._wrapperFields.type.copyRefObject(this._refObject.type);
    this._class_typeIntialized = true;
    this._wrapperFields.class_type.copyRefObject(this._refObject.class_type);
    this._wrapperFields.state.copyRefObject(this._refObject.state);
    this._wrapperFields.command_interfaces.copyRefObject(this._refObject.command_interfaces);
    this._wrapperFields.state_interfaces.copyRefObject(this._refObject.state_interfaces);
  }
  copy(other) {
    this._refObject = new HardwareComponentStateRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._typeIntialized = true;
    this._wrapperFields.type.copy(other._wrapperFields.type);
    this._class_typeIntialized = true;
    this._wrapperFields.class_type.copy(other._wrapperFields.class_type);
    this._wrapperFields.state.copy(other._wrapperFields.state);
    this._wrapperFields.command_interfaces.copy(other._wrapperFields.command_interfaces);
    this._wrapperFields.state_interfaces.copy(other._wrapperFields.state_interfaces);
  }
  static get classType() {
    return HardwareComponentStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"class_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"lifecycle_msgs","type":"State","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"command_interfaces","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"controller_manager_msgs","type":"HardwareInterface","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"state_interfaces","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"controller_manager_msgs","type":"HardwareInterface","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"HardwareComponentState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HardwareComponentState"};
  }
  hasMember(name) {
    let memberNames = ["name","type","class_type","state","command_interfaces","state_interfaces"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HardwareComponentStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HardwareComponentStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HardwareComponentStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HardwareComponentStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HardwareComponentStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HardwareComponentStateRefArray(size);
    this._refObject = new HardwareComponentStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HardwareComponentStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HardwareComponentStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HardwareComponentStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HardwareComponentStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HardwareComponentStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HardwareComponentStateArrayWrapper;
  }
}
module.exports = HardwareComponentStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "class_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "lifecycle_msgs",
        "type": "State",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "command_interfaces",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "controller_manager_msgs",
        "type": "HardwareInterface",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "state_interfaces",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "controller_manager_msgs",
        "type": "HardwareInterface",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "controller_manager_msgs",
    "type": "HardwareComponentState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HardwareComponentState"
}
*/
