// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__PlanningSceneWorld.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CollisionObjectWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CollisionObject.js');
const OctomapWithPoseWrapper = require('../../generated/octomap_msgs/octomap_msgs__msg__OctomapWithPose.js');
const PlanningSceneWorldRefStruct = StructType({
  collision_objects: CollisionObjectWrapper.refObjectArrayType,
  octomap: OctomapWithPoseWrapper.refObjectType,
});
const PlanningSceneWorldRefArray = ArrayType(PlanningSceneWorldRefStruct);
const PlanningSceneWorldRefStructArray = StructType({
  data: PlanningSceneWorldRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PlanningSceneWorldWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PlanningSceneWorldRefStruct(other._refObject.toObject());
      this._wrapperFields.collision_objects = CollisionObjectWrapper.createArray();
      this._wrapperFields.collision_objects.copy(other._wrapperFields.collision_objects);
      this._wrapperFields.octomap =  new OctomapWithPoseWrapper(other._wrapperFields.octomap);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PlanningSceneWorldRefStruct();
    this._wrapperFields.collision_objects = CollisionObjectWrapper.createArray();
    this._wrapperFields.octomap = new OctomapWithPoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PlanningSceneWorldWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PlanningSceneWorldArrayWrapper;
  }
  static get ArrayType() {
    return PlanningSceneWorldArrayWrapper;
  }
  static get refObjectArrayType() {
    return PlanningSceneWorldRefStructArray
  }
  static get refObjectType() {
    return PlanningSceneWorldRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.collision_objects.freeze(own, checkConsistency);
    this._refObject.collision_objects = this._wrapperFields.collision_objects.refObject;
    this._wrapperFields.octomap.freeze(own, checkConsistency);
    this._refObject.octomap = this._wrapperFields.octomap.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.collision_objects.copyRefObject(refObject.collision_objects);
    this._wrapperFields.octomap.copyRefObject(refObject.octomap);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.collision_objects.size != 0) {
      CollisionObjectWrapper.ArrayType.freeArray(refObject.collision_objects);
      if (CollisionObjectWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.collision_objects, CollisionObjectWrapper.refObjectArrayType, 'data');
      }
    }
    OctomapWithPoseWrapper.freeStruct(refObject.octomap);
  }
  static destoryRawROS(msg) {
    PlanningSceneWorldWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'PlanningSceneWorld'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get collision_objects() {
    return this._wrapperFields.collision_objects;
  }
  set collision_objects(value) {
    this._wrapperFields.collision_objects.fill(value);
  }
  get octomap() {
    return this._wrapperFields.octomap;
  }
  set octomap(value) {
    if (value instanceof OctomapWithPoseWrapper) {
      this._wrapperFields.octomap.copy(value);
    } else {
      this._wrapperFields.octomap.copy(new OctomapWithPoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new PlanningSceneWorldRefStruct(refObject.toObject());
    this._wrapperFields.collision_objects.copyRefObject(this._refObject.collision_objects);
    this._wrapperFields.octomap.copyRefObject(this._refObject.octomap);
  }
  copy(other) {
    this._refObject = new PlanningSceneWorldRefStruct(other._refObject.toObject());
    this._wrapperFields.collision_objects.copy(other._wrapperFields.collision_objects);
    this._wrapperFields.octomap.copy(other._wrapperFields.octomap);
  }
  static get classType() {
    return PlanningSceneWorldWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"collision_objects","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"CollisionObject","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"octomap","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"octomap_msgs","type":"OctomapWithPose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"PlanningSceneWorld","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PlanningSceneWorld"};
  }
  hasMember(name) {
    let memberNames = ["collision_objects","octomap"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PlanningSceneWorldArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PlanningSceneWorldWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PlanningSceneWorldWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlanningSceneWorldArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlanningSceneWorldArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PlanningSceneWorldRefArray(size);
    this._refObject = new PlanningSceneWorldRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PlanningSceneWorldWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PlanningSceneWorldArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PlanningSceneWorldArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PlanningSceneWorldWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PlanningSceneWorldWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PlanningSceneWorldArrayWrapper;
  }
}
module.exports = PlanningSceneWorldWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "collision_objects",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "CollisionObject",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "octomap",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "octomap_msgs",
        "type": "OctomapWithPose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "PlanningSceneWorld",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PlanningSceneWorld"
}
*/
