// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ChangeControlDimensions_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ChangeControlDimensions_RequestRefStruct = StructType({
  control_x_translation: primitiveTypes.bool,
  control_y_translation: primitiveTypes.bool,
  control_z_translation: primitiveTypes.bool,
  control_x_rotation: primitiveTypes.bool,
  control_y_rotation: primitiveTypes.bool,
  control_z_rotation: primitiveTypes.bool,
});
const ChangeControlDimensions_RequestRefArray = ArrayType(ChangeControlDimensions_RequestRefStruct);
const ChangeControlDimensions_RequestRefStructArray = StructType({
  data: ChangeControlDimensions_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ChangeControlDimensions_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ChangeControlDimensions_RequestRefStruct(other._refObject.toObject());
      this._control_x_translationIntialized = true;
      this._control_y_translationIntialized = true;
      this._control_z_translationIntialized = true;
      this._control_x_rotationIntialized = true;
      this._control_y_rotationIntialized = true;
      this._control_z_rotationIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ChangeControlDimensions_RequestRefStruct();
    this._control_x_translationIntialized = false;
    this._control_y_translationIntialized = false;
    this._control_z_translationIntialized = false;
    this._control_x_rotationIntialized = false;
    this._control_y_rotationIntialized = false;
    this._control_z_rotationIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ChangeControlDimensions_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ChangeControlDimensions_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ChangeControlDimensions_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ChangeControlDimensions_RequestRefStructArray
  }
  static get refObjectType() {
    return ChangeControlDimensions_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._control_x_translationIntialized) {
        throw new TypeError('Invalid argument: control_x_translation in ChangeControlDimensions_Request');
      }
      if (checkConsistency && !this._control_y_translationIntialized) {
        throw new TypeError('Invalid argument: control_y_translation in ChangeControlDimensions_Request');
      }
      if (checkConsistency && !this._control_z_translationIntialized) {
        throw new TypeError('Invalid argument: control_z_translation in ChangeControlDimensions_Request');
      }
      if (checkConsistency && !this._control_x_rotationIntialized) {
        throw new TypeError('Invalid argument: control_x_rotation in ChangeControlDimensions_Request');
      }
      if (checkConsistency && !this._control_y_rotationIntialized) {
        throw new TypeError('Invalid argument: control_y_rotation in ChangeControlDimensions_Request');
      }
      if (checkConsistency && !this._control_z_rotationIntialized) {
        throw new TypeError('Invalid argument: control_z_rotation in ChangeControlDimensions_Request');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._control_x_translationIntialized = true;
    this._refObject.control_x_translation = refObject.control_x_translation;
    this._control_y_translationIntialized = true;
    this._refObject.control_y_translation = refObject.control_y_translation;
    this._control_z_translationIntialized = true;
    this._refObject.control_z_translation = refObject.control_z_translation;
    this._control_x_rotationIntialized = true;
    this._refObject.control_x_rotation = refObject.control_x_rotation;
    this._control_y_rotationIntialized = true;
    this._refObject.control_y_rotation = refObject.control_y_rotation;
    this._control_z_rotationIntialized = true;
    this._refObject.control_z_rotation = refObject.control_z_rotation;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ChangeControlDimensions_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ChangeControlDimensions_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get control_x_translation() {
    if (!this._control_x_translationIntialized) {
      return undefined;
    }
    return this._refObject.control_x_translation;
  }
  set control_x_translation(value) {
    this._control_x_translationIntialized = true;
    this._refObject.control_x_translation = value;
  }
  get control_y_translation() {
    if (!this._control_y_translationIntialized) {
      return undefined;
    }
    return this._refObject.control_y_translation;
  }
  set control_y_translation(value) {
    this._control_y_translationIntialized = true;
    this._refObject.control_y_translation = value;
  }
  get control_z_translation() {
    if (!this._control_z_translationIntialized) {
      return undefined;
    }
    return this._refObject.control_z_translation;
  }
  set control_z_translation(value) {
    this._control_z_translationIntialized = true;
    this._refObject.control_z_translation = value;
  }
  get control_x_rotation() {
    if (!this._control_x_rotationIntialized) {
      return undefined;
    }
    return this._refObject.control_x_rotation;
  }
  set control_x_rotation(value) {
    this._control_x_rotationIntialized = true;
    this._refObject.control_x_rotation = value;
  }
  get control_y_rotation() {
    if (!this._control_y_rotationIntialized) {
      return undefined;
    }
    return this._refObject.control_y_rotation;
  }
  set control_y_rotation(value) {
    this._control_y_rotationIntialized = true;
    this._refObject.control_y_rotation = value;
  }
  get control_z_rotation() {
    if (!this._control_z_rotationIntialized) {
      return undefined;
    }
    return this._refObject.control_z_rotation;
  }
  set control_z_rotation(value) {
    this._control_z_rotationIntialized = true;
    this._refObject.control_z_rotation = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ChangeControlDimensions_RequestRefStruct(refObject.toObject());
    this._control_x_translationIntialized = true;
    this._control_y_translationIntialized = true;
    this._control_z_translationIntialized = true;
    this._control_x_rotationIntialized = true;
    this._control_y_rotationIntialized = true;
    this._control_z_rotationIntialized = true;
  }
  copy(other) {
    this._refObject = new ChangeControlDimensions_RequestRefStruct(other._refObject.toObject());
    this._control_x_translationIntialized = true;
    this._control_y_translationIntialized = true;
    this._control_z_translationIntialized = true;
    this._control_x_rotationIntialized = true;
    this._control_y_rotationIntialized = true;
    this._control_z_rotationIntialized = true;
  }
  static get classType() {
    return ChangeControlDimensions_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"control_x_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_y_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_z_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_x_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_y_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_z_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ChangeControlDimensions_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ChangeControlDimensions_Request"};
  }
  hasMember(name) {
    let memberNames = ["control_x_translation","control_y_translation","control_z_translation","control_x_rotation","control_y_rotation","control_z_rotation"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ChangeControlDimensions_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ChangeControlDimensions_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ChangeControlDimensions_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeControlDimensions_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeControlDimensions_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ChangeControlDimensions_RequestRefArray(size);
    this._refObject = new ChangeControlDimensions_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ChangeControlDimensions_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ChangeControlDimensions_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ChangeControlDimensions_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ChangeControlDimensions_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ChangeControlDimensions_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ChangeControlDimensions_RequestArrayWrapper;
  }
}
module.exports = ChangeControlDimensions_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "control_x_translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_y_translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_z_translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_x_rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_y_rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_z_rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ChangeControlDimensions_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ChangeControlDimensions_Request"
}
*/
