// rclnodejs/generated/moveit_msgs/moveit_msgs__action__HybridPlanner_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MotionSequenceRequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionSequenceRequest.js');
const HybridPlanner_GoalRefStruct = StructType({
  planning_group: primitiveTypes.string,
  motion_sequence: MotionSequenceRequestWrapper.refObjectType,
});
const HybridPlanner_GoalRefArray = ArrayType(HybridPlanner_GoalRefStruct);
const HybridPlanner_GoalRefStructArray = StructType({
  data: HybridPlanner_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HybridPlanner_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HybridPlanner_GoalRefStruct(other._refObject.toObject());
      this._planning_groupIntialized = true;
      this._wrapperFields.planning_group =  new StringWrapper(other._wrapperFields.planning_group);
      this._wrapperFields.motion_sequence =  new MotionSequenceRequestWrapper(other._wrapperFields.motion_sequence);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HybridPlanner_GoalRefStruct();
    this._planning_groupIntialized = false;
    this._wrapperFields.planning_group = new StringWrapper();
    this._wrapperFields.motion_sequence = new MotionSequenceRequestWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new HybridPlanner_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HybridPlanner_GoalArrayWrapper;
  }
  static get ArrayType() {
    return HybridPlanner_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return HybridPlanner_GoalRefStructArray
  }
  static get refObjectType() {
    return HybridPlanner_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._planning_groupIntialized) {
        throw new TypeError('Invalid argument: planning_group in HybridPlanner_Goal');
      }
    if (own) {
      this._wrapperFields.planning_group.freeze(own, checkConsistency);
    }
    this._refObject.planning_group = this._wrapperFields.planning_group.refObject;
    this._wrapperFields.motion_sequence.freeze(own, checkConsistency);
    this._refObject.motion_sequence = this._wrapperFields.motion_sequence.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._planning_groupIntialized = true;
    this._wrapperFields.planning_group.data = refObject.planning_group.data;
    this._wrapperFields.motion_sequence.copyRefObject(refObject.motion_sequence);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.planning_group);
    MotionSequenceRequestWrapper.freeStruct(refObject.motion_sequence);
  }
  static destoryRawROS(msg) {
    HybridPlanner_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'HybridPlanner_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get planning_group() {
    if (!this._planning_groupIntialized) {
      return undefined;
    }
    return this._wrapperFields.planning_group.data;
  }
  set planning_group(value) {
    this._planning_groupIntialized = true;
    this._wrapperFields.planning_group.data = value;
  }
  get motion_sequence() {
    return this._wrapperFields.motion_sequence;
  }
  set motion_sequence(value) {
    if (value instanceof MotionSequenceRequestWrapper) {
      this._wrapperFields.motion_sequence.copy(value);
    } else {
      this._wrapperFields.motion_sequence.copy(new MotionSequenceRequestWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new HybridPlanner_GoalRefStruct(refObject.toObject());
    this._planning_groupIntialized = true;
    this._wrapperFields.planning_group.copyRefObject(this._refObject.planning_group);
    this._wrapperFields.motion_sequence.copyRefObject(this._refObject.motion_sequence);
  }
  copy(other) {
    this._refObject = new HybridPlanner_GoalRefStruct(other._refObject.toObject());
    this._planning_groupIntialized = true;
    this._wrapperFields.planning_group.copy(other._wrapperFields.planning_group);
    this._wrapperFields.motion_sequence.copy(other._wrapperFields.motion_sequence);
  }
  static get classType() {
    return HybridPlanner_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"planning_group","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"motion_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MotionSequenceRequest","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"HybridPlanner_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HybridPlanner_Goal"};
  }
  hasMember(name) {
    let memberNames = ["planning_group","motion_sequence"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HybridPlanner_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HybridPlanner_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HybridPlanner_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HybridPlanner_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HybridPlanner_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HybridPlanner_GoalRefArray(size);
    this._refObject = new HybridPlanner_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HybridPlanner_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HybridPlanner_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HybridPlanner_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HybridPlanner_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HybridPlanner_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HybridPlanner_GoalArrayWrapper;
  }
}
module.exports = HybridPlanner_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "planning_group",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "motion_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MotionSequenceRequest",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "HybridPlanner_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HybridPlanner_Goal"
}
*/
