// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__MotionPlanDetailedResponse.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const MotionPlanDetailedResponseRefStruct = StructType({
  trajectory_start: RobotStateWrapper.refObjectType,
  group_name: primitiveTypes.string,
  trajectory: RobotTrajectoryWrapper.refObjectArrayType,
  description: StringWrapper.refObjectArrayType,
  processing_time: Float64Wrapper.refObjectArrayType,
  error_code: MoveItErrorCodesWrapper.refObjectType,
});
const MotionPlanDetailedResponseRefArray = ArrayType(MotionPlanDetailedResponseRefStruct);
const MotionPlanDetailedResponseRefStructArray = StructType({
  data: MotionPlanDetailedResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MotionPlanDetailedResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._descriptionArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionPlanDetailedResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.trajectory_start =  new RobotStateWrapper(other._wrapperFields.trajectory_start);
      this._group_nameIntialized = true;
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
      this._wrapperFields.trajectory = RobotTrajectoryWrapper.createArray();
      this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
      this._wrapperFields.description = StringWrapper.createArray();
      this._wrapperFields.description.copy(other._wrapperFields.description);
      this.description = other.description;
      this._wrapperFields.processing_time = Float64Wrapper.createArray();
      this._wrapperFields.processing_time.copy(other._wrapperFields.processing_time);
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MotionPlanDetailedResponseRefStruct();
    this._wrapperFields.trajectory_start = new RobotStateWrapper();
    this._group_nameIntialized = false;
    this._wrapperFields.group_name = new StringWrapper();
    this._wrapperFields.trajectory = RobotTrajectoryWrapper.createArray();
    this._wrapperFields.description = StringWrapper.createArray();
    this._wrapperFields.processing_time = Float64Wrapper.createArray();
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MotionPlanDetailedResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionPlanDetailedResponseArrayWrapper;
  }
  static get ArrayType() {
    return MotionPlanDetailedResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionPlanDetailedResponseRefStructArray
  }
  static get refObjectType() {
    return MotionPlanDetailedResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._group_nameIntialized) {
        throw new TypeError('Invalid argument: group_name in MotionPlanDetailedResponse');
      }
    this._wrapperFields.trajectory_start.freeze(own, checkConsistency);
    this._refObject.trajectory_start = this._wrapperFields.trajectory_start.refObject;
    if (own) {
      this._wrapperFields.group_name.freeze(own, checkConsistency);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
    if (!own) {
      this._wrapperFields.description.fill(this._descriptionArray);
      this._wrapperFields.description.freeze(own, checkConsistency);
      this._refObject.description = this._wrapperFields.description.refObject;
    } else {
      this._wrapperFields.description.fill([]);
      this._wrapperFields.description.freeze(own, checkConsistency);
      this._refObject.description = this._wrapperFields.description.refObject;
    }
    this._wrapperFields.processing_time.freeze(own, checkConsistency);
    this._refObject.processing_time = this._wrapperFields.processing_time.refObject;
    if (own) {
      this._wrapperFields.processing_time.fill(Float64Array.from([]));
      this._wrapperFields.processing_time.freeze(own, checkConsistency);
      this._refObject.processing_time = this._wrapperFields.processing_time.refObject;
    }
    this._wrapperFields.error_code.freeze(own, checkConsistency);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.trajectory_start.copyRefObject(refObject.trajectory_start);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.data = refObject.group_name.data;
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
    refObject.description.data.length = refObject.description.size;
    for (let index = 0; index < refObject.description.size; index++) {
      this._descriptionArray[index] = refObject.description.data[index].data;
    }
    this._wrapperFields.processing_time.copyRefObject(refObject.processing_time);
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RobotStateWrapper.freeStruct(refObject.trajectory_start);
    StringWrapper.freeStruct(refObject.group_name);
    if (refObject.trajectory.size != 0) {
      RobotTrajectoryWrapper.ArrayType.freeArray(refObject.trajectory);
      if (RobotTrajectoryWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.trajectory, RobotTrajectoryWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.description.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.description);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.description, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.processing_time.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.processing_time);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.processing_time, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
  }
  static destoryRawROS(msg) {
    MotionPlanDetailedResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'MotionPlanDetailedResponse'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory_start() {
    return this._wrapperFields.trajectory_start;
  }
  set trajectory_start(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.trajectory_start.copy(value);
    } else {
      this._wrapperFields.trajectory_start.copy(new RobotStateWrapper(value));
    }
  }
  get group_name() {
    if (!this._group_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.data = value;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    this._wrapperFields.trajectory.fill(value);
  }
  get description() {
    return this._descriptionArray;
  }
  set description(value) {
    this._descriptionArray = value;
  }
  get processing_time() {
    return this._wrapperFields['processing_time'].data;
  }
  set processing_time(value) {
    this._wrapperFields['processing_time'].fill(value);
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MotionPlanDetailedResponseRefStruct(refObject.toObject());
    this._wrapperFields.trajectory_start.copyRefObject(this._refObject.trajectory_start);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
    refObject.description.data.length = refObject.description.size;
    for (let index = 0; index < refObject.description.size; index++) {
      this._descriptionArray[index] = refObject.description.data[index].data;
    }
    this._wrapperFields.processing_time.copyRefObject(this._refObject.processing_time);
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
  }
  copy(other) {
    this._refObject = new MotionPlanDetailedResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.trajectory_start.copy(other._wrapperFields.trajectory_start);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
    this._descriptionArray = other._descriptionArray.slice();
    this._wrapperFields.processing_time.copy(other._wrapperFields.processing_time);
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
  }
  static get classType() {
    return MotionPlanDetailedResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trajectory","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"description","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"processing_time","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MotionPlanDetailedResponse","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionPlanDetailedResponse"};
  }
  hasMember(name) {
    let memberNames = ["trajectory_start","group_name","trajectory","description","processing_time","error_code"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MotionPlanDetailedResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionPlanDetailedResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionPlanDetailedResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPlanDetailedResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPlanDetailedResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MotionPlanDetailedResponseRefArray(size);
    this._refObject = new MotionPlanDetailedResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionPlanDetailedResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionPlanDetailedResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionPlanDetailedResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionPlanDetailedResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionPlanDetailedResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionPlanDetailedResponseArrayWrapper;
  }
}
module.exports = MotionPlanDetailedResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "group_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trajectory",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "description",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "processing_time",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MoveItErrorCodes",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "MotionPlanDetailedResponse",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MotionPlanDetailedResponse"
}
*/
