// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetMotionSequence_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionSequenceRequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionSequenceRequest.js');
const GetMotionSequence_RequestRefStruct = StructType({
  request: MotionSequenceRequestWrapper.refObjectType,
});
const GetMotionSequence_RequestRefArray = ArrayType(GetMotionSequence_RequestRefStruct);
const GetMotionSequence_RequestRefStructArray = StructType({
  data: GetMotionSequence_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetMotionSequence_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetMotionSequence_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.request =  new MotionSequenceRequestWrapper(other._wrapperFields.request);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetMotionSequence_RequestRefStruct();
    this._wrapperFields.request = new MotionSequenceRequestWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetMotionSequence_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetMotionSequence_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetMotionSequence_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetMotionSequence_RequestRefStructArray
  }
  static get refObjectType() {
    return GetMotionSequence_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.request.freeze(own, checkConsistency);
    this._refObject.request = this._wrapperFields.request.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.request.copyRefObject(refObject.request);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MotionSequenceRequestWrapper.freeStruct(refObject.request);
  }
  static destoryRawROS(msg) {
    GetMotionSequence_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetMotionSequence_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get request() {
    return this._wrapperFields.request;
  }
  set request(value) {
    if (value instanceof MotionSequenceRequestWrapper) {
      this._wrapperFields.request.copy(value);
    } else {
      this._wrapperFields.request.copy(new MotionSequenceRequestWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetMotionSequence_RequestRefStruct(refObject.toObject());
    this._wrapperFields.request.copyRefObject(this._refObject.request);
  }
  copy(other) {
    this._refObject = new GetMotionSequence_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.request.copy(other._wrapperFields.request);
  }
  static get classType() {
    return GetMotionSequence_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"request","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MotionSequenceRequest","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetMotionSequence_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetMotionSequence_Request"};
  }
  hasMember(name) {
    let memberNames = ["request"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetMotionSequence_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetMotionSequence_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetMotionSequence_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetMotionSequence_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetMotionSequence_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetMotionSequence_RequestRefArray(size);
    this._refObject = new GetMotionSequence_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetMotionSequence_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetMotionSequence_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetMotionSequence_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetMotionSequence_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetMotionSequence_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetMotionSequence_RequestArrayWrapper;
  }
}
module.exports = GetMotionSequence_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "request",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MotionSequenceRequest",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetMotionSequence_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetMotionSequence_Request"
}
*/
