// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__ObjectColor.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ColorRGBAWrapper = require('../../generated/std_msgs/std_msgs__msg__ColorRGBA.js');
const ObjectColorRefStruct = StructType({
  id: primitiveTypes.string,
  color: ColorRGBAWrapper.refObjectType,
});
const ObjectColorRefArray = ArrayType(ObjectColorRefStruct);
const ObjectColorRefStructArray = StructType({
  data: ObjectColorRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObjectColorWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectColorRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
      this._wrapperFields.color =  new ColorRGBAWrapper(other._wrapperFields.color);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObjectColorRefStruct();
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
    this._wrapperFields.color = new ColorRGBAWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectColorWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectColorArrayWrapper;
  }
  static get ArrayType() {
    return ObjectColorArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectColorRefStructArray
  }
  static get refObjectType() {
    return ObjectColorRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in ObjectColor');
      }
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
    this._wrapperFields.color.freeze(own, checkConsistency);
    this._refObject.color = this._wrapperFields.color.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
    this._wrapperFields.color.copyRefObject(refObject.color);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.id);
    ColorRGBAWrapper.freeStruct(refObject.color);
  }
  static destoryRawROS(msg) {
    ObjectColorWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'ObjectColor'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  get color() {
    return this._wrapperFields.color;
  }
  set color(value) {
    if (value instanceof ColorRGBAWrapper) {
      this._wrapperFields.color.copy(value);
    } else {
      this._wrapperFields.color.copy(new ColorRGBAWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectColorRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._wrapperFields.color.copyRefObject(this._refObject.color);
  }
  copy(other) {
    this._refObject = new ObjectColorRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._wrapperFields.color.copy(other._wrapperFields.color);
  }
  static get classType() {
    return ObjectColorWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"color","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"ColorRGBA","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ObjectColor","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectColor"};
  }
  hasMember(name) {
    let memberNames = ["id","color"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObjectColorArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectColorWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectColorWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectColorArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectColorArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObjectColorRefArray(size);
    this._refObject = new ObjectColorRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectColorWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectColorArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectColorArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectColorWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectColorWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectColorArrayWrapper;
  }
}
module.exports = ObjectColorWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "color",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "ColorRGBA",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ObjectColor",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObjectColor"
}
*/
