// rclnodejs/generated/moveit_msgs/moveit_msgs__action__MoveGroup_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const MoveGroup_ResultRefStruct = StructType({
  error_code: MoveItErrorCodesWrapper.refObjectType,
  trajectory_start: RobotStateWrapper.refObjectType,
  planned_trajectory: RobotTrajectoryWrapper.refObjectType,
  executed_trajectory: RobotTrajectoryWrapper.refObjectType,
  planning_time: primitiveTypes.float64,
});
const MoveGroup_ResultRefArray = ArrayType(MoveGroup_ResultRefStruct);
const MoveGroup_ResultRefStructArray = StructType({
  data: MoveGroup_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MoveGroup_ResultWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MoveGroup_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
      this._wrapperFields.trajectory_start =  new RobotStateWrapper(other._wrapperFields.trajectory_start);
      this._wrapperFields.planned_trajectory =  new RobotTrajectoryWrapper(other._wrapperFields.planned_trajectory);
      this._wrapperFields.executed_trajectory =  new RobotTrajectoryWrapper(other._wrapperFields.executed_trajectory);
      this._planning_timeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MoveGroup_ResultRefStruct();
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
    this._wrapperFields.trajectory_start = new RobotStateWrapper();
    this._wrapperFields.planned_trajectory = new RobotTrajectoryWrapper();
    this._wrapperFields.executed_trajectory = new RobotTrajectoryWrapper();
    this._planning_timeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MoveGroup_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MoveGroup_ResultArrayWrapper;
  }
  static get ArrayType() {
    return MoveGroup_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return MoveGroup_ResultRefStructArray
  }
  static get refObjectType() {
    return MoveGroup_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._planning_timeIntialized) {
        throw new TypeError('Invalid argument: planning_time in MoveGroup_Result');
      }
    this._wrapperFields.error_code.freeze(own, checkConsistency);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
    this._wrapperFields.trajectory_start.freeze(own, checkConsistency);
    this._refObject.trajectory_start = this._wrapperFields.trajectory_start.refObject;
    this._wrapperFields.planned_trajectory.freeze(own, checkConsistency);
    this._refObject.planned_trajectory = this._wrapperFields.planned_trajectory.refObject;
    this._wrapperFields.executed_trajectory.freeze(own, checkConsistency);
    this._refObject.executed_trajectory = this._wrapperFields.executed_trajectory.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
    this._wrapperFields.trajectory_start.copyRefObject(refObject.trajectory_start);
    this._wrapperFields.planned_trajectory.copyRefObject(refObject.planned_trajectory);
    this._wrapperFields.executed_trajectory.copyRefObject(refObject.executed_trajectory);
    this._planning_timeIntialized = true;
    this._refObject.planning_time = refObject.planning_time;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
    RobotStateWrapper.freeStruct(refObject.trajectory_start);
    RobotTrajectoryWrapper.freeStruct(refObject.planned_trajectory);
    RobotTrajectoryWrapper.freeStruct(refObject.executed_trajectory);
  }
  static destoryRawROS(msg) {
    MoveGroup_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'MoveGroup_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  get trajectory_start() {
    return this._wrapperFields.trajectory_start;
  }
  set trajectory_start(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.trajectory_start.copy(value);
    } else {
      this._wrapperFields.trajectory_start.copy(new RobotStateWrapper(value));
    }
  }
  get planned_trajectory() {
    return this._wrapperFields.planned_trajectory;
  }
  set planned_trajectory(value) {
    if (value instanceof RobotTrajectoryWrapper) {
      this._wrapperFields.planned_trajectory.copy(value);
    } else {
      this._wrapperFields.planned_trajectory.copy(new RobotTrajectoryWrapper(value));
    }
  }
  get executed_trajectory() {
    return this._wrapperFields.executed_trajectory;
  }
  set executed_trajectory(value) {
    if (value instanceof RobotTrajectoryWrapper) {
      this._wrapperFields.executed_trajectory.copy(value);
    } else {
      this._wrapperFields.executed_trajectory.copy(new RobotTrajectoryWrapper(value));
    }
  }
  get planning_time() {
    if (!this._planning_timeIntialized) {
      return undefined;
    }
    return this._refObject.planning_time;
  }
  set planning_time(value) {
    this._planning_timeIntialized = true;
    this._refObject.planning_time = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MoveGroup_ResultRefStruct(refObject.toObject());
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
    this._wrapperFields.trajectory_start.copyRefObject(this._refObject.trajectory_start);
    this._wrapperFields.planned_trajectory.copyRefObject(this._refObject.planned_trajectory);
    this._wrapperFields.executed_trajectory.copyRefObject(this._refObject.executed_trajectory);
    this._planning_timeIntialized = true;
  }
  copy(other) {
    this._refObject = new MoveGroup_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
    this._wrapperFields.trajectory_start.copy(other._wrapperFields.trajectory_start);
    this._wrapperFields.planned_trajectory.copy(other._wrapperFields.planned_trajectory);
    this._wrapperFields.executed_trajectory.copy(other._wrapperFields.executed_trajectory);
    this._planning_timeIntialized = true;
  }
  static get classType() {
    return MoveGroup_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"trajectory_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planned_trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"executed_trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planning_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MoveGroup_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MoveGroup_Result"};
  }
  hasMember(name) {
    let memberNames = ["error_code","trajectory_start","planned_trajectory","executed_trajectory","planning_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MoveGroup_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MoveGroup_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MoveGroup_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveGroup_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveGroup_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MoveGroup_ResultRefArray(size);
    this._refObject = new MoveGroup_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MoveGroup_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MoveGroup_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MoveGroup_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MoveGroup_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MoveGroup_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MoveGroup_ResultArrayWrapper;
  }
}
module.exports = MoveGroup_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "error_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MoveItErrorCodes",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "trajectory_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planned_trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "executed_trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planning_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "MoveGroup_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MoveGroup_Result"
}
*/
