// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__KinematicSolverInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointLimitsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__JointLimits.js');
const KinematicSolverInfoRefStruct = StructType({
  joint_names: StringWrapper.refObjectArrayType,
  limits: JointLimitsWrapper.refObjectArrayType,
  link_names: StringWrapper.refObjectArrayType,
});
const KinematicSolverInfoRefArray = ArrayType(KinematicSolverInfoRefStruct);
const KinematicSolverInfoRefStructArray = StructType({
  data: KinematicSolverInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class KinematicSolverInfoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._joint_namesArray = [];
    this._link_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new KinematicSolverInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.joint_names = StringWrapper.createArray();
      this._wrapperFields.joint_names.copy(other._wrapperFields.joint_names);
      this.joint_names = other.joint_names;
      this._wrapperFields.limits = JointLimitsWrapper.createArray();
      this._wrapperFields.limits.copy(other._wrapperFields.limits);
      this._wrapperFields.link_names = StringWrapper.createArray();
      this._wrapperFields.link_names.copy(other._wrapperFields.link_names);
      this.link_names = other.link_names;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new KinematicSolverInfoRefStruct();
    this._wrapperFields.joint_names = StringWrapper.createArray();
    this._wrapperFields.limits = JointLimitsWrapper.createArray();
    this._wrapperFields.link_names = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new KinematicSolverInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new KinematicSolverInfoArrayWrapper;
  }
  static get ArrayType() {
    return KinematicSolverInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return KinematicSolverInfoRefStructArray
  }
  static get refObjectType() {
    return KinematicSolverInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (!own) {
      this._wrapperFields.joint_names.fill(this._joint_namesArray);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    } else {
      this._wrapperFields.joint_names.fill([]);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    }
    this._wrapperFields.limits.freeze(own, checkConsistency);
    this._refObject.limits = this._wrapperFields.limits.refObject;
    if (!own) {
      this._wrapperFields.link_names.fill(this._link_namesArray);
      this._wrapperFields.link_names.freeze(own, checkConsistency);
      this._refObject.link_names = this._wrapperFields.link_names.refObject;
    } else {
      this._wrapperFields.link_names.fill([]);
      this._wrapperFields.link_names.freeze(own, checkConsistency);
      this._refObject.link_names = this._wrapperFields.link_names.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.limits.copyRefObject(refObject.limits);
    refObject.link_names.data.length = refObject.link_names.size;
    for (let index = 0; index < refObject.link_names.size; index++) {
      this._link_namesArray[index] = refObject.link_names.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.joint_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.joint_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.joint_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.limits.size != 0) {
      JointLimitsWrapper.ArrayType.freeArray(refObject.limits);
      if (JointLimitsWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.limits, JointLimitsWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.link_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.link_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.link_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    KinematicSolverInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'KinematicSolverInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get joint_names() {
    return this._joint_namesArray;
  }
  set joint_names(value) {
    this._joint_namesArray = value;
  }
  get limits() {
    return this._wrapperFields.limits;
  }
  set limits(value) {
    this._wrapperFields.limits.fill(value);
  }
  get link_names() {
    return this._link_namesArray;
  }
  set link_names(value) {
    this._link_namesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new KinematicSolverInfoRefStruct(refObject.toObject());
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.limits.copyRefObject(this._refObject.limits);
    refObject.link_names.data.length = refObject.link_names.size;
    for (let index = 0; index < refObject.link_names.size; index++) {
      this._link_namesArray[index] = refObject.link_names.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new KinematicSolverInfoRefStruct(other._refObject.toObject());
    this._joint_namesArray = other._joint_namesArray.slice();
    this._wrapperFields.limits.copy(other._wrapperFields.limits);
    this._link_namesArray = other._link_namesArray.slice();
  }
  static get classType() {
    return KinematicSolverInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"joint_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"limits","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"JointLimits","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"link_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"KinematicSolverInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"KinematicSolverInfo"};
  }
  hasMember(name) {
    let memberNames = ["joint_names","limits","link_names"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class KinematicSolverInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof KinematicSolverInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new KinematicSolverInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to KinematicSolverInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to KinematicSolverInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new KinematicSolverInfoRefArray(size);
    this._refObject = new KinematicSolverInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new KinematicSolverInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof KinematicSolverInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "KinematicSolverInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      KinematicSolverInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return KinematicSolverInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return KinematicSolverInfoArrayWrapper;
  }
}
module.exports = KinematicSolverInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "joint_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "limits",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "JointLimits",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "link_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "KinematicSolverInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "KinematicSolverInfo"
}
*/
