// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__CartesianTrajectoryPoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CartesianPointWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CartesianPoint.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const CartesianTrajectoryPointRefStruct = StructType({
  point: CartesianPointWrapper.refObjectType,
  time_from_start: DurationWrapper.refObjectType,
});
const CartesianTrajectoryPointRefArray = ArrayType(CartesianTrajectoryPointRefStruct);
const CartesianTrajectoryPointRefStructArray = StructType({
  data: CartesianTrajectoryPointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CartesianTrajectoryPointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CartesianTrajectoryPointRefStruct(other._refObject.toObject());
      this._wrapperFields.point =  new CartesianPointWrapper(other._wrapperFields.point);
      this._wrapperFields.time_from_start =  new DurationWrapper(other._wrapperFields.time_from_start);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CartesianTrajectoryPointRefStruct();
    this._wrapperFields.point = new CartesianPointWrapper();
    this._wrapperFields.time_from_start = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CartesianTrajectoryPointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CartesianTrajectoryPointArrayWrapper;
  }
  static get ArrayType() {
    return CartesianTrajectoryPointArrayWrapper;
  }
  static get refObjectArrayType() {
    return CartesianTrajectoryPointRefStructArray
  }
  static get refObjectType() {
    return CartesianTrajectoryPointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.point.freeze(own, checkConsistency);
    this._refObject.point = this._wrapperFields.point.refObject;
    this._wrapperFields.time_from_start.freeze(own, checkConsistency);
    this._refObject.time_from_start = this._wrapperFields.time_from_start.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.point.copyRefObject(refObject.point);
    this._wrapperFields.time_from_start.copyRefObject(refObject.time_from_start);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    CartesianPointWrapper.freeStruct(refObject.point);
    DurationWrapper.freeStruct(refObject.time_from_start);
  }
  static destoryRawROS(msg) {
    CartesianTrajectoryPointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'CartesianTrajectoryPoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get point() {
    return this._wrapperFields.point;
  }
  set point(value) {
    if (value instanceof CartesianPointWrapper) {
      this._wrapperFields.point.copy(value);
    } else {
      this._wrapperFields.point.copy(new CartesianPointWrapper(value));
    }
  }
  get time_from_start() {
    return this._wrapperFields.time_from_start;
  }
  set time_from_start(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.time_from_start.copy(value);
    } else {
      this._wrapperFields.time_from_start.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CartesianTrajectoryPointRefStruct(refObject.toObject());
    this._wrapperFields.point.copyRefObject(this._refObject.point);
    this._wrapperFields.time_from_start.copyRefObject(this._refObject.time_from_start);
  }
  copy(other) {
    this._refObject = new CartesianTrajectoryPointRefStruct(other._refObject.toObject());
    this._wrapperFields.point.copy(other._wrapperFields.point);
    this._wrapperFields.time_from_start.copy(other._wrapperFields.time_from_start);
  }
  static get classType() {
    return CartesianTrajectoryPointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"point","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"CartesianPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"time_from_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"CartesianTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CartesianTrajectoryPoint"};
  }
  hasMember(name) {
    let memberNames = ["point","time_from_start"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CartesianTrajectoryPointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CartesianTrajectoryPointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CartesianTrajectoryPointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianTrajectoryPointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianTrajectoryPointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CartesianTrajectoryPointRefArray(size);
    this._refObject = new CartesianTrajectoryPointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CartesianTrajectoryPointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CartesianTrajectoryPointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CartesianTrajectoryPointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CartesianTrajectoryPointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CartesianTrajectoryPointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CartesianTrajectoryPointArrayWrapper;
  }
}
module.exports = CartesianTrajectoryPointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "point",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "CartesianPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "time_from_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "CartesianTrajectoryPoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CartesianTrajectoryPoint"
}
*/
