// rclnodejs/generated/moveit_msgs/moveit_msgs__action__MoveGroupSequence_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionSequenceRequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionSequenceRequest.js');
const PlanningOptionsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlanningOptions.js');
const MoveGroupSequence_GoalRefStruct = StructType({
  request: MotionSequenceRequestWrapper.refObjectType,
  planning_options: PlanningOptionsWrapper.refObjectType,
});
const MoveGroupSequence_GoalRefArray = ArrayType(MoveGroupSequence_GoalRefStruct);
const MoveGroupSequence_GoalRefStructArray = StructType({
  data: MoveGroupSequence_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MoveGroupSequence_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MoveGroupSequence_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.request =  new MotionSequenceRequestWrapper(other._wrapperFields.request);
      this._wrapperFields.planning_options =  new PlanningOptionsWrapper(other._wrapperFields.planning_options);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MoveGroupSequence_GoalRefStruct();
    this._wrapperFields.request = new MotionSequenceRequestWrapper();
    this._wrapperFields.planning_options = new PlanningOptionsWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MoveGroupSequence_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MoveGroupSequence_GoalArrayWrapper;
  }
  static get ArrayType() {
    return MoveGroupSequence_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return MoveGroupSequence_GoalRefStructArray
  }
  static get refObjectType() {
    return MoveGroupSequence_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.request.freeze(own, checkConsistency);
    this._refObject.request = this._wrapperFields.request.refObject;
    this._wrapperFields.planning_options.freeze(own, checkConsistency);
    this._refObject.planning_options = this._wrapperFields.planning_options.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.request.copyRefObject(refObject.request);
    this._wrapperFields.planning_options.copyRefObject(refObject.planning_options);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MotionSequenceRequestWrapper.freeStruct(refObject.request);
    PlanningOptionsWrapper.freeStruct(refObject.planning_options);
  }
  static destoryRawROS(msg) {
    MoveGroupSequence_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'MoveGroupSequence_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get request() {
    return this._wrapperFields.request;
  }
  set request(value) {
    if (value instanceof MotionSequenceRequestWrapper) {
      this._wrapperFields.request.copy(value);
    } else {
      this._wrapperFields.request.copy(new MotionSequenceRequestWrapper(value));
    }
  }
  get planning_options() {
    return this._wrapperFields.planning_options;
  }
  set planning_options(value) {
    if (value instanceof PlanningOptionsWrapper) {
      this._wrapperFields.planning_options.copy(value);
    } else {
      this._wrapperFields.planning_options.copy(new PlanningOptionsWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MoveGroupSequence_GoalRefStruct(refObject.toObject());
    this._wrapperFields.request.copyRefObject(this._refObject.request);
    this._wrapperFields.planning_options.copyRefObject(this._refObject.planning_options);
  }
  copy(other) {
    this._refObject = new MoveGroupSequence_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.request.copy(other._wrapperFields.request);
    this._wrapperFields.planning_options.copy(other._wrapperFields.planning_options);
  }
  static get classType() {
    return MoveGroupSequence_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"request","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MotionSequenceRequest","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planning_options","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlanningOptions","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MoveGroupSequence_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MoveGroupSequence_Goal"};
  }
  hasMember(name) {
    let memberNames = ["request","planning_options"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MoveGroupSequence_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MoveGroupSequence_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MoveGroupSequence_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveGroupSequence_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveGroupSequence_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MoveGroupSequence_GoalRefArray(size);
    this._refObject = new MoveGroupSequence_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MoveGroupSequence_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MoveGroupSequence_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MoveGroupSequence_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MoveGroupSequence_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MoveGroupSequence_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MoveGroupSequence_GoalArrayWrapper;
  }
}
module.exports = MoveGroupSequence_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "request",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MotionSequenceRequest",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planning_options",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "PlanningOptions",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "MoveGroupSequence_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MoveGroupSequence_Goal"
}
*/
