// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetRobotStateFromWarehouse_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const GetRobotStateFromWarehouse_ResponseRefStruct = StructType({
  state: RobotStateWrapper.refObjectType,
});
const GetRobotStateFromWarehouse_ResponseRefArray = ArrayType(GetRobotStateFromWarehouse_ResponseRefStruct);
const GetRobotStateFromWarehouse_ResponseRefStructArray = StructType({
  data: GetRobotStateFromWarehouse_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetRobotStateFromWarehouse_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetRobotStateFromWarehouse_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.state =  new RobotStateWrapper(other._wrapperFields.state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetRobotStateFromWarehouse_ResponseRefStruct();
    this._wrapperFields.state = new RobotStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetRobotStateFromWarehouse_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetRobotStateFromWarehouse_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetRobotStateFromWarehouse_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetRobotStateFromWarehouse_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetRobotStateFromWarehouse_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.state.copyRefObject(refObject.state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RobotStateWrapper.freeStruct(refObject.state);
  }
  static destoryRawROS(msg) {
    GetRobotStateFromWarehouse_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetRobotStateFromWarehouse_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new RobotStateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetRobotStateFromWarehouse_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.state.copyRefObject(this._refObject.state);
  }
  copy(other) {
    this._refObject = new GetRobotStateFromWarehouse_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
  }
  static get classType() {
    return GetRobotStateFromWarehouse_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetRobotStateFromWarehouse_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetRobotStateFromWarehouse_Response"};
  }
  hasMember(name) {
    let memberNames = ["state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetRobotStateFromWarehouse_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetRobotStateFromWarehouse_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetRobotStateFromWarehouse_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetRobotStateFromWarehouse_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetRobotStateFromWarehouse_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetRobotStateFromWarehouse_ResponseRefArray(size);
    this._refObject = new GetRobotStateFromWarehouse_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetRobotStateFromWarehouse_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetRobotStateFromWarehouse_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetRobotStateFromWarehouse_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetRobotStateFromWarehouse_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetRobotStateFromWarehouse_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetRobotStateFromWarehouse_ResponseArrayWrapper;
  }
}
module.exports = GetRobotStateFromWarehouse_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetRobotStateFromWarehouse_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetRobotStateFromWarehouse_Response"
}
*/
