// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__MotionSequenceResponse.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const MotionSequenceResponseRefStruct = StructType({
  error_code: MoveItErrorCodesWrapper.refObjectType,
  sequence_start: RobotStateWrapper.refObjectType,
  planned_trajectories: RobotTrajectoryWrapper.refObjectArrayType,
  planning_time: primitiveTypes.float64,
});
const MotionSequenceResponseRefArray = ArrayType(MotionSequenceResponseRefStruct);
const MotionSequenceResponseRefStructArray = StructType({
  data: MotionSequenceResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MotionSequenceResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionSequenceResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
      this._wrapperFields.sequence_start =  new RobotStateWrapper(other._wrapperFields.sequence_start);
      this._wrapperFields.planned_trajectories = RobotTrajectoryWrapper.createArray();
      this._wrapperFields.planned_trajectories.copy(other._wrapperFields.planned_trajectories);
      this._planning_timeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MotionSequenceResponseRefStruct();
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
    this._wrapperFields.sequence_start = new RobotStateWrapper();
    this._wrapperFields.planned_trajectories = RobotTrajectoryWrapper.createArray();
    this._planning_timeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MotionSequenceResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionSequenceResponseArrayWrapper;
  }
  static get ArrayType() {
    return MotionSequenceResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionSequenceResponseRefStructArray
  }
  static get refObjectType() {
    return MotionSequenceResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._planning_timeIntialized) {
        throw new TypeError('Invalid argument: planning_time in MotionSequenceResponse');
      }
    this._wrapperFields.error_code.freeze(own, checkConsistency);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
    this._wrapperFields.sequence_start.freeze(own, checkConsistency);
    this._refObject.sequence_start = this._wrapperFields.sequence_start.refObject;
    this._wrapperFields.planned_trajectories.freeze(own, checkConsistency);
    this._refObject.planned_trajectories = this._wrapperFields.planned_trajectories.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
    this._wrapperFields.sequence_start.copyRefObject(refObject.sequence_start);
    this._wrapperFields.planned_trajectories.copyRefObject(refObject.planned_trajectories);
    this._planning_timeIntialized = true;
    this._refObject.planning_time = refObject.planning_time;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
    RobotStateWrapper.freeStruct(refObject.sequence_start);
    if (refObject.planned_trajectories.size != 0) {
      RobotTrajectoryWrapper.ArrayType.freeArray(refObject.planned_trajectories);
      if (RobotTrajectoryWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.planned_trajectories, RobotTrajectoryWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MotionSequenceResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'MotionSequenceResponse'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  get sequence_start() {
    return this._wrapperFields.sequence_start;
  }
  set sequence_start(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.sequence_start.copy(value);
    } else {
      this._wrapperFields.sequence_start.copy(new RobotStateWrapper(value));
    }
  }
  get planned_trajectories() {
    return this._wrapperFields.planned_trajectories;
  }
  set planned_trajectories(value) {
    this._wrapperFields.planned_trajectories.fill(value);
  }
  get planning_time() {
    if (!this._planning_timeIntialized) {
      return undefined;
    }
    return this._refObject.planning_time;
  }
  set planning_time(value) {
    this._planning_timeIntialized = true;
    this._refObject.planning_time = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MotionSequenceResponseRefStruct(refObject.toObject());
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
    this._wrapperFields.sequence_start.copyRefObject(this._refObject.sequence_start);
    this._wrapperFields.planned_trajectories.copyRefObject(this._refObject.planned_trajectories);
    this._planning_timeIntialized = true;
  }
  copy(other) {
    this._refObject = new MotionSequenceResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
    this._wrapperFields.sequence_start.copy(other._wrapperFields.sequence_start);
    this._wrapperFields.planned_trajectories.copy(other._wrapperFields.planned_trajectories);
    this._planning_timeIntialized = true;
  }
  static get classType() {
    return MotionSequenceResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"sequence_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planned_trajectories","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planning_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MotionSequenceResponse","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionSequenceResponse"};
  }
  hasMember(name) {
    let memberNames = ["error_code","sequence_start","planned_trajectories","planning_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MotionSequenceResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionSequenceResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionSequenceResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionSequenceResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionSequenceResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MotionSequenceResponseRefArray(size);
    this._refObject = new MotionSequenceResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionSequenceResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionSequenceResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionSequenceResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionSequenceResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionSequenceResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionSequenceResponseArrayWrapper;
  }
}
module.exports = MotionSequenceResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "error_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MoveItErrorCodes",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "sequence_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planned_trajectories",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planning_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "MotionSequenceResponse",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MotionSequenceResponse"
}
*/
