// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetStateValidity_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ContactInformationWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__ContactInformation.js');
const CostSourceWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CostSource.js');
const ConstraintEvalResultWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__ConstraintEvalResult.js');
const GetStateValidity_ResponseRefStruct = StructType({
  valid: primitiveTypes.bool,
  contacts: ContactInformationWrapper.refObjectArrayType,
  cost_sources: CostSourceWrapper.refObjectArrayType,
  constraint_result: ConstraintEvalResultWrapper.refObjectArrayType,
});
const GetStateValidity_ResponseRefArray = ArrayType(GetStateValidity_ResponseRefStruct);
const GetStateValidity_ResponseRefStructArray = StructType({
  data: GetStateValidity_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetStateValidity_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetStateValidity_ResponseRefStruct(other._refObject.toObject());
      this._validIntialized = true;
      this._wrapperFields.contacts = ContactInformationWrapper.createArray();
      this._wrapperFields.contacts.copy(other._wrapperFields.contacts);
      this._wrapperFields.cost_sources = CostSourceWrapper.createArray();
      this._wrapperFields.cost_sources.copy(other._wrapperFields.cost_sources);
      this._wrapperFields.constraint_result = ConstraintEvalResultWrapper.createArray();
      this._wrapperFields.constraint_result.copy(other._wrapperFields.constraint_result);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetStateValidity_ResponseRefStruct();
    this._validIntialized = false;
    this._wrapperFields.contacts = ContactInformationWrapper.createArray();
    this._wrapperFields.cost_sources = CostSourceWrapper.createArray();
    this._wrapperFields.constraint_result = ConstraintEvalResultWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetStateValidity_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetStateValidity_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetStateValidity_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetStateValidity_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetStateValidity_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._validIntialized) {
        throw new TypeError('Invalid argument: valid in GetStateValidity_Response');
      }
    this._wrapperFields.contacts.freeze(own, checkConsistency);
    this._refObject.contacts = this._wrapperFields.contacts.refObject;
    this._wrapperFields.cost_sources.freeze(own, checkConsistency);
    this._refObject.cost_sources = this._wrapperFields.cost_sources.refObject;
    this._wrapperFields.constraint_result.freeze(own, checkConsistency);
    this._refObject.constraint_result = this._wrapperFields.constraint_result.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._validIntialized = true;
    this._refObject.valid = refObject.valid;
    this._wrapperFields.contacts.copyRefObject(refObject.contacts);
    this._wrapperFields.cost_sources.copyRefObject(refObject.cost_sources);
    this._wrapperFields.constraint_result.copyRefObject(refObject.constraint_result);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.contacts.size != 0) {
      ContactInformationWrapper.ArrayType.freeArray(refObject.contacts);
      if (ContactInformationWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.contacts, ContactInformationWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.cost_sources.size != 0) {
      CostSourceWrapper.ArrayType.freeArray(refObject.cost_sources);
      if (CostSourceWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.cost_sources, CostSourceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.constraint_result.size != 0) {
      ConstraintEvalResultWrapper.ArrayType.freeArray(refObject.constraint_result);
      if (ConstraintEvalResultWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.constraint_result, ConstraintEvalResultWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetStateValidity_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetStateValidity_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get valid() {
    if (!this._validIntialized) {
      return undefined;
    }
    return this._refObject.valid;
  }
  set valid(value) {
    this._validIntialized = true;
    this._refObject.valid = value;
  }
  get contacts() {
    return this._wrapperFields.contacts;
  }
  set contacts(value) {
    this._wrapperFields.contacts.fill(value);
  }
  get cost_sources() {
    return this._wrapperFields.cost_sources;
  }
  set cost_sources(value) {
    this._wrapperFields.cost_sources.fill(value);
  }
  get constraint_result() {
    return this._wrapperFields.constraint_result;
  }
  set constraint_result(value) {
    this._wrapperFields.constraint_result.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetStateValidity_ResponseRefStruct(refObject.toObject());
    this._validIntialized = true;
    this._wrapperFields.contacts.copyRefObject(this._refObject.contacts);
    this._wrapperFields.cost_sources.copyRefObject(this._refObject.cost_sources);
    this._wrapperFields.constraint_result.copyRefObject(this._refObject.constraint_result);
  }
  copy(other) {
    this._refObject = new GetStateValidity_ResponseRefStruct(other._refObject.toObject());
    this._validIntialized = true;
    this._wrapperFields.contacts.copy(other._wrapperFields.contacts);
    this._wrapperFields.cost_sources.copy(other._wrapperFields.cost_sources);
    this._wrapperFields.constraint_result.copy(other._wrapperFields.constraint_result);
  }
  static get classType() {
    return GetStateValidity_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"contacts","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"ContactInformation","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cost_sources","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"CostSource","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"constraint_result","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"ConstraintEvalResult","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetStateValidity_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetStateValidity_Response"};
  }
  hasMember(name) {
    let memberNames = ["valid","contacts","cost_sources","constraint_result"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetStateValidity_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetStateValidity_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetStateValidity_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetStateValidity_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetStateValidity_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetStateValidity_ResponseRefArray(size);
    this._refObject = new GetStateValidity_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetStateValidity_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetStateValidity_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetStateValidity_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetStateValidity_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetStateValidity_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetStateValidity_ResponseArrayWrapper;
  }
}
module.exports = GetStateValidity_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "contacts",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "ContactInformation",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "cost_sources",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "CostSource",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "constraint_result",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "ConstraintEvalResult",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetStateValidity_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetStateValidity_Response"
}
*/
