// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__PlannerInterfaceDescription.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PlannerInterfaceDescriptionRefStruct = StructType({
  name: primitiveTypes.string,
  pipeline_id: primitiveTypes.string,
  planner_ids: StringWrapper.refObjectArrayType,
});
const PlannerInterfaceDescriptionRefArray = ArrayType(PlannerInterfaceDescriptionRefStruct);
const PlannerInterfaceDescriptionRefStructArray = StructType({
  data: PlannerInterfaceDescriptionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PlannerInterfaceDescriptionWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._planner_idsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PlannerInterfaceDescriptionRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._pipeline_idIntialized = true;
      this._wrapperFields.pipeline_id =  new StringWrapper(other._wrapperFields.pipeline_id);
      this._wrapperFields.planner_ids = StringWrapper.createArray();
      this._wrapperFields.planner_ids.copy(other._wrapperFields.planner_ids);
      this.planner_ids = other.planner_ids;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PlannerInterfaceDescriptionRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._pipeline_idIntialized = false;
    this._wrapperFields.pipeline_id = new StringWrapper();
    this._wrapperFields.planner_ids = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PlannerInterfaceDescriptionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PlannerInterfaceDescriptionArrayWrapper;
  }
  static get ArrayType() {
    return PlannerInterfaceDescriptionArrayWrapper;
  }
  static get refObjectArrayType() {
    return PlannerInterfaceDescriptionRefStructArray
  }
  static get refObjectType() {
    return PlannerInterfaceDescriptionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in PlannerInterfaceDescription');
      }
      if (checkConsistency && !this._pipeline_idIntialized) {
        throw new TypeError('Invalid argument: pipeline_id in PlannerInterfaceDescription');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.pipeline_id.freeze(own, checkConsistency);
    }
    this._refObject.pipeline_id = this._wrapperFields.pipeline_id.refObject;
    if (!own) {
      this._wrapperFields.planner_ids.fill(this._planner_idsArray);
      this._wrapperFields.planner_ids.freeze(own, checkConsistency);
      this._refObject.planner_ids = this._wrapperFields.planner_ids.refObject;
    } else {
      this._wrapperFields.planner_ids.fill([]);
      this._wrapperFields.planner_ids.freeze(own, checkConsistency);
      this._refObject.planner_ids = this._wrapperFields.planner_ids.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.data = refObject.pipeline_id.data;
    refObject.planner_ids.data.length = refObject.planner_ids.size;
    for (let index = 0; index < refObject.planner_ids.size; index++) {
      this._planner_idsArray[index] = refObject.planner_ids.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    StringWrapper.freeStruct(refObject.pipeline_id);
    if (refObject.planner_ids.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.planner_ids);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.planner_ids, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    PlannerInterfaceDescriptionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'PlannerInterfaceDescription'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get pipeline_id() {
    if (!this._pipeline_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.pipeline_id.data;
  }
  set pipeline_id(value) {
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.data = value;
  }
  get planner_ids() {
    return this._planner_idsArray;
  }
  set planner_ids(value) {
    this._planner_idsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PlannerInterfaceDescriptionRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.copyRefObject(this._refObject.pipeline_id);
    refObject.planner_ids.data.length = refObject.planner_ids.size;
    for (let index = 0; index < refObject.planner_ids.size; index++) {
      this._planner_idsArray[index] = refObject.planner_ids.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new PlannerInterfaceDescriptionRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.copy(other._wrapperFields.pipeline_id);
    this._planner_idsArray = other._planner_idsArray.slice();
  }
  static get classType() {
    return PlannerInterfaceDescriptionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pipeline_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"planner_ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"PlannerInterfaceDescription","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PlannerInterfaceDescription"};
  }
  hasMember(name) {
    let memberNames = ["name","pipeline_id","planner_ids"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PlannerInterfaceDescriptionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PlannerInterfaceDescriptionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PlannerInterfaceDescriptionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlannerInterfaceDescriptionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlannerInterfaceDescriptionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PlannerInterfaceDescriptionRefArray(size);
    this._refObject = new PlannerInterfaceDescriptionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PlannerInterfaceDescriptionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PlannerInterfaceDescriptionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PlannerInterfaceDescriptionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PlannerInterfaceDescriptionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PlannerInterfaceDescriptionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PlannerInterfaceDescriptionArrayWrapper;
  }
}
module.exports = PlannerInterfaceDescriptionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pipeline_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "planner_ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "PlannerInterfaceDescription",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PlannerInterfaceDescription"
}
*/
