// rclnodejs/generated/moveit_msgs/moveit_msgs__action__LocalPlanner_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ConstraintsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__Constraints.js');
const LocalPlanner_GoalRefStruct = StructType({
  local_constraints: ConstraintsWrapper.refObjectArrayType,
});
const LocalPlanner_GoalRefArray = ArrayType(LocalPlanner_GoalRefStruct);
const LocalPlanner_GoalRefStructArray = StructType({
  data: LocalPlanner_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LocalPlanner_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LocalPlanner_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.local_constraints = ConstraintsWrapper.createArray();
      this._wrapperFields.local_constraints.copy(other._wrapperFields.local_constraints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LocalPlanner_GoalRefStruct();
    this._wrapperFields.local_constraints = ConstraintsWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LocalPlanner_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LocalPlanner_GoalArrayWrapper;
  }
  static get ArrayType() {
    return LocalPlanner_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return LocalPlanner_GoalRefStructArray
  }
  static get refObjectType() {
    return LocalPlanner_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.local_constraints.freeze(own, checkConsistency);
    this._refObject.local_constraints = this._wrapperFields.local_constraints.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.local_constraints.copyRefObject(refObject.local_constraints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.local_constraints.size != 0) {
      ConstraintsWrapper.ArrayType.freeArray(refObject.local_constraints);
      if (ConstraintsWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.local_constraints, ConstraintsWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LocalPlanner_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'LocalPlanner_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get local_constraints() {
    return this._wrapperFields.local_constraints;
  }
  set local_constraints(value) {
    this._wrapperFields.local_constraints.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LocalPlanner_GoalRefStruct(refObject.toObject());
    this._wrapperFields.local_constraints.copyRefObject(this._refObject.local_constraints);
  }
  copy(other) {
    this._refObject = new LocalPlanner_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.local_constraints.copy(other._wrapperFields.local_constraints);
  }
  static get classType() {
    return LocalPlanner_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"local_constraints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"Constraints","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"LocalPlanner_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LocalPlanner_Goal"};
  }
  hasMember(name) {
    let memberNames = ["local_constraints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LocalPlanner_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LocalPlanner_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LocalPlanner_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocalPlanner_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocalPlanner_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LocalPlanner_GoalRefArray(size);
    this._refObject = new LocalPlanner_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LocalPlanner_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LocalPlanner_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LocalPlanner_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LocalPlanner_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LocalPlanner_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LocalPlanner_GoalArrayWrapper;
  }
}
module.exports = LocalPlanner_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "local_constraints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "Constraints",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "LocalPlanner_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LocalPlanner_Goal"
}
*/
