// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectory.js');
const MultiDOFJointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__MultiDOFJointTrajectory.js');
const RobotTrajectoryRefStruct = StructType({
  joint_trajectory: JointTrajectoryWrapper.refObjectType,
  multi_dof_joint_trajectory: MultiDOFJointTrajectoryWrapper.refObjectType,
});
const RobotTrajectoryRefArray = ArrayType(RobotTrajectoryRefStruct);
const RobotTrajectoryRefStructArray = StructType({
  data: RobotTrajectoryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RobotTrajectoryWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RobotTrajectoryRefStruct(other._refObject.toObject());
      this._wrapperFields.joint_trajectory =  new JointTrajectoryWrapper(other._wrapperFields.joint_trajectory);
      this._wrapperFields.multi_dof_joint_trajectory =  new MultiDOFJointTrajectoryWrapper(other._wrapperFields.multi_dof_joint_trajectory);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RobotTrajectoryRefStruct();
    this._wrapperFields.joint_trajectory = new JointTrajectoryWrapper();
    this._wrapperFields.multi_dof_joint_trajectory = new MultiDOFJointTrajectoryWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RobotTrajectoryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RobotTrajectoryArrayWrapper;
  }
  static get ArrayType() {
    return RobotTrajectoryArrayWrapper;
  }
  static get refObjectArrayType() {
    return RobotTrajectoryRefStructArray
  }
  static get refObjectType() {
    return RobotTrajectoryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.joint_trajectory.freeze(own, checkConsistency);
    this._refObject.joint_trajectory = this._wrapperFields.joint_trajectory.refObject;
    this._wrapperFields.multi_dof_joint_trajectory.freeze(own, checkConsistency);
    this._refObject.multi_dof_joint_trajectory = this._wrapperFields.multi_dof_joint_trajectory.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.joint_trajectory.copyRefObject(refObject.joint_trajectory);
    this._wrapperFields.multi_dof_joint_trajectory.copyRefObject(refObject.multi_dof_joint_trajectory);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    JointTrajectoryWrapper.freeStruct(refObject.joint_trajectory);
    MultiDOFJointTrajectoryWrapper.freeStruct(refObject.multi_dof_joint_trajectory);
  }
  static destoryRawROS(msg) {
    RobotTrajectoryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'RobotTrajectory'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get joint_trajectory() {
    return this._wrapperFields.joint_trajectory;
  }
  set joint_trajectory(value) {
    if (value instanceof JointTrajectoryWrapper) {
      this._wrapperFields.joint_trajectory.copy(value);
    } else {
      this._wrapperFields.joint_trajectory.copy(new JointTrajectoryWrapper(value));
    }
  }
  get multi_dof_joint_trajectory() {
    return this._wrapperFields.multi_dof_joint_trajectory;
  }
  set multi_dof_joint_trajectory(value) {
    if (value instanceof MultiDOFJointTrajectoryWrapper) {
      this._wrapperFields.multi_dof_joint_trajectory.copy(value);
    } else {
      this._wrapperFields.multi_dof_joint_trajectory.copy(new MultiDOFJointTrajectoryWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new RobotTrajectoryRefStruct(refObject.toObject());
    this._wrapperFields.joint_trajectory.copyRefObject(this._refObject.joint_trajectory);
    this._wrapperFields.multi_dof_joint_trajectory.copyRefObject(this._refObject.multi_dof_joint_trajectory);
  }
  copy(other) {
    this._refObject = new RobotTrajectoryRefStruct(other._refObject.toObject());
    this._wrapperFields.joint_trajectory.copy(other._wrapperFields.joint_trajectory);
    this._wrapperFields.multi_dof_joint_trajectory.copy(other._wrapperFields.multi_dof_joint_trajectory);
  }
  static get classType() {
    return RobotTrajectoryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"joint_trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"multi_dof_joint_trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"MultiDOFJointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RobotTrajectory"};
  }
  hasMember(name) {
    let memberNames = ["joint_trajectory","multi_dof_joint_trajectory"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RobotTrajectoryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RobotTrajectoryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RobotTrajectoryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotTrajectoryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotTrajectoryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RobotTrajectoryRefArray(size);
    this._refObject = new RobotTrajectoryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RobotTrajectoryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RobotTrajectoryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RobotTrajectoryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RobotTrajectoryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RobotTrajectoryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RobotTrajectoryArrayWrapper;
  }
}
module.exports = RobotTrajectoryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "joint_trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "multi_dof_joint_trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "MultiDOFJointTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "RobotTrajectory",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RobotTrajectory"
}
*/
