// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetPositionFK_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const GetPositionFK_RequestRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  fk_link_names: StringWrapper.refObjectArrayType,
  robot_state: RobotStateWrapper.refObjectType,
});
const GetPositionFK_RequestRefArray = ArrayType(GetPositionFK_RequestRefStruct);
const GetPositionFK_RequestRefStructArray = StructType({
  data: GetPositionFK_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPositionFK_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._fk_link_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPositionFK_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.fk_link_names = StringWrapper.createArray();
      this._wrapperFields.fk_link_names.copy(other._wrapperFields.fk_link_names);
      this.fk_link_names = other.fk_link_names;
      this._wrapperFields.robot_state =  new RobotStateWrapper(other._wrapperFields.robot_state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPositionFK_RequestRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.fk_link_names = StringWrapper.createArray();
    this._wrapperFields.robot_state = new RobotStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPositionFK_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPositionFK_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetPositionFK_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPositionFK_RequestRefStructArray
  }
  static get refObjectType() {
    return GetPositionFK_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (!own) {
      this._wrapperFields.fk_link_names.fill(this._fk_link_namesArray);
      this._wrapperFields.fk_link_names.freeze(own, checkConsistency);
      this._refObject.fk_link_names = this._wrapperFields.fk_link_names.refObject;
    } else {
      this._wrapperFields.fk_link_names.fill([]);
      this._wrapperFields.fk_link_names.freeze(own, checkConsistency);
      this._refObject.fk_link_names = this._wrapperFields.fk_link_names.refObject;
    }
    this._wrapperFields.robot_state.freeze(own, checkConsistency);
    this._refObject.robot_state = this._wrapperFields.robot_state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.fk_link_names.data.length = refObject.fk_link_names.size;
    for (let index = 0; index < refObject.fk_link_names.size; index++) {
      this._fk_link_namesArray[index] = refObject.fk_link_names.data[index].data;
    }
    this._wrapperFields.robot_state.copyRefObject(refObject.robot_state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.fk_link_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.fk_link_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.fk_link_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    RobotStateWrapper.freeStruct(refObject.robot_state);
  }
  static destoryRawROS(msg) {
    GetPositionFK_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetPositionFK_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get fk_link_names() {
    return this._fk_link_namesArray;
  }
  set fk_link_names(value) {
    this._fk_link_namesArray = value;
  }
  get robot_state() {
    return this._wrapperFields.robot_state;
  }
  set robot_state(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.robot_state.copy(value);
    } else {
      this._wrapperFields.robot_state.copy(new RobotStateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetPositionFK_RequestRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.fk_link_names.data.length = refObject.fk_link_names.size;
    for (let index = 0; index < refObject.fk_link_names.size; index++) {
      this._fk_link_namesArray[index] = refObject.fk_link_names.data[index].data;
    }
    this._wrapperFields.robot_state.copyRefObject(this._refObject.robot_state);
  }
  copy(other) {
    this._refObject = new GetPositionFK_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._fk_link_namesArray = other._fk_link_namesArray.slice();
    this._wrapperFields.robot_state.copy(other._wrapperFields.robot_state);
  }
  static get classType() {
    return GetPositionFK_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fk_link_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetPositionFK_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPositionFK_Request"};
  }
  hasMember(name) {
    let memberNames = ["header","fk_link_names","robot_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPositionFK_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPositionFK_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPositionFK_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPositionFK_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPositionFK_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPositionFK_RequestRefArray(size);
    this._refObject = new GetPositionFK_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPositionFK_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPositionFK_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPositionFK_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPositionFK_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPositionFK_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPositionFK_RequestArrayWrapper;
  }
}
module.exports = GetPositionFK_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "fk_link_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "robot_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetPositionFK_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPositionFK_Request"
}
*/
