// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__BoundingVolume.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SolidPrimitiveWrapper = require('../../generated/shape_msgs/shape_msgs__msg__SolidPrimitive.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const MeshWrapper = require('../../generated/shape_msgs/shape_msgs__msg__Mesh.js');
const BoundingVolumeRefStruct = StructType({
  primitives: SolidPrimitiveWrapper.refObjectArrayType,
  primitive_poses: PoseWrapper.refObjectArrayType,
  meshes: MeshWrapper.refObjectArrayType,
  mesh_poses: PoseWrapper.refObjectArrayType,
});
const BoundingVolumeRefArray = ArrayType(BoundingVolumeRefStruct);
const BoundingVolumeRefStructArray = StructType({
  data: BoundingVolumeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BoundingVolumeWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BoundingVolumeRefStruct(other._refObject.toObject());
      this._wrapperFields.primitives = SolidPrimitiveWrapper.createArray();
      this._wrapperFields.primitives.copy(other._wrapperFields.primitives);
      this._wrapperFields.primitive_poses = PoseWrapper.createArray();
      this._wrapperFields.primitive_poses.copy(other._wrapperFields.primitive_poses);
      this._wrapperFields.meshes = MeshWrapper.createArray();
      this._wrapperFields.meshes.copy(other._wrapperFields.meshes);
      this._wrapperFields.mesh_poses = PoseWrapper.createArray();
      this._wrapperFields.mesh_poses.copy(other._wrapperFields.mesh_poses);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BoundingVolumeRefStruct();
    this._wrapperFields.primitives = SolidPrimitiveWrapper.createArray();
    this._wrapperFields.primitive_poses = PoseWrapper.createArray();
    this._wrapperFields.meshes = MeshWrapper.createArray();
    this._wrapperFields.mesh_poses = PoseWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new BoundingVolumeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BoundingVolumeArrayWrapper;
  }
  static get ArrayType() {
    return BoundingVolumeArrayWrapper;
  }
  static get refObjectArrayType() {
    return BoundingVolumeRefStructArray
  }
  static get refObjectType() {
    return BoundingVolumeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.primitives.freeze(own, checkConsistency);
    this._refObject.primitives = this._wrapperFields.primitives.refObject;
    this._wrapperFields.primitive_poses.freeze(own, checkConsistency);
    this._refObject.primitive_poses = this._wrapperFields.primitive_poses.refObject;
    this._wrapperFields.meshes.freeze(own, checkConsistency);
    this._refObject.meshes = this._wrapperFields.meshes.refObject;
    this._wrapperFields.mesh_poses.freeze(own, checkConsistency);
    this._refObject.mesh_poses = this._wrapperFields.mesh_poses.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.primitives.copyRefObject(refObject.primitives);
    this._wrapperFields.primitive_poses.copyRefObject(refObject.primitive_poses);
    this._wrapperFields.meshes.copyRefObject(refObject.meshes);
    this._wrapperFields.mesh_poses.copyRefObject(refObject.mesh_poses);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.primitives.size != 0) {
      SolidPrimitiveWrapper.ArrayType.freeArray(refObject.primitives);
      if (SolidPrimitiveWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.primitives, SolidPrimitiveWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.primitive_poses.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.primitive_poses);
      if (PoseWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.primitive_poses, PoseWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.meshes.size != 0) {
      MeshWrapper.ArrayType.freeArray(refObject.meshes);
      if (MeshWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.meshes, MeshWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.mesh_poses.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.mesh_poses);
      if (PoseWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.mesh_poses, PoseWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    BoundingVolumeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'BoundingVolume'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get primitives() {
    return this._wrapperFields.primitives;
  }
  set primitives(value) {
    this._wrapperFields.primitives.fill(value);
  }
  get primitive_poses() {
    return this._wrapperFields.primitive_poses;
  }
  set primitive_poses(value) {
    this._wrapperFields.primitive_poses.fill(value);
  }
  get meshes() {
    return this._wrapperFields.meshes;
  }
  set meshes(value) {
    this._wrapperFields.meshes.fill(value);
  }
  get mesh_poses() {
    return this._wrapperFields.mesh_poses;
  }
  set mesh_poses(value) {
    this._wrapperFields.mesh_poses.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new BoundingVolumeRefStruct(refObject.toObject());
    this._wrapperFields.primitives.copyRefObject(this._refObject.primitives);
    this._wrapperFields.primitive_poses.copyRefObject(this._refObject.primitive_poses);
    this._wrapperFields.meshes.copyRefObject(this._refObject.meshes);
    this._wrapperFields.mesh_poses.copyRefObject(this._refObject.mesh_poses);
  }
  copy(other) {
    this._refObject = new BoundingVolumeRefStruct(other._refObject.toObject());
    this._wrapperFields.primitives.copy(other._wrapperFields.primitives);
    this._wrapperFields.primitive_poses.copy(other._wrapperFields.primitive_poses);
    this._wrapperFields.meshes.copy(other._wrapperFields.meshes);
    this._wrapperFields.mesh_poses.copy(other._wrapperFields.mesh_poses);
  }
  static get classType() {
    return BoundingVolumeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"primitives","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"shape_msgs","type":"SolidPrimitive","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"primitive_poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"meshes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"shape_msgs","type":"Mesh","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"mesh_poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"BoundingVolume","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BoundingVolume"};
  }
  hasMember(name) {
    let memberNames = ["primitives","primitive_poses","meshes","mesh_poses"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BoundingVolumeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BoundingVolumeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BoundingVolumeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingVolumeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingVolumeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BoundingVolumeRefArray(size);
    this._refObject = new BoundingVolumeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BoundingVolumeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BoundingVolumeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BoundingVolumeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BoundingVolumeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BoundingVolumeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BoundingVolumeArrayWrapper;
  }
}
module.exports = BoundingVolumeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "primitives",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "shape_msgs",
        "type": "SolidPrimitive",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "primitive_poses",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "meshes",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "shape_msgs",
        "type": "Mesh",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "mesh_poses",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "BoundingVolume",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BoundingVolume"
}
*/
