// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ExecuteKnownTrajectory_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const ExecuteKnownTrajectory_ResponseRefStruct = StructType({
  error_code: MoveItErrorCodesWrapper.refObjectType,
});
const ExecuteKnownTrajectory_ResponseRefArray = ArrayType(ExecuteKnownTrajectory_ResponseRefStruct);
const ExecuteKnownTrajectory_ResponseRefStructArray = StructType({
  data: ExecuteKnownTrajectory_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ExecuteKnownTrajectory_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ExecuteKnownTrajectory_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ExecuteKnownTrajectory_ResponseRefStruct();
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ExecuteKnownTrajectory_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ExecuteKnownTrajectory_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ExecuteKnownTrajectory_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ExecuteKnownTrajectory_ResponseRefStructArray
  }
  static get refObjectType() {
    return ExecuteKnownTrajectory_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.error_code.freeze(own, checkConsistency);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
  }
  static destoryRawROS(msg) {
    ExecuteKnownTrajectory_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ExecuteKnownTrajectory_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ExecuteKnownTrajectory_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
  }
  copy(other) {
    this._refObject = new ExecuteKnownTrajectory_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
  }
  static get classType() {
    return ExecuteKnownTrajectory_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ExecuteKnownTrajectory_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ExecuteKnownTrajectory_Response"};
  }
  hasMember(name) {
    let memberNames = ["error_code"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ExecuteKnownTrajectory_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ExecuteKnownTrajectory_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ExecuteKnownTrajectory_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExecuteKnownTrajectory_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExecuteKnownTrajectory_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ExecuteKnownTrajectory_ResponseRefArray(size);
    this._refObject = new ExecuteKnownTrajectory_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ExecuteKnownTrajectory_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ExecuteKnownTrajectory_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ExecuteKnownTrajectory_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ExecuteKnownTrajectory_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ExecuteKnownTrajectory_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ExecuteKnownTrajectory_ResponseArrayWrapper;
  }
}
module.exports = ExecuteKnownTrajectory_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "error_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MoveItErrorCodes",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ExecuteKnownTrajectory_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ExecuteKnownTrajectory_Response"
}
*/
