// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__DisplayRobotState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const ObjectColorWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__ObjectColor.js');
const DisplayRobotStateRefStruct = StructType({
  state: RobotStateWrapper.refObjectType,
  highlight_links: ObjectColorWrapper.refObjectArrayType,
  hide: primitiveTypes.bool,
});
const DisplayRobotStateRefArray = ArrayType(DisplayRobotStateRefStruct);
const DisplayRobotStateRefStructArray = StructType({
  data: DisplayRobotStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DisplayRobotStateWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DisplayRobotStateRefStruct(other._refObject.toObject());
      this._wrapperFields.state =  new RobotStateWrapper(other._wrapperFields.state);
      this._wrapperFields.highlight_links = ObjectColorWrapper.createArray();
      this._wrapperFields.highlight_links.copy(other._wrapperFields.highlight_links);
      this._hideIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DisplayRobotStateRefStruct();
    this._wrapperFields.state = new RobotStateWrapper();
    this._wrapperFields.highlight_links = ObjectColorWrapper.createArray();
    this._hideIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new DisplayRobotStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DisplayRobotStateArrayWrapper;
  }
  static get ArrayType() {
    return DisplayRobotStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return DisplayRobotStateRefStructArray
  }
  static get refObjectType() {
    return DisplayRobotStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._hideIntialized) {
        throw new TypeError('Invalid argument: hide in DisplayRobotState');
      }
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
    this._wrapperFields.highlight_links.freeze(own, checkConsistency);
    this._refObject.highlight_links = this._wrapperFields.highlight_links.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.state.copyRefObject(refObject.state);
    this._wrapperFields.highlight_links.copyRefObject(refObject.highlight_links);
    this._hideIntialized = true;
    this._refObject.hide = refObject.hide;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RobotStateWrapper.freeStruct(refObject.state);
    if (refObject.highlight_links.size != 0) {
      ObjectColorWrapper.ArrayType.freeArray(refObject.highlight_links);
      if (ObjectColorWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.highlight_links, ObjectColorWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    DisplayRobotStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'DisplayRobotState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new RobotStateWrapper(value));
    }
  }
  get highlight_links() {
    return this._wrapperFields.highlight_links;
  }
  set highlight_links(value) {
    this._wrapperFields.highlight_links.fill(value);
  }
  get hide() {
    if (!this._hideIntialized) {
      return undefined;
    }
    return this._refObject.hide;
  }
  set hide(value) {
    this._hideIntialized = true;
    this._refObject.hide = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DisplayRobotStateRefStruct(refObject.toObject());
    this._wrapperFields.state.copyRefObject(this._refObject.state);
    this._wrapperFields.highlight_links.copyRefObject(this._refObject.highlight_links);
    this._hideIntialized = true;
  }
  copy(other) {
    this._refObject = new DisplayRobotStateRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
    this._wrapperFields.highlight_links.copy(other._wrapperFields.highlight_links);
    this._hideIntialized = true;
  }
  static get classType() {
    return DisplayRobotStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"highlight_links","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"ObjectColor","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"hide","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"DisplayRobotState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DisplayRobotState"};
  }
  hasMember(name) {
    let memberNames = ["state","highlight_links","hide"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DisplayRobotStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DisplayRobotStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DisplayRobotStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DisplayRobotStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DisplayRobotStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DisplayRobotStateRefArray(size);
    this._refObject = new DisplayRobotStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DisplayRobotStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DisplayRobotStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DisplayRobotStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DisplayRobotStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DisplayRobotStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DisplayRobotStateArrayWrapper;
  }
}
module.exports = DisplayRobotStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "highlight_links",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "ObjectColor",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "hide",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "DisplayRobotState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DisplayRobotState"
}
*/
