// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__SetPlannerParams_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PlannerParamsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlannerParams.js');
const SetPlannerParams_RequestRefStruct = StructType({
  pipeline_id: primitiveTypes.string,
  planner_config: primitiveTypes.string,
  group: primitiveTypes.string,
  params: PlannerParamsWrapper.refObjectType,
  replace: primitiveTypes.bool,
});
const SetPlannerParams_RequestRefArray = ArrayType(SetPlannerParams_RequestRefStruct);
const SetPlannerParams_RequestRefStructArray = StructType({
  data: SetPlannerParams_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetPlannerParams_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetPlannerParams_RequestRefStruct(other._refObject.toObject());
      this._pipeline_idIntialized = true;
      this._wrapperFields.pipeline_id =  new StringWrapper(other._wrapperFields.pipeline_id);
      this._planner_configIntialized = true;
      this._wrapperFields.planner_config =  new StringWrapper(other._wrapperFields.planner_config);
      this._groupIntialized = true;
      this._wrapperFields.group =  new StringWrapper(other._wrapperFields.group);
      this._wrapperFields.params =  new PlannerParamsWrapper(other._wrapperFields.params);
      this._replaceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetPlannerParams_RequestRefStruct();
    this._pipeline_idIntialized = false;
    this._wrapperFields.pipeline_id = new StringWrapper();
    this._planner_configIntialized = false;
    this._wrapperFields.planner_config = new StringWrapper();
    this._groupIntialized = false;
    this._wrapperFields.group = new StringWrapper();
    this._wrapperFields.params = new PlannerParamsWrapper();
    this._replaceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetPlannerParams_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetPlannerParams_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetPlannerParams_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetPlannerParams_RequestRefStructArray
  }
  static get refObjectType() {
    return SetPlannerParams_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._pipeline_idIntialized) {
        throw new TypeError('Invalid argument: pipeline_id in SetPlannerParams_Request');
      }
      if (checkConsistency && !this._planner_configIntialized) {
        throw new TypeError('Invalid argument: planner_config in SetPlannerParams_Request');
      }
      if (checkConsistency && !this._groupIntialized) {
        throw new TypeError('Invalid argument: group in SetPlannerParams_Request');
      }
      if (checkConsistency && !this._replaceIntialized) {
        throw new TypeError('Invalid argument: replace in SetPlannerParams_Request');
      }
    if (own) {
      this._wrapperFields.pipeline_id.freeze(own, checkConsistency);
    }
    this._refObject.pipeline_id = this._wrapperFields.pipeline_id.refObject;
    if (own) {
      this._wrapperFields.planner_config.freeze(own, checkConsistency);
    }
    this._refObject.planner_config = this._wrapperFields.planner_config.refObject;
    if (own) {
      this._wrapperFields.group.freeze(own, checkConsistency);
    }
    this._refObject.group = this._wrapperFields.group.refObject;
    this._wrapperFields.params.freeze(own, checkConsistency);
    this._refObject.params = this._wrapperFields.params.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.data = refObject.pipeline_id.data;
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.data = refObject.planner_config.data;
    this._groupIntialized = true;
    this._wrapperFields.group.data = refObject.group.data;
    this._wrapperFields.params.copyRefObject(refObject.params);
    this._replaceIntialized = true;
    this._refObject.replace = refObject.replace;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.pipeline_id);
    StringWrapper.freeStruct(refObject.planner_config);
    StringWrapper.freeStruct(refObject.group);
    PlannerParamsWrapper.freeStruct(refObject.params);
  }
  static destoryRawROS(msg) {
    SetPlannerParams_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'SetPlannerParams_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pipeline_id() {
    if (!this._pipeline_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.pipeline_id.data;
  }
  set pipeline_id(value) {
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.data = value;
  }
  get planner_config() {
    if (!this._planner_configIntialized) {
      return undefined;
    }
    return this._wrapperFields.planner_config.data;
  }
  set planner_config(value) {
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.data = value;
  }
  get group() {
    if (!this._groupIntialized) {
      return undefined;
    }
    return this._wrapperFields.group.data;
  }
  set group(value) {
    this._groupIntialized = true;
    this._wrapperFields.group.data = value;
  }
  get params() {
    return this._wrapperFields.params;
  }
  set params(value) {
    if (value instanceof PlannerParamsWrapper) {
      this._wrapperFields.params.copy(value);
    } else {
      this._wrapperFields.params.copy(new PlannerParamsWrapper(value));
    }
  }
  get replace() {
    if (!this._replaceIntialized) {
      return undefined;
    }
    return this._refObject.replace;
  }
  set replace(value) {
    this._replaceIntialized = true;
    this._refObject.replace = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetPlannerParams_RequestRefStruct(refObject.toObject());
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.copyRefObject(this._refObject.pipeline_id);
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.copyRefObject(this._refObject.planner_config);
    this._groupIntialized = true;
    this._wrapperFields.group.copyRefObject(this._refObject.group);
    this._wrapperFields.params.copyRefObject(this._refObject.params);
    this._replaceIntialized = true;
  }
  copy(other) {
    this._refObject = new SetPlannerParams_RequestRefStruct(other._refObject.toObject());
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.copy(other._wrapperFields.pipeline_id);
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.copy(other._wrapperFields.planner_config);
    this._groupIntialized = true;
    this._wrapperFields.group.copy(other._wrapperFields.group);
    this._wrapperFields.params.copy(other._wrapperFields.params);
    this._replaceIntialized = true;
  }
  static get classType() {
    return SetPlannerParams_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pipeline_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"planner_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"group","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"params","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlannerParams","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"replace","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"SetPlannerParams_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetPlannerParams_Request"};
  }
  hasMember(name) {
    let memberNames = ["pipeline_id","planner_config","group","params","replace"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetPlannerParams_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetPlannerParams_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetPlannerParams_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPlannerParams_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPlannerParams_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetPlannerParams_RequestRefArray(size);
    this._refObject = new SetPlannerParams_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetPlannerParams_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetPlannerParams_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetPlannerParams_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetPlannerParams_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetPlannerParams_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetPlannerParams_RequestArrayWrapper;
  }
}
module.exports = SetPlannerParams_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pipeline_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "planner_config",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "group",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "params",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "PlannerParams",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "replace",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "SetPlannerParams_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetPlannerParams_Request"
}
*/
