// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__PlannerParams.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PlannerParamsRefStruct = StructType({
  keys: StringWrapper.refObjectArrayType,
  values: StringWrapper.refObjectArrayType,
  descriptions: StringWrapper.refObjectArrayType,
});
const PlannerParamsRefArray = ArrayType(PlannerParamsRefStruct);
const PlannerParamsRefStructArray = StructType({
  data: PlannerParamsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PlannerParamsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._keysArray = [];
    this._valuesArray = [];
    this._descriptionsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PlannerParamsRefStruct(other._refObject.toObject());
      this._wrapperFields.keys = StringWrapper.createArray();
      this._wrapperFields.keys.copy(other._wrapperFields.keys);
      this.keys = other.keys;
      this._wrapperFields.values = StringWrapper.createArray();
      this._wrapperFields.values.copy(other._wrapperFields.values);
      this.values = other.values;
      this._wrapperFields.descriptions = StringWrapper.createArray();
      this._wrapperFields.descriptions.copy(other._wrapperFields.descriptions);
      this.descriptions = other.descriptions;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PlannerParamsRefStruct();
    this._wrapperFields.keys = StringWrapper.createArray();
    this._wrapperFields.values = StringWrapper.createArray();
    this._wrapperFields.descriptions = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PlannerParamsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PlannerParamsArrayWrapper;
  }
  static get ArrayType() {
    return PlannerParamsArrayWrapper;
  }
  static get refObjectArrayType() {
    return PlannerParamsRefStructArray
  }
  static get refObjectType() {
    return PlannerParamsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (!own) {
      this._wrapperFields.keys.fill(this._keysArray);
      this._wrapperFields.keys.freeze(own, checkConsistency);
      this._refObject.keys = this._wrapperFields.keys.refObject;
    } else {
      this._wrapperFields.keys.fill([]);
      this._wrapperFields.keys.freeze(own, checkConsistency);
      this._refObject.keys = this._wrapperFields.keys.refObject;
    }
    if (!own) {
      this._wrapperFields.values.fill(this._valuesArray);
      this._wrapperFields.values.freeze(own, checkConsistency);
      this._refObject.values = this._wrapperFields.values.refObject;
    } else {
      this._wrapperFields.values.fill([]);
      this._wrapperFields.values.freeze(own, checkConsistency);
      this._refObject.values = this._wrapperFields.values.refObject;
    }
    if (!own) {
      this._wrapperFields.descriptions.fill(this._descriptionsArray);
      this._wrapperFields.descriptions.freeze(own, checkConsistency);
      this._refObject.descriptions = this._wrapperFields.descriptions.refObject;
    } else {
      this._wrapperFields.descriptions.fill([]);
      this._wrapperFields.descriptions.freeze(own, checkConsistency);
      this._refObject.descriptions = this._wrapperFields.descriptions.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.keys.data.length = refObject.keys.size;
    for (let index = 0; index < refObject.keys.size; index++) {
      this._keysArray[index] = refObject.keys.data[index].data;
    }
    refObject.values.data.length = refObject.values.size;
    for (let index = 0; index < refObject.values.size; index++) {
      this._valuesArray[index] = refObject.values.data[index].data;
    }
    refObject.descriptions.data.length = refObject.descriptions.size;
    for (let index = 0; index < refObject.descriptions.size; index++) {
      this._descriptionsArray[index] = refObject.descriptions.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.keys.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.keys);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.keys, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.values.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.values);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.values, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.descriptions.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.descriptions);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.descriptions, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    PlannerParamsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'PlannerParams'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get keys() {
    return this._keysArray;
  }
  set keys(value) {
    this._keysArray = value;
  }
  get values() {
    return this._valuesArray;
  }
  set values(value) {
    this._valuesArray = value;
  }
  get descriptions() {
    return this._descriptionsArray;
  }
  set descriptions(value) {
    this._descriptionsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PlannerParamsRefStruct(refObject.toObject());
    refObject.keys.data.length = refObject.keys.size;
    for (let index = 0; index < refObject.keys.size; index++) {
      this._keysArray[index] = refObject.keys.data[index].data;
    }
    refObject.values.data.length = refObject.values.size;
    for (let index = 0; index < refObject.values.size; index++) {
      this._valuesArray[index] = refObject.values.data[index].data;
    }
    refObject.descriptions.data.length = refObject.descriptions.size;
    for (let index = 0; index < refObject.descriptions.size; index++) {
      this._descriptionsArray[index] = refObject.descriptions.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new PlannerParamsRefStruct(other._refObject.toObject());
    this._keysArray = other._keysArray.slice();
    this._valuesArray = other._valuesArray.slice();
    this._descriptionsArray = other._descriptionsArray.slice();
  }
  static get classType() {
    return PlannerParamsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"keys","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"descriptions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"PlannerParams","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PlannerParams"};
  }
  hasMember(name) {
    let memberNames = ["keys","values","descriptions"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PlannerParamsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PlannerParamsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PlannerParamsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlannerParamsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlannerParamsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PlannerParamsRefArray(size);
    this._refObject = new PlannerParamsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PlannerParamsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PlannerParamsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PlannerParamsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PlannerParamsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PlannerParamsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PlannerParamsArrayWrapper;
  }
}
module.exports = PlannerParamsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "keys",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "values",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "descriptions",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "PlannerParams",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PlannerParams"
}
*/
