// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__CostSource.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const CostSourceRefStruct = StructType({
  cost_density: primitiveTypes.float64,
  aabb_min: Vector3Wrapper.refObjectType,
  aabb_max: Vector3Wrapper.refObjectType,
});
const CostSourceRefArray = ArrayType(CostSourceRefStruct);
const CostSourceRefStructArray = StructType({
  data: CostSourceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CostSourceWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CostSourceRefStruct(other._refObject.toObject());
      this._cost_densityIntialized = true;
      this._wrapperFields.aabb_min =  new Vector3Wrapper(other._wrapperFields.aabb_min);
      this._wrapperFields.aabb_max =  new Vector3Wrapper(other._wrapperFields.aabb_max);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CostSourceRefStruct();
    this._cost_densityIntialized = false;
    this._wrapperFields.aabb_min = new Vector3Wrapper();
    this._wrapperFields.aabb_max = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CostSourceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CostSourceArrayWrapper;
  }
  static get ArrayType() {
    return CostSourceArrayWrapper;
  }
  static get refObjectArrayType() {
    return CostSourceRefStructArray
  }
  static get refObjectType() {
    return CostSourceRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._cost_densityIntialized) {
        throw new TypeError('Invalid argument: cost_density in CostSource');
      }
    this._wrapperFields.aabb_min.freeze(own, checkConsistency);
    this._refObject.aabb_min = this._wrapperFields.aabb_min.refObject;
    this._wrapperFields.aabb_max.freeze(own, checkConsistency);
    this._refObject.aabb_max = this._wrapperFields.aabb_max.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._cost_densityIntialized = true;
    this._refObject.cost_density = refObject.cost_density;
    this._wrapperFields.aabb_min.copyRefObject(refObject.aabb_min);
    this._wrapperFields.aabb_max.copyRefObject(refObject.aabb_max);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.aabb_min);
    Vector3Wrapper.freeStruct(refObject.aabb_max);
  }
  static destoryRawROS(msg) {
    CostSourceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'CostSource'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get cost_density() {
    if (!this._cost_densityIntialized) {
      return undefined;
    }
    return this._refObject.cost_density;
  }
  set cost_density(value) {
    this._cost_densityIntialized = true;
    this._refObject.cost_density = value;
  }
  get aabb_min() {
    return this._wrapperFields.aabb_min;
  }
  set aabb_min(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.aabb_min.copy(value);
    } else {
      this._wrapperFields.aabb_min.copy(new Vector3Wrapper(value));
    }
  }
  get aabb_max() {
    return this._wrapperFields.aabb_max;
  }
  set aabb_max(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.aabb_max.copy(value);
    } else {
      this._wrapperFields.aabb_max.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CostSourceRefStruct(refObject.toObject());
    this._cost_densityIntialized = true;
    this._wrapperFields.aabb_min.copyRefObject(this._refObject.aabb_min);
    this._wrapperFields.aabb_max.copyRefObject(this._refObject.aabb_max);
  }
  copy(other) {
    this._refObject = new CostSourceRefStruct(other._refObject.toObject());
    this._cost_densityIntialized = true;
    this._wrapperFields.aabb_min.copy(other._wrapperFields.aabb_min);
    this._wrapperFields.aabb_max.copy(other._wrapperFields.aabb_max);
  }
  static get classType() {
    return CostSourceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"cost_density","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aabb_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"aabb_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"CostSource","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CostSource"};
  }
  hasMember(name) {
    let memberNames = ["cost_density","aabb_min","aabb_max"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CostSourceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CostSourceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CostSourceWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CostSourceArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CostSourceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CostSourceRefArray(size);
    this._refObject = new CostSourceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CostSourceWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CostSourceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CostSourceArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CostSourceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CostSourceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CostSourceArrayWrapper;
  }
}
module.exports = CostSourceWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "cost_density",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aabb_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "aabb_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "CostSource",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CostSource"
}
*/
