// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetPositionIK_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PositionIKRequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PositionIKRequest.js');
const GetPositionIK_RequestRefStruct = StructType({
  ik_request: PositionIKRequestWrapper.refObjectType,
});
const GetPositionIK_RequestRefArray = ArrayType(GetPositionIK_RequestRefStruct);
const GetPositionIK_RequestRefStructArray = StructType({
  data: GetPositionIK_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPositionIK_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPositionIK_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.ik_request =  new PositionIKRequestWrapper(other._wrapperFields.ik_request);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPositionIK_RequestRefStruct();
    this._wrapperFields.ik_request = new PositionIKRequestWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPositionIK_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPositionIK_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetPositionIK_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPositionIK_RequestRefStructArray
  }
  static get refObjectType() {
    return GetPositionIK_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.ik_request.freeze(own, checkConsistency);
    this._refObject.ik_request = this._wrapperFields.ik_request.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.ik_request.copyRefObject(refObject.ik_request);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PositionIKRequestWrapper.freeStruct(refObject.ik_request);
  }
  static destoryRawROS(msg) {
    GetPositionIK_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetPositionIK_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ik_request() {
    return this._wrapperFields.ik_request;
  }
  set ik_request(value) {
    if (value instanceof PositionIKRequestWrapper) {
      this._wrapperFields.ik_request.copy(value);
    } else {
      this._wrapperFields.ik_request.copy(new PositionIKRequestWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetPositionIK_RequestRefStruct(refObject.toObject());
    this._wrapperFields.ik_request.copyRefObject(this._refObject.ik_request);
  }
  copy(other) {
    this._refObject = new GetPositionIK_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.ik_request.copy(other._wrapperFields.ik_request);
  }
  static get classType() {
    return GetPositionIK_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ik_request","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PositionIKRequest","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetPositionIK_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPositionIK_Request"};
  }
  hasMember(name) {
    let memberNames = ["ik_request"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPositionIK_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPositionIK_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPositionIK_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPositionIK_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPositionIK_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPositionIK_RequestRefArray(size);
    this._refObject = new GetPositionIK_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPositionIK_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPositionIK_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPositionIK_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPositionIK_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPositionIK_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPositionIK_RequestArrayWrapper;
  }
}
module.exports = GetPositionIK_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ik_request",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "PositionIKRequest",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetPositionIK_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPositionIK_Request"
}
*/
