// rclnodejs/generated/moveit_msgs/moveit_msgs__action__GlobalPlanner.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class moveit_msgs__action__GlobalPlanner {
  static get Goal() {
    return require('./moveit_msgs__action__GlobalPlanner_Goal.js');
  }
  static get Result() {
    return require('./moveit_msgs__action__GlobalPlanner_Result.js');
  }
  static get Feedback() {
    return require('./moveit_msgs__action__GlobalPlanner_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./moveit_msgs__action__GlobalPlanner_SendGoal.js');
      },
      get GetResultService() {
        return require('./moveit_msgs__action__GlobalPlanner_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./moveit_msgs__action__GlobalPlanner_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'GlobalPlanner'};
  }
}
module.exports = moveit_msgs__action__GlobalPlanner;
