// rclnodejs/generated/moveit_msgs/moveit_msgs__action__LocalPlanner_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LocalPlanner_FeedbackRefStruct = StructType({
  feedback: primitiveTypes.string,
});
const LocalPlanner_FeedbackRefArray = ArrayType(LocalPlanner_FeedbackRefStruct);
const LocalPlanner_FeedbackRefStructArray = StructType({
  data: LocalPlanner_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LocalPlanner_FeedbackWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LocalPlanner_FeedbackRefStruct(other._refObject.toObject());
      this._feedbackIntialized = true;
      this._wrapperFields.feedback =  new StringWrapper(other._wrapperFields.feedback);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LocalPlanner_FeedbackRefStruct();
    this._feedbackIntialized = false;
    this._wrapperFields.feedback = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LocalPlanner_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LocalPlanner_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return LocalPlanner_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return LocalPlanner_FeedbackRefStructArray
  }
  static get refObjectType() {
    return LocalPlanner_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._feedbackIntialized) {
        throw new TypeError('Invalid argument: feedback in LocalPlanner_Feedback');
      }
    if (own) {
      this._wrapperFields.feedback.freeze(own, checkConsistency);
    }
    this._refObject.feedback = this._wrapperFields.feedback.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._feedbackIntialized = true;
    this._wrapperFields.feedback.data = refObject.feedback.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.feedback);
  }
  static destoryRawROS(msg) {
    LocalPlanner_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'LocalPlanner_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get feedback() {
    if (!this._feedbackIntialized) {
      return undefined;
    }
    return this._wrapperFields.feedback.data;
  }
  set feedback(value) {
    this._feedbackIntialized = true;
    this._wrapperFields.feedback.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LocalPlanner_FeedbackRefStruct(refObject.toObject());
    this._feedbackIntialized = true;
    this._wrapperFields.feedback.copyRefObject(this._refObject.feedback);
  }
  copy(other) {
    this._refObject = new LocalPlanner_FeedbackRefStruct(other._refObject.toObject());
    this._feedbackIntialized = true;
    this._wrapperFields.feedback.copy(other._wrapperFields.feedback);
  }
  static get classType() {
    return LocalPlanner_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"feedback","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"LocalPlanner_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LocalPlanner_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["feedback"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LocalPlanner_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LocalPlanner_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LocalPlanner_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocalPlanner_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocalPlanner_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LocalPlanner_FeedbackRefArray(size);
    this._refObject = new LocalPlanner_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LocalPlanner_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LocalPlanner_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LocalPlanner_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LocalPlanner_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LocalPlanner_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LocalPlanner_FeedbackArrayWrapper;
  }
}
module.exports = LocalPlanner_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "feedback",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "LocalPlanner_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LocalPlanner_Feedback"
}
*/
