// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ChangeDriftDimensions_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TransformWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Transform.js');
const ChangeDriftDimensions_RequestRefStruct = StructType({
  drift_x_translation: primitiveTypes.bool,
  drift_y_translation: primitiveTypes.bool,
  drift_z_translation: primitiveTypes.bool,
  drift_x_rotation: primitiveTypes.bool,
  drift_y_rotation: primitiveTypes.bool,
  drift_z_rotation: primitiveTypes.bool,
  transform_jog_frame_to_drift_frame: TransformWrapper.refObjectType,
});
const ChangeDriftDimensions_RequestRefArray = ArrayType(ChangeDriftDimensions_RequestRefStruct);
const ChangeDriftDimensions_RequestRefStructArray = StructType({
  data: ChangeDriftDimensions_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ChangeDriftDimensions_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ChangeDriftDimensions_RequestRefStruct(other._refObject.toObject());
      this._drift_x_translationIntialized = true;
      this._drift_y_translationIntialized = true;
      this._drift_z_translationIntialized = true;
      this._drift_x_rotationIntialized = true;
      this._drift_y_rotationIntialized = true;
      this._drift_z_rotationIntialized = true;
      this._wrapperFields.transform_jog_frame_to_drift_frame =  new TransformWrapper(other._wrapperFields.transform_jog_frame_to_drift_frame);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ChangeDriftDimensions_RequestRefStruct();
    this._drift_x_translationIntialized = false;
    this._drift_y_translationIntialized = false;
    this._drift_z_translationIntialized = false;
    this._drift_x_rotationIntialized = false;
    this._drift_y_rotationIntialized = false;
    this._drift_z_rotationIntialized = false;
    this._wrapperFields.transform_jog_frame_to_drift_frame = new TransformWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ChangeDriftDimensions_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ChangeDriftDimensions_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ChangeDriftDimensions_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ChangeDriftDimensions_RequestRefStructArray
  }
  static get refObjectType() {
    return ChangeDriftDimensions_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._drift_x_translationIntialized) {
        throw new TypeError('Invalid argument: drift_x_translation in ChangeDriftDimensions_Request');
      }
      if (checkConsistency && !this._drift_y_translationIntialized) {
        throw new TypeError('Invalid argument: drift_y_translation in ChangeDriftDimensions_Request');
      }
      if (checkConsistency && !this._drift_z_translationIntialized) {
        throw new TypeError('Invalid argument: drift_z_translation in ChangeDriftDimensions_Request');
      }
      if (checkConsistency && !this._drift_x_rotationIntialized) {
        throw new TypeError('Invalid argument: drift_x_rotation in ChangeDriftDimensions_Request');
      }
      if (checkConsistency && !this._drift_y_rotationIntialized) {
        throw new TypeError('Invalid argument: drift_y_rotation in ChangeDriftDimensions_Request');
      }
      if (checkConsistency && !this._drift_z_rotationIntialized) {
        throw new TypeError('Invalid argument: drift_z_rotation in ChangeDriftDimensions_Request');
      }
    this._wrapperFields.transform_jog_frame_to_drift_frame.freeze(own, checkConsistency);
    this._refObject.transform_jog_frame_to_drift_frame = this._wrapperFields.transform_jog_frame_to_drift_frame.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._drift_x_translationIntialized = true;
    this._refObject.drift_x_translation = refObject.drift_x_translation;
    this._drift_y_translationIntialized = true;
    this._refObject.drift_y_translation = refObject.drift_y_translation;
    this._drift_z_translationIntialized = true;
    this._refObject.drift_z_translation = refObject.drift_z_translation;
    this._drift_x_rotationIntialized = true;
    this._refObject.drift_x_rotation = refObject.drift_x_rotation;
    this._drift_y_rotationIntialized = true;
    this._refObject.drift_y_rotation = refObject.drift_y_rotation;
    this._drift_z_rotationIntialized = true;
    this._refObject.drift_z_rotation = refObject.drift_z_rotation;
    this._wrapperFields.transform_jog_frame_to_drift_frame.copyRefObject(refObject.transform_jog_frame_to_drift_frame);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TransformWrapper.freeStruct(refObject.transform_jog_frame_to_drift_frame);
  }
  static destoryRawROS(msg) {
    ChangeDriftDimensions_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ChangeDriftDimensions_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get drift_x_translation() {
    if (!this._drift_x_translationIntialized) {
      return undefined;
    }
    return this._refObject.drift_x_translation;
  }
  set drift_x_translation(value) {
    this._drift_x_translationIntialized = true;
    this._refObject.drift_x_translation = value;
  }
  get drift_y_translation() {
    if (!this._drift_y_translationIntialized) {
      return undefined;
    }
    return this._refObject.drift_y_translation;
  }
  set drift_y_translation(value) {
    this._drift_y_translationIntialized = true;
    this._refObject.drift_y_translation = value;
  }
  get drift_z_translation() {
    if (!this._drift_z_translationIntialized) {
      return undefined;
    }
    return this._refObject.drift_z_translation;
  }
  set drift_z_translation(value) {
    this._drift_z_translationIntialized = true;
    this._refObject.drift_z_translation = value;
  }
  get drift_x_rotation() {
    if (!this._drift_x_rotationIntialized) {
      return undefined;
    }
    return this._refObject.drift_x_rotation;
  }
  set drift_x_rotation(value) {
    this._drift_x_rotationIntialized = true;
    this._refObject.drift_x_rotation = value;
  }
  get drift_y_rotation() {
    if (!this._drift_y_rotationIntialized) {
      return undefined;
    }
    return this._refObject.drift_y_rotation;
  }
  set drift_y_rotation(value) {
    this._drift_y_rotationIntialized = true;
    this._refObject.drift_y_rotation = value;
  }
  get drift_z_rotation() {
    if (!this._drift_z_rotationIntialized) {
      return undefined;
    }
    return this._refObject.drift_z_rotation;
  }
  set drift_z_rotation(value) {
    this._drift_z_rotationIntialized = true;
    this._refObject.drift_z_rotation = value;
  }
  get transform_jog_frame_to_drift_frame() {
    return this._wrapperFields.transform_jog_frame_to_drift_frame;
  }
  set transform_jog_frame_to_drift_frame(value) {
    if (value instanceof TransformWrapper) {
      this._wrapperFields.transform_jog_frame_to_drift_frame.copy(value);
    } else {
      this._wrapperFields.transform_jog_frame_to_drift_frame.copy(new TransformWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ChangeDriftDimensions_RequestRefStruct(refObject.toObject());
    this._drift_x_translationIntialized = true;
    this._drift_y_translationIntialized = true;
    this._drift_z_translationIntialized = true;
    this._drift_x_rotationIntialized = true;
    this._drift_y_rotationIntialized = true;
    this._drift_z_rotationIntialized = true;
    this._wrapperFields.transform_jog_frame_to_drift_frame.copyRefObject(this._refObject.transform_jog_frame_to_drift_frame);
  }
  copy(other) {
    this._refObject = new ChangeDriftDimensions_RequestRefStruct(other._refObject.toObject());
    this._drift_x_translationIntialized = true;
    this._drift_y_translationIntialized = true;
    this._drift_z_translationIntialized = true;
    this._drift_x_rotationIntialized = true;
    this._drift_y_rotationIntialized = true;
    this._drift_z_rotationIntialized = true;
    this._wrapperFields.transform_jog_frame_to_drift_frame.copy(other._wrapperFields.transform_jog_frame_to_drift_frame);
  }
  static get classType() {
    return ChangeDriftDimensions_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"drift_x_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drift_y_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drift_z_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drift_x_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drift_y_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drift_z_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"transform_jog_frame_to_drift_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Transform","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ChangeDriftDimensions_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ChangeDriftDimensions_Request"};
  }
  hasMember(name) {
    let memberNames = ["drift_x_translation","drift_y_translation","drift_z_translation","drift_x_rotation","drift_y_rotation","drift_z_rotation","transform_jog_frame_to_drift_frame"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ChangeDriftDimensions_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ChangeDriftDimensions_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ChangeDriftDimensions_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeDriftDimensions_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeDriftDimensions_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ChangeDriftDimensions_RequestRefArray(size);
    this._refObject = new ChangeDriftDimensions_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ChangeDriftDimensions_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ChangeDriftDimensions_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ChangeDriftDimensions_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ChangeDriftDimensions_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ChangeDriftDimensions_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ChangeDriftDimensions_RequestArrayWrapper;
  }
}
module.exports = ChangeDriftDimensions_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "drift_x_translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drift_y_translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drift_z_translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drift_x_rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drift_y_rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drift_z_rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "transform_jog_frame_to_drift_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Transform",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ChangeDriftDimensions_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ChangeDriftDimensions_Request"
}
*/
