// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__ConstraintEvalResult.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ConstraintEvalResultRefStruct = StructType({
  result: primitiveTypes.bool,
  distance: primitiveTypes.float64,
});
const ConstraintEvalResultRefArray = ArrayType(ConstraintEvalResultRefStruct);
const ConstraintEvalResultRefStructArray = StructType({
  data: ConstraintEvalResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ConstraintEvalResultWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ConstraintEvalResultRefStruct(other._refObject.toObject());
      this._resultIntialized = true;
      this._distanceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ConstraintEvalResultRefStruct();
    this._resultIntialized = false;
    this._distanceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ConstraintEvalResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ConstraintEvalResultArrayWrapper;
  }
  static get ArrayType() {
    return ConstraintEvalResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return ConstraintEvalResultRefStructArray
  }
  static get refObjectType() {
    return ConstraintEvalResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._resultIntialized) {
        throw new TypeError('Invalid argument: result in ConstraintEvalResult');
      }
      if (checkConsistency && !this._distanceIntialized) {
        throw new TypeError('Invalid argument: distance in ConstraintEvalResult');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._resultIntialized = true;
    this._refObject.result = refObject.result;
    this._distanceIntialized = true;
    this._refObject.distance = refObject.distance;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ConstraintEvalResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'ConstraintEvalResult'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get result() {
    if (!this._resultIntialized) {
      return undefined;
    }
    return this._refObject.result;
  }
  set result(value) {
    this._resultIntialized = true;
    this._refObject.result = value;
  }
  get distance() {
    if (!this._distanceIntialized) {
      return undefined;
    }
    return this._refObject.distance;
  }
  set distance(value) {
    this._distanceIntialized = true;
    this._refObject.distance = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ConstraintEvalResultRefStruct(refObject.toObject());
    this._resultIntialized = true;
    this._distanceIntialized = true;
  }
  copy(other) {
    this._refObject = new ConstraintEvalResultRefStruct(other._refObject.toObject());
    this._resultIntialized = true;
    this._distanceIntialized = true;
  }
  static get classType() {
    return ConstraintEvalResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ConstraintEvalResult","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ConstraintEvalResult"};
  }
  hasMember(name) {
    let memberNames = ["result","distance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ConstraintEvalResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ConstraintEvalResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ConstraintEvalResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConstraintEvalResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConstraintEvalResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ConstraintEvalResultRefArray(size);
    this._refObject = new ConstraintEvalResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ConstraintEvalResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ConstraintEvalResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ConstraintEvalResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ConstraintEvalResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ConstraintEvalResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ConstraintEvalResultArrayWrapper;
  }
}
module.exports = ConstraintEvalResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "result",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ConstraintEvalResult",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ConstraintEvalResult"
}
*/
