// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__AllowedCollisionMatrix.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const AllowedCollisionEntryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__AllowedCollisionEntry.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const AllowedCollisionMatrixRefStruct = StructType({
  entry_names: StringWrapper.refObjectArrayType,
  entry_values: AllowedCollisionEntryWrapper.refObjectArrayType,
  default_entry_names: StringWrapper.refObjectArrayType,
  default_entry_values: BoolWrapper.refObjectArrayType,
});
const AllowedCollisionMatrixRefArray = ArrayType(AllowedCollisionMatrixRefStruct);
const AllowedCollisionMatrixRefStructArray = StructType({
  data: AllowedCollisionMatrixRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AllowedCollisionMatrixWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._entry_namesArray = [];
    this._default_entry_namesArray = [];
    this._default_entry_valuesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AllowedCollisionMatrixRefStruct(other._refObject.toObject());
      this._wrapperFields.entry_names = StringWrapper.createArray();
      this._wrapperFields.entry_names.copy(other._wrapperFields.entry_names);
      this.entry_names = other.entry_names;
      this._wrapperFields.entry_values = AllowedCollisionEntryWrapper.createArray();
      this._wrapperFields.entry_values.copy(other._wrapperFields.entry_values);
      this._wrapperFields.default_entry_names = StringWrapper.createArray();
      this._wrapperFields.default_entry_names.copy(other._wrapperFields.default_entry_names);
      this.default_entry_names = other.default_entry_names;
      this._wrapperFields.default_entry_values = BoolWrapper.createArray();
      this._wrapperFields.default_entry_values.copy(other._wrapperFields.default_entry_values);
      this.default_entry_values = other.default_entry_values;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AllowedCollisionMatrixRefStruct();
    this._wrapperFields.entry_names = StringWrapper.createArray();
    this._wrapperFields.entry_values = AllowedCollisionEntryWrapper.createArray();
    this._wrapperFields.default_entry_names = StringWrapper.createArray();
    this._wrapperFields.default_entry_values = BoolWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new AllowedCollisionMatrixWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AllowedCollisionMatrixArrayWrapper;
  }
  static get ArrayType() {
    return AllowedCollisionMatrixArrayWrapper;
  }
  static get refObjectArrayType() {
    return AllowedCollisionMatrixRefStructArray
  }
  static get refObjectType() {
    return AllowedCollisionMatrixRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (!own) {
      this._wrapperFields.entry_names.fill(this._entry_namesArray);
      this._wrapperFields.entry_names.freeze(own, checkConsistency);
      this._refObject.entry_names = this._wrapperFields.entry_names.refObject;
    } else {
      this._wrapperFields.entry_names.fill([]);
      this._wrapperFields.entry_names.freeze(own, checkConsistency);
      this._refObject.entry_names = this._wrapperFields.entry_names.refObject;
    }
    this._wrapperFields.entry_values.freeze(own, checkConsistency);
    this._refObject.entry_values = this._wrapperFields.entry_values.refObject;
    if (!own) {
      this._wrapperFields.default_entry_names.fill(this._default_entry_namesArray);
      this._wrapperFields.default_entry_names.freeze(own, checkConsistency);
      this._refObject.default_entry_names = this._wrapperFields.default_entry_names.refObject;
    } else {
      this._wrapperFields.default_entry_names.fill([]);
      this._wrapperFields.default_entry_names.freeze(own, checkConsistency);
      this._refObject.default_entry_names = this._wrapperFields.default_entry_names.refObject;
    }
    if (!own) {
      this._wrapperFields.default_entry_values.fill(this._default_entry_valuesArray);
      this._wrapperFields.default_entry_values.freeze(own, checkConsistency);
      this._refObject.default_entry_values = this._wrapperFields.default_entry_values.refObject;
    } else {
      this._wrapperFields.default_entry_values.fill([]);
      this._wrapperFields.default_entry_values.freeze(own, checkConsistency);
      this._refObject.default_entry_values = this._wrapperFields.default_entry_values.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.entry_names.data.length = refObject.entry_names.size;
    for (let index = 0; index < refObject.entry_names.size; index++) {
      this._entry_namesArray[index] = refObject.entry_names.data[index].data;
    }
    this._wrapperFields.entry_values.copyRefObject(refObject.entry_values);
    refObject.default_entry_names.data.length = refObject.default_entry_names.size;
    for (let index = 0; index < refObject.default_entry_names.size; index++) {
      this._default_entry_namesArray[index] = refObject.default_entry_names.data[index].data;
    }
    refObject.default_entry_values.data.length = refObject.default_entry_values.size;
    for (let index = 0; index < refObject.default_entry_values.size; index++) {
      this._default_entry_valuesArray[index] = refObject.default_entry_values.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.entry_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.entry_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.entry_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.entry_values.size != 0) {
      AllowedCollisionEntryWrapper.ArrayType.freeArray(refObject.entry_values);
      if (AllowedCollisionEntryWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.entry_values, AllowedCollisionEntryWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.default_entry_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.default_entry_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.default_entry_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.default_entry_values.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.default_entry_values);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.default_entry_values, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    AllowedCollisionMatrixWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'AllowedCollisionMatrix'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get entry_names() {
    return this._entry_namesArray;
  }
  set entry_names(value) {
    this._entry_namesArray = value;
  }
  get entry_values() {
    return this._wrapperFields.entry_values;
  }
  set entry_values(value) {
    this._wrapperFields.entry_values.fill(value);
  }
  get default_entry_names() {
    return this._default_entry_namesArray;
  }
  set default_entry_names(value) {
    this._default_entry_namesArray = value;
  }
  get default_entry_values() {
    return this._default_entry_valuesArray;
  }
  set default_entry_values(value) {
    this._default_entry_valuesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AllowedCollisionMatrixRefStruct(refObject.toObject());
    refObject.entry_names.data.length = refObject.entry_names.size;
    for (let index = 0; index < refObject.entry_names.size; index++) {
      this._entry_namesArray[index] = refObject.entry_names.data[index].data;
    }
    this._wrapperFields.entry_values.copyRefObject(this._refObject.entry_values);
    refObject.default_entry_names.data.length = refObject.default_entry_names.size;
    for (let index = 0; index < refObject.default_entry_names.size; index++) {
      this._default_entry_namesArray[index] = refObject.default_entry_names.data[index].data;
    }
    refObject.default_entry_values.data.length = refObject.default_entry_values.size;
    for (let index = 0; index < refObject.default_entry_values.size; index++) {
      this._default_entry_valuesArray[index] = refObject.default_entry_values.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new AllowedCollisionMatrixRefStruct(other._refObject.toObject());
    this._entry_namesArray = other._entry_namesArray.slice();
    this._wrapperFields.entry_values.copy(other._wrapperFields.entry_values);
    this._default_entry_namesArray = other._default_entry_namesArray.slice();
    this._default_entry_valuesArray = other._default_entry_valuesArray.slice();
  }
  static get classType() {
    return AllowedCollisionMatrixWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"entry_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"entry_values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"AllowedCollisionEntry","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"default_entry_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"default_entry_values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"AllowedCollisionMatrix","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AllowedCollisionMatrix"};
  }
  hasMember(name) {
    let memberNames = ["entry_names","entry_values","default_entry_names","default_entry_values"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AllowedCollisionMatrixArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AllowedCollisionMatrixWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AllowedCollisionMatrixWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AllowedCollisionMatrixArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AllowedCollisionMatrixArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AllowedCollisionMatrixRefArray(size);
    this._refObject = new AllowedCollisionMatrixRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AllowedCollisionMatrixWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AllowedCollisionMatrixArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AllowedCollisionMatrixArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AllowedCollisionMatrixWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AllowedCollisionMatrixWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AllowedCollisionMatrixArrayWrapper;
  }
}
module.exports = AllowedCollisionMatrixWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "entry_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "entry_values",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "AllowedCollisionEntry",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "default_entry_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "default_entry_values",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "AllowedCollisionMatrix",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AllowedCollisionMatrix"
}
*/
