// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__DeleteRobotStateFromWarehouse_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DeleteRobotStateFromWarehouse_RequestRefStruct = StructType({
  name: primitiveTypes.string,
  robot: primitiveTypes.string,
});
const DeleteRobotStateFromWarehouse_RequestRefArray = ArrayType(DeleteRobotStateFromWarehouse_RequestRefStruct);
const DeleteRobotStateFromWarehouse_RequestRefStructArray = StructType({
  data: DeleteRobotStateFromWarehouse_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DeleteRobotStateFromWarehouse_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DeleteRobotStateFromWarehouse_RequestRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._robotIntialized = true;
      this._wrapperFields.robot =  new StringWrapper(other._wrapperFields.robot);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DeleteRobotStateFromWarehouse_RequestRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._robotIntialized = false;
    this._wrapperFields.robot = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new DeleteRobotStateFromWarehouse_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DeleteRobotStateFromWarehouse_RequestArrayWrapper;
  }
  static get ArrayType() {
    return DeleteRobotStateFromWarehouse_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return DeleteRobotStateFromWarehouse_RequestRefStructArray
  }
  static get refObjectType() {
    return DeleteRobotStateFromWarehouse_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in DeleteRobotStateFromWarehouse_Request');
      }
      if (checkConsistency && !this._robotIntialized) {
        throw new TypeError('Invalid argument: robot in DeleteRobotStateFromWarehouse_Request');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.robot.freeze(own, checkConsistency);
    }
    this._refObject.robot = this._wrapperFields.robot.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._robotIntialized = true;
    this._wrapperFields.robot.data = refObject.robot.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    StringWrapper.freeStruct(refObject.robot);
  }
  static destoryRawROS(msg) {
    DeleteRobotStateFromWarehouse_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'DeleteRobotStateFromWarehouse_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get robot() {
    if (!this._robotIntialized) {
      return undefined;
    }
    return this._wrapperFields.robot.data;
  }
  set robot(value) {
    this._robotIntialized = true;
    this._wrapperFields.robot.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DeleteRobotStateFromWarehouse_RequestRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._robotIntialized = true;
    this._wrapperFields.robot.copyRefObject(this._refObject.robot);
  }
  copy(other) {
    this._refObject = new DeleteRobotStateFromWarehouse_RequestRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._robotIntialized = true;
    this._wrapperFields.robot.copy(other._wrapperFields.robot);
  }
  static get classType() {
    return DeleteRobotStateFromWarehouse_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"DeleteRobotStateFromWarehouse_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DeleteRobotStateFromWarehouse_Request"};
  }
  hasMember(name) {
    let memberNames = ["name","robot"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DeleteRobotStateFromWarehouse_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DeleteRobotStateFromWarehouse_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DeleteRobotStateFromWarehouse_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DeleteRobotStateFromWarehouse_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DeleteRobotStateFromWarehouse_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DeleteRobotStateFromWarehouse_RequestRefArray(size);
    this._refObject = new DeleteRobotStateFromWarehouse_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DeleteRobotStateFromWarehouse_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DeleteRobotStateFromWarehouse_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DeleteRobotStateFromWarehouse_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DeleteRobotStateFromWarehouse_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DeleteRobotStateFromWarehouse_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DeleteRobotStateFromWarehouse_RequestArrayWrapper;
  }
}
module.exports = DeleteRobotStateFromWarehouse_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "robot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "DeleteRobotStateFromWarehouse_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DeleteRobotStateFromWarehouse_Request"
}
*/
