// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__PositionConstraint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const BoundingVolumeWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__BoundingVolume.js');
const PositionConstraintRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  link_name: primitiveTypes.string,
  target_point_offset: Vector3Wrapper.refObjectType,
  constraint_region: BoundingVolumeWrapper.refObjectType,
  weight: primitiveTypes.float64,
});
const PositionConstraintRefArray = ArrayType(PositionConstraintRefStruct);
const PositionConstraintRefStructArray = StructType({
  data: PositionConstraintRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PositionConstraintWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PositionConstraintRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._link_nameIntialized = true;
      this._wrapperFields.link_name =  new StringWrapper(other._wrapperFields.link_name);
      this._wrapperFields.target_point_offset =  new Vector3Wrapper(other._wrapperFields.target_point_offset);
      this._wrapperFields.constraint_region =  new BoundingVolumeWrapper(other._wrapperFields.constraint_region);
      this._weightIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PositionConstraintRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._link_nameIntialized = false;
    this._wrapperFields.link_name = new StringWrapper();
    this._wrapperFields.target_point_offset = new Vector3Wrapper();
    this._wrapperFields.constraint_region = new BoundingVolumeWrapper();
    this._weightIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PositionConstraintWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PositionConstraintArrayWrapper;
  }
  static get ArrayType() {
    return PositionConstraintArrayWrapper;
  }
  static get refObjectArrayType() {
    return PositionConstraintRefStructArray
  }
  static get refObjectType() {
    return PositionConstraintRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._link_nameIntialized) {
        throw new TypeError('Invalid argument: link_name in PositionConstraint');
      }
      if (checkConsistency && !this._weightIntialized) {
        throw new TypeError('Invalid argument: weight in PositionConstraint');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.link_name.freeze(own, checkConsistency);
    }
    this._refObject.link_name = this._wrapperFields.link_name.refObject;
    this._wrapperFields.target_point_offset.freeze(own, checkConsistency);
    this._refObject.target_point_offset = this._wrapperFields.target_point_offset.refObject;
    this._wrapperFields.constraint_region.freeze(own, checkConsistency);
    this._refObject.constraint_region = this._wrapperFields.constraint_region.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.data = refObject.link_name.data;
    this._wrapperFields.target_point_offset.copyRefObject(refObject.target_point_offset);
    this._wrapperFields.constraint_region.copyRefObject(refObject.constraint_region);
    this._weightIntialized = true;
    this._refObject.weight = refObject.weight;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.link_name);
    Vector3Wrapper.freeStruct(refObject.target_point_offset);
    BoundingVolumeWrapper.freeStruct(refObject.constraint_region);
  }
  static destoryRawROS(msg) {
    PositionConstraintWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'PositionConstraint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get link_name() {
    if (!this._link_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.link_name.data;
  }
  set link_name(value) {
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.data = value;
  }
  get target_point_offset() {
    return this._wrapperFields.target_point_offset;
  }
  set target_point_offset(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.target_point_offset.copy(value);
    } else {
      this._wrapperFields.target_point_offset.copy(new Vector3Wrapper(value));
    }
  }
  get constraint_region() {
    return this._wrapperFields.constraint_region;
  }
  set constraint_region(value) {
    if (value instanceof BoundingVolumeWrapper) {
      this._wrapperFields.constraint_region.copy(value);
    } else {
      this._wrapperFields.constraint_region.copy(new BoundingVolumeWrapper(value));
    }
  }
  get weight() {
    if (!this._weightIntialized) {
      return undefined;
    }
    return this._refObject.weight;
  }
  set weight(value) {
    this._weightIntialized = true;
    this._refObject.weight = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PositionConstraintRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.copyRefObject(this._refObject.link_name);
    this._wrapperFields.target_point_offset.copyRefObject(this._refObject.target_point_offset);
    this._wrapperFields.constraint_region.copyRefObject(this._refObject.constraint_region);
    this._weightIntialized = true;
  }
  copy(other) {
    this._refObject = new PositionConstraintRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.copy(other._wrapperFields.link_name);
    this._wrapperFields.target_point_offset.copy(other._wrapperFields.target_point_offset);
    this._wrapperFields.constraint_region.copy(other._wrapperFields.constraint_region);
    this._weightIntialized = true;
  }
  static get classType() {
    return PositionConstraintWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"link_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_point_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"constraint_region","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"BoundingVolume","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"weight","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"PositionConstraint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PositionConstraint"};
  }
  hasMember(name) {
    let memberNames = ["header","link_name","target_point_offset","constraint_region","weight"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PositionConstraintArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PositionConstraintWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PositionConstraintWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PositionConstraintArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PositionConstraintArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PositionConstraintRefArray(size);
    this._refObject = new PositionConstraintRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PositionConstraintWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PositionConstraintArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PositionConstraintArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PositionConstraintWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PositionConstraintWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PositionConstraintArrayWrapper;
  }
}
module.exports = PositionConstraintWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "link_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_point_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "constraint_region",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "BoundingVolume",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "weight",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "PositionConstraint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PositionConstraint"
}
*/
