// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetStateValidity_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ConstraintsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__Constraints.js');
const GetStateValidity_RequestRefStruct = StructType({
  robot_state: RobotStateWrapper.refObjectType,
  group_name: primitiveTypes.string,
  constraints: ConstraintsWrapper.refObjectType,
});
const GetStateValidity_RequestRefArray = ArrayType(GetStateValidity_RequestRefStruct);
const GetStateValidity_RequestRefStructArray = StructType({
  data: GetStateValidity_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetStateValidity_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetStateValidity_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.robot_state =  new RobotStateWrapper(other._wrapperFields.robot_state);
      this._group_nameIntialized = true;
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
      this._wrapperFields.constraints =  new ConstraintsWrapper(other._wrapperFields.constraints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetStateValidity_RequestRefStruct();
    this._wrapperFields.robot_state = new RobotStateWrapper();
    this._group_nameIntialized = false;
    this._wrapperFields.group_name = new StringWrapper();
    this._wrapperFields.constraints = new ConstraintsWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetStateValidity_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetStateValidity_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetStateValidity_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetStateValidity_RequestRefStructArray
  }
  static get refObjectType() {
    return GetStateValidity_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._group_nameIntialized) {
        throw new TypeError('Invalid argument: group_name in GetStateValidity_Request');
      }
    this._wrapperFields.robot_state.freeze(own, checkConsistency);
    this._refObject.robot_state = this._wrapperFields.robot_state.refObject;
    if (own) {
      this._wrapperFields.group_name.freeze(own, checkConsistency);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
    this._wrapperFields.constraints.freeze(own, checkConsistency);
    this._refObject.constraints = this._wrapperFields.constraints.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.robot_state.copyRefObject(refObject.robot_state);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.data = refObject.group_name.data;
    this._wrapperFields.constraints.copyRefObject(refObject.constraints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RobotStateWrapper.freeStruct(refObject.robot_state);
    StringWrapper.freeStruct(refObject.group_name);
    ConstraintsWrapper.freeStruct(refObject.constraints);
  }
  static destoryRawROS(msg) {
    GetStateValidity_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetStateValidity_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get robot_state() {
    return this._wrapperFields.robot_state;
  }
  set robot_state(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.robot_state.copy(value);
    } else {
      this._wrapperFields.robot_state.copy(new RobotStateWrapper(value));
    }
  }
  get group_name() {
    if (!this._group_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.data = value;
  }
  get constraints() {
    return this._wrapperFields.constraints;
  }
  set constraints(value) {
    if (value instanceof ConstraintsWrapper) {
      this._wrapperFields.constraints.copy(value);
    } else {
      this._wrapperFields.constraints.copy(new ConstraintsWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetStateValidity_RequestRefStruct(refObject.toObject());
    this._wrapperFields.robot_state.copyRefObject(this._refObject.robot_state);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
    this._wrapperFields.constraints.copyRefObject(this._refObject.constraints);
  }
  copy(other) {
    this._refObject = new GetStateValidity_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.robot_state.copy(other._wrapperFields.robot_state);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
    this._wrapperFields.constraints.copy(other._wrapperFields.constraints);
  }
  static get classType() {
    return GetStateValidity_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"robot_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"constraints","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"Constraints","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetStateValidity_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetStateValidity_Request"};
  }
  hasMember(name) {
    let memberNames = ["robot_state","group_name","constraints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetStateValidity_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetStateValidity_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetStateValidity_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetStateValidity_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetStateValidity_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetStateValidity_RequestRefArray(size);
    this._refObject = new GetStateValidity_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetStateValidity_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetStateValidity_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetStateValidity_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetStateValidity_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetStateValidity_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetStateValidity_RequestArrayWrapper;
  }
}
module.exports = GetStateValidity_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "robot_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "group_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "constraints",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "Constraints",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetStateValidity_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetStateValidity_Request"
}
*/
