// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__CartesianPoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const AccelWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Accel.js');
const CartesianPointRefStruct = StructType({
  pose: PoseWrapper.refObjectType,
  velocity: TwistWrapper.refObjectType,
  acceleration: AccelWrapper.refObjectType,
});
const CartesianPointRefArray = ArrayType(CartesianPointRefStruct);
const CartesianPointRefStructArray = StructType({
  data: CartesianPointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CartesianPointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CartesianPointRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._wrapperFields.velocity =  new TwistWrapper(other._wrapperFields.velocity);
      this._wrapperFields.acceleration =  new AccelWrapper(other._wrapperFields.acceleration);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CartesianPointRefStruct();
    this._wrapperFields.pose = new PoseWrapper();
    this._wrapperFields.velocity = new TwistWrapper();
    this._wrapperFields.acceleration = new AccelWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CartesianPointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CartesianPointArrayWrapper;
  }
  static get ArrayType() {
    return CartesianPointArrayWrapper;
  }
  static get refObjectArrayType() {
    return CartesianPointRefStructArray
  }
  static get refObjectType() {
    return CartesianPointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.velocity.freeze(own, checkConsistency);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
    this._wrapperFields.acceleration.freeze(own, checkConsistency);
    this._refObject.acceleration = this._wrapperFields.acceleration.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
    this._wrapperFields.acceleration.copyRefObject(refObject.acceleration);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.pose);
    TwistWrapper.freeStruct(refObject.velocity);
    AccelWrapper.freeStruct(refObject.acceleration);
  }
  static destoryRawROS(msg) {
    CartesianPointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'CartesianPoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get velocity() {
    return this._wrapperFields.velocity;
  }
  set velocity(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.velocity.copy(value);
    } else {
      this._wrapperFields.velocity.copy(new TwistWrapper(value));
    }
  }
  get acceleration() {
    return this._wrapperFields.acceleration;
  }
  set acceleration(value) {
    if (value instanceof AccelWrapper) {
      this._wrapperFields.acceleration.copy(value);
    } else {
      this._wrapperFields.acceleration.copy(new AccelWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CartesianPointRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
    this._wrapperFields.acceleration.copyRefObject(this._refObject.acceleration);
  }
  copy(other) {
    this._refObject = new CartesianPointRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
  }
  static get classType() {
    return CartesianPointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Accel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"CartesianPoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CartesianPoint"};
  }
  hasMember(name) {
    let memberNames = ["pose","velocity","acceleration"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CartesianPointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CartesianPointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CartesianPointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianPointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianPointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CartesianPointRefArray(size);
    this._refObject = new CartesianPointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CartesianPointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CartesianPointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CartesianPointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CartesianPointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CartesianPointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CartesianPointArrayWrapper;
  }
}
module.exports = CartesianPointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "acceleration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Accel",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "CartesianPoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CartesianPoint"
}
*/
