// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ApplyPlanningScene_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PlanningSceneWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlanningScene.js');
const ApplyPlanningScene_RequestRefStruct = StructType({
  scene: PlanningSceneWrapper.refObjectType,
});
const ApplyPlanningScene_RequestRefArray = ArrayType(ApplyPlanningScene_RequestRefStruct);
const ApplyPlanningScene_RequestRefStructArray = StructType({
  data: ApplyPlanningScene_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ApplyPlanningScene_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ApplyPlanningScene_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.scene =  new PlanningSceneWrapper(other._wrapperFields.scene);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ApplyPlanningScene_RequestRefStruct();
    this._wrapperFields.scene = new PlanningSceneWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ApplyPlanningScene_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ApplyPlanningScene_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ApplyPlanningScene_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ApplyPlanningScene_RequestRefStructArray
  }
  static get refObjectType() {
    return ApplyPlanningScene_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.scene.freeze(own, checkConsistency);
    this._refObject.scene = this._wrapperFields.scene.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.scene.copyRefObject(refObject.scene);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PlanningSceneWrapper.freeStruct(refObject.scene);
  }
  static destoryRawROS(msg) {
    ApplyPlanningScene_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ApplyPlanningScene_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get scene() {
    return this._wrapperFields.scene;
  }
  set scene(value) {
    if (value instanceof PlanningSceneWrapper) {
      this._wrapperFields.scene.copy(value);
    } else {
      this._wrapperFields.scene.copy(new PlanningSceneWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ApplyPlanningScene_RequestRefStruct(refObject.toObject());
    this._wrapperFields.scene.copyRefObject(this._refObject.scene);
  }
  copy(other) {
    this._refObject = new ApplyPlanningScene_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.scene.copy(other._wrapperFields.scene);
  }
  static get classType() {
    return ApplyPlanningScene_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"scene","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlanningScene","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ApplyPlanningScene_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ApplyPlanningScene_Request"};
  }
  hasMember(name) {
    let memberNames = ["scene"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ApplyPlanningScene_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ApplyPlanningScene_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ApplyPlanningScene_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplyPlanningScene_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplyPlanningScene_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ApplyPlanningScene_RequestRefArray(size);
    this._refObject = new ApplyPlanningScene_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ApplyPlanningScene_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ApplyPlanningScene_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ApplyPlanningScene_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ApplyPlanningScene_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ApplyPlanningScene_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ApplyPlanningScene_RequestArrayWrapper;
  }
}
module.exports = ApplyPlanningScene_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "scene",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "PlanningScene",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ApplyPlanningScene_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ApplyPlanningScene_Request"
}
*/
