// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ExecuteKnownTrajectory_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const ExecuteKnownTrajectory_RequestRefStruct = StructType({
  trajectory: RobotTrajectoryWrapper.refObjectType,
  wait_for_execution: primitiveTypes.bool,
});
const ExecuteKnownTrajectory_RequestRefArray = ArrayType(ExecuteKnownTrajectory_RequestRefStruct);
const ExecuteKnownTrajectory_RequestRefStructArray = StructType({
  data: ExecuteKnownTrajectory_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ExecuteKnownTrajectory_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ExecuteKnownTrajectory_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.trajectory =  new RobotTrajectoryWrapper(other._wrapperFields.trajectory);
      this._wait_for_executionIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ExecuteKnownTrajectory_RequestRefStruct();
    this._wrapperFields.trajectory = new RobotTrajectoryWrapper();
    this._wait_for_executionIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ExecuteKnownTrajectory_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ExecuteKnownTrajectory_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ExecuteKnownTrajectory_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ExecuteKnownTrajectory_RequestRefStructArray
  }
  static get refObjectType() {
    return ExecuteKnownTrajectory_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._wait_for_executionIntialized) {
        throw new TypeError('Invalid argument: wait_for_execution in ExecuteKnownTrajectory_Request');
      }
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
    this._wait_for_executionIntialized = true;
    this._refObject.wait_for_execution = refObject.wait_for_execution;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RobotTrajectoryWrapper.freeStruct(refObject.trajectory);
  }
  static destoryRawROS(msg) {
    ExecuteKnownTrajectory_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ExecuteKnownTrajectory_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    if (value instanceof RobotTrajectoryWrapper) {
      this._wrapperFields.trajectory.copy(value);
    } else {
      this._wrapperFields.trajectory.copy(new RobotTrajectoryWrapper(value));
    }
  }
  get wait_for_execution() {
    if (!this._wait_for_executionIntialized) {
      return undefined;
    }
    return this._refObject.wait_for_execution;
  }
  set wait_for_execution(value) {
    this._wait_for_executionIntialized = true;
    this._refObject.wait_for_execution = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ExecuteKnownTrajectory_RequestRefStruct(refObject.toObject());
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
    this._wait_for_executionIntialized = true;
  }
  copy(other) {
    this._refObject = new ExecuteKnownTrajectory_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
    this._wait_for_executionIntialized = true;
  }
  static get classType() {
    return ExecuteKnownTrajectory_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"wait_for_execution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ExecuteKnownTrajectory_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ExecuteKnownTrajectory_Request"};
  }
  hasMember(name) {
    let memberNames = ["trajectory","wait_for_execution"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ExecuteKnownTrajectory_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ExecuteKnownTrajectory_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ExecuteKnownTrajectory_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExecuteKnownTrajectory_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ExecuteKnownTrajectory_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ExecuteKnownTrajectory_RequestRefArray(size);
    this._refObject = new ExecuteKnownTrajectory_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ExecuteKnownTrajectory_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ExecuteKnownTrajectory_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ExecuteKnownTrajectory_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ExecuteKnownTrajectory_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ExecuteKnownTrajectory_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ExecuteKnownTrajectory_RequestArrayWrapper;
  }
}
module.exports = ExecuteKnownTrajectory_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "wait_for_execution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ExecuteKnownTrajectory_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ExecuteKnownTrajectory_Request"
}
*/
