// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ListRobotStatesInWarehouse_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ListRobotStatesInWarehouse_RequestRefStruct = StructType({
  regex: primitiveTypes.string,
  robot: primitiveTypes.string,
});
const ListRobotStatesInWarehouse_RequestRefArray = ArrayType(ListRobotStatesInWarehouse_RequestRefStruct);
const ListRobotStatesInWarehouse_RequestRefStructArray = StructType({
  data: ListRobotStatesInWarehouse_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ListRobotStatesInWarehouse_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListRobotStatesInWarehouse_RequestRefStruct(other._refObject.toObject());
      this._regexIntialized = true;
      this._wrapperFields.regex =  new StringWrapper(other._wrapperFields.regex);
      this._robotIntialized = true;
      this._wrapperFields.robot =  new StringWrapper(other._wrapperFields.robot);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ListRobotStatesInWarehouse_RequestRefStruct();
    this._regexIntialized = false;
    this._wrapperFields.regex = new StringWrapper();
    this._robotIntialized = false;
    this._wrapperFields.robot = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ListRobotStatesInWarehouse_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListRobotStatesInWarehouse_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ListRobotStatesInWarehouse_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListRobotStatesInWarehouse_RequestRefStructArray
  }
  static get refObjectType() {
    return ListRobotStatesInWarehouse_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._regexIntialized) {
        throw new TypeError('Invalid argument: regex in ListRobotStatesInWarehouse_Request');
      }
      if (checkConsistency && !this._robotIntialized) {
        throw new TypeError('Invalid argument: robot in ListRobotStatesInWarehouse_Request');
      }
    if (own) {
      this._wrapperFields.regex.freeze(own, checkConsistency);
    }
    this._refObject.regex = this._wrapperFields.regex.refObject;
    if (own) {
      this._wrapperFields.robot.freeze(own, checkConsistency);
    }
    this._refObject.robot = this._wrapperFields.robot.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._regexIntialized = true;
    this._wrapperFields.regex.data = refObject.regex.data;
    this._robotIntialized = true;
    this._wrapperFields.robot.data = refObject.robot.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.regex);
    StringWrapper.freeStruct(refObject.robot);
  }
  static destoryRawROS(msg) {
    ListRobotStatesInWarehouse_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ListRobotStatesInWarehouse_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get regex() {
    if (!this._regexIntialized) {
      return undefined;
    }
    return this._wrapperFields.regex.data;
  }
  set regex(value) {
    this._regexIntialized = true;
    this._wrapperFields.regex.data = value;
  }
  get robot() {
    if (!this._robotIntialized) {
      return undefined;
    }
    return this._wrapperFields.robot.data;
  }
  set robot(value) {
    this._robotIntialized = true;
    this._wrapperFields.robot.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ListRobotStatesInWarehouse_RequestRefStruct(refObject.toObject());
    this._regexIntialized = true;
    this._wrapperFields.regex.copyRefObject(this._refObject.regex);
    this._robotIntialized = true;
    this._wrapperFields.robot.copyRefObject(this._refObject.robot);
  }
  copy(other) {
    this._refObject = new ListRobotStatesInWarehouse_RequestRefStruct(other._refObject.toObject());
    this._regexIntialized = true;
    this._wrapperFields.regex.copy(other._wrapperFields.regex);
    this._robotIntialized = true;
    this._wrapperFields.robot.copy(other._wrapperFields.robot);
  }
  static get classType() {
    return ListRobotStatesInWarehouse_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"regex","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ListRobotStatesInWarehouse_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListRobotStatesInWarehouse_Request"};
  }
  hasMember(name) {
    let memberNames = ["regex","robot"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ListRobotStatesInWarehouse_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListRobotStatesInWarehouse_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListRobotStatesInWarehouse_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListRobotStatesInWarehouse_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListRobotStatesInWarehouse_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ListRobotStatesInWarehouse_RequestRefArray(size);
    this._refObject = new ListRobotStatesInWarehouse_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListRobotStatesInWarehouse_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListRobotStatesInWarehouse_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListRobotStatesInWarehouse_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListRobotStatesInWarehouse_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListRobotStatesInWarehouse_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListRobotStatesInWarehouse_RequestArrayWrapper;
  }
}
module.exports = ListRobotStatesInWarehouse_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "regex",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "robot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "ListRobotStatesInWarehouse_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ListRobotStatesInWarehouse_Request"
}
*/
