// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetPlannerParams_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetPlannerParams_RequestRefStruct = StructType({
  pipeline_id: primitiveTypes.string,
  planner_config: primitiveTypes.string,
  group: primitiveTypes.string,
});
const GetPlannerParams_RequestRefArray = ArrayType(GetPlannerParams_RequestRefStruct);
const GetPlannerParams_RequestRefStructArray = StructType({
  data: GetPlannerParams_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPlannerParams_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPlannerParams_RequestRefStruct(other._refObject.toObject());
      this._pipeline_idIntialized = true;
      this._wrapperFields.pipeline_id =  new StringWrapper(other._wrapperFields.pipeline_id);
      this._planner_configIntialized = true;
      this._wrapperFields.planner_config =  new StringWrapper(other._wrapperFields.planner_config);
      this._groupIntialized = true;
      this._wrapperFields.group =  new StringWrapper(other._wrapperFields.group);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPlannerParams_RequestRefStruct();
    this._pipeline_idIntialized = false;
    this._wrapperFields.pipeline_id = new StringWrapper();
    this._planner_configIntialized = false;
    this._wrapperFields.planner_config = new StringWrapper();
    this._groupIntialized = false;
    this._wrapperFields.group = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPlannerParams_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPlannerParams_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetPlannerParams_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPlannerParams_RequestRefStructArray
  }
  static get refObjectType() {
    return GetPlannerParams_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._pipeline_idIntialized) {
        throw new TypeError('Invalid argument: pipeline_id in GetPlannerParams_Request');
      }
      if (checkConsistency && !this._planner_configIntialized) {
        throw new TypeError('Invalid argument: planner_config in GetPlannerParams_Request');
      }
      if (checkConsistency && !this._groupIntialized) {
        throw new TypeError('Invalid argument: group in GetPlannerParams_Request');
      }
    if (own) {
      this._wrapperFields.pipeline_id.freeze(own, checkConsistency);
    }
    this._refObject.pipeline_id = this._wrapperFields.pipeline_id.refObject;
    if (own) {
      this._wrapperFields.planner_config.freeze(own, checkConsistency);
    }
    this._refObject.planner_config = this._wrapperFields.planner_config.refObject;
    if (own) {
      this._wrapperFields.group.freeze(own, checkConsistency);
    }
    this._refObject.group = this._wrapperFields.group.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.data = refObject.pipeline_id.data;
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.data = refObject.planner_config.data;
    this._groupIntialized = true;
    this._wrapperFields.group.data = refObject.group.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.pipeline_id);
    StringWrapper.freeStruct(refObject.planner_config);
    StringWrapper.freeStruct(refObject.group);
  }
  static destoryRawROS(msg) {
    GetPlannerParams_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetPlannerParams_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pipeline_id() {
    if (!this._pipeline_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.pipeline_id.data;
  }
  set pipeline_id(value) {
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.data = value;
  }
  get planner_config() {
    if (!this._planner_configIntialized) {
      return undefined;
    }
    return this._wrapperFields.planner_config.data;
  }
  set planner_config(value) {
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.data = value;
  }
  get group() {
    if (!this._groupIntialized) {
      return undefined;
    }
    return this._wrapperFields.group.data;
  }
  set group(value) {
    this._groupIntialized = true;
    this._wrapperFields.group.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetPlannerParams_RequestRefStruct(refObject.toObject());
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.copyRefObject(this._refObject.pipeline_id);
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.copyRefObject(this._refObject.planner_config);
    this._groupIntialized = true;
    this._wrapperFields.group.copyRefObject(this._refObject.group);
  }
  copy(other) {
    this._refObject = new GetPlannerParams_RequestRefStruct(other._refObject.toObject());
    this._pipeline_idIntialized = true;
    this._wrapperFields.pipeline_id.copy(other._wrapperFields.pipeline_id);
    this._planner_configIntialized = true;
    this._wrapperFields.planner_config.copy(other._wrapperFields.planner_config);
    this._groupIntialized = true;
    this._wrapperFields.group.copy(other._wrapperFields.group);
  }
  static get classType() {
    return GetPlannerParams_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pipeline_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"planner_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"group","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetPlannerParams_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPlannerParams_Request"};
  }
  hasMember(name) {
    let memberNames = ["pipeline_id","planner_config","group"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPlannerParams_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPlannerParams_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPlannerParams_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPlannerParams_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPlannerParams_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPlannerParams_RequestRefArray(size);
    this._refObject = new GetPlannerParams_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPlannerParams_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPlannerParams_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPlannerParams_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPlannerParams_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPlannerParams_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPlannerParams_RequestArrayWrapper;
  }
}
module.exports = GetPlannerParams_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pipeline_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "planner_config",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "group",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "GetPlannerParams_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPlannerParams_Request"
}
*/
