// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__CartesianTrajectory.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CartesianTrajectoryPointWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CartesianTrajectoryPoint.js');
const CartesianTrajectoryRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  tracked_frame: primitiveTypes.string,
  points: CartesianTrajectoryPointWrapper.refObjectArrayType,
});
const CartesianTrajectoryRefArray = ArrayType(CartesianTrajectoryRefStruct);
const CartesianTrajectoryRefStructArray = StructType({
  data: CartesianTrajectoryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CartesianTrajectoryWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CartesianTrajectoryRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._tracked_frameIntialized = true;
      this._wrapperFields.tracked_frame =  new StringWrapper(other._wrapperFields.tracked_frame);
      this._wrapperFields.points = CartesianTrajectoryPointWrapper.createArray();
      this._wrapperFields.points.copy(other._wrapperFields.points);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CartesianTrajectoryRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._tracked_frameIntialized = false;
    this._wrapperFields.tracked_frame = new StringWrapper();
    this._wrapperFields.points = CartesianTrajectoryPointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new CartesianTrajectoryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CartesianTrajectoryArrayWrapper;
  }
  static get ArrayType() {
    return CartesianTrajectoryArrayWrapper;
  }
  static get refObjectArrayType() {
    return CartesianTrajectoryRefStructArray
  }
  static get refObjectType() {
    return CartesianTrajectoryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._tracked_frameIntialized) {
        throw new TypeError('Invalid argument: tracked_frame in CartesianTrajectory');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.tracked_frame.freeze(own, checkConsistency);
    }
    this._refObject.tracked_frame = this._wrapperFields.tracked_frame.refObject;
    this._wrapperFields.points.freeze(own, checkConsistency);
    this._refObject.points = this._wrapperFields.points.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._tracked_frameIntialized = true;
    this._wrapperFields.tracked_frame.data = refObject.tracked_frame.data;
    this._wrapperFields.points.copyRefObject(refObject.points);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.tracked_frame);
    if (refObject.points.size != 0) {
      CartesianTrajectoryPointWrapper.ArrayType.freeArray(refObject.points);
      if (CartesianTrajectoryPointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.points, CartesianTrajectoryPointWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    CartesianTrajectoryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'CartesianTrajectory'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get tracked_frame() {
    if (!this._tracked_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.tracked_frame.data;
  }
  set tracked_frame(value) {
    this._tracked_frameIntialized = true;
    this._wrapperFields.tracked_frame.data = value;
  }
  get points() {
    return this._wrapperFields.points;
  }
  set points(value) {
    this._wrapperFields.points.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new CartesianTrajectoryRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._tracked_frameIntialized = true;
    this._wrapperFields.tracked_frame.copyRefObject(this._refObject.tracked_frame);
    this._wrapperFields.points.copyRefObject(this._refObject.points);
  }
  copy(other) {
    this._refObject = new CartesianTrajectoryRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._tracked_frameIntialized = true;
    this._wrapperFields.tracked_frame.copy(other._wrapperFields.tracked_frame);
    this._wrapperFields.points.copy(other._wrapperFields.points);
  }
  static get classType() {
    return CartesianTrajectoryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"tracked_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"points","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"CartesianTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"CartesianTrajectory","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CartesianTrajectory"};
  }
  hasMember(name) {
    let memberNames = ["header","tracked_frame","points"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CartesianTrajectoryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CartesianTrajectoryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CartesianTrajectoryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianTrajectoryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianTrajectoryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CartesianTrajectoryRefArray(size);
    this._refObject = new CartesianTrajectoryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CartesianTrajectoryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CartesianTrajectoryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CartesianTrajectoryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CartesianTrajectoryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CartesianTrajectoryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CartesianTrajectoryArrayWrapper;
  }
}
module.exports = CartesianTrajectoryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "tracked_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "points",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "CartesianTrajectoryPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "CartesianTrajectory",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CartesianTrajectory"
}
*/
