// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__DisplayTrajectory.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const DisplayTrajectoryRefStruct = StructType({
  model_id: primitiveTypes.string,
  trajectory: RobotTrajectoryWrapper.refObjectArrayType,
  trajectory_start: RobotStateWrapper.refObjectType,
});
const DisplayTrajectoryRefArray = ArrayType(DisplayTrajectoryRefStruct);
const DisplayTrajectoryRefStructArray = StructType({
  data: DisplayTrajectoryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DisplayTrajectoryWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DisplayTrajectoryRefStruct(other._refObject.toObject());
      this._model_idIntialized = true;
      this._wrapperFields.model_id =  new StringWrapper(other._wrapperFields.model_id);
      this._wrapperFields.trajectory = RobotTrajectoryWrapper.createArray();
      this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
      this._wrapperFields.trajectory_start =  new RobotStateWrapper(other._wrapperFields.trajectory_start);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DisplayTrajectoryRefStruct();
    this._model_idIntialized = false;
    this._wrapperFields.model_id = new StringWrapper();
    this._wrapperFields.trajectory = RobotTrajectoryWrapper.createArray();
    this._wrapperFields.trajectory_start = new RobotStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new DisplayTrajectoryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DisplayTrajectoryArrayWrapper;
  }
  static get ArrayType() {
    return DisplayTrajectoryArrayWrapper;
  }
  static get refObjectArrayType() {
    return DisplayTrajectoryRefStructArray
  }
  static get refObjectType() {
    return DisplayTrajectoryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._model_idIntialized) {
        throw new TypeError('Invalid argument: model_id in DisplayTrajectory');
      }
    if (own) {
      this._wrapperFields.model_id.freeze(own, checkConsistency);
    }
    this._refObject.model_id = this._wrapperFields.model_id.refObject;
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
    this._wrapperFields.trajectory_start.freeze(own, checkConsistency);
    this._refObject.trajectory_start = this._wrapperFields.trajectory_start.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._model_idIntialized = true;
    this._wrapperFields.model_id.data = refObject.model_id.data;
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
    this._wrapperFields.trajectory_start.copyRefObject(refObject.trajectory_start);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.model_id);
    if (refObject.trajectory.size != 0) {
      RobotTrajectoryWrapper.ArrayType.freeArray(refObject.trajectory);
      if (RobotTrajectoryWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.trajectory, RobotTrajectoryWrapper.refObjectArrayType, 'data');
      }
    }
    RobotStateWrapper.freeStruct(refObject.trajectory_start);
  }
  static destoryRawROS(msg) {
    DisplayTrajectoryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'DisplayTrajectory'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get model_id() {
    if (!this._model_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.model_id.data;
  }
  set model_id(value) {
    this._model_idIntialized = true;
    this._wrapperFields.model_id.data = value;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    this._wrapperFields.trajectory.fill(value);
  }
  get trajectory_start() {
    return this._wrapperFields.trajectory_start;
  }
  set trajectory_start(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.trajectory_start.copy(value);
    } else {
      this._wrapperFields.trajectory_start.copy(new RobotStateWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new DisplayTrajectoryRefStruct(refObject.toObject());
    this._model_idIntialized = true;
    this._wrapperFields.model_id.copyRefObject(this._refObject.model_id);
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
    this._wrapperFields.trajectory_start.copyRefObject(this._refObject.trajectory_start);
  }
  copy(other) {
    this._refObject = new DisplayTrajectoryRefStruct(other._refObject.toObject());
    this._model_idIntialized = true;
    this._wrapperFields.model_id.copy(other._wrapperFields.model_id);
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
    this._wrapperFields.trajectory_start.copy(other._wrapperFields.trajectory_start);
  }
  static get classType() {
    return DisplayTrajectoryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"model_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trajectory","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"trajectory_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"DisplayTrajectory","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DisplayTrajectory"};
  }
  hasMember(name) {
    let memberNames = ["model_id","trajectory","trajectory_start"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DisplayTrajectoryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DisplayTrajectoryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DisplayTrajectoryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DisplayTrajectoryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DisplayTrajectoryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DisplayTrajectoryRefArray(size);
    this._refObject = new DisplayTrajectoryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DisplayTrajectoryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DisplayTrajectoryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DisplayTrajectoryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DisplayTrajectoryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DisplayTrajectoryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DisplayTrajectoryArrayWrapper;
  }
}
module.exports = DisplayTrajectoryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "model_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trajectory",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "trajectory_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "DisplayTrajectory",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DisplayTrajectory"
}
*/
