// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__Constraints.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointConstraintWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__JointConstraint.js');
const PositionConstraintWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PositionConstraint.js');
const OrientationConstraintWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__OrientationConstraint.js');
const VisibilityConstraintWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__VisibilityConstraint.js');
const ConstraintsRefStruct = StructType({
  name: primitiveTypes.string,
  joint_constraints: JointConstraintWrapper.refObjectArrayType,
  position_constraints: PositionConstraintWrapper.refObjectArrayType,
  orientation_constraints: OrientationConstraintWrapper.refObjectArrayType,
  visibility_constraints: VisibilityConstraintWrapper.refObjectArrayType,
});
const ConstraintsRefArray = ArrayType(ConstraintsRefStruct);
const ConstraintsRefStructArray = StructType({
  data: ConstraintsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ConstraintsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ConstraintsRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.joint_constraints = JointConstraintWrapper.createArray();
      this._wrapperFields.joint_constraints.copy(other._wrapperFields.joint_constraints);
      this._wrapperFields.position_constraints = PositionConstraintWrapper.createArray();
      this._wrapperFields.position_constraints.copy(other._wrapperFields.position_constraints);
      this._wrapperFields.orientation_constraints = OrientationConstraintWrapper.createArray();
      this._wrapperFields.orientation_constraints.copy(other._wrapperFields.orientation_constraints);
      this._wrapperFields.visibility_constraints = VisibilityConstraintWrapper.createArray();
      this._wrapperFields.visibility_constraints.copy(other._wrapperFields.visibility_constraints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ConstraintsRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.joint_constraints = JointConstraintWrapper.createArray();
    this._wrapperFields.position_constraints = PositionConstraintWrapper.createArray();
    this._wrapperFields.orientation_constraints = OrientationConstraintWrapper.createArray();
    this._wrapperFields.visibility_constraints = VisibilityConstraintWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ConstraintsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ConstraintsArrayWrapper;
  }
  static get ArrayType() {
    return ConstraintsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ConstraintsRefStructArray
  }
  static get refObjectType() {
    return ConstraintsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in Constraints');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    this._wrapperFields.joint_constraints.freeze(own, checkConsistency);
    this._refObject.joint_constraints = this._wrapperFields.joint_constraints.refObject;
    this._wrapperFields.position_constraints.freeze(own, checkConsistency);
    this._refObject.position_constraints = this._wrapperFields.position_constraints.refObject;
    this._wrapperFields.orientation_constraints.freeze(own, checkConsistency);
    this._refObject.orientation_constraints = this._wrapperFields.orientation_constraints.refObject;
    this._wrapperFields.visibility_constraints.freeze(own, checkConsistency);
    this._refObject.visibility_constraints = this._wrapperFields.visibility_constraints.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._wrapperFields.joint_constraints.copyRefObject(refObject.joint_constraints);
    this._wrapperFields.position_constraints.copyRefObject(refObject.position_constraints);
    this._wrapperFields.orientation_constraints.copyRefObject(refObject.orientation_constraints);
    this._wrapperFields.visibility_constraints.copyRefObject(refObject.visibility_constraints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    if (refObject.joint_constraints.size != 0) {
      JointConstraintWrapper.ArrayType.freeArray(refObject.joint_constraints);
      if (JointConstraintWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.joint_constraints, JointConstraintWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.position_constraints.size != 0) {
      PositionConstraintWrapper.ArrayType.freeArray(refObject.position_constraints);
      if (PositionConstraintWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.position_constraints, PositionConstraintWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.orientation_constraints.size != 0) {
      OrientationConstraintWrapper.ArrayType.freeArray(refObject.orientation_constraints);
      if (OrientationConstraintWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.orientation_constraints, OrientationConstraintWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.visibility_constraints.size != 0) {
      VisibilityConstraintWrapper.ArrayType.freeArray(refObject.visibility_constraints);
      if (VisibilityConstraintWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.visibility_constraints, VisibilityConstraintWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ConstraintsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'Constraints'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get joint_constraints() {
    return this._wrapperFields.joint_constraints;
  }
  set joint_constraints(value) {
    this._wrapperFields.joint_constraints.fill(value);
  }
  get position_constraints() {
    return this._wrapperFields.position_constraints;
  }
  set position_constraints(value) {
    this._wrapperFields.position_constraints.fill(value);
  }
  get orientation_constraints() {
    return this._wrapperFields.orientation_constraints;
  }
  set orientation_constraints(value) {
    this._wrapperFields.orientation_constraints.fill(value);
  }
  get visibility_constraints() {
    return this._wrapperFields.visibility_constraints;
  }
  set visibility_constraints(value) {
    this._wrapperFields.visibility_constraints.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ConstraintsRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._wrapperFields.joint_constraints.copyRefObject(this._refObject.joint_constraints);
    this._wrapperFields.position_constraints.copyRefObject(this._refObject.position_constraints);
    this._wrapperFields.orientation_constraints.copyRefObject(this._refObject.orientation_constraints);
    this._wrapperFields.visibility_constraints.copyRefObject(this._refObject.visibility_constraints);
  }
  copy(other) {
    this._refObject = new ConstraintsRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._wrapperFields.joint_constraints.copy(other._wrapperFields.joint_constraints);
    this._wrapperFields.position_constraints.copy(other._wrapperFields.position_constraints);
    this._wrapperFields.orientation_constraints.copy(other._wrapperFields.orientation_constraints);
    this._wrapperFields.visibility_constraints.copy(other._wrapperFields.visibility_constraints);
  }
  static get classType() {
    return ConstraintsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"joint_constraints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"JointConstraint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"position_constraints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"PositionConstraint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"orientation_constraints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"OrientationConstraint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"visibility_constraints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"VisibilityConstraint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"Constraints","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Constraints"};
  }
  hasMember(name) {
    let memberNames = ["name","joint_constraints","position_constraints","orientation_constraints","visibility_constraints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ConstraintsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ConstraintsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ConstraintsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConstraintsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConstraintsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ConstraintsRefArray(size);
    this._refObject = new ConstraintsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ConstraintsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ConstraintsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ConstraintsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ConstraintsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ConstraintsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ConstraintsArrayWrapper;
  }
}
module.exports = ConstraintsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "joint_constraints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "JointConstraint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "position_constraints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "PositionConstraint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "orientation_constraints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "OrientationConstraint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "visibility_constraints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "moveit_msgs",
        "type": "VisibilityConstraint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "Constraints",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Constraints"
}
*/
