// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__AllowedCollisionEntry.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const AllowedCollisionEntryRefStruct = StructType({
  enabled: BoolWrapper.refObjectArrayType,
});
const AllowedCollisionEntryRefArray = ArrayType(AllowedCollisionEntryRefStruct);
const AllowedCollisionEntryRefStructArray = StructType({
  data: AllowedCollisionEntryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AllowedCollisionEntryWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._enabledArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AllowedCollisionEntryRefStruct(other._refObject.toObject());
      this._wrapperFields.enabled = BoolWrapper.createArray();
      this._wrapperFields.enabled.copy(other._wrapperFields.enabled);
      this.enabled = other.enabled;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AllowedCollisionEntryRefStruct();
    this._wrapperFields.enabled = BoolWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new AllowedCollisionEntryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AllowedCollisionEntryArrayWrapper;
  }
  static get ArrayType() {
    return AllowedCollisionEntryArrayWrapper;
  }
  static get refObjectArrayType() {
    return AllowedCollisionEntryRefStructArray
  }
  static get refObjectType() {
    return AllowedCollisionEntryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (!own) {
      this._wrapperFields.enabled.fill(this._enabledArray);
      this._wrapperFields.enabled.freeze(own, checkConsistency);
      this._refObject.enabled = this._wrapperFields.enabled.refObject;
    } else {
      this._wrapperFields.enabled.fill([]);
      this._wrapperFields.enabled.freeze(own, checkConsistency);
      this._refObject.enabled = this._wrapperFields.enabled.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.enabled.data.length = refObject.enabled.size;
    for (let index = 0; index < refObject.enabled.size; index++) {
      this._enabledArray[index] = refObject.enabled.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.enabled.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.enabled);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.enabled, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    AllowedCollisionEntryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'AllowedCollisionEntry'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get enabled() {
    return this._enabledArray;
  }
  set enabled(value) {
    this._enabledArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AllowedCollisionEntryRefStruct(refObject.toObject());
    refObject.enabled.data.length = refObject.enabled.size;
    for (let index = 0; index < refObject.enabled.size; index++) {
      this._enabledArray[index] = refObject.enabled.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new AllowedCollisionEntryRefStruct(other._refObject.toObject());
    this._enabledArray = other._enabledArray.slice();
  }
  static get classType() {
    return AllowedCollisionEntryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"enabled","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"AllowedCollisionEntry","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AllowedCollisionEntry"};
  }
  hasMember(name) {
    let memberNames = ["enabled"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AllowedCollisionEntryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AllowedCollisionEntryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AllowedCollisionEntryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AllowedCollisionEntryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AllowedCollisionEntryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AllowedCollisionEntryRefArray(size);
    this._refObject = new AllowedCollisionEntryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AllowedCollisionEntryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AllowedCollisionEntryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AllowedCollisionEntryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AllowedCollisionEntryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AllowedCollisionEntryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AllowedCollisionEntryArrayWrapper;
  }
}
module.exports = AllowedCollisionEntryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "enabled",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "AllowedCollisionEntry",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AllowedCollisionEntry"
}
*/
