// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__MotionPlanResponse.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const MotionPlanResponseRefStruct = StructType({
  trajectory_start: RobotStateWrapper.refObjectType,
  group_name: primitiveTypes.string,
  trajectory: RobotTrajectoryWrapper.refObjectType,
  planning_time: primitiveTypes.float64,
  error_code: MoveItErrorCodesWrapper.refObjectType,
});
const MotionPlanResponseRefArray = ArrayType(MotionPlanResponseRefStruct);
const MotionPlanResponseRefStructArray = StructType({
  data: MotionPlanResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MotionPlanResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionPlanResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.trajectory_start =  new RobotStateWrapper(other._wrapperFields.trajectory_start);
      this._group_nameIntialized = true;
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
      this._wrapperFields.trajectory =  new RobotTrajectoryWrapper(other._wrapperFields.trajectory);
      this._planning_timeIntialized = true;
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MotionPlanResponseRefStruct();
    this._wrapperFields.trajectory_start = new RobotStateWrapper();
    this._group_nameIntialized = false;
    this._wrapperFields.group_name = new StringWrapper();
    this._wrapperFields.trajectory = new RobotTrajectoryWrapper();
    this._planning_timeIntialized = false;
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MotionPlanResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionPlanResponseArrayWrapper;
  }
  static get ArrayType() {
    return MotionPlanResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionPlanResponseRefStructArray
  }
  static get refObjectType() {
    return MotionPlanResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._group_nameIntialized) {
        throw new TypeError('Invalid argument: group_name in MotionPlanResponse');
      }
      if (checkConsistency && !this._planning_timeIntialized) {
        throw new TypeError('Invalid argument: planning_time in MotionPlanResponse');
      }
    this._wrapperFields.trajectory_start.freeze(own, checkConsistency);
    this._refObject.trajectory_start = this._wrapperFields.trajectory_start.refObject;
    if (own) {
      this._wrapperFields.group_name.freeze(own, checkConsistency);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
    this._wrapperFields.error_code.freeze(own, checkConsistency);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.trajectory_start.copyRefObject(refObject.trajectory_start);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.data = refObject.group_name.data;
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
    this._planning_timeIntialized = true;
    this._refObject.planning_time = refObject.planning_time;
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RobotStateWrapper.freeStruct(refObject.trajectory_start);
    StringWrapper.freeStruct(refObject.group_name);
    RobotTrajectoryWrapper.freeStruct(refObject.trajectory);
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
  }
  static destoryRawROS(msg) {
    MotionPlanResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'MotionPlanResponse'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory_start() {
    return this._wrapperFields.trajectory_start;
  }
  set trajectory_start(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.trajectory_start.copy(value);
    } else {
      this._wrapperFields.trajectory_start.copy(new RobotStateWrapper(value));
    }
  }
  get group_name() {
    if (!this._group_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.data = value;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    if (value instanceof RobotTrajectoryWrapper) {
      this._wrapperFields.trajectory.copy(value);
    } else {
      this._wrapperFields.trajectory.copy(new RobotTrajectoryWrapper(value));
    }
  }
  get planning_time() {
    if (!this._planning_timeIntialized) {
      return undefined;
    }
    return this._refObject.planning_time;
  }
  set planning_time(value) {
    this._planning_timeIntialized = true;
    this._refObject.planning_time = value;
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MotionPlanResponseRefStruct(refObject.toObject());
    this._wrapperFields.trajectory_start.copyRefObject(this._refObject.trajectory_start);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
    this._planning_timeIntialized = true;
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
  }
  copy(other) {
    this._refObject = new MotionPlanResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.trajectory_start.copy(other._wrapperFields.trajectory_start);
    this._group_nameIntialized = true;
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
    this._planning_timeIntialized = true;
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
  }
  static get classType() {
    return MotionPlanResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planning_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MotionPlanResponse","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionPlanResponse"};
  }
  hasMember(name) {
    let memberNames = ["trajectory_start","group_name","trajectory","planning_time","error_code"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MotionPlanResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionPlanResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionPlanResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPlanResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPlanResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MotionPlanResponseRefArray(size);
    this._refObject = new MotionPlanResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionPlanResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionPlanResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionPlanResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionPlanResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionPlanResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionPlanResponseArrayWrapper;
  }
}
module.exports = MotionPlanResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "group_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "RobotTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planning_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "moveit_msgs",
        "type": "MoveItErrorCodes",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "moveit_msgs",
    "type": "MotionPlanResponse",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MotionPlanResponse"
}
*/
