// rclnodejs/generated/moveit_msgs/moveit_msgs__action__LocalPlanner.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class moveit_msgs__action__LocalPlanner {
  static get Goal() {
    return require('./moveit_msgs__action__LocalPlanner_Goal.js');
  }
  static get Result() {
    return require('./moveit_msgs__action__LocalPlanner_Result.js');
  }
  static get Feedback() {
    return require('./moveit_msgs__action__LocalPlanner_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./moveit_msgs__action__LocalPlanner_SendGoal.js');
      },
      get GetResultService() {
        return require('./moveit_msgs__action__LocalPlanner_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./moveit_msgs__action__LocalPlanner_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'LocalPlanner'};
  }
}
module.exports = moveit_msgs__action__LocalPlanner;
