// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__msg__Entity.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const EntityRefStruct = StructType({
  id: primitiveTypes.uint64,
  name: primitiveTypes.string,
  type: primitiveTypes.uint8,
});
const EntityRefArray = ArrayType(EntityRefStruct);
const EntityRefStructArray = StructType({
  data: EntityRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EntityWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EntityRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EntityRefStruct();
    this._idIntialized = false;
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EntityWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EntityArrayWrapper;
  }
  static get ArrayType() {
    return EntityArrayWrapper;
  }
  static get refObjectArrayType() {
    return EntityRefStructArray
  }
  static get refObjectType() {
    return EntityRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in Entity');
      }
      if (checkConsistency && !this._nameIntialized) {
        throw new TypeError('Invalid argument: name in Entity');
      }
      if (checkConsistency && !this._typeIntialized) {
        throw new TypeError('Invalid argument: type in Entity');
      }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destoryRawROS(msg) {
    EntityWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'msg', interfaceName: 'Entity'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EntityRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._typeIntialized = true;
  }
  copy(other) {
    this._refObject = new EntityRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._typeIntialized = true;
  }
  static get classType() {
    return EntityWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"NONE","value":0},{"type":"uint8","name":"LIGHT","value":1},{"type":"uint8","name":"MODEL","value":2},{"type":"uint8","name":"LINK","value":3},{"type":"uint8","name":"VISUAL","value":4},{"type":"uint8","name":"COLLISION","value":5},{"type":"uint8","name":"SENSOR","value":6},{"type":"uint8","name":"JOINT","value":7}],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ros_gz_interfaces","type":"Entity","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Entity"};
  }
  hasMember(name) {
    let memberNames = ["id","name","type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EntityArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EntityWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EntityWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EntityArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EntityArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EntityRefArray(size);
    this._refObject = new EntityRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EntityWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EntityArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EntityArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EntityWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EntityWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EntityArrayWrapper;
  }
}
// Define constants (8 in total)
Object.defineProperty(EntityWrapper, "NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "LIGHT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "MODEL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "LINK", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "VISUAL", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "COLLISION", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "SENSOR", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EntityWrapper, "JOINT", {value: 7, writable: false, enumerable: true, configurable: true});
module.exports = EntityWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "LIGHT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MODEL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "LINK",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "VISUAL",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "COLLISION",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "SENSOR",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "JOINT",
      "value": 7
    }
  ],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "Entity",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Entity"
}
*/
