// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__msg__JointWrench.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const JointWrenchRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  body_1_name: StringWrapper.refObjectType,
  body_1_id: UInt32Wrapper.refObjectType,
  body_2_name: StringWrapper.refObjectType,
  body_2_id: UInt32Wrapper.refObjectType,
  body_1_wrench: WrenchWrapper.refObjectType,
  body_2_wrench: WrenchWrapper.refObjectType,
});
const JointWrenchRefArray = ArrayType(JointWrenchRefStruct);
const JointWrenchRefStructArray = StructType({
  data: JointWrenchRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class JointWrenchWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointWrenchRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.body_1_name =  new StringWrapper(other._wrapperFields.body_1_name);
      this._wrapperFields.body_1_id =  new UInt32Wrapper(other._wrapperFields.body_1_id);
      this._wrapperFields.body_2_name =  new StringWrapper(other._wrapperFields.body_2_name);
      this._wrapperFields.body_2_id =  new UInt32Wrapper(other._wrapperFields.body_2_id);
      this._wrapperFields.body_1_wrench =  new WrenchWrapper(other._wrapperFields.body_1_wrench);
      this._wrapperFields.body_2_wrench =  new WrenchWrapper(other._wrapperFields.body_2_wrench);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new JointWrenchRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.body_1_name = new StringWrapper();
    this._wrapperFields.body_1_id = new UInt32Wrapper();
    this._wrapperFields.body_2_name = new StringWrapper();
    this._wrapperFields.body_2_id = new UInt32Wrapper();
    this._wrapperFields.body_1_wrench = new WrenchWrapper();
    this._wrapperFields.body_2_wrench = new WrenchWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointWrenchWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointWrenchArrayWrapper;
  }
  static get ArrayType() {
    return JointWrenchArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointWrenchRefStructArray
  }
  static get refObjectType() {
    return JointWrenchRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.body_1_name.freeze(own, checkConsistency);
    this._refObject.body_1_name = this._wrapperFields.body_1_name.refObject;
    this._wrapperFields.body_1_id.freeze(own, checkConsistency);
    this._refObject.body_1_id = this._wrapperFields.body_1_id.refObject;
    this._wrapperFields.body_2_name.freeze(own, checkConsistency);
    this._refObject.body_2_name = this._wrapperFields.body_2_name.refObject;
    this._wrapperFields.body_2_id.freeze(own, checkConsistency);
    this._refObject.body_2_id = this._wrapperFields.body_2_id.refObject;
    this._wrapperFields.body_1_wrench.freeze(own, checkConsistency);
    this._refObject.body_1_wrench = this._wrapperFields.body_1_wrench.refObject;
    this._wrapperFields.body_2_wrench.freeze(own, checkConsistency);
    this._refObject.body_2_wrench = this._wrapperFields.body_2_wrench.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.body_1_name.copyRefObject(refObject.body_1_name);
    this._wrapperFields.body_1_id.copyRefObject(refObject.body_1_id);
    this._wrapperFields.body_2_name.copyRefObject(refObject.body_2_name);
    this._wrapperFields.body_2_id.copyRefObject(refObject.body_2_id);
    this._wrapperFields.body_1_wrench.copyRefObject(refObject.body_1_wrench);
    this._wrapperFields.body_2_wrench.copyRefObject(refObject.body_2_wrench);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.body_1_name);
    UInt32Wrapper.freeStruct(refObject.body_1_id);
    StringWrapper.freeStruct(refObject.body_2_name);
    UInt32Wrapper.freeStruct(refObject.body_2_id);
    WrenchWrapper.freeStruct(refObject.body_1_wrench);
    WrenchWrapper.freeStruct(refObject.body_2_wrench);
  }
  static destoryRawROS(msg) {
    JointWrenchWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'msg', interfaceName: 'JointWrench'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get body_1_name() {
    return this._wrapperFields.body_1_name;
  }
  set body_1_name(value) {
    if (value instanceof StringWrapper) {
      this._wrapperFields.body_1_name.copy(value);
    } else {
      this._wrapperFields.body_1_name.copy(new StringWrapper(value));
    }
  }
  get body_1_id() {
    return this._wrapperFields.body_1_id;
  }
  set body_1_id(value) {
    if (value instanceof UInt32Wrapper) {
      this._wrapperFields.body_1_id.copy(value);
    } else {
      this._wrapperFields.body_1_id.copy(new UInt32Wrapper(value));
    }
  }
  get body_2_name() {
    return this._wrapperFields.body_2_name;
  }
  set body_2_name(value) {
    if (value instanceof StringWrapper) {
      this._wrapperFields.body_2_name.copy(value);
    } else {
      this._wrapperFields.body_2_name.copy(new StringWrapper(value));
    }
  }
  get body_2_id() {
    return this._wrapperFields.body_2_id;
  }
  set body_2_id(value) {
    if (value instanceof UInt32Wrapper) {
      this._wrapperFields.body_2_id.copy(value);
    } else {
      this._wrapperFields.body_2_id.copy(new UInt32Wrapper(value));
    }
  }
  get body_1_wrench() {
    return this._wrapperFields.body_1_wrench;
  }
  set body_1_wrench(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.body_1_wrench.copy(value);
    } else {
      this._wrapperFields.body_1_wrench.copy(new WrenchWrapper(value));
    }
  }
  get body_2_wrench() {
    return this._wrapperFields.body_2_wrench;
  }
  set body_2_wrench(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.body_2_wrench.copy(value);
    } else {
      this._wrapperFields.body_2_wrench.copy(new WrenchWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new JointWrenchRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.body_1_name.copyRefObject(this._refObject.body_1_name);
    this._wrapperFields.body_1_id.copyRefObject(this._refObject.body_1_id);
    this._wrapperFields.body_2_name.copyRefObject(this._refObject.body_2_name);
    this._wrapperFields.body_2_id.copyRefObject(this._refObject.body_2_id);
    this._wrapperFields.body_1_wrench.copyRefObject(this._refObject.body_1_wrench);
    this._wrapperFields.body_2_wrench.copyRefObject(this._refObject.body_2_wrench);
  }
  copy(other) {
    this._refObject = new JointWrenchRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.body_1_name.copy(other._wrapperFields.body_1_name);
    this._wrapperFields.body_1_id.copy(other._wrapperFields.body_1_id);
    this._wrapperFields.body_2_name.copy(other._wrapperFields.body_2_name);
    this._wrapperFields.body_2_id.copy(other._wrapperFields.body_2_id);
    this._wrapperFields.body_1_wrench.copy(other._wrapperFields.body_1_wrench);
    this._wrapperFields.body_2_wrench.copy(other._wrapperFields.body_2_wrench);
  }
  static get classType() {
    return JointWrenchWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_1_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"String","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_1_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"UInt32","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_2_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"String","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_2_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"UInt32","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_1_wrench","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"body_2_wrench","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ros_gz_interfaces","type":"JointWrench","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointWrench"};
  }
  hasMember(name) {
    let memberNames = ["header","body_1_name","body_1_id","body_2_name","body_2_id","body_1_wrench","body_2_wrench"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class JointWrenchArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointWrenchWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointWrenchWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointWrenchArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointWrenchArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new JointWrenchRefArray(size);
    this._refObject = new JointWrenchRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointWrenchWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointWrenchArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointWrenchArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointWrenchWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointWrenchWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointWrenchArrayWrapper;
  }
}
module.exports = JointWrenchWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_1_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "String",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_1_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "UInt32",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_2_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "String",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_2_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "UInt32",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_1_wrench",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "body_2_wrench",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "JointWrench",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "JointWrench"
}
*/
