// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__msg__Dataframe.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const DataframeRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  src_address: primitiveTypes.string,
  dst_address: primitiveTypes.string,
  data: UInt8Wrapper.refObjectArrayType,
  rssi: primitiveTypes.float64,
});
const DataframeRefArray = ArrayType(DataframeRefStruct);
const DataframeRefStructArray = StructType({
  data: DataframeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DataframeWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DataframeRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._src_addressIntialized = true;
      this._wrapperFields.src_address =  new StringWrapper(other._wrapperFields.src_address);
      this._dst_addressIntialized = true;
      this._wrapperFields.dst_address =  new StringWrapper(other._wrapperFields.dst_address);
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
      this._rssiIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DataframeRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._src_addressIntialized = false;
    this._wrapperFields.src_address = new StringWrapper();
    this._dst_addressIntialized = false;
    this._wrapperFields.dst_address = new StringWrapper();
    this._wrapperFields.data = UInt8Wrapper.createArray();
    this._rssiIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new DataframeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DataframeArrayWrapper;
  }
  static get ArrayType() {
    return DataframeArrayWrapper;
  }
  static get refObjectArrayType() {
    return DataframeRefStructArray
  }
  static get refObjectType() {
    return DataframeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._src_addressIntialized) {
        throw new TypeError('Invalid argument: src_address in Dataframe');
      }
      if (checkConsistency && !this._dst_addressIntialized) {
        throw new TypeError('Invalid argument: dst_address in Dataframe');
      }
      if (checkConsistency && !this._rssiIntialized) {
        throw new TypeError('Invalid argument: rssi in Dataframe');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.src_address.freeze(own, checkConsistency);
    }
    this._refObject.src_address = this._wrapperFields.src_address.refObject;
    if (own) {
      this._wrapperFields.dst_address.freeze(own, checkConsistency);
    }
    this._refObject.dst_address = this._wrapperFields.dst_address.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Uint8Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._src_addressIntialized = true;
    this._wrapperFields.src_address.data = refObject.src_address.data;
    this._dst_addressIntialized = true;
    this._wrapperFields.dst_address.data = refObject.dst_address.data;
    this._wrapperFields.data.copyRefObject(refObject.data);
    this._rssiIntialized = true;
    this._refObject.rssi = refObject.rssi;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.src_address);
    StringWrapper.freeStruct(refObject.dst_address);
    if (refObject.data.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.data);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    DataframeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'msg', interfaceName: 'Dataframe'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get src_address() {
    if (!this._src_addressIntialized) {
      return undefined;
    }
    return this._wrapperFields.src_address.data;
  }
  set src_address(value) {
    this._src_addressIntialized = true;
    this._wrapperFields.src_address.data = value;
  }
  get dst_address() {
    if (!this._dst_addressIntialized) {
      return undefined;
    }
    return this._wrapperFields.dst_address.data;
  }
  set dst_address(value) {
    this._dst_addressIntialized = true;
    this._wrapperFields.dst_address.data = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  get rssi() {
    if (!this._rssiIntialized) {
      return undefined;
    }
    return this._refObject.rssi;
  }
  set rssi(value) {
    this._rssiIntialized = true;
    this._refObject.rssi = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DataframeRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._src_addressIntialized = true;
    this._wrapperFields.src_address.copyRefObject(this._refObject.src_address);
    this._dst_addressIntialized = true;
    this._wrapperFields.dst_address.copyRefObject(this._refObject.dst_address);
    this._wrapperFields.data.copyRefObject(this._refObject.data);
    this._rssiIntialized = true;
  }
  copy(other) {
    this._refObject = new DataframeRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._src_addressIntialized = true;
    this._wrapperFields.src_address.copy(other._wrapperFields.src_address);
    this._dst_addressIntialized = true;
    this._wrapperFields.dst_address.copy(other._wrapperFields.dst_address);
    this._wrapperFields.data.copy(other._wrapperFields.data);
    this._rssiIntialized = true;
  }
  static get classType() {
    return DataframeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"src_address","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dst_address","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rssi","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ros_gz_interfaces","type":"Dataframe","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Dataframe"};
  }
  hasMember(name) {
    let memberNames = ["header","src_address","dst_address","data","rssi"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DataframeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DataframeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DataframeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DataframeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DataframeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DataframeRefArray(size);
    this._refObject = new DataframeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DataframeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DataframeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DataframeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DataframeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DataframeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DataframeArrayWrapper;
  }
}
module.exports = DataframeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "src_address",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dst_address",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rssi",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "Dataframe",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Dataframe"
}
*/
