// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__msg__VideoRecord.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VideoRecordRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  start: primitiveTypes.bool,
  stop: primitiveTypes.bool,
  format: primitiveTypes.string,
  save_filename: primitiveTypes.string,
});
const VideoRecordRefArray = ArrayType(VideoRecordRefStruct);
const VideoRecordRefStructArray = StructType({
  data: VideoRecordRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VideoRecordWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VideoRecordRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._startIntialized = true;
      this._stopIntialized = true;
      this._formatIntialized = true;
      this._wrapperFields.format =  new StringWrapper(other._wrapperFields.format);
      this._save_filenameIntialized = true;
      this._wrapperFields.save_filename =  new StringWrapper(other._wrapperFields.save_filename);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VideoRecordRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._startIntialized = false;
    this._stopIntialized = false;
    this._formatIntialized = false;
    this._wrapperFields.format = new StringWrapper();
    this._save_filenameIntialized = false;
    this._wrapperFields.save_filename = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new VideoRecordWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VideoRecordArrayWrapper;
  }
  static get ArrayType() {
    return VideoRecordArrayWrapper;
  }
  static get refObjectArrayType() {
    return VideoRecordRefStructArray
  }
  static get refObjectType() {
    return VideoRecordRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._startIntialized) {
        throw new TypeError('Invalid argument: start in VideoRecord');
      }
      if (checkConsistency && !this._stopIntialized) {
        throw new TypeError('Invalid argument: stop in VideoRecord');
      }
      if (checkConsistency && !this._formatIntialized) {
        throw new TypeError('Invalid argument: format in VideoRecord');
      }
      if (checkConsistency && !this._save_filenameIntialized) {
        throw new TypeError('Invalid argument: save_filename in VideoRecord');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.format.freeze(own, checkConsistency);
    }
    this._refObject.format = this._wrapperFields.format.refObject;
    if (own) {
      this._wrapperFields.save_filename.freeze(own, checkConsistency);
    }
    this._refObject.save_filename = this._wrapperFields.save_filename.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._startIntialized = true;
    this._refObject.start = refObject.start;
    this._stopIntialized = true;
    this._refObject.stop = refObject.stop;
    this._formatIntialized = true;
    this._wrapperFields.format.data = refObject.format.data;
    this._save_filenameIntialized = true;
    this._wrapperFields.save_filename.data = refObject.save_filename.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.format);
    StringWrapper.freeStruct(refObject.save_filename);
  }
  static destoryRawROS(msg) {
    VideoRecordWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'msg', interfaceName: 'VideoRecord'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get start() {
    if (!this._startIntialized) {
      return undefined;
    }
    return this._refObject.start;
  }
  set start(value) {
    this._startIntialized = true;
    this._refObject.start = value;
  }
  get stop() {
    if (!this._stopIntialized) {
      return undefined;
    }
    return this._refObject.stop;
  }
  set stop(value) {
    this._stopIntialized = true;
    this._refObject.stop = value;
  }
  get format() {
    if (!this._formatIntialized) {
      return undefined;
    }
    return this._wrapperFields.format.data;
  }
  set format(value) {
    this._formatIntialized = true;
    this._wrapperFields.format.data = value;
  }
  get save_filename() {
    if (!this._save_filenameIntialized) {
      return undefined;
    }
    return this._wrapperFields.save_filename.data;
  }
  set save_filename(value) {
    this._save_filenameIntialized = true;
    this._wrapperFields.save_filename.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VideoRecordRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._startIntialized = true;
    this._stopIntialized = true;
    this._formatIntialized = true;
    this._wrapperFields.format.copyRefObject(this._refObject.format);
    this._save_filenameIntialized = true;
    this._wrapperFields.save_filename.copyRefObject(this._refObject.save_filename);
  }
  copy(other) {
    this._refObject = new VideoRecordRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._startIntialized = true;
    this._stopIntialized = true;
    this._formatIntialized = true;
    this._wrapperFields.format.copy(other._wrapperFields.format);
    this._save_filenameIntialized = true;
    this._wrapperFields.save_filename.copy(other._wrapperFields.save_filename);
  }
  static get classType() {
    return VideoRecordWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"format","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"save_filename","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ros_gz_interfaces","type":"VideoRecord","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VideoRecord"};
  }
  hasMember(name) {
    let memberNames = ["header","start","stop","format","save_filename"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VideoRecordArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VideoRecordWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VideoRecordWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VideoRecordArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VideoRecordArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VideoRecordRefArray(size);
    this._refObject = new VideoRecordRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VideoRecordWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VideoRecordArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VideoRecordArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VideoRecordWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VideoRecordWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VideoRecordArrayWrapper;
  }
}
module.exports = VideoRecordWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "format",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "save_filename",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "VideoRecord",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VideoRecord"
}
*/
