// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__msg__WorldReset.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WorldResetRefStruct = StructType({
  all: primitiveTypes.bool,
  time_only: primitiveTypes.bool,
  model_only: primitiveTypes.bool,
});
const WorldResetRefArray = ArrayType(WorldResetRefStruct);
const WorldResetRefStructArray = StructType({
  data: WorldResetRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WorldResetWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WorldResetRefStruct(other._refObject.toObject());
      this._allIntialized = true;
      this._time_onlyIntialized = true;
      this._model_onlyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WorldResetRefStruct();
    this._refObject.all = false;
    this._allIntialized = true;
    this._refObject.time_only = false;
    this._time_onlyIntialized = true;
    this._refObject.model_only = false;
    this._model_onlyIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new WorldResetWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WorldResetArrayWrapper;
  }
  static get ArrayType() {
    return WorldResetArrayWrapper;
  }
  static get refObjectArrayType() {
    return WorldResetRefStructArray
  }
  static get refObjectType() {
    return WorldResetRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._allIntialized) {
        throw new TypeError('Invalid argument: all in WorldReset');
      }
      if (checkConsistency && !this._time_onlyIntialized) {
        throw new TypeError('Invalid argument: time_only in WorldReset');
      }
      if (checkConsistency && !this._model_onlyIntialized) {
        throw new TypeError('Invalid argument: model_only in WorldReset');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._allIntialized = true;
    this._refObject.all = refObject.all;
    this._time_onlyIntialized = true;
    this._refObject.time_only = refObject.time_only;
    this._model_onlyIntialized = true;
    this._refObject.model_only = refObject.model_only;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    WorldResetWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'msg', interfaceName: 'WorldReset'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get all() {
    if (!this._allIntialized) {
      return undefined;
    }
    return this._refObject.all;
  }
  set all(value) {
    this._allIntialized = true;
    this._refObject.all = value;
  }
  get time_only() {
    if (!this._time_onlyIntialized) {
      return undefined;
    }
    return this._refObject.time_only;
  }
  set time_only(value) {
    this._time_onlyIntialized = true;
    this._refObject.time_only = value;
  }
  get model_only() {
    if (!this._model_onlyIntialized) {
      return undefined;
    }
    return this._refObject.model_only;
  }
  set model_only(value) {
    this._model_onlyIntialized = true;
    this._refObject.model_only = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WorldResetRefStruct(refObject.toObject());
    this._allIntialized = true;
    this._time_onlyIntialized = true;
    this._model_onlyIntialized = true;
  }
  copy(other) {
    this._refObject = new WorldResetRefStruct(other._refObject.toObject());
    this._allIntialized = true;
    this._time_onlyIntialized = true;
    this._model_onlyIntialized = true;
  }
  static get classType() {
    return WorldResetWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"all","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"time_only","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"model_only","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false}],"baseType":{"pkgName":"ros_gz_interfaces","type":"WorldReset","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WorldReset"};
  }
  hasMember(name) {
    let memberNames = ["all","time_only","model_only"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WorldResetArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WorldResetWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WorldResetWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WorldResetArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WorldResetArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WorldResetRefArray(size);
    this._refObject = new WorldResetRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WorldResetWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WorldResetArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WorldResetArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WorldResetWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WorldResetWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WorldResetArrayWrapper;
  }
}
module.exports = WorldResetWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "all",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": false
    },
    {
      "name": "time_only",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": false
    },
    {
      "name": "model_only",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": false
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "WorldReset",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WorldReset"
}
*/
