// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__srv__SetEntityPose_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EntityWrapper = require('../../generated/ros_gz_interfaces/ros_gz_interfaces__msg__Entity.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const SetEntityPose_RequestRefStruct = StructType({
  entity: EntityWrapper.refObjectType,
  pose: PoseWrapper.refObjectType,
});
const SetEntityPose_RequestRefArray = ArrayType(SetEntityPose_RequestRefStruct);
const SetEntityPose_RequestRefStructArray = StructType({
  data: SetEntityPose_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetEntityPose_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetEntityPose_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.entity =  new EntityWrapper(other._wrapperFields.entity);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetEntityPose_RequestRefStruct();
    this._wrapperFields.entity = new EntityWrapper();
    this._wrapperFields.pose = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetEntityPose_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetEntityPose_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetEntityPose_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetEntityPose_RequestRefStructArray
  }
  static get refObjectType() {
    return SetEntityPose_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.entity.freeze(own, checkConsistency);
    this._refObject.entity = this._wrapperFields.entity.refObject;
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.entity.copyRefObject(refObject.entity);
    this._wrapperFields.pose.copyRefObject(refObject.pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    EntityWrapper.freeStruct(refObject.entity);
    PoseWrapper.freeStruct(refObject.pose);
  }
  static destoryRawROS(msg) {
    SetEntityPose_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'srv', interfaceName: 'SetEntityPose_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get entity() {
    return this._wrapperFields.entity;
  }
  set entity(value) {
    if (value instanceof EntityWrapper) {
      this._wrapperFields.entity.copy(value);
    } else {
      this._wrapperFields.entity.copy(new EntityWrapper(value));
    }
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetEntityPose_RequestRefStruct(refObject.toObject());
    this._wrapperFields.entity.copyRefObject(this._refObject.entity);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new SetEntityPose_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.entity.copy(other._wrapperFields.entity);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return SetEntityPose_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"entity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"ros_gz_interfaces","type":"Entity","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ros_gz_interfaces","type":"SetEntityPose_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetEntityPose_Request"};
  }
  hasMember(name) {
    let memberNames = ["entity","pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetEntityPose_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetEntityPose_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetEntityPose_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetEntityPose_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetEntityPose_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetEntityPose_RequestRefArray(size);
    this._refObject = new SetEntityPose_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetEntityPose_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetEntityPose_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetEntityPose_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetEntityPose_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetEntityPose_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetEntityPose_RequestArrayWrapper;
  }
}
module.exports = SetEntityPose_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "entity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "ros_gz_interfaces",
        "type": "Entity",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "SetEntityPose_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetEntityPose_Request"
}
*/
