// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__Dock_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Dock_ResultRefStruct = StructType({
  is_docked: primitiveTypes.bool,
});
const Dock_ResultRefArray = ArrayType(Dock_ResultRefStruct);
const Dock_ResultRefStructArray = StructType({
  data: Dock_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Dock_ResultWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Dock_ResultRefStruct(other._refObject.toObject());
      this._is_dockedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Dock_ResultRefStruct();
    this._is_dockedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new Dock_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Dock_ResultArrayWrapper;
  }
  static get ArrayType() {
    return Dock_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return Dock_ResultRefStructArray
  }
  static get refObjectType() {
    return Dock_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._is_dockedIntialized) {
        throw new TypeError('Invalid argument: is_docked in Dock_Result');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._is_dockedIntialized = true;
    this._refObject.is_docked = refObject.is_docked;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    Dock_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'Dock_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get is_docked() {
    if (!this._is_dockedIntialized) {
      return undefined;
    }
    return this._refObject.is_docked;
  }
  set is_docked(value) {
    this._is_dockedIntialized = true;
    this._refObject.is_docked = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Dock_ResultRefStruct(refObject.toObject());
    this._is_dockedIntialized = true;
  }
  copy(other) {
    this._refObject = new Dock_ResultRefStruct(other._refObject.toObject());
    this._is_dockedIntialized = true;
  }
  static get classType() {
    return Dock_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"is_docked","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"Dock_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Dock_Result"};
  }
  hasMember(name) {
    let memberNames = ["is_docked"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Dock_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Dock_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Dock_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Dock_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Dock_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Dock_ResultRefArray(size);
    this._refObject = new Dock_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Dock_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Dock_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Dock_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Dock_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Dock_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Dock_ResultArrayWrapper;
  }
}
module.exports = Dock_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "is_docked",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "Dock_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Dock_Result"
}
*/
