// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__DriveArc_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DriveArc_GoalRefStruct = StructType({
  translate_direction: primitiveTypes.int8,
  angle: primitiveTypes.float32,
  radius: primitiveTypes.float32,
  max_translation_speed: primitiveTypes.float32,
});
const DriveArc_GoalRefArray = ArrayType(DriveArc_GoalRefStruct);
const DriveArc_GoalRefStructArray = StructType({
  data: DriveArc_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DriveArc_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DriveArc_GoalRefStruct(other._refObject.toObject());
      this._translate_directionIntialized = true;
      this._angleIntialized = true;
      this._radiusIntialized = true;
      this._max_translation_speedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DriveArc_GoalRefStruct();
    this._translate_directionIntialized = false;
    this._angleIntialized = false;
    this._radiusIntialized = false;
    this._refObject.max_translation_speed = 0.3;
    this._max_translation_speedIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new DriveArc_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DriveArc_GoalArrayWrapper;
  }
  static get ArrayType() {
    return DriveArc_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return DriveArc_GoalRefStructArray
  }
  static get refObjectType() {
    return DriveArc_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._translate_directionIntialized) {
        throw new TypeError('Invalid argument: translate_direction in DriveArc_Goal');
      }
      if (checkConsistency && !this._angleIntialized) {
        throw new TypeError('Invalid argument: angle in DriveArc_Goal');
      }
      if (checkConsistency && !this._radiusIntialized) {
        throw new TypeError('Invalid argument: radius in DriveArc_Goal');
      }
      if (checkConsistency && !this._max_translation_speedIntialized) {
        throw new TypeError('Invalid argument: max_translation_speed in DriveArc_Goal');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._translate_directionIntialized = true;
    this._refObject.translate_direction = refObject.translate_direction;
    this._angleIntialized = true;
    this._refObject.angle = refObject.angle;
    this._radiusIntialized = true;
    this._refObject.radius = refObject.radius;
    this._max_translation_speedIntialized = true;
    this._refObject.max_translation_speed = refObject.max_translation_speed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    DriveArc_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'DriveArc_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get translate_direction() {
    if (!this._translate_directionIntialized) {
      return undefined;
    }
    return this._refObject.translate_direction;
  }
  set translate_direction(value) {
    this._translate_directionIntialized = true;
    this._refObject.translate_direction = value;
  }
  get angle() {
    if (!this._angleIntialized) {
      return undefined;
    }
    return this._refObject.angle;
  }
  set angle(value) {
    this._angleIntialized = true;
    this._refObject.angle = value;
  }
  get radius() {
    if (!this._radiusIntialized) {
      return undefined;
    }
    return this._refObject.radius;
  }
  set radius(value) {
    this._radiusIntialized = true;
    this._refObject.radius = value;
  }
  get max_translation_speed() {
    if (!this._max_translation_speedIntialized) {
      return undefined;
    }
    return this._refObject.max_translation_speed;
  }
  set max_translation_speed(value) {
    this._max_translation_speedIntialized = true;
    this._refObject.max_translation_speed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DriveArc_GoalRefStruct(refObject.toObject());
    this._translate_directionIntialized = true;
    this._angleIntialized = true;
    this._radiusIntialized = true;
    this._max_translation_speedIntialized = true;
  }
  copy(other) {
    this._refObject = new DriveArc_GoalRefStruct(other._refObject.toObject());
    this._translate_directionIntialized = true;
    this._angleIntialized = true;
    this._radiusIntialized = true;
    this._max_translation_speedIntialized = true;
  }
  static get classType() {
    return DriveArc_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"TRANSLATE_FORWARD","value":1},{"type":"int8","name":"TRANSLATE_BACKWARD","value":-1}],"fields":[{"name":"translate_direction","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_translation_speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":0.3}],"baseType":{"pkgName":"irobot_create_msgs","type":"DriveArc_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DriveArc_Goal"};
  }
  hasMember(name) {
    let memberNames = ["translate_direction","angle","radius","max_translation_speed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DriveArc_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DriveArc_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DriveArc_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DriveArc_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DriveArc_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DriveArc_GoalRefArray(size);
    this._refObject = new DriveArc_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DriveArc_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DriveArc_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DriveArc_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DriveArc_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DriveArc_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DriveArc_GoalArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(DriveArc_GoalWrapper, "TRANSLATE_FORWARD", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DriveArc_GoalWrapper, "TRANSLATE_BACKWARD", {value: -1, writable: false, enumerable: true, configurable: true});
module.exports = DriveArc_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int8",
      "name": "TRANSLATE_FORWARD",
      "value": 1
    },
    {
      "type": "int8",
      "name": "TRANSLATE_BACKWARD",
      "value": -1
    }
  ],
  "fields": [
    {
      "name": "translate_direction",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "radius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_translation_speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 0.3
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "DriveArc_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DriveArc_Goal"
}
*/
