// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__WheelTicks.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const WheelTicksRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  ticks_left: primitiveTypes.int32,
  ticks_right: primitiveTypes.int32,
});
const WheelTicksRefArray = ArrayType(WheelTicksRefStruct);
const WheelTicksRefStructArray = StructType({
  data: WheelTicksRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WheelTicksWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WheelTicksRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._ticks_leftIntialized = true;
      this._ticks_rightIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WheelTicksRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._ticks_leftIntialized = false;
    this._ticks_rightIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new WheelTicksWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WheelTicksArrayWrapper;
  }
  static get ArrayType() {
    return WheelTicksArrayWrapper;
  }
  static get refObjectArrayType() {
    return WheelTicksRefStructArray
  }
  static get refObjectType() {
    return WheelTicksRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._ticks_leftIntialized) {
        throw new TypeError('Invalid argument: ticks_left in WheelTicks');
      }
      if (checkConsistency && !this._ticks_rightIntialized) {
        throw new TypeError('Invalid argument: ticks_right in WheelTicks');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._ticks_leftIntialized = true;
    this._refObject.ticks_left = refObject.ticks_left;
    this._ticks_rightIntialized = true;
    this._refObject.ticks_right = refObject.ticks_right;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    WheelTicksWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'WheelTicks'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get ticks_left() {
    if (!this._ticks_leftIntialized) {
      return undefined;
    }
    return this._refObject.ticks_left;
  }
  set ticks_left(value) {
    this._ticks_leftIntialized = true;
    this._refObject.ticks_left = value;
  }
  get ticks_right() {
    if (!this._ticks_rightIntialized) {
      return undefined;
    }
    return this._refObject.ticks_right;
  }
  set ticks_right(value) {
    this._ticks_rightIntialized = true;
    this._refObject.ticks_right = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WheelTicksRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._ticks_leftIntialized = true;
    this._ticks_rightIntialized = true;
  }
  copy(other) {
    this._refObject = new WheelTicksRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._ticks_leftIntialized = true;
    this._ticks_rightIntialized = true;
  }
  static get classType() {
    return WheelTicksWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"ticks_left","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ticks_right","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"WheelTicks","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WheelTicks"};
  }
  hasMember(name) {
    let memberNames = ["header","ticks_left","ticks_right"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WheelTicksArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WheelTicksWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WheelTicksWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WheelTicksArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WheelTicksArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WheelTicksRefArray(size);
    this._refObject = new WheelTicksRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WheelTicksWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WheelTicksArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WheelTicksArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WheelTicksWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WheelTicksWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WheelTicksArrayWrapper;
  }
}
module.exports = WheelTicksWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "ticks_left",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ticks_right",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "WheelTicks",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WheelTicks"
}
*/
