// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__LedColor.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LedColorRefStruct = StructType({
  red: primitiveTypes.uint8,
  green: primitiveTypes.uint8,
  blue: primitiveTypes.uint8,
});
const LedColorRefArray = ArrayType(LedColorRefStruct);
const LedColorRefStructArray = StructType({
  data: LedColorRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LedColorWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LedColorRefStruct(other._refObject.toObject());
      this._redIntialized = true;
      this._greenIntialized = true;
      this._blueIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LedColorRefStruct();
    this._redIntialized = false;
    this._greenIntialized = false;
    this._blueIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LedColorWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LedColorArrayWrapper;
  }
  static get ArrayType() {
    return LedColorArrayWrapper;
  }
  static get refObjectArrayType() {
    return LedColorRefStructArray
  }
  static get refObjectType() {
    return LedColorRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._redIntialized) {
        throw new TypeError('Invalid argument: red in LedColor');
      }
      if (checkConsistency && !this._greenIntialized) {
        throw new TypeError('Invalid argument: green in LedColor');
      }
      if (checkConsistency && !this._blueIntialized) {
        throw new TypeError('Invalid argument: blue in LedColor');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._redIntialized = true;
    this._refObject.red = refObject.red;
    this._greenIntialized = true;
    this._refObject.green = refObject.green;
    this._blueIntialized = true;
    this._refObject.blue = refObject.blue;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LedColorWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'LedColor'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get red() {
    if (!this._redIntialized) {
      return undefined;
    }
    return this._refObject.red;
  }
  set red(value) {
    this._redIntialized = true;
    this._refObject.red = value;
  }
  get green() {
    if (!this._greenIntialized) {
      return undefined;
    }
    return this._refObject.green;
  }
  set green(value) {
    this._greenIntialized = true;
    this._refObject.green = value;
  }
  get blue() {
    if (!this._blueIntialized) {
      return undefined;
    }
    return this._refObject.blue;
  }
  set blue(value) {
    this._blueIntialized = true;
    this._refObject.blue = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LedColorRefStruct(refObject.toObject());
    this._redIntialized = true;
    this._greenIntialized = true;
    this._blueIntialized = true;
  }
  copy(other) {
    this._refObject = new LedColorRefStruct(other._refObject.toObject());
    this._redIntialized = true;
    this._greenIntialized = true;
    this._blueIntialized = true;
  }
  static get classType() {
    return LedColorWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"red","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"green","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"blue","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"LedColor","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LedColor"};
  }
  hasMember(name) {
    let memberNames = ["red","green","blue"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LedColorArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LedColorWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LedColorWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LedColorArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LedColorArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LedColorRefArray(size);
    this._refObject = new LedColorRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LedColorWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LedColorArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LedColorArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LedColorWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LedColorWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LedColorArrayWrapper;
  }
}
module.exports = LedColorWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "red",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "green",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "blue",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "LedColor",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LedColor"
}
*/
