// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__LightringLeds.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const LedColorWrapper = require('../../generated/irobot_create_msgs/irobot_create_msgs__msg__LedColor.js');
const LightringLedsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  leds: ArrayType(LedColorWrapper.refObjectType, 6),
  override_system: primitiveTypes.bool,
});
const LightringLedsRefArray = ArrayType(LightringLedsRefStruct);
const LightringLedsRefStructArray = StructType({
  data: LightringLedsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LightringLedsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LightringLedsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.leds = LedColorWrapper.createArray();
      this._wrapperFields.leds.copy(other._wrapperFields.leds);
      this._override_systemIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LightringLedsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.leds = LedColorWrapper.createArray();
    this._override_systemIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LightringLedsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LightringLedsArrayWrapper;
  }
  static get ArrayType() {
    return LightringLedsArrayWrapper;
  }
  static get refObjectArrayType() {
    return LightringLedsRefStructArray
  }
  static get refObjectType() {
    return LightringLedsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._override_systemIntialized) {
        throw new TypeError('Invalid argument: override_system in LightringLeds');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < 6; i++) {
      if (this._wrapperFields.leds.data[i]) {
        this._wrapperFields.leds.data[i].freeze(own, checkConsistency);
        this._refObject.leds[i] = this._wrapperFields.leds.data[i].refObject;
      }
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.leds = refObject.leds;
    this._wrapperFields.leds.size = 6
    for (let i = 0; i < 6; i++) {
      this._wrapperFields.leds.data[i].copyRefObject(refObject.leds[i]);
    }
    this._override_systemIntialized = true;
    this._refObject.override_system = refObject.override_system;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    for (let i = 0; i < 6; i++) {
      LedColorWrapper.freeStruct(refObject.leds[i]);
    }
  }
  static destoryRawROS(msg) {
    LightringLedsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'LightringLeds'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get leds() {
    return this._wrapperFields.leds;
  }
  set leds(value) {
    if (value.length !== 6) {
      throw new RangeError('The length of the array must be 6.');
    }
    this._wrapperFields.leds.fill(value);
  }
  get override_system() {
    if (!this._override_systemIntialized) {
      return undefined;
    }
    return this._refObject.override_system;
  }
  set override_system(value) {
    this._override_systemIntialized = true;
    this._refObject.override_system = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LightringLedsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._refObject.leds = refObject.leds;
    this._wrapperFields.leds.size = 6
    for (let i = 0; i < 6; i++) {
      this._wrapperFields.leds.data[i].copyRefObject(refObject.leds[i]);
    }
    this._override_systemIntialized = true;
  }
  copy(other) {
    this._refObject = new LightringLedsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.leds.copy(other._wrapperFields.leds);
    this._override_systemIntialized = true;
  }
  static get classType() {
    return LightringLedsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"leds","type":{"isArray":true,"arraySize":6,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"irobot_create_msgs","type":"LedColor","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"override_system","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"LightringLeds","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LightringLeds"};
  }
  hasMember(name) {
    let memberNames = ["header","leds","override_system"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LightringLedsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LightringLedsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LightringLedsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LightringLedsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LightringLedsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LightringLedsRefArray(size);
    this._refObject = new LightringLedsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LightringLedsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LightringLedsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LightringLedsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LightringLedsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LightringLedsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LightringLedsArrayWrapper;
  }
}
module.exports = LightringLedsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "leds",
      "type": {
        "isArray": true,
        "arraySize": 6,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": "irobot_create_msgs",
        "type": "LedColor",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "override_system",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "LightringLeds",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LightringLeds"
}
*/
