// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__Button.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const ButtonRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  is_pressed: primitiveTypes.bool,
  last_start_pressed_time: TimeWrapper.refObjectType,
  last_pressed_duration: DurationWrapper.refObjectType,
});
const ButtonRefArray = ArrayType(ButtonRefStruct);
const ButtonRefStructArray = StructType({
  data: ButtonRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ButtonWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ButtonRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._is_pressedIntialized = true;
      this._wrapperFields.last_start_pressed_time =  new TimeWrapper(other._wrapperFields.last_start_pressed_time);
      this._wrapperFields.last_pressed_duration =  new DurationWrapper(other._wrapperFields.last_pressed_duration);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ButtonRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._is_pressedIntialized = false;
    this._wrapperFields.last_start_pressed_time = new TimeWrapper();
    this._wrapperFields.last_pressed_duration = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ButtonWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ButtonArrayWrapper;
  }
  static get ArrayType() {
    return ButtonArrayWrapper;
  }
  static get refObjectArrayType() {
    return ButtonRefStructArray
  }
  static get refObjectType() {
    return ButtonRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._is_pressedIntialized) {
        throw new TypeError('Invalid argument: is_pressed in Button');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.last_start_pressed_time.freeze(own, checkConsistency);
    this._refObject.last_start_pressed_time = this._wrapperFields.last_start_pressed_time.refObject;
    this._wrapperFields.last_pressed_duration.freeze(own, checkConsistency);
    this._refObject.last_pressed_duration = this._wrapperFields.last_pressed_duration.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._is_pressedIntialized = true;
    this._refObject.is_pressed = refObject.is_pressed;
    this._wrapperFields.last_start_pressed_time.copyRefObject(refObject.last_start_pressed_time);
    this._wrapperFields.last_pressed_duration.copyRefObject(refObject.last_pressed_duration);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    TimeWrapper.freeStruct(refObject.last_start_pressed_time);
    DurationWrapper.freeStruct(refObject.last_pressed_duration);
  }
  static destoryRawROS(msg) {
    ButtonWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'Button'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get is_pressed() {
    if (!this._is_pressedIntialized) {
      return undefined;
    }
    return this._refObject.is_pressed;
  }
  set is_pressed(value) {
    this._is_pressedIntialized = true;
    this._refObject.is_pressed = value;
  }
  get last_start_pressed_time() {
    return this._wrapperFields.last_start_pressed_time;
  }
  set last_start_pressed_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.last_start_pressed_time.copy(value);
    } else {
      this._wrapperFields.last_start_pressed_time.copy(new TimeWrapper(value));
    }
  }
  get last_pressed_duration() {
    return this._wrapperFields.last_pressed_duration;
  }
  set last_pressed_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.last_pressed_duration.copy(value);
    } else {
      this._wrapperFields.last_pressed_duration.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ButtonRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._is_pressedIntialized = true;
    this._wrapperFields.last_start_pressed_time.copyRefObject(this._refObject.last_start_pressed_time);
    this._wrapperFields.last_pressed_duration.copyRefObject(this._refObject.last_pressed_duration);
  }
  copy(other) {
    this._refObject = new ButtonRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._is_pressedIntialized = true;
    this._wrapperFields.last_start_pressed_time.copy(other._wrapperFields.last_start_pressed_time);
    this._wrapperFields.last_pressed_duration.copy(other._wrapperFields.last_pressed_duration);
  }
  static get classType() {
    return ButtonWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"is_pressed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"last_start_pressed_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"last_pressed_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"Button","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Button"};
  }
  hasMember(name) {
    let memberNames = ["header","is_pressed","last_start_pressed_time","last_pressed_duration"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ButtonArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ButtonWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ButtonWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ButtonArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ButtonArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ButtonRefArray(size);
    this._refObject = new ButtonRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ButtonWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ButtonArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ButtonArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ButtonWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ButtonWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ButtonArrayWrapper;
  }
}
module.exports = ButtonWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "is_pressed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "last_start_pressed_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "last_pressed_duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "Button",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Button"
}
*/
