// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__DriveDistance_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const DriveDistance_ResultRefStruct = StructType({
  pose: PoseStampedWrapper.refObjectType,
});
const DriveDistance_ResultRefArray = ArrayType(DriveDistance_ResultRefStruct);
const DriveDistance_ResultRefStructArray = StructType({
  data: DriveDistance_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DriveDistance_ResultWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DriveDistance_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new PoseStampedWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DriveDistance_ResultRefStruct();
    this._wrapperFields.pose = new PoseStampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new DriveDistance_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DriveDistance_ResultArrayWrapper;
  }
  static get ArrayType() {
    return DriveDistance_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return DriveDistance_ResultRefStructArray
  }
  static get refObjectType() {
    return DriveDistance_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose.copyRefObject(refObject.pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.pose);
  }
  static destoryRawROS(msg) {
    DriveDistance_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'DriveDistance_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseStampedWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new DriveDistance_ResultRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new DriveDistance_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return DriveDistance_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"DriveDistance_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DriveDistance_Result"};
  }
  hasMember(name) {
    let memberNames = ["pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DriveDistance_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DriveDistance_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DriveDistance_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DriveDistance_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DriveDistance_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DriveDistance_ResultRefArray(size);
    this._refObject = new DriveDistance_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DriveDistance_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DriveDistance_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DriveDistance_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DriveDistance_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DriveDistance_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DriveDistance_ResultArrayWrapper;
  }
}
module.exports = DriveDistance_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "DriveDistance_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DriveDistance_Result"
}
*/
