// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__WheelStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const WheelStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  current_ma_left: primitiveTypes.int16,
  current_ma_right: primitiveTypes.int16,
  pwm_left: primitiveTypes.int16,
  pwm_right: primitiveTypes.int16,
  wheels_enabled: primitiveTypes.bool,
});
const WheelStatusRefArray = ArrayType(WheelStatusRefStruct);
const WheelStatusRefStructArray = StructType({
  data: WheelStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WheelStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WheelStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._current_ma_leftIntialized = true;
      this._current_ma_rightIntialized = true;
      this._pwm_leftIntialized = true;
      this._pwm_rightIntialized = true;
      this._wheels_enabledIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WheelStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._current_ma_leftIntialized = false;
    this._current_ma_rightIntialized = false;
    this._pwm_leftIntialized = false;
    this._pwm_rightIntialized = false;
    this._wheels_enabledIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new WheelStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WheelStatusArrayWrapper;
  }
  static get ArrayType() {
    return WheelStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return WheelStatusRefStructArray
  }
  static get refObjectType() {
    return WheelStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._current_ma_leftIntialized) {
        throw new TypeError('Invalid argument: current_ma_left in WheelStatus');
      }
      if (checkConsistency && !this._current_ma_rightIntialized) {
        throw new TypeError('Invalid argument: current_ma_right in WheelStatus');
      }
      if (checkConsistency && !this._pwm_leftIntialized) {
        throw new TypeError('Invalid argument: pwm_left in WheelStatus');
      }
      if (checkConsistency && !this._pwm_rightIntialized) {
        throw new TypeError('Invalid argument: pwm_right in WheelStatus');
      }
      if (checkConsistency && !this._wheels_enabledIntialized) {
        throw new TypeError('Invalid argument: wheels_enabled in WheelStatus');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._current_ma_leftIntialized = true;
    this._refObject.current_ma_left = refObject.current_ma_left;
    this._current_ma_rightIntialized = true;
    this._refObject.current_ma_right = refObject.current_ma_right;
    this._pwm_leftIntialized = true;
    this._refObject.pwm_left = refObject.pwm_left;
    this._pwm_rightIntialized = true;
    this._refObject.pwm_right = refObject.pwm_right;
    this._wheels_enabledIntialized = true;
    this._refObject.wheels_enabled = refObject.wheels_enabled;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    WheelStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'WheelStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get current_ma_left() {
    if (!this._current_ma_leftIntialized) {
      return undefined;
    }
    return this._refObject.current_ma_left;
  }
  set current_ma_left(value) {
    this._current_ma_leftIntialized = true;
    this._refObject.current_ma_left = value;
  }
  get current_ma_right() {
    if (!this._current_ma_rightIntialized) {
      return undefined;
    }
    return this._refObject.current_ma_right;
  }
  set current_ma_right(value) {
    this._current_ma_rightIntialized = true;
    this._refObject.current_ma_right = value;
  }
  get pwm_left() {
    if (!this._pwm_leftIntialized) {
      return undefined;
    }
    return this._refObject.pwm_left;
  }
  set pwm_left(value) {
    this._pwm_leftIntialized = true;
    this._refObject.pwm_left = value;
  }
  get pwm_right() {
    if (!this._pwm_rightIntialized) {
      return undefined;
    }
    return this._refObject.pwm_right;
  }
  set pwm_right(value) {
    this._pwm_rightIntialized = true;
    this._refObject.pwm_right = value;
  }
  get wheels_enabled() {
    if (!this._wheels_enabledIntialized) {
      return undefined;
    }
    return this._refObject.wheels_enabled;
  }
  set wheels_enabled(value) {
    this._wheels_enabledIntialized = true;
    this._refObject.wheels_enabled = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WheelStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._current_ma_leftIntialized = true;
    this._current_ma_rightIntialized = true;
    this._pwm_leftIntialized = true;
    this._pwm_rightIntialized = true;
    this._wheels_enabledIntialized = true;
  }
  copy(other) {
    this._refObject = new WheelStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._current_ma_leftIntialized = true;
    this._current_ma_rightIntialized = true;
    this._pwm_leftIntialized = true;
    this._pwm_rightIntialized = true;
    this._wheels_enabledIntialized = true;
  }
  static get classType() {
    return WheelStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"current_ma_left","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_ma_right","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pwm_left","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pwm_right","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wheels_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"WheelStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WheelStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","current_ma_left","current_ma_right","pwm_left","pwm_right","wheels_enabled"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WheelStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WheelStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WheelStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WheelStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WheelStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WheelStatusRefArray(size);
    this._refObject = new WheelStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WheelStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WheelStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WheelStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WheelStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WheelStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WheelStatusArrayWrapper;
  }
}
module.exports = WheelStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "current_ma_left",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_ma_right",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pwm_left",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pwm_right",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wheels_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "WheelStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WheelStatus"
}
*/
