// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__DriveDistance.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class irobot_create_msgs__action__DriveDistance {
  static get Goal() {
    return require('./irobot_create_msgs__action__DriveDistance_Goal.js');
  }
  static get Result() {
    return require('./irobot_create_msgs__action__DriveDistance_Result.js');
  }
  static get Feedback() {
    return require('./irobot_create_msgs__action__DriveDistance_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./irobot_create_msgs__action__DriveDistance_SendGoal.js');
      },
      get GetResultService() {
        return require('./irobot_create_msgs__action__DriveDistance_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./irobot_create_msgs__action__DriveDistance_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'DriveDistance'};
  }
}
module.exports = irobot_create_msgs__action__DriveDistance;
