// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__Mouse.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const MouseRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  integrated_x: primitiveTypes.float32,
  integrated_y: primitiveTypes.float32,
  frame_id: primitiveTypes.uint32,
  last_squal: primitiveTypes.uint8,
});
const MouseRefArray = ArrayType(MouseRefStruct);
const MouseRefStructArray = StructType({
  data: MouseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MouseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MouseRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._integrated_xIntialized = true;
      this._integrated_yIntialized = true;
      this._frame_idIntialized = true;
      this._last_squalIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MouseRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._integrated_xIntialized = false;
    this._integrated_yIntialized = false;
    this._frame_idIntialized = false;
    this._last_squalIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MouseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MouseArrayWrapper;
  }
  static get ArrayType() {
    return MouseArrayWrapper;
  }
  static get refObjectArrayType() {
    return MouseRefStructArray
  }
  static get refObjectType() {
    return MouseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._integrated_xIntialized) {
        throw new TypeError('Invalid argument: integrated_x in Mouse');
      }
      if (checkConsistency && !this._integrated_yIntialized) {
        throw new TypeError('Invalid argument: integrated_y in Mouse');
      }
      if (checkConsistency && !this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in Mouse');
      }
      if (checkConsistency && !this._last_squalIntialized) {
        throw new TypeError('Invalid argument: last_squal in Mouse');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._integrated_xIntialized = true;
    this._refObject.integrated_x = refObject.integrated_x;
    this._integrated_yIntialized = true;
    this._refObject.integrated_y = refObject.integrated_y;
    this._frame_idIntialized = true;
    this._refObject.frame_id = refObject.frame_id;
    this._last_squalIntialized = true;
    this._refObject.last_squal = refObject.last_squal;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    MouseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'Mouse'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get integrated_x() {
    if (!this._integrated_xIntialized) {
      return undefined;
    }
    return this._refObject.integrated_x;
  }
  set integrated_x(value) {
    this._integrated_xIntialized = true;
    this._refObject.integrated_x = value;
  }
  get integrated_y() {
    if (!this._integrated_yIntialized) {
      return undefined;
    }
    return this._refObject.integrated_y;
  }
  set integrated_y(value) {
    this._integrated_yIntialized = true;
    this._refObject.integrated_y = value;
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._refObject.frame_id;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._refObject.frame_id = value;
  }
  get last_squal() {
    if (!this._last_squalIntialized) {
      return undefined;
    }
    return this._refObject.last_squal;
  }
  set last_squal(value) {
    this._last_squalIntialized = true;
    this._refObject.last_squal = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MouseRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._integrated_xIntialized = true;
    this._integrated_yIntialized = true;
    this._frame_idIntialized = true;
    this._last_squalIntialized = true;
  }
  copy(other) {
    this._refObject = new MouseRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._integrated_xIntialized = true;
    this._integrated_yIntialized = true;
    this._frame_idIntialized = true;
    this._last_squalIntialized = true;
  }
  static get classType() {
    return MouseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"integrated_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integrated_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"last_squal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"Mouse","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Mouse"};
  }
  hasMember(name) {
    let memberNames = ["header","integrated_x","integrated_y","frame_id","last_squal"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MouseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MouseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MouseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MouseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MouseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MouseRefArray(size);
    this._refObject = new MouseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MouseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MouseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MouseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MouseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MouseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MouseArrayWrapper;
  }
}
module.exports = MouseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "integrated_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integrated_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "last_squal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "Mouse",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Mouse"
}
*/
