// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__RotateAngle_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RotateAngle_GoalRefStruct = StructType({
  angle: primitiveTypes.float32,
  max_rotation_speed: primitiveTypes.float32,
});
const RotateAngle_GoalRefArray = ArrayType(RotateAngle_GoalRefStruct);
const RotateAngle_GoalRefStructArray = StructType({
  data: RotateAngle_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RotateAngle_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RotateAngle_GoalRefStruct(other._refObject.toObject());
      this._angleIntialized = true;
      this._max_rotation_speedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RotateAngle_GoalRefStruct();
    this._angleIntialized = false;
    this._refObject.max_rotation_speed = 1.9;
    this._max_rotation_speedIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new RotateAngle_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RotateAngle_GoalArrayWrapper;
  }
  static get ArrayType() {
    return RotateAngle_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return RotateAngle_GoalRefStructArray
  }
  static get refObjectType() {
    return RotateAngle_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._angleIntialized) {
        throw new TypeError('Invalid argument: angle in RotateAngle_Goal');
      }
      if (checkConsistency && !this._max_rotation_speedIntialized) {
        throw new TypeError('Invalid argument: max_rotation_speed in RotateAngle_Goal');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._angleIntialized = true;
    this._refObject.angle = refObject.angle;
    this._max_rotation_speedIntialized = true;
    this._refObject.max_rotation_speed = refObject.max_rotation_speed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    RotateAngle_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'RotateAngle_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get angle() {
    if (!this._angleIntialized) {
      return undefined;
    }
    return this._refObject.angle;
  }
  set angle(value) {
    this._angleIntialized = true;
    this._refObject.angle = value;
  }
  get max_rotation_speed() {
    if (!this._max_rotation_speedIntialized) {
      return undefined;
    }
    return this._refObject.max_rotation_speed;
  }
  set max_rotation_speed(value) {
    this._max_rotation_speedIntialized = true;
    this._refObject.max_rotation_speed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RotateAngle_GoalRefStruct(refObject.toObject());
    this._angleIntialized = true;
    this._max_rotation_speedIntialized = true;
  }
  copy(other) {
    this._refObject = new RotateAngle_GoalRefStruct(other._refObject.toObject());
    this._angleIntialized = true;
    this._max_rotation_speedIntialized = true;
  }
  static get classType() {
    return RotateAngle_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_rotation_speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":1.9}],"baseType":{"pkgName":"irobot_create_msgs","type":"RotateAngle_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RotateAngle_Goal"};
  }
  hasMember(name) {
    let memberNames = ["angle","max_rotation_speed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RotateAngle_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RotateAngle_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RotateAngle_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAngle_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAngle_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RotateAngle_GoalRefArray(size);
    this._refObject = new RotateAngle_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RotateAngle_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RotateAngle_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RotateAngle_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RotateAngle_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RotateAngle_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RotateAngle_GoalArrayWrapper;
  }
}
module.exports = RotateAngle_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_rotation_speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 1.9
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "RotateAngle_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RotateAngle_Goal"
}
*/
