// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__AudioNoteSequence_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const AudioNoteSequence_FeedbackRefStruct = StructType({
  iterations_played: primitiveTypes.int32,
  current_runtime: DurationWrapper.refObjectType,
});
const AudioNoteSequence_FeedbackRefArray = ArrayType(AudioNoteSequence_FeedbackRefStruct);
const AudioNoteSequence_FeedbackRefStructArray = StructType({
  data: AudioNoteSequence_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AudioNoteSequence_FeedbackWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AudioNoteSequence_FeedbackRefStruct(other._refObject.toObject());
      this._iterations_playedIntialized = true;
      this._wrapperFields.current_runtime =  new DurationWrapper(other._wrapperFields.current_runtime);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AudioNoteSequence_FeedbackRefStruct();
    this._iterations_playedIntialized = false;
    this._wrapperFields.current_runtime = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new AudioNoteSequence_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AudioNoteSequence_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return AudioNoteSequence_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return AudioNoteSequence_FeedbackRefStructArray
  }
  static get refObjectType() {
    return AudioNoteSequence_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._iterations_playedIntialized) {
        throw new TypeError('Invalid argument: iterations_played in AudioNoteSequence_Feedback');
      }
    this._wrapperFields.current_runtime.freeze(own, checkConsistency);
    this._refObject.current_runtime = this._wrapperFields.current_runtime.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._iterations_playedIntialized = true;
    this._refObject.iterations_played = refObject.iterations_played;
    this._wrapperFields.current_runtime.copyRefObject(refObject.current_runtime);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.current_runtime);
  }
  static destoryRawROS(msg) {
    AudioNoteSequence_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'AudioNoteSequence_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get iterations_played() {
    if (!this._iterations_playedIntialized) {
      return undefined;
    }
    return this._refObject.iterations_played;
  }
  set iterations_played(value) {
    this._iterations_playedIntialized = true;
    this._refObject.iterations_played = value;
  }
  get current_runtime() {
    return this._wrapperFields.current_runtime;
  }
  set current_runtime(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.current_runtime.copy(value);
    } else {
      this._wrapperFields.current_runtime.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new AudioNoteSequence_FeedbackRefStruct(refObject.toObject());
    this._iterations_playedIntialized = true;
    this._wrapperFields.current_runtime.copyRefObject(this._refObject.current_runtime);
  }
  copy(other) {
    this._refObject = new AudioNoteSequence_FeedbackRefStruct(other._refObject.toObject());
    this._iterations_playedIntialized = true;
    this._wrapperFields.current_runtime.copy(other._wrapperFields.current_runtime);
  }
  static get classType() {
    return AudioNoteSequence_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"iterations_played","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_runtime","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"AudioNoteSequence_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AudioNoteSequence_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["iterations_played","current_runtime"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AudioNoteSequence_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AudioNoteSequence_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AudioNoteSequence_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AudioNoteSequence_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AudioNoteSequence_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AudioNoteSequence_FeedbackRefArray(size);
    this._refObject = new AudioNoteSequence_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AudioNoteSequence_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AudioNoteSequence_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AudioNoteSequence_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AudioNoteSequence_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AudioNoteSequence_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AudioNoteSequence_FeedbackArrayWrapper;
  }
}
module.exports = AudioNoteSequence_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "iterations_played",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_runtime",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "AudioNoteSequence_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AudioNoteSequence_Feedback"
}
*/
