// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__HazardDetection.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const HazardDetectionRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  type: primitiveTypes.uint8,
});
const HazardDetectionRefArray = ArrayType(HazardDetectionRefStruct);
const HazardDetectionRefStructArray = StructType({
  data: HazardDetectionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HazardDetectionWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HazardDetectionRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HazardDetectionRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HazardDetectionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HazardDetectionArrayWrapper;
  }
  static get ArrayType() {
    return HazardDetectionArrayWrapper;
  }
  static get refObjectArrayType() {
    return HazardDetectionRefStructArray
  }
  static get refObjectType() {
    return HazardDetectionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._typeIntialized) {
        throw new TypeError('Invalid argument: type in HazardDetection');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    HazardDetectionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'HazardDetection'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HazardDetectionRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._typeIntialized = true;
  }
  copy(other) {
    this._refObject = new HazardDetectionRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._typeIntialized = true;
  }
  static get classType() {
    return HazardDetectionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"BACKUP_LIMIT","value":0},{"type":"uint8","name":"BUMP","value":1},{"type":"uint8","name":"CLIFF","value":2},{"type":"uint8","name":"STALL","value":3},{"type":"uint8","name":"WHEEL_DROP","value":4},{"type":"uint8","name":"OBJECT_PROXIMITY","value":5}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"HazardDetection","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HazardDetection"};
  }
  hasMember(name) {
    let memberNames = ["header","type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HazardDetectionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HazardDetectionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HazardDetectionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HazardDetectionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HazardDetectionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HazardDetectionRefArray(size);
    this._refObject = new HazardDetectionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HazardDetectionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HazardDetectionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HazardDetectionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HazardDetectionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HazardDetectionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HazardDetectionArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(HazardDetectionWrapper, "BACKUP_LIMIT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(HazardDetectionWrapper, "BUMP", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(HazardDetectionWrapper, "CLIFF", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(HazardDetectionWrapper, "STALL", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(HazardDetectionWrapper, "WHEEL_DROP", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(HazardDetectionWrapper, "OBJECT_PROXIMITY", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = HazardDetectionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "BACKUP_LIMIT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "BUMP",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "CLIFF",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "STALL",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "WHEEL_DROP",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "OBJECT_PROXIMITY",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "HazardDetection",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HazardDetection"
}
*/
