// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__RotateAngle_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RotateAngle_FeedbackRefStruct = StructType({
  remaining_angle_travel: primitiveTypes.float32,
});
const RotateAngle_FeedbackRefArray = ArrayType(RotateAngle_FeedbackRefStruct);
const RotateAngle_FeedbackRefStructArray = StructType({
  data: RotateAngle_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RotateAngle_FeedbackWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RotateAngle_FeedbackRefStruct(other._refObject.toObject());
      this._remaining_angle_travelIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RotateAngle_FeedbackRefStruct();
    this._remaining_angle_travelIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RotateAngle_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RotateAngle_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return RotateAngle_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return RotateAngle_FeedbackRefStructArray
  }
  static get refObjectType() {
    return RotateAngle_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._remaining_angle_travelIntialized) {
        throw new TypeError('Invalid argument: remaining_angle_travel in RotateAngle_Feedback');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._remaining_angle_travelIntialized = true;
    this._refObject.remaining_angle_travel = refObject.remaining_angle_travel;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    RotateAngle_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'RotateAngle_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get remaining_angle_travel() {
    if (!this._remaining_angle_travelIntialized) {
      return undefined;
    }
    return this._refObject.remaining_angle_travel;
  }
  set remaining_angle_travel(value) {
    this._remaining_angle_travelIntialized = true;
    this._refObject.remaining_angle_travel = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RotateAngle_FeedbackRefStruct(refObject.toObject());
    this._remaining_angle_travelIntialized = true;
  }
  copy(other) {
    this._refObject = new RotateAngle_FeedbackRefStruct(other._refObject.toObject());
    this._remaining_angle_travelIntialized = true;
  }
  static get classType() {
    return RotateAngle_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"remaining_angle_travel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"RotateAngle_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RotateAngle_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["remaining_angle_travel"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RotateAngle_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RotateAngle_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RotateAngle_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAngle_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAngle_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RotateAngle_FeedbackRefArray(size);
    this._refObject = new RotateAngle_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RotateAngle_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RotateAngle_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RotateAngle_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RotateAngle_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RotateAngle_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RotateAngle_FeedbackArrayWrapper;
  }
}
module.exports = RotateAngle_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "remaining_angle_travel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "RotateAngle_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RotateAngle_Feedback"
}
*/
