// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__srv__EStop_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EStop_RequestRefStruct = StructType({
  e_stop_on: primitiveTypes.bool,
});
const EStop_RequestRefArray = ArrayType(EStop_RequestRefStruct);
const EStop_RequestRefStructArray = StructType({
  data: EStop_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EStop_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EStop_RequestRefStruct(other._refObject.toObject());
      this._e_stop_onIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EStop_RequestRefStruct();
    this._e_stop_onIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EStop_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EStop_RequestArrayWrapper;
  }
  static get ArrayType() {
    return EStop_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return EStop_RequestRefStructArray
  }
  static get refObjectType() {
    return EStop_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._e_stop_onIntialized) {
        throw new TypeError('Invalid argument: e_stop_on in EStop_Request');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._e_stop_onIntialized = true;
    this._refObject.e_stop_on = refObject.e_stop_on;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EStop_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'srv', interfaceName: 'EStop_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get e_stop_on() {
    if (!this._e_stop_onIntialized) {
      return undefined;
    }
    return this._refObject.e_stop_on;
  }
  set e_stop_on(value) {
    this._e_stop_onIntialized = true;
    this._refObject.e_stop_on = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EStop_RequestRefStruct(refObject.toObject());
    this._e_stop_onIntialized = true;
  }
  copy(other) {
    this._refObject = new EStop_RequestRefStruct(other._refObject.toObject());
    this._e_stop_onIntialized = true;
  }
  static get classType() {
    return EStop_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"e_stop_on","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"EStop_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EStop_Request"};
  }
  hasMember(name) {
    let memberNames = ["e_stop_on"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EStop_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EStop_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EStop_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EStop_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EStop_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EStop_RequestRefArray(size);
    this._refObject = new EStop_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EStop_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EStop_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EStop_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EStop_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EStop_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EStop_RequestArrayWrapper;
  }
}
module.exports = EStop_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "e_stop_on",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "EStop_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EStop_Request"
}
*/
