// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__InterfaceButtons.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const ButtonWrapper = require('../../generated/irobot_create_msgs/irobot_create_msgs__msg__Button.js');
const InterfaceButtonsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  button_1: ButtonWrapper.refObjectType,
  button_power: ButtonWrapper.refObjectType,
  button_2: ButtonWrapper.refObjectType,
});
const InterfaceButtonsRefArray = ArrayType(InterfaceButtonsRefStruct);
const InterfaceButtonsRefStructArray = StructType({
  data: InterfaceButtonsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class InterfaceButtonsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new InterfaceButtonsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.button_1 =  new ButtonWrapper(other._wrapperFields.button_1);
      this._wrapperFields.button_power =  new ButtonWrapper(other._wrapperFields.button_power);
      this._wrapperFields.button_2 =  new ButtonWrapper(other._wrapperFields.button_2);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new InterfaceButtonsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.button_1 = new ButtonWrapper();
    this._wrapperFields.button_power = new ButtonWrapper();
    this._wrapperFields.button_2 = new ButtonWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new InterfaceButtonsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new InterfaceButtonsArrayWrapper;
  }
  static get ArrayType() {
    return InterfaceButtonsArrayWrapper;
  }
  static get refObjectArrayType() {
    return InterfaceButtonsRefStructArray
  }
  static get refObjectType() {
    return InterfaceButtonsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.button_1.freeze(own, checkConsistency);
    this._refObject.button_1 = this._wrapperFields.button_1.refObject;
    this._wrapperFields.button_power.freeze(own, checkConsistency);
    this._refObject.button_power = this._wrapperFields.button_power.refObject;
    this._wrapperFields.button_2.freeze(own, checkConsistency);
    this._refObject.button_2 = this._wrapperFields.button_2.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.button_1.copyRefObject(refObject.button_1);
    this._wrapperFields.button_power.copyRefObject(refObject.button_power);
    this._wrapperFields.button_2.copyRefObject(refObject.button_2);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    ButtonWrapper.freeStruct(refObject.button_1);
    ButtonWrapper.freeStruct(refObject.button_power);
    ButtonWrapper.freeStruct(refObject.button_2);
  }
  static destoryRawROS(msg) {
    InterfaceButtonsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'InterfaceButtons'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get button_1() {
    return this._wrapperFields.button_1;
  }
  set button_1(value) {
    if (value instanceof ButtonWrapper) {
      this._wrapperFields.button_1.copy(value);
    } else {
      this._wrapperFields.button_1.copy(new ButtonWrapper(value));
    }
  }
  get button_power() {
    return this._wrapperFields.button_power;
  }
  set button_power(value) {
    if (value instanceof ButtonWrapper) {
      this._wrapperFields.button_power.copy(value);
    } else {
      this._wrapperFields.button_power.copy(new ButtonWrapper(value));
    }
  }
  get button_2() {
    return this._wrapperFields.button_2;
  }
  set button_2(value) {
    if (value instanceof ButtonWrapper) {
      this._wrapperFields.button_2.copy(value);
    } else {
      this._wrapperFields.button_2.copy(new ButtonWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new InterfaceButtonsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.button_1.copyRefObject(this._refObject.button_1);
    this._wrapperFields.button_power.copyRefObject(this._refObject.button_power);
    this._wrapperFields.button_2.copyRefObject(this._refObject.button_2);
  }
  copy(other) {
    this._refObject = new InterfaceButtonsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.button_1.copy(other._wrapperFields.button_1);
    this._wrapperFields.button_power.copy(other._wrapperFields.button_power);
    this._wrapperFields.button_2.copy(other._wrapperFields.button_2);
  }
  static get classType() {
    return InterfaceButtonsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"button_1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"irobot_create_msgs","type":"Button","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"button_power","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"irobot_create_msgs","type":"Button","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"button_2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"irobot_create_msgs","type":"Button","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"InterfaceButtons","stringUpperBound":null,"isPrimitiveType":false},"msgName":"InterfaceButtons"};
  }
  hasMember(name) {
    let memberNames = ["header","button_1","button_power","button_2"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class InterfaceButtonsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof InterfaceButtonsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new InterfaceButtonsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InterfaceButtonsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InterfaceButtonsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new InterfaceButtonsRefArray(size);
    this._refObject = new InterfaceButtonsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new InterfaceButtonsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof InterfaceButtonsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "InterfaceButtonsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      InterfaceButtonsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return InterfaceButtonsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return InterfaceButtonsArrayWrapper;
  }
}
module.exports = InterfaceButtonsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "button_1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "irobot_create_msgs",
        "type": "Button",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "button_power",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "irobot_create_msgs",
        "type": "Button",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "button_2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "irobot_create_msgs",
        "type": "Button",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "InterfaceButtons",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "InterfaceButtons"
}
*/
