// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__AudioNote.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const AudioNoteRefStruct = StructType({
  frequency: primitiveTypes.uint16,
  max_runtime: DurationWrapper.refObjectType,
});
const AudioNoteRefArray = ArrayType(AudioNoteRefStruct);
const AudioNoteRefStructArray = StructType({
  data: AudioNoteRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AudioNoteWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AudioNoteRefStruct(other._refObject.toObject());
      this._frequencyIntialized = true;
      this._wrapperFields.max_runtime =  new DurationWrapper(other._wrapperFields.max_runtime);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AudioNoteRefStruct();
    this._frequencyIntialized = false;
    this._wrapperFields.max_runtime = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new AudioNoteWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AudioNoteArrayWrapper;
  }
  static get ArrayType() {
    return AudioNoteArrayWrapper;
  }
  static get refObjectArrayType() {
    return AudioNoteRefStructArray
  }
  static get refObjectType() {
    return AudioNoteRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._frequencyIntialized) {
        throw new TypeError('Invalid argument: frequency in AudioNote');
      }
    this._wrapperFields.max_runtime.freeze(own, checkConsistency);
    this._refObject.max_runtime = this._wrapperFields.max_runtime.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._frequencyIntialized = true;
    this._refObject.frequency = refObject.frequency;
    this._wrapperFields.max_runtime.copyRefObject(refObject.max_runtime);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.max_runtime);
  }
  static destoryRawROS(msg) {
    AudioNoteWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'AudioNote'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frequency() {
    if (!this._frequencyIntialized) {
      return undefined;
    }
    return this._refObject.frequency;
  }
  set frequency(value) {
    this._frequencyIntialized = true;
    this._refObject.frequency = value;
  }
  get max_runtime() {
    return this._wrapperFields.max_runtime;
  }
  set max_runtime(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.max_runtime.copy(value);
    } else {
      this._wrapperFields.max_runtime.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new AudioNoteRefStruct(refObject.toObject());
    this._frequencyIntialized = true;
    this._wrapperFields.max_runtime.copyRefObject(this._refObject.max_runtime);
  }
  copy(other) {
    this._refObject = new AudioNoteRefStruct(other._refObject.toObject());
    this._frequencyIntialized = true;
    this._wrapperFields.max_runtime.copy(other._wrapperFields.max_runtime);
  }
  static get classType() {
    return AudioNoteWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"frequency","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_runtime","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"AudioNote","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AudioNote"};
  }
  hasMember(name) {
    let memberNames = ["frequency","max_runtime"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AudioNoteArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AudioNoteWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AudioNoteWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AudioNoteArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AudioNoteArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AudioNoteRefArray(size);
    this._refObject = new AudioNoteRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AudioNoteWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AudioNoteArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AudioNoteArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AudioNoteWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AudioNoteWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AudioNoteArrayWrapper;
  }
}
module.exports = AudioNoteWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "frequency",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_runtime",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "AudioNote",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AudioNote"
}
*/
