// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__msg__IrOpcode.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const IrOpcodeRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  opcode: primitiveTypes.uint8,
  sensor: primitiveTypes.uint8,
});
const IrOpcodeRefArray = ArrayType(IrOpcodeRefStruct);
const IrOpcodeRefStructArray = StructType({
  data: IrOpcodeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IrOpcodeWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IrOpcodeRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._opcodeIntialized = true;
      this._sensorIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IrOpcodeRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._opcodeIntialized = false;
    this._sensorIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new IrOpcodeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IrOpcodeArrayWrapper;
  }
  static get ArrayType() {
    return IrOpcodeArrayWrapper;
  }
  static get refObjectArrayType() {
    return IrOpcodeRefStructArray
  }
  static get refObjectType() {
    return IrOpcodeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._opcodeIntialized) {
        throw new TypeError('Invalid argument: opcode in IrOpcode');
      }
      if (checkConsistency && !this._sensorIntialized) {
        throw new TypeError('Invalid argument: sensor in IrOpcode');
      }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._opcodeIntialized = true;
    this._refObject.opcode = refObject.opcode;
    this._sensorIntialized = true;
    this._refObject.sensor = refObject.sensor;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    IrOpcodeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'msg', interfaceName: 'IrOpcode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get opcode() {
    if (!this._opcodeIntialized) {
      return undefined;
    }
    return this._refObject.opcode;
  }
  set opcode(value) {
    this._opcodeIntialized = true;
    this._refObject.opcode = value;
  }
  get sensor() {
    if (!this._sensorIntialized) {
      return undefined;
    }
    return this._refObject.sensor;
  }
  set sensor(value) {
    this._sensorIntialized = true;
    this._refObject.sensor = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IrOpcodeRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._opcodeIntialized = true;
    this._sensorIntialized = true;
  }
  copy(other) {
    this._refObject = new IrOpcodeRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._opcodeIntialized = true;
    this._sensorIntialized = true;
  }
  static get classType() {
    return IrOpcodeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"CODE_IR_FORCE_FIELD","value":161},{"type":"uint8","name":"CODE_IR_VIRTUAL_WALL","value":162},{"type":"uint8","name":"CODE_IR_BUOY_GREEN","value":164},{"type":"uint8","name":"CODE_IR_BUOY_RED","value":168},{"type":"uint8","name":"CODE_IR_BUOY_BOTH","value":172},{"type":"uint8","name":"CODE_IR_EVAC_GREEN_FIELD","value":244},{"type":"uint8","name":"CODE_IR_EVAC_RED_FIELD","value":248},{"type":"uint8","name":"CODE_IR_EVAC_BOTH_FIELD","value":252},{"type":"uint8","name":"SENSOR_OMNI","value":0},{"type":"uint8","name":"SENSOR_DIRECTIONAL_FRONT","value":1}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"opcode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"irobot_create_msgs","type":"IrOpcode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IrOpcode"};
  }
  hasMember(name) {
    let memberNames = ["header","opcode","sensor"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IrOpcodeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IrOpcodeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IrOpcodeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IrOpcodeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IrOpcodeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IrOpcodeRefArray(size);
    this._refObject = new IrOpcodeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IrOpcodeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IrOpcodeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IrOpcodeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IrOpcodeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IrOpcodeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IrOpcodeArrayWrapper;
  }
}
// Define constants (10 in total)
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_FORCE_FIELD", {value: 161, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_VIRTUAL_WALL", {value: 162, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_BUOY_GREEN", {value: 164, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_BUOY_RED", {value: 168, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_BUOY_BOTH", {value: 172, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_EVAC_GREEN_FIELD", {value: 244, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_EVAC_RED_FIELD", {value: 248, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "CODE_IR_EVAC_BOTH_FIELD", {value: 252, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "SENSOR_OMNI", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(IrOpcodeWrapper, "SENSOR_DIRECTIONAL_FRONT", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = IrOpcodeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "CODE_IR_FORCE_FIELD",
      "value": 161
    },
    {
      "type": "uint8",
      "name": "CODE_IR_VIRTUAL_WALL",
      "value": 162
    },
    {
      "type": "uint8",
      "name": "CODE_IR_BUOY_GREEN",
      "value": 164
    },
    {
      "type": "uint8",
      "name": "CODE_IR_BUOY_RED",
      "value": 168
    },
    {
      "type": "uint8",
      "name": "CODE_IR_BUOY_BOTH",
      "value": 172
    },
    {
      "type": "uint8",
      "name": "CODE_IR_EVAC_GREEN_FIELD",
      "value": 244
    },
    {
      "type": "uint8",
      "name": "CODE_IR_EVAC_RED_FIELD",
      "value": 248
    },
    {
      "type": "uint8",
      "name": "CODE_IR_EVAC_BOTH_FIELD",
      "value": 252
    },
    {
      "type": "uint8",
      "name": "SENSOR_OMNI",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "SENSOR_DIRECTIONAL_FRONT",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "opcode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensor",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "IrOpcode",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IrOpcode"
}
*/
