// rclnodejs/generated/irobot_create_msgs/irobot_create_msgs__action__NavigateToPosition_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const NavigateToPosition_GoalRefStruct = StructType({
  goal_pose: PoseStampedWrapper.refObjectType,
  achieve_goal_heading: primitiveTypes.bool,
  max_translation_speed: primitiveTypes.float32,
  max_rotation_speed: primitiveTypes.float32,
});
const NavigateToPosition_GoalRefArray = ArrayType(NavigateToPosition_GoalRefStruct);
const NavigateToPosition_GoalRefStructArray = StructType({
  data: NavigateToPosition_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavigateToPosition_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavigateToPosition_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.goal_pose =  new PoseStampedWrapper(other._wrapperFields.goal_pose);
      this._achieve_goal_headingIntialized = true;
      this._max_translation_speedIntialized = true;
      this._max_rotation_speedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavigateToPosition_GoalRefStruct();
    this._wrapperFields.goal_pose = new PoseStampedWrapper();
    this._achieve_goal_headingIntialized = false;
    this._refObject.max_translation_speed = 0.3;
    this._max_translation_speedIntialized = true;
    this._refObject.max_rotation_speed = 1.9;
    this._max_rotation_speedIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new NavigateToPosition_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavigateToPosition_GoalArrayWrapper;
  }
  static get ArrayType() {
    return NavigateToPosition_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavigateToPosition_GoalRefStructArray
  }
  static get refObjectType() {
    return NavigateToPosition_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._achieve_goal_headingIntialized) {
        throw new TypeError('Invalid argument: achieve_goal_heading in NavigateToPosition_Goal');
      }
      if (checkConsistency && !this._max_translation_speedIntialized) {
        throw new TypeError('Invalid argument: max_translation_speed in NavigateToPosition_Goal');
      }
      if (checkConsistency && !this._max_rotation_speedIntialized) {
        throw new TypeError('Invalid argument: max_rotation_speed in NavigateToPosition_Goal');
      }
    this._wrapperFields.goal_pose.freeze(own, checkConsistency);
    this._refObject.goal_pose = this._wrapperFields.goal_pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goal_pose.copyRefObject(refObject.goal_pose);
    this._achieve_goal_headingIntialized = true;
    this._refObject.achieve_goal_heading = refObject.achieve_goal_heading;
    this._max_translation_speedIntialized = true;
    this._refObject.max_translation_speed = refObject.max_translation_speed;
    this._max_rotation_speedIntialized = true;
    this._refObject.max_rotation_speed = refObject.max_rotation_speed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.goal_pose);
  }
  static destoryRawROS(msg) {
    NavigateToPosition_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'irobot_create_msgs', subFolder: 'action', interfaceName: 'NavigateToPosition_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goal_pose() {
    return this._wrapperFields.goal_pose;
  }
  set goal_pose(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.goal_pose.copy(value);
    } else {
      this._wrapperFields.goal_pose.copy(new PoseStampedWrapper(value));
    }
  }
  get achieve_goal_heading() {
    if (!this._achieve_goal_headingIntialized) {
      return undefined;
    }
    return this._refObject.achieve_goal_heading;
  }
  set achieve_goal_heading(value) {
    this._achieve_goal_headingIntialized = true;
    this._refObject.achieve_goal_heading = value;
  }
  get max_translation_speed() {
    if (!this._max_translation_speedIntialized) {
      return undefined;
    }
    return this._refObject.max_translation_speed;
  }
  set max_translation_speed(value) {
    this._max_translation_speedIntialized = true;
    this._refObject.max_translation_speed = value;
  }
  get max_rotation_speed() {
    if (!this._max_rotation_speedIntialized) {
      return undefined;
    }
    return this._refObject.max_rotation_speed;
  }
  set max_rotation_speed(value) {
    this._max_rotation_speedIntialized = true;
    this._refObject.max_rotation_speed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NavigateToPosition_GoalRefStruct(refObject.toObject());
    this._wrapperFields.goal_pose.copyRefObject(this._refObject.goal_pose);
    this._achieve_goal_headingIntialized = true;
    this._max_translation_speedIntialized = true;
    this._max_rotation_speedIntialized = true;
  }
  copy(other) {
    this._refObject = new NavigateToPosition_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.goal_pose.copy(other._wrapperFields.goal_pose);
    this._achieve_goal_headingIntialized = true;
    this._max_translation_speedIntialized = true;
    this._max_rotation_speedIntialized = true;
  }
  static get classType() {
    return NavigateToPosition_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"goal_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"achieve_goal_heading","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_translation_speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":0.3},{"name":"max_rotation_speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":1.9}],"baseType":{"pkgName":"irobot_create_msgs","type":"NavigateToPosition_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavigateToPosition_Goal"};
  }
  hasMember(name) {
    let memberNames = ["goal_pose","achieve_goal_heading","max_translation_speed","max_rotation_speed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavigateToPosition_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavigateToPosition_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavigateToPosition_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateToPosition_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateToPosition_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavigateToPosition_GoalRefArray(size);
    this._refObject = new NavigateToPosition_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavigateToPosition_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavigateToPosition_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavigateToPosition_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavigateToPosition_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavigateToPosition_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavigateToPosition_GoalArrayWrapper;
  }
}
module.exports = NavigateToPosition_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "goal_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "achieve_goal_heading",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_translation_speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 0.3
    },
    {
      "name": "max_rotation_speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 1.9
    }
  ],
  "baseType": {
    "pkgName": "irobot_create_msgs",
    "type": "NavigateToPosition_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavigateToPosition_Goal"
}
*/
